{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Graph module\n",
    "\n",
    "> Provides main class and helpers classes and functions to handle pangenome graph and related data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *\n",
    "from nbdev.export import notebook2script\n",
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exporti\n",
    "\n",
    "import os\n",
    "import glob\n",
    "import json\n",
    "import itertools\n",
    "import warnings\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from skbio.io import read as skbio_read\n",
    "from skbio.metadata import IntervalMetadata\n",
    "from skbio.sequence import DNA\n",
    "\n",
    "from dnasim.IO import writeFASTA\n",
    "from dnasim.simulation import inverseSequence\n",
    "\n",
    "from pangraph_constructor.synteny import processAccession,generatePathsLinks\n",
    "from pangraph_constructor.utils import bidict\n",
    "from pangraph_constructor.tree import TremauxTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "class LinkGetter:\n",
    "    def __init__(self,nodes,links):\n",
    "        self.links = links\n",
    "        self.nodes = nodes\n",
    "    \n",
    "    def __getitem__(self,nodeID):\n",
    "        try:\n",
    "            node = self.links[nodeID]\n",
    "            childrenList = list(itertools.chain(*list(node.values())))\n",
    "            return list(set([el[0] for el in childrenList]))\n",
    "        except KeyError:\n",
    "            if nodeID>0 and nodeID<=len(self.nodes):\n",
    "                return []\n",
    "            else:\n",
    "                raise KeyError(f'There is no node with ID {nodeID}')\n",
    "    def __call__(self):\n",
    "        allLinks = []\n",
    "        for node in range(1,len(self.nodes)+1):\n",
    "            res = self[node]\n",
    "            if res is not None:\n",
    "                allLinks.extend([(node,el) for el in res])\n",
    "        return allLinks\n",
    "#         return list(self)\n",
    "    def __iter__(self):\n",
    "        for i in range(len(self.nodes)):\n",
    "            res = self[i+1]\n",
    "            if res is not None:\n",
    "                yield (i+1,res)\n",
    "            else:\n",
    "                yield (i+1,[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "class GenomeGraph:\n",
    "    def __init__(self,gfaPath=None,\n",
    "                 paths=None,\n",
    "                 nodes=None,nodesData=None,links=None,\n",
    "                 sequenceFiles=None,annotationFiles=None,\n",
    "                 doBack=False,**kwargs):\n",
    "        self.nodes = []\n",
    "        self.nodesData = []\n",
    "        self.forwardLinks = {}\n",
    "        self.paths = []\n",
    "        self.accessions = []\n",
    "        self._nxGraph = None\n",
    "        self.tremauxTree = None\n",
    "        \n",
    "        self.isBack = doBack\n",
    "        \n",
    "        if gfaPath is not None:\n",
    "            self._loadGFA(gfaPath,isGFASeq=kwargs.get('isGFASeq',True))\n",
    "        elif nodes is not None and links is not None and paths is not None:\n",
    "            self.nodes = nodes\n",
    "            if nodesData is not None:\n",
    "                self.nodesData = nodesData\n",
    "            else:\n",
    "                self.nodesData = ['']*len(self.nodes)\n",
    "            self.forwardLinks = links\n",
    "            self.paths = paths\n",
    "        elif paths is not None and sequenceFiles is not None:\n",
    "            self.nodes,self.nodesData,self.forwardLinks = self.graphFromPaths(paths,sequenceFiles) # sequenceFiles can be None\n",
    "            self.paths = paths\n",
    "        elif annotationFiles is not None:\n",
    "            self._graphFromAnnotation(annotationFiles,sequenceFiles,**kwargs)\n",
    "        \n",
    "        self.order = list(range(1,len(self.nodes)+1))\n",
    "        \n",
    "        self.pureForwardlinks = LinkGetter(self.nodes,self.forwardLinks)\n",
    "        \n",
    "        self._pathCount()\n",
    "        \n",
    "        if doBack:\n",
    "            self.backLinks = self._revertLinks(self.forwardLinks)\n",
    "            self.pureBackLinks = LinkGetter(self.nodes,self.backLinks)\n",
    "    \n",
    "    # check edgePaths calculation. Incorrectly calculates the numbers \n",
    "    # (one too many in one of group of incoming edges to each node)\n",
    "    def _pathCount(self):\n",
    "        # self.inPath = []\n",
    "        # self.outPath = []\n",
    "        self.nodePass = [0]*len(self.nodes)\n",
    "        self.nodeStrandPaths = [[0,0] for _ in range(len(self.nodes))]\n",
    "        self.pathStarts = [0]*len(self.nodes)\n",
    "        self.inPath = [0]*len(self.nodes)\n",
    "        self.outPath = [0]*len(self.nodes)\n",
    "        self.edgePaths = {}\n",
    "        \n",
    "        for path in self.paths:\n",
    "            \n",
    "            previousNode = None\n",
    "            for nodeStrand in path:\n",
    "                pathNode = self.nodeNameToID[nodeStrand[:-1]]-1\n",
    "                pathStrand = int(nodeStrand[-1]=='-') # 0 if \"+\", 1 if \"-\"\n",
    "                self.nodePass[pathNode] += 1\n",
    "                self.nodeStrandPaths[pathNode][pathStrand] += 1\n",
    "                self.outPath[pathNode] += 1\n",
    "                self.inPath[pathNode] += 1\n",
    "                if previousNode is not None:\n",
    "                    self.edgePaths.setdefault((previousNode+1,pathNode+1),0)\n",
    "                    self.edgePaths[(previousNode+1,pathNode+1)] += 1\n",
    "                else:\n",
    "                    self.pathStarts[pathNode] += 1\n",
    "                    self.inPath[pathNode] -= 1\n",
    "                previousNode = pathNode\n",
    "            self.outPath[previousNode] -= 1\n",
    "    \n",
    "    def _revertLinks(self,forwardLinks):\n",
    "        backLinks = {}\n",
    "        for fromNome,links in forwardLinks.items():\n",
    "            for fromStrand,strandLinks in links.items():\n",
    "                for item in strandLinks:\n",
    "                    if item[0] not in backLinks.keys():\n",
    "                        backLinks[item[0]] = {}\n",
    "                        backLinks[item[0]][item[1]] = []\n",
    "\n",
    "                    elif item[1] not in forwardLinks[item[0]].keys():\n",
    "                        backLinks[item[0]][item[1]] = []\n",
    "\n",
    "                    backLinks[item[0]][item[1]].append([fromNode,fromStrand])\n",
    "        return backLinks\n",
    "    \n",
    "    def _loadGFA(self,gfaFile,isGFASeq=True):\n",
    "        baseName = os.path.splitext(os.path.basename(gfaFile))[0]\n",
    "        dirPath = os.path.dirname(gfaFile)\n",
    "        jsonFile = f'{dirPath}{os.path.sep}{baseName}.json'\n",
    "        \n",
    "        if os.path.exists(jsonFile):\n",
    "            nodeNames = json.load(open(jsonFile,mode='r'))\n",
    "        else:\n",
    "            nodeNames = None\n",
    "        \n",
    "        prevPathSegment = None\n",
    "        \n",
    "        gfaList = open(gfaFile,mode='r').readlines()\n",
    "        \n",
    "        if gfaList[0][7:-1] == '2.0':\n",
    "            raise NotImplementedError('At the moment only GFA v1 is supported')\n",
    "        elif gfaList[0][7:-1] != '1.0':\n",
    "            warnings.warn('Cannot identify version of GFA. Assuming it is GFA v1. If it is not, expect unpredictable results.',category=ResourceWarning)\n",
    "       \n",
    "        segmentList = [el for el in gfaList if el.lower().startswith('s')]\n",
    "        linkList = [el for el in gfaList if el.lower().startswith('l')]\n",
    "        pathStringsList = [el for el in gfaList if el.lower().startswith('p')]\n",
    "\n",
    "        self.nodeNameToID = {}\n",
    "        \n",
    "        for nodeID,node in enumerate(segmentList):\n",
    "            _,segID,segGFAData = node.rstrip().split(sep='\\t')\n",
    "            if isGFASeq:\n",
    "                segSeq = segGFAData\n",
    "            else:\n",
    "                segSeq = ''\n",
    "            \n",
    "            if nodeNames is not None:\n",
    "                segName = nodeNames[int(segID)-1]\n",
    "            elif isGFASeq:\n",
    "                segName = str(segID)\n",
    "            else:\n",
    "                segName = segGFAData\n",
    "            \n",
    "            self.nodeNameToID[segID] = len(self.nodes)+1\n",
    "            \n",
    "            self.nodes.append(segName)\n",
    "            self.nodesData.append(segSeq)\n",
    "            \n",
    "        for link in linkList:\n",
    "            _,fromNodeID,fromStrand,toNodeID,toStrand,_ = link.rstrip().split(sep='\\t')\n",
    "            fromNode = self.nodeNameToID[fromNodeID]\n",
    "            toNode = self.nodeNameToID[toNodeID]\n",
    "            \n",
    "            curLink = self.forwardLinks.setdefault(fromNode,{})\n",
    "            curStrand = curLink.setdefault(fromStrand,[])\n",
    "            curStrand.append((toNode,toStrand))\n",
    "        \n",
    "        for pathString in pathStringsList:\n",
    "            _,seqID,path,_ = pathString.rstrip().split(sep='\\t')\n",
    "            self.paths.append(path.split(','))\n",
    "            self.accessions.append(seqID)\n",
    "                 \n",
    "    def _graphFromNodesLinks(self,nodes,links):\n",
    "        raise NotImplementedError('Generating graph from nodes and links is not yet implemented.')\n",
    "    \n",
    "    def _graphFromPaths(self,paths,sequenceFiles=None):\n",
    "        raise NotImplementedError('Generating graphs from paths is not implemented yet.')\n",
    "    \n",
    "    def _graphFromAnnotation(self,annotationFiles,sequenceFiles=None,**kwargs):\n",
    "        \n",
    "        try:\n",
    "            fileOrder = kwargs['fileOrder']\n",
    "        except KeyError:\n",
    "            fileOrder = list(range(len(annotationFiles)))\n",
    "            \n",
    "        doUS = kwargs['doUS']\n",
    "        self.usCounter = 0\n",
    "        self.OGList = []\n",
    "        links = self._linksDictToSet(self.forwardLinks)\n",
    "        \n",
    "        for fileNum in fileOrder:\n",
    "            if sequenceFiles is not None:\n",
    "                seqFile = sequenceFiles[fileNum]\n",
    "            else:\n",
    "                seqFile = None\n",
    "            accessionID,genes,sequences = processAccession(annotationFiles[fileNum],seqFile)\n",
    "            try:\n",
    "                seqList = [f'{accessionID}{kwargs[\"seqSuffix\"]}']\n",
    "            except KeyError:\n",
    "                seqList = genes.sequenceID.unique().tolist()\n",
    "                seqList.sort()\n",
    "            path = []\n",
    "            for seqID in seqList:\n",
    "                \n",
    "                p,cigar,usCounter = generatePathsLinks(genes.loc[genes.sequenceID==seqID],sequences,self.OGList,\n",
    "                                                       self.nodes,links,\n",
    "                                                       self.usCounter,doUS=doUS,segmentData=self.nodesData)\n",
    "                path = path + p\n",
    "            self.paths.append(path)\n",
    "            self.accessions.append(accessionID)\n",
    "            \n",
    "        \n",
    "        self.forwardLinks = self._linksSetToDict(links)\n",
    "    \n",
    "    def _linksDictToSet(self,dictLinks):\n",
    "        setLinks = {}\n",
    "        for fromNode,strandLinks in dictLinks.items():\n",
    "            for fromStrand,linkList in strandLinks.items():\n",
    "                for toNode,toStrand in linkList:\n",
    "                    fromNodeStrand = f'{fromNode}{fromStrand}'\n",
    "                    toNodeStrand = f'{toNode}{toStrand}'\n",
    "                    if fromNodeStrand not in setLinks.keys():\n",
    "                        setLinks[fromNodeStrand] = set()\n",
    "                    setLinks[fromNodeStrand].add(toNodeStrand)\n",
    "        return setLinks\n",
    "    \n",
    "    def _linksSetToDict(self,setLinks):\n",
    "        dictLinks = {}\n",
    "        for fromNodeStrand,links in setLinks.items():\n",
    "            fromNode = int(fromNodeStrand[:-2])\n",
    "            fromStrand = fromNodeStrand[-1]\n",
    "            for toNodeStrand in links:\n",
    "                toNode = int(toNodeStrand[:-2])\n",
    "                toStrand = toNodeStrand[-1]\n",
    "                \n",
    "                if fromNode not in dictLinks.keys():\n",
    "                    dictLinks[fromNode] = {}\n",
    "                    dictLinks[fromNode][fromStrand] = []\n",
    "\n",
    "                elif fromStrand not in dictLinks[fromNode].keys():\n",
    "                    dictLinks[fromNode][fromStrand] = []\n",
    "\n",
    "                dictLinks[fromNode][fromStrand].append((toNode,toStrand))\n",
    "        return dictLinks\n",
    "    \n",
    "    def generateTremauxTree(self,byPath=True):\n",
    "        _nxGraph = nx.DiGraph()\n",
    "        \n",
    "        for nodeID,toList in self.pureForwardlinks:\n",
    "            if toList is not None:\n",
    "                edgeList = [(nodeID,toItem) for toItem in toList]\n",
    "                _nxGraph.add_edges_from(edgeList)\n",
    "            \n",
    "        self.tremauxTree = TremauxTree(_nxGraph,self,byPath)\n",
    "        \n",
    "    def addAccessionAnnotation(self,annotationFile,sequenceFile=None):\n",
    "        '''\n",
    "        Ideally, a function should be able to add one accesstion to existing graph. \n",
    "        When implemented, `_graphFromAnnotation` should be using this function.\n",
    "        \n",
    "        '''\n",
    "        pass\n",
    "    \n",
    "    def addLink(self,fromNode,fromStrand,toNode,toStrand):\n",
    "        # Need testing\n",
    "        if fromNode not in self.nodes:\n",
    "            raise IndexError(f'Node {fromNode} is not in nodes list')\n",
    "        if toNode not in self.nodes:\n",
    "            raise IndexError(f'Node {toNode} is not in nodes list')\n",
    "        \n",
    "        if fromNode not in self.forwardLinks.keys():\n",
    "            self.forwardLinks[fromNode] = {}\n",
    "            self.forwardLinks[fromNode][fromStrand] = []\n",
    "            \n",
    "        elif fromStrand not in self.forwardLinks[fromNode].keys():\n",
    "            self.forwardLinks[fromNode][fromStrand] = []\n",
    "\n",
    "        self.forwardLinks[fromNode][fromStrand].append([toNode,toStrand])\n",
    "\n",
    "        if self.isBack:\n",
    "            if toNode not in self.backLinks.keys():\n",
    "                self.backLinks[toNode] = {}\n",
    "                self.backLinks[toNode][toStrand] = []\n",
    "\n",
    "            elif toStrand not in self.backLinks[toNode].keys():\n",
    "                self.backLinks[toNode][toStrand] = []\n",
    "\n",
    "            self.backLinks[toNode][toStrand].append([fromNode,fromStrand])\n",
    "    \n",
    "    def addNode(self,nodeID,data=None):\n",
    "        # Need testing\n",
    "        if nodeID not in self.nodes:\n",
    "            self.nodes.append(nodeID)\n",
    "            self.nodesData[nodeID] = data\n",
    "        else:\n",
    "            warnings.warn(f'You attempted to add node with {nodeID}, but it already exists in the graph. The addition was ignored.')\n",
    "        pass\n",
    "    \n",
    "    def _getStartNodes(self,bubbleNode):\n",
    "        allPaths = list(nx.shortest_path(self.tremauxTree,None,bubbleNode).values())\n",
    "        startList = []\n",
    "        for path in allPaths:\n",
    "            if len(startList)>0:\n",
    "                for node in startList.copy():\n",
    "                    if nx.has_path(self.tremauxTree,node,path[0]):\n",
    "                        continue\n",
    "                    if nx.has_path(self.tremauxTree,path[0],node):\n",
    "                        startList.remove(node)\n",
    "\n",
    "            startList.append(path[0])\n",
    "\n",
    "        return startList\n",
    "    \n",
    "    def treeSort(self,byPath=True,bubblePriorityThreshold=0.5):\n",
    "        \n",
    "        print('Constructing Tremaux tree')\n",
    "        self.generateTremauxTree(byPath)\n",
    "        print('Done!')\n",
    "        queue = []\n",
    "        processed = []\n",
    "        self.order = []\n",
    "        \n",
    "        print('Getting root nodes')\n",
    "        rootNodes = self.tremauxTree.getRootNodes()[::-1]\n",
    "#         self.order.append(rootNode)\n",
    "        startNode = rootNodes.pop()\n",
    "        self.order.append(startNode)\n",
    "        descendantsToAdd = list(self.tremauxTree.edges(startNode))\n",
    "        descendantsToAdd.sort(key=lambda edge: self.edgePaths.get(edge,0))\n",
    "        queue.extend(descendantsToAdd)\n",
    "        \n",
    "        print('Start Loop...')\n",
    "        while len(queue)>0 or len(rootNodes)>0:\n",
    "#             pdb.set_trace()\n",
    "            if len(queue)==0:\n",
    "                try:\n",
    "                    while len(queue)==0:\n",
    "                        startNode = rootNodes.pop()\n",
    "                        descendantsToAdd = list(self.tremauxTree.edges(startNode))\n",
    "                        if len(descendantsToAdd)>0:\n",
    "                            descendantsToAdd.sort(key=lambda edge: self.edgePaths.get(edge,0))\n",
    "                            queue.extend([edge for edge in descendantsToAdd if edge not in processed])\n",
    "                        if startNode not in self.order:\n",
    "                            self.order.append(startNode)\n",
    "                except IndexError:\n",
    "                    break\n",
    "            \n",
    "            startNode,endNode = queue.pop()\n",
    "            \n",
    "            print(f'{len(queue)} - {len(processed)} - {len(self.order)} - {startNode} - {endNode}')\n",
    "            \n",
    "            if (startNode,endNode) in processed:\n",
    "                continue\n",
    "            if (startNode in self.order and endNode in self.order):\n",
    "                if not ((startNode,endNode) in processed):\n",
    "                    processed.append((startNode,endNode))\n",
    "                continue\n",
    "            \n",
    "            endNodeAdded = False\n",
    "            \n",
    "            bubblePriorityQueue = []\n",
    "            bubbleEdges = self.tremauxTree.bubbleEdges.inverse.get(endNode,[])\n",
    "            if len(bubbleEdges)>0:\n",
    "                if byPath:\n",
    "                    bubbleEdgesOrder = np.argsort([self.edgePaths.get((edge,endNode),0) for edge in bubbleEdges])[::-1]\n",
    "                else:\n",
    "                    bubbleEdgesOrder = list(range(len(bubbleEdges)))\n",
    "                # Chech that bubble node is not in order already\n",
    "                # Check edgePaths for all outgoing links (including non-bubbles but not loops), use this bubble \n",
    "                # only if the bubble edge has maximum number of paths.\n",
    "                # if this bubble should be used compare edgePaths for bubble edge and for normal edge to \n",
    "                # current node. If normal edge is larger, then add current node to order and node past LCA \n",
    "                # on alternative path(s) to the queue (to be processed next) and after them the descendandts of \n",
    "                # current node. Otherwise, do not add current node to order, but add nodes past LCA and then (to processed later) current node to queue.\n",
    "                \n",
    "                for bEdge in bubbleEdgesOrder:\n",
    "                    bubbleNode = bubbleEdges[bEdge]\n",
    "                    if bubbleNode in self.order:\n",
    "                        continue\n",
    "                    if byPath:\n",
    "                        curBubbleEdgePaths = self.edgePaths.get((bubbleNode,endNode),0)\n",
    "                        bubbleNodeOutEdges = list(self.tremauxTree.out_edges(bubbleNode)) + \\\n",
    "                                            [(bubbleNode,node) for node in self.tremauxTree.bubbleEdges[bubbleNode]]\n",
    "                        \n",
    "                        bubbleNodeEdgePathsMax = max([self.edgePaths.get(edge,0) for edge in bubbleNodeOutEdges])\n",
    "                        if bubbleNodeEdgePathsMax>curBubbleEdgePaths:\n",
    "                            continue\n",
    "                    \n",
    "                        if self.edgePaths[(bubbleNode,endNode)]/self.outPath[bubbleNode-1]>=bubblePriorityThreshold:\n",
    "                            bubblePriority = True\n",
    "                        else:\n",
    "                            bubblePriority = False\n",
    "                        \n",
    "                    lca = nx.lowest_common_ancestor(self.tremauxTree,endNode,bubbleNode)\n",
    "                    if byPath:\n",
    "                        if lca is None:\n",
    "                            startBubbleNodes = self._getStartNodes(bubbleNode)\n",
    "                            startBubbleNodes.sort(key=lambda node: self.nodePass[node-1])\n",
    "                            lcaDescendantsToProcess = []\n",
    "                            for startBubbleNode in startBubbleNodes: \n",
    "                                snDescendantsToProcess = [(startBubbleNode,node) for node in list(self.tremauxTree[startBubbleNode])\\\n",
    "                                                          if node not in self.order and nx.has_path(self.tremauxTree,node,bubbleNode)]\n",
    "                                snDescendantsToProcess.sort(key=lambda node: self.edgePaths.get((startBubbleNode,node),0))\n",
    "                                lcaDescendantsToProcess.extend(snDescendantsToProcess)\n",
    "                        else: \n",
    "                            lcaDescendantsToProcess = [(lca,node) for node in list(self.tremauxTree[lca]) \\\n",
    "                                                       if node not in self.order and nx.has_path(self.tremauxTree,node,bubbleNode)]\n",
    "                            lcaDescendantsToProcess.sort(key=lambda node: self.edgePaths.get((lca,node),0))\n",
    "                        \n",
    "                        for edge in lcaDescendantsToProcess:\n",
    "                            if edge in queue:\n",
    "                                queue.remove(edge)\n",
    "                        if len(lcaDescendantsToProcess)>0:\n",
    "                            if bubblePriority:\n",
    "                                bubblePriorityQueue.append((bubbleNode,[edge for edge in lcaDescendantsToProcess if edge not in processed]))\n",
    "                            else:\n",
    "                                queue.extend([edge for edge in lcaDescendantsToProcess if edge not in processed])\n",
    "                    else:\n",
    "                        # TODO!!!\n",
    "                        # rewrite pure graph sorting to process alternative branches by adding the top nodes \n",
    "                        # of each branch to the queue for immediate processing.\n",
    "                        pathToAdd = nx.shortest_path(self.tremauxTree,lca,bubbleNode)\n",
    "                        if lca is None:\n",
    "                            allPaths = list(pathToAdd.values())\n",
    "                            allPaths.sort(key=len)\n",
    "                            pathToAdd = allPaths[-1]\n",
    "                        else:\n",
    "                            if endNode not in self.order and byPath:\n",
    "                                self.order.append(endNode)\n",
    "                                endNodeAdded = True\n",
    "                        # WRONG!\n",
    "                        self.order.extend([node for node in pathToAdd if node not in self.order])\n",
    "\n",
    "            \n",
    "            \n",
    "            if len(bubblePriorityQueue)>0:\n",
    "                queue.append((startNode,endNode))#Why would I add the processed block again? Should I add descendant edges of the end node here? \n",
    "                # It should save a lot of wasted time and speed up the process.\n",
    "                bubblePriorityQueue.sort(key=lambda b: self.edgePaths[b[0],endNode])\n",
    "                while len(bubblePriorityQueue)>0:\n",
    "                    bNode,bToAdd = bubblePriorityQueue.pop()\n",
    "                    queue.extend(bToAdd)\n",
    "            else:\n",
    "                processed.append((startNode,endNode))\n",
    "                if endNode not in self.order:\n",
    "                    self.order.append(endNode)\n",
    "                    endNodeAdded = True\n",
    "            \n",
    "                edgesToAdd = [edge for edge in list(self.tremauxTree.edges(endNode)) if edge not in processed and edge not in queue]\n",
    "                edgesToAdd.sort(key=lambda edge: self.edgePaths.get(edge,0))\n",
    "                queue.extend(edgesToAdd)\n",
    "             \n",
    "    def _generateIsInversionDict(self):\n",
    "#         setLinks = self._linksDictToSet(self.forwardLinks)\n",
    "#         uniqueNodeStrands = list(set(list(setLinks.keys())).union(*list(setLinks.values())))\n",
    "        \n",
    "#         uniqueNodeStrands.sort(key=lambda val: int(val[:-1])) # Is it really necessary if we build a dict\n",
    "        \n",
    "#         nodeStrandArray = np.array([(v[:-1],v[-1]) for v in uniqueNodeStrands])\n",
    "#         numStrandsPerNode = np.unique(nodeStrandArray[:,0],return_counts=True)\n",
    "        \n",
    "#         isInversionDict = dict([(int(node),numStrands<=1) for node,numStrands in zip(*numStrandsPerNode)])\n",
    "        \n",
    "        isInversionDict = dict([(int(node+1),(self.nodeStrandPaths[node][1]>self.nodeStrandPaths[node][0])) for node in range(len(self.nodes))])\n",
    "    \n",
    "        return isInversionDict\n",
    "    \n",
    "    def toGFA(self,gfaFile,doSeq=True):\n",
    "        baseName = os.path.splitext(os.path.basename(gfaFile))[0]\n",
    "        dirPath = os.path.dirname(gfaFile)\n",
    "        jsonFile = f'{dirPath}{os.path.sep}nodeNames_{baseName}.json'\n",
    "        \n",
    "        gfaWriter = open(gfaFile,mode='w')\n",
    "        \n",
    "        gfaWriter.write('H\\tVN:Z:1.0\\n')\n",
    "        \n",
    "        translator = {}\n",
    "        nodeNameList = []\n",
    "        \n",
    "        for i,nodeID in enumerate(self.order):\n",
    "            nodeSeq = self.nodesData[nodeID-1]\n",
    "            nodeName = self.nodes[nodeID-1]\n",
    "            if doSeq and len(nodeSeq)>0:\n",
    "                gfaWriter.write(f'S\\t{i+1}\\t{nodeSeq}\\n')\n",
    "                nodeNameList.append(nodeName)\n",
    "            else:\n",
    "                gfaWriter.write(f'S\\t{i+1}\\t{nodeName}\\n')\n",
    "            \n",
    "            translator[nodeID] = i+1\n",
    "        \n",
    "        isInversionDict = self._generateIsInversionDict()\n",
    "        strandInversionDict = {'+':'-','-':'+'}\n",
    "        \n",
    "        for fromNode,strandLinks in self.forwardLinks.items():\n",
    "            for fromStrand,toLinkStrands in strandLinks.items():\n",
    "                for toNode,toStrand in toLinkStrands:\n",
    "                    \n",
    "                    if isInversionDict.get(fromNode,False):\n",
    "                        _fromStrand = strandInversionDict[fromStrand]\n",
    "                    else:\n",
    "                        _fromStrand = fromStrand\n",
    "\n",
    "                    if isInversionDict.get(toNode,False):\n",
    "                        _toStrand = strandInversionDict[toStrand]\n",
    "                    else:\n",
    "                        _toStrand = toStrand\n",
    "                    gfaWriter.write(f'L\\t{translator[fromNode]}\\t{_fromStrand}\\t{translator[toNode]}\\t{_toStrand}\\t*\\n')\n",
    "        \n",
    "        for path,accessionID in zip(self.paths,self.accessions):\n",
    "            newPath = []\n",
    "            for nodeStrand in path:\n",
    "                node = self.nodeNameToID[nodeStrand[:-1]]\n",
    "                if isInversionDict.get(node,False):\n",
    "                    strand = strandInversionDict[nodeStrand[-1]]\n",
    "                else:\n",
    "                    strand = nodeStrand[-1]\n",
    "                newPath.append(f'{translator[node]}{strand}')\n",
    "            gfaWriter.write(f'P\\t{accessionID}\\t{\",\".join(newPath)}\\t*\\n')\n",
    "                \n",
    "        gfaWriter.write('\\n')\n",
    "        gfaWriter.close()\n",
    "        \n",
    "        if len(nodeNameList)==len(self.nodes):\n",
    "            with open(jsonFile,'w') as jsf:\n",
    "                json.dump(nodeNameList,jsf)\n",
    "        \n",
    "    def __iter__(self,forward=True):\n",
    "        print(forward)\n",
    "        for i in range(len(self.nodes)):\n",
    "            yield i+1,self.forwardLinks[i+1]\n",
    "    \n",
    "    def __getitem__(self,ind):\n",
    "        if isinstance(ind,tuple):\n",
    "            nodeID = ind[0]\n",
    "            forward = ind[1]\n",
    "        else:\n",
    "            nodeID = ind\n",
    "            forward = True\n",
    "            \n",
    "        if nodeID>len(self.nodes):\n",
    "            raise IndexError('The nodeID index {nodeID} is out of range')\n",
    "            \n",
    "        if not forward and self.isBack:\n",
    "            try:\n",
    "                return self.backLinks[nodeID]\n",
    "            except KeyError:\n",
    "                return None\n",
    "        else:\n",
    "            try:\n",
    "                return self.forwardLinks[nodeID]\n",
    "            except KeyError:\n",
    "                return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
