{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "description: Provides functionality to construct special synteny graph from pangenome\n",
    "  annotation (and actual sequences).\n",
    "output-file: synteny.html\n",
    "title: Synteny module\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp synteny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| include: false\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| include: false\n",
    "from nbdev import nbdev_export\n",
    "nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exporti\n",
    "import time\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from skbio.io import read as skbio_read\n",
    "from skbio.metadata import IntervalMetadata\n",
    "from skbio.sequence import DNA\n",
    "\n",
    "from pygengraph.utils import bidict,inverseSequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def readTransMap(transMapFile,ATaccessionName='araport',similarityGroupColName='Orthogroup'):\n",
    "    '''\n",
    "    Read trandmap file which is pretty much a tsv file with first column containing orthogroup, or similarity group ID,\n",
    "    and the rest of columns correspond to each accession with unique gene/interval IDs in each columns for the given orthogroup.\n",
    "    If one accession has several genes in the given orthogroup, then they will be in a comma separated list.\n",
    "    \n",
    "    Parameters\n",
    "    ==========\n",
    "    transMapFile: path to the file to read\n",
    "    \n",
    "    ATaccessionName: Which accession to read genes from? Usually used to read reference gene IDs, but can be any accession.\n",
    "                     If it is given a `bidict` object of two way relationship between similarity group IDs and this accession\n",
    "                     gene IDs is returned. Otherwise, a Pandas DataFrame read from the file is returned.\n",
    "    \n",
    "    similarityGroupColName: Column name with similarity group IDs (index column).\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    # reading and preparing transmap in pandas format\n",
    "    transMap = pd.read_csv(transMapFile,delimiter='\\t')\n",
    "    transMap.fillna('',inplace=True)\n",
    "    transMap.rename(columns={similarityGroupColName:'orthogroup'},inplace=True)\n",
    "    transMap['orthogroup'] = transMap.orthogroup.str.rstrip(':')\n",
    "    transMap.set_index('orthogroup',inplace=True)\n",
    "    if ATaccessionName is not None:\n",
    "        ATTransMap = transMap[ATaccessionName]\n",
    "        return bidict({og:ATTransMap[og].split(', ') for og in ATTransMap.index})\n",
    "    else:\n",
    "        return transMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def generateOrder(files, priorityAccession='TIAR10'):\n",
    "    '''\n",
    "    Genrate a list of files and float a file which contains specific priority accession name (or any given string)\n",
    "    on top of the list. \n",
    "    \n",
    "    It is not used any more, but left if will be needed in the future.\n",
    "    '''\n",
    "    idxList = list(range(len(files)))\n",
    "    if priorityAccession is not None:\n",
    "        ind = [idx for idx, file in enumerate(files) if priorityAccession in file][0]\n",
    "        del idxList[ind]\n",
    "        idxList = [ind] + idxList\n",
    "    return idxList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def getIDs(iterator):\n",
    "    idList = []\n",
    "    for interval in iterator:\n",
    "        idList.append(interval.metadata['ID'])\n",
    "    \n",
    "    return idList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def readGFF(gffFile: str):\n",
    "    '''\n",
    "    Function which reads GFF3 file into a dict structure. For large GFF files random search is extremely slow and is impossible to use.\n",
    "    This greatly speed up the process (from hours to seconds).\n",
    "    '''\n",
    "    annotation = skbio_read(gffFile,format='gff3')\n",
    "    annDict = {}\n",
    "\n",
    "    for seqid,intList in annotation:\n",
    "        annDict.setdefault(seqid,IntervalMetadata(None)).merge(intList)\n",
    "        \n",
    "    return annDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def addPangenomePositions(pangenomeFiles):\n",
    "    if pangenomeFiles is None:\n",
    "        return {}\n",
    "    \n",
    "    numFiles = len(pangenomeFiles)\n",
    "    \n",
    "    for pangenomeFile in pangenomeFiles:\n",
    "        pangenAnnDict = readGFF(pangenomeFile)\n",
    "        print(f'Pangenome annotation file {pangenomeFile} read.')\n",
    "\n",
    "        pangenGeneDict = {}\n",
    "        numSeqs = len(pangenAnnDict)\n",
    "\n",
    "        for si,[seqID,pangenAnn] in enumerate(pangenAnnDict.items()):\n",
    "\n",
    "            print(f'Processing pangenome sequence {si+1}/{numSeqs}')\n",
    "            chromosome = ''.join(seqID.split('_')[-2:]) # This specifies how to extract molecule name from pangenome sequence ID. If it changes, should also be changed.\n",
    "            pangenChrDict = pangenGeneDict.setdefault(chromosome,{})\n",
    "            \n",
    "            for gi,gene in enumerate(pangenAnn.query(metadata={'type': 'gene'})):\n",
    "                print(f'\\rProcessing gene number {gi}', end = '')\n",
    "                pangenChrDict[gene.metadata['ID']] = gene.bounds[0]\n",
    "\n",
    "            print()\n",
    "\n",
    "    return pangenGeneDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def processAccessions(annotationFiles, ATmap = None, pangenomeDict = None,\n",
    "                      similarityIDKey = None, similarityIDAssignment = None, \n",
    "                      pangenomeFiles=None, sequenceFilesDict=None, \n",
    "                      seqidJoinSym = '_', ATsplitSym = ','):\n",
    "    '''\n",
    "    This function process only custom annotation for a group of accession with similarity IDs. It also creates separate ATmap and \n",
    "    returns the last number of unmatchedID for consistency. There is a separate function for processing reference annotation.\n",
    "    \n",
    "    `sequenceFilesDict` can either be None or dict with accession IDs as keys and paths to FASTA files with sequences of annotated elements as values.\n",
    "    \n",
    "    `pangenomeFiles`: list[str] or None. If list of strings is provided, then its length should be the same length as `annotationFiles` and should be\n",
    "                    in the same order as in `annotationFiles`, i.e. `pangenomeFiles[i]` should correspond to `annotationFiles[i]`\n",
    "    \n",
    "    `similarityIDAssignment`: str. Can be either 'gene' or 'mRNA'. If it is not any of these an error will be raised.\n",
    "    '''\n",
    "    \n",
    "    genes = {}\n",
    "    unmatchedID = 0\n",
    "    anngrTosimgr = {}\n",
    "    def generateUID(anngr):\n",
    "        nonlocal unmatchedID\n",
    "        nonlocal anngrTosimgr\n",
    "        if anngr in anngrTosimgr:\n",
    "            return anngrTosimgr[anngr]\n",
    "        else:\n",
    "            UID = f'UID{unmatchedID:010d}'\n",
    "            anngrTosimgr[anngr] = UID\n",
    "            unmatchedID += 1\n",
    "            return UID\n",
    "    # id generator f'UID{q:010d}'\n",
    "    \n",
    "    sequenceDict = None\n",
    "    \n",
    "    if ATmap is None:\n",
    "        _ATmap = bidict()\n",
    "    chromosomes = set()\n",
    "    \n",
    "    if pangenomeDict is None:\n",
    "        pangenomeDict = addPangenomePositions(pangenomeFiles)\n",
    "    \n",
    "    overalltime = time.time()\n",
    "    \n",
    "    for fileID,annotationFile in enumerate(annotationFiles):\n",
    "        filetime = time.time()\n",
    "        \n",
    "        print(f'Processing file {annotationFile}...')\n",
    "        \n",
    "        annotationDict = readGFF(annotationFile)\n",
    "        print(f'Annotation file {annotationFile} read.')\n",
    "\n",
    "        accessionID = None\n",
    "        changeAccID = True\n",
    "        \n",
    "        for seqID,annotation in sorted(annotationDict.items()):\n",
    "            \n",
    "            seqtime = time.time()\n",
    "            \n",
    "            proposedAccID,chromosome = seqID.split(seqidJoinSym)\n",
    "            \n",
    "            chromosomes.add(chromosome)\n",
    "            \n",
    "            if accessionID != proposedAccID:\n",
    "                accessionID = proposedAccID\n",
    "                changeAccID = True\n",
    "            \n",
    "            if changeAccID:\n",
    "                sequenceDict = None\n",
    "                if sequenceFilesDict is not None and accessionID in sequenceFilesDict:\n",
    "                    sequenceDict = {}\n",
    "                    sequenceGen = skbio_read(sequenceFileDict[accessionID],format='fasta')\n",
    "                    for seq in sequenceGen:\n",
    "                        sequenceDict[seq.metadata['id']] = bytearray(seq.values).decode()\n",
    "                changeAccID = False\n",
    "\n",
    "            geneInts = list(annotation.query(metadata={'type':'gene'}))\n",
    "            genenum = len(geneInts)\n",
    "\n",
    "            for gi,gene in enumerate(geneInts):\n",
    "                geneID = gene.metadata['ID']\n",
    "                \n",
    "                if gene.metadata['source'] == 'Consensus_tair10':\n",
    "                    orthogroup = geneID\n",
    "                    atNameList = geneID\n",
    "                else:\n",
    "                    if similarityIDAssignment=='gene':\n",
    "                        orthogroup = gene.metadata[similarityIDKey]\n",
    "                    elif similarityIDAssignment=='mRNA':\n",
    "                        mrna = list(annotation.query(metadata={'type': 'mRNA', 'Parent': geneID}))\n",
    "                        # FIX! If no mRNA, do not record gene?\n",
    "                        if len(mrna)>0:\n",
    "                            orthogroup = mrna[0].metadata.get('simgr',generateUID(geneID))\n",
    "                        else:\n",
    "                            orthogroup = generateUID(geneID)\n",
    "                    elif ATmap is not None:\n",
    "                        orthogroup = ATmap.inverse.get(geneID,generateUID(geneID))\n",
    "                    else:\n",
    "                        raise ValueError(f'`similarityIDAssignment` supports only \"gene\" or \"mRNA\", but {similarityIDAssignment} was given and ATmap is not available.')\n",
    "\n",
    "                    if orthogroup is None:\n",
    "                        continue\n",
    "\n",
    "                \n",
    "                if ATmap is None:\n",
    "                    \n",
    "                        if 'AT' in gene.metadata:\n",
    "                            atNameList = gene.metadata['AT'].split(ATsplitSym)\n",
    "                        elif 'Name' in gene.metadata:\n",
    "                            atNameList = gene.metadata['Name'].split(ATsplitSym)\n",
    "                        else:\n",
    "                            atNameList = ['']\n",
    "                \n",
    "                forward = gene.metadata['strand']=='+'\n",
    "                start,end = gene.bounds[0]\n",
    "                \n",
    "                pStart,pEnd = pangenomeDict.get(chromosome,{}).get(geneID,[-2,-1])\n",
    "                \n",
    "                if sequenceDict is not None:\n",
    "                    geneSeq = sequenceDict[seqID][start:end+1]\n",
    "                else:\n",
    "                    geneSeq = ''\n",
    "                \n",
    "                overlaps = getIDs(annotation.query(bounds=[(start,end)],metadata={'type':'gene'}))\n",
    "                genes.setdefault(accessionID,[]).append([geneID,orthogroup,seqID,accessionID,chromosome,forward,start+1,end,pStart+1,pEnd,geneSeq,overlaps])\n",
    "                if ATmap is None:\n",
    "                    if len(atNameList[0])>0:\n",
    "                        _ATmap.add(orthogroup,atNameList)\n",
    "        print(f'Processing file {annotationFile} finished in {time.time() - filetime} seconds.')\n",
    "    \n",
    "    print('Converting gene tables into pandas DataFrames.')\n",
    "    \n",
    "    for accID in genes.keys():\n",
    "        genes[accID] = pd.DataFrame(genes[accID],columns=['geneID','orthogroup','sequenceID','accessionID','chromosome','forward','start','end','pstart','pend','geneSeq','overlapGenes'])\n",
    "        genes[accID].sort_values(by=['chromosome','start'],inplace=True)\n",
    "\n",
    "    print(f'Reading annotation files finished in {time.time() - overalltime} seconds')\n",
    "    \n",
    "    if ATmap is None:\n",
    "        return genes,chromosomes,_ATmap,pangenomeDict,sequenceDict\n",
    "    else:\n",
    "        return genes,chromosomes,pangenomeDict,sequenceDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def recordSegment(name,segmentIDs,segmentIDToNumDict,sequence=None,gfaFile=None,segmentData=None):\n",
    "    segmentIDs.append(name)\n",
    "    \n",
    "    segmentIDToNumDict[name] = len(segmentIDs)-1\n",
    "    segID = len(segmentIDs)\n",
    "    \n",
    "    if segmentData is not None and sequence is not None:\n",
    "        segmentData.append(sequence)\n",
    "    \n",
    "    if gfaFile is not None:\n",
    "        if sequence is not None:\n",
    "            gfaFile.write(f'S\\t{segID}\\t{sequence}\\n')        \n",
    "        else:\n",
    "            gfaFile.write(f'S\\t{segID}\\t{name}\\n')\n",
    "    return segID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def recordAnnotation(nodeID,accessionID,sequenceID,chrID,start,end,og,atList,sequence,nodesMetadata,pstart=-1,pend=-1):\n",
    "\n",
    "    if len(nodesMetadata)==nodeID-1:\n",
    "        nodesMetadata.append({})\n",
    "\n",
    "    geneLen = 1\n",
    "    if len(sequence)>0:\n",
    "        geneLen = len(sequence)\n",
    "\n",
    "    nodesMetadata[nodeID-1].setdefault(accessionID, {}).setdefault('posisions',[]).append({'gen':{'chr':chrID, 'genomePosition':[start,end]}})\n",
    "    \n",
    "    if pstart != -1 and pend != -1:\n",
    "        nodesMetadata[nodeID-1][accessionID]['positions'][-1]['pangen'] = {'chr':chrID, 'genomePosition':[pstart,pend]}\n",
    "    \n",
    "    nodesMetadata[nodeID-1][accessionID].setdefault('annotation',{}).update({og: [(0,geneLen-1)]})\n",
    "    nodesMetadata[nodeID-1][accessionID]['annotation'].update({at:[(0,geneLen-1)] for at in atList})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def recordAltChr(nodeID,accessionID,chrID,start,end,nodesMetadata):\n",
    "    nodesMetadata[nodeID-1][accessionID].setdefault('altChrGenPos',[]).append({'chr':chrID, 'genomePosition':[start,end]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def addLink(links,prevPathSegment,name,forward):\n",
    "    '''\n",
    "    `links`: mutable\n",
    "    `prevPathSegment`: mutable\n",
    "    '''\n",
    "    if prevPathSegment is not None:\n",
    "        links[prevPathSegment].add(f'{name}\\t{\"+\" if forward else \"-\"}')\n",
    "    return f'{name}\\t{\"+\" if forward else \"-\"}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def generatePathsLinks(genesAll,ATmap,accessionID,\n",
    "                       sequences,OGList,segmentIDs,\n",
    "                       nodesMetadata,\n",
    "                       segmentIDToNumDict,links,usCounter,\n",
    "                       chromosomeID = None,\n",
    "                       doUS=True,segmentData=None,gfaFile=None):\n",
    "    '''\n",
    "    This function takes a list of genes in specific format (`genesAll`) and some extra data and pretty much generates a graph \n",
    "    (gene graph from annotations).\n",
    "    \n",
    "    `gfaFile`: file handle to write segments to GFA file\n",
    "    `OGList`: mutable\n",
    "    `links`: mutable\n",
    "    `usCounter`: mutable\n",
    "    \n",
    "    '''\n",
    "    print('Converting gene table into graph...')\n",
    "    overalltime = time.time()\n",
    "    \n",
    "    genes = genesAll.loc[genesAll.chromosome == chromosomeID]\n",
    "    \n",
    "    path = []\n",
    "    cigar = []\n",
    "    prevEnd = 0\n",
    "    prevPathSegment = None\n",
    "    curSeqID = ''\n",
    "    numGenes = len(genes)\n",
    "    for gi,[generow,gene] in enumerate(genes.iterrows()):\n",
    "        print(f'\\rProcessing genes {gi+1}/{numGenes}', end='')\n",
    "        og = gene.orthogroup\n",
    "        \n",
    "        geneSeqID = gene.sequenceID\n",
    "        if curSeqID != geneSeqID:\n",
    "            curSeqID = geneSeqID\n",
    "        \n",
    "        atStr = ATmap.get(og,[''])\n",
    "        if len(atStr[0])>0:\n",
    "            atList = atStr\n",
    "        else:\n",
    "            atList = []\n",
    "        \n",
    "        geneChr = gene.chromosome\n",
    "        geneStart = gene.start\n",
    "        geneEnd = gene.end\n",
    "        genePStart = gene.pstart\n",
    "        genePEnd = gene.pend\n",
    "        geneForward = gene.forward\n",
    "        \n",
    "        if sequences is not None:\n",
    "            if geneForward:\n",
    "                geneSeq = sequences[geneSeqID][geneStart-1:geneEnd]\n",
    "            else:\n",
    "                geneSeq = inverseSequence(sequences[geneSeqID][geneStart-1:geneEnd])\n",
    "        else:\n",
    "            geneSeq = ''\n",
    "        \n",
    "        if doUS:\n",
    "        \n",
    "            if sequences is not None:\n",
    "                usSeq = sequences[geneSeqID][prevEnd-1:geneStart]\n",
    "            else:\n",
    "                usSeq = ''\n",
    "\n",
    "            if len(usSeq)>0:\n",
    "                isUS = True\n",
    "                us = f'US{usCounter:07d}'\n",
    "                usCounter += 1\n",
    "            else:\n",
    "                isUS = False\n",
    "\n",
    "            if isUS:\n",
    "                usID = recordSegment(us,segmentIDs,segmentIDToNumDict,usSeq,gfaFile=gfaFile,segmentData=segmentData)\n",
    "                recordAnnotation(usID,accessionID,geneSeqID,geneChr,prevEnd,geneStart-1,us,[],usSeq,nodesMetadata)\n",
    "        \n",
    "        if og not in OGList:\n",
    "            ogID = recordSegment(og,segmentIDs,segmentIDToNumDict,geneSeq,gfaFile=gfaFile,segmentData=segmentData)\n",
    "            OGList.append(og)\n",
    "        else:\n",
    "            ogID = segmentIDs.index(og)+1\n",
    "        \n",
    "        recordAnnotation(ogID,accessionID,geneSeqID,geneChr,geneStart,geneEnd,og,atList,geneSeq,nodesMetadata)\n",
    "        if chromosomeID is not None:\n",
    "            altPos = genesAll.loc[(genesAll.orthogroup == og) & (genesAll.chromosome != chromosomeID)]\n",
    "            for altrow,altChrOG in altPos.iterrows():\n",
    "                recordAltChr(ogID,accessionID,altChrOG.chromosome,altChrOG.start,altChrOG.end,nodesMetadata)\n",
    "        \n",
    "        pathAdd = [f'{ogID}{\"+\" if geneForward else \"-\"}']\n",
    "        if doUS and isUS:\n",
    "            pathAdd.insert(0,f'{usID}+')\n",
    "            \n",
    "        path.extend(pathAdd)\n",
    "        \n",
    "        if len(cigar)>0 and doUS and isUS:\n",
    "            cigar.extend(['0M','0M']) # with previous block and between two current blocks\n",
    "        else:\n",
    "            cigar.append('0M') # only between current blocks or between previous and current gene\n",
    "                               # without unrelated sequence (intergenic) block.\n",
    "        \n",
    "        if doUS and isUS:\n",
    "            prevPathSegment = addLink(links,prevPathSegment,usID,True)\n",
    "            links[prevPathSegment] = set()\n",
    "        \n",
    "        prevPathSegment = addLink(links,prevPathSegment,ogID,geneForward)\n",
    "        if prevPathSegment not in links:\n",
    "            links[prevPathSegment] = set()\n",
    "        \n",
    "        prevEnd = geneEnd+1\n",
    "    \n",
    "    if doUS:\n",
    "        \n",
    "        if sequences is not None:\n",
    "            usSeq = sequences[curSeqID][prevEnd-1:]\n",
    "        else:\n",
    "            usSeq = ''\n",
    "\n",
    "        if len(usSeq)>0:\n",
    "            us = f'US{usCounter:07d}'\n",
    "            usID = recordSegment(us,segmentIDs,segmentIDToNumDict,usSeq,gfaFile=gfaFile,segmentData=segmentData)\n",
    "            recordAnnotation(usID,accessionID,geneSeqID,prevEnd,len(sequences[curSeqID]),us,[],nodesMetadata)\n",
    "            usCounter += 1\n",
    "            path.append(f'{usID}+')\n",
    "            cigar.append('0M')\n",
    "            prevPathSegment = addLink(links,prevPathSegment,usID,True)\n",
    "    \n",
    "    print(f'\\nProcessing gene table finished in {time.time() - overalltime}')\n",
    "    \n",
    "    return path,cigar,usCounter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def writeLinks(gfaFile,links,doCigars=True):\n",
    "    for linkLeft,linksRight in links.items():\n",
    "        for linkRight in linksRight:\n",
    "            if doCigars:\n",
    "                gfaFile.write(f'L\\t{linkLeft}\\t{linkRight}\\t0M\\n')\n",
    "            else:\n",
    "                gfaFile.write(f'L\\t{linkLeft}\\t{linkRight}\\t*\\n')\n",
    "                \n",
    "def writePath(gfaFile,AccessionID,path,cigar,doCigars):   \n",
    "    if doCigars:\n",
    "        cigarString = \",\".join(cigar)\n",
    "    else:\n",
    "        cigarString = \"*\"\n",
    "        \n",
    "    gfaFile.write(f'P\\t{AccessionID}\\t{\",\".join(path)}\\t{cigarString}\\n')\n",
    "    \n",
    "def writeSegmentIDs(path,segmentIDs):\n",
    "    with open(path,'w') as jsf:\n",
    "        json.dump(segmentIDs,jsf)\n",
    "        \n",
    "def readSegmentIDs(path):\n",
    "    with open(path,'r') as jsf:\n",
    "        return json.load(jsf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
