{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp synteny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Synteny module\n",
    "\n",
    "> Provides functionality to construct special synteny graph from pangenome annotation (and actual sequences)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted 05_export.ipynb.\n",
      "Converted deBruijnGraphProcessing.ipynb.\n",
      "Converted graphTesting.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *\n",
    "from nbdev.export import notebook2script\n",
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exporti\n",
    "import os\n",
    "import glob\n",
    "import json\n",
    "import itertools\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from skbio.io import read as skbio_read\n",
    "from skbio.metadata import IntervalMetadata\n",
    "from skbio.sequence import DNA\n",
    "\n",
    "from dnasim.IO import writeFASTA\n",
    "from dnasim.simulation import inverseSequence\n",
    "from pangraph_constructor.utils import bidict"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "datadir = '../../1001G/annotation'\n",
    "gfadir = '../../1001G/pantograph/data'\n",
    "\n",
    "annotationFiles = sorted(glob.glob(f'{datadir}{os.path.sep}*.gff'))\n",
    "sequenceFiles = sorted(glob.glob(f'{datadir}{os.path.sep}sequences{os.path.sep}*.fasta'))\n",
    "transMapFile = f'{datadir}{os.path.sep}stats{os.path.sep}TransMap.map'"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#Parameters\n",
    "doSequences = False\n",
    "doCigars = False\n",
    "doUS = False\n",
    "gfaFilename = 'AT_Chr1_OGOnly.gfa'\n",
    "seqSuffix = '_Chr1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def readTransMap(transMapFile,ATaccessionName='araport'):\n",
    "    # reading and preparing transmap in pandas format\n",
    "    transMap = pd.read_csv(transMapFile,delimiter='\\t')\n",
    "    transMap.fillna('',inplace=True)\n",
    "    transMap.rename(columns={'Orthogroup:':'orthogroup'},inplace=True)\n",
    "    transMap['orthogroup'] = transMap.orthogroup.str.rstrip(':')\n",
    "    transMap.set_index('orthogroup',inplace=True)\n",
    "#     transMap.fillna('',inplace=True)\n",
    "    if ATaccessionName is not None:\n",
    "        ATTransMap = transMap[ATaccessionName]\n",
    "        return bidict({og:ATTransMap[og].split(', ') for og in ATTransMap.index})\n",
    "    else:\n",
    "        return transMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def generateOrder(files, priorityAccession='TIAR10'):\n",
    "    idxList = list(range(len(files)))\n",
    "    if priorityAccession is not None:\n",
    "        ind = [idx for idx, file in enumerate(files) if priorityAccession in file][0]\n",
    "        del idxList[ind]\n",
    "        idxList = [ind] + idxList\n",
    "    return idxList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def getIDs(iterator):\n",
    "    idList = []\n",
    "    for interval in iterator:\n",
    "        idList.append(interval.metadata['ID'])\n",
    "    \n",
    "    return idList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def getChromosomeFromSeqID(seqID):\n",
    "    # This function should be modified if chromosome is encoded differently in gff3 file.\n",
    "    return seqID.split('_')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def processAccession(annotationFile, sequenceFile=None,\n",
    "                     ATmap=None, isRef=False, accID=None):\n",
    "    if accID is None:\n",
    "        accessionID = os.path.splitext(os.path.basename(annotationFile))[0]\n",
    "    else:\n",
    "        accessionID = accID\n",
    "    annotationGen = skbio_read(annotationFile, format='gff3')\n",
    "    \n",
    "    sequenceDict = None\n",
    "    \n",
    "    if sequenceFile is not None:\n",
    "        sequenceDict = {}\n",
    "        sequenceGen = skbio_read(sequenceFile,format='fasta')\n",
    "        for seq in sequenceGen:\n",
    "            sequenceDict[seq.metadata['id']] = bytearray(seq.values).decode()\n",
    "    \n",
    "    genes = []\n",
    "    for seqID,annotation in annotationGen:\n",
    "        geneInts = annotation.query(metadata={'type':'gene'})\n",
    "        \n",
    "        for gene in geneInts:\n",
    "            geneID = gene.metadata['ID'].split('.')[-1]\n",
    "            if isRef and ATmap is not None:\n",
    "                orthogroup = ATmap.inverse.get(geneID,[None])[0]\n",
    "            elif isRef:\n",
    "                raise ValueError(\"If reference is provided, then the TransMap (bidict) with bidirectional relation between reference and annotation IDs should be provided.\")\n",
    "                #orthogroup = gene.metadata['ID']\n",
    "            else:\n",
    "                orthogroup = gene.metadata['OG']\n",
    "            \n",
    "            if orthogroup is None:\n",
    "                continue\n",
    "            \n",
    "            if ATmap is not None:\n",
    "                atNamesStr = ATmap.get(orthogroup,[''])\n",
    "            else:\n",
    "                atNameStr = ['']\n",
    "            forward = gene.metadata['strand']=='+'\n",
    "            start,end = gene.bounds[0]\n",
    "            \n",
    "            if sequenceDict is not None:\n",
    "                geneSeq = sequenceDict[seqID][start:end+1]\n",
    "            else:\n",
    "                geneSeq = ''\n",
    "            if isRef:\n",
    "                pass\n",
    "            overlaps = getIDs(annotation.query(bounds=[(start,end)],metadata={'type':'gene'}))\n",
    "            genes.append([geneID,orthogroup,seqID,accessionID,getChromosomeFromSeqID(seqID),forward,start+1,end,atNamesStr,geneSeq,overlaps])\n",
    "    \n",
    "    genes = pd.DataFrame(genes,columns=['geneID','orthogroup','sequenceID','accessionID','chromosome','forward','start','end','AT_str','geneSeq','overlapGenes'])\n",
    "    genes.sort_values(by=['sequenceID','start'],inplace=True)\n",
    "        \n",
    "    return accessionID,genes,sequenceDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def recordSegment(name,segmentIDs,segmentIDToNumDict,sequence=None,gfaFile=None,segmentData=None):\n",
    "    segmentIDs.append(name)\n",
    "    \n",
    "    segmentIDToNumDict[name] = len(segmentIDs)-1\n",
    "    segID = len(segmentIDs)\n",
    "    \n",
    "    if segmentData is not None and sequence is not None:\n",
    "        segmentData.append(sequence)\n",
    "    \n",
    "    if gfaFile is not None:\n",
    "        if sequence is not None:\n",
    "            gfaFile.write(f'S\\t{segID}\\t{sequence}\\n')        \n",
    "        else:\n",
    "            gfaFile.write(f'S\\t{segID}\\t{name}\\n')\n",
    "    return segID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def recordAnnotation(nodeID,accessionID,sequenceID,chrID,start,end,og,atList,sequence,nodesMetadata):\n",
    "    # if len(nodesAnnotation)==nodeID-1:\n",
    "    #     nodesAnnotation.append({})\n",
    "\n",
    "    if len(nodesMetadata)==nodeID-1:\n",
    "        nodesMetadata.append({})\n",
    "\n",
    "    geneLen = 1\n",
    "    if len(sequence)>0:\n",
    "        geneLen = len(sequence)\n",
    "\n",
    "    # nodesAnnotation[nodeID-1].setdefault(accessionID,{})[og] = [(0,geneLen-1)]#[(0,len(og)-1)]\n",
    "    # for at in atList:\n",
    "    #     nodesAnnotation[nodeID-1].setdefault(accessionID,{})[at] = [(0,geneLen-1)]#[(0,len(at)-1)]\n",
    "\n",
    "    nodesMetadata[nodeID-1].setdefault(accessionID, {}).setdefault('genPos',[]).append({'chr':chrID, 'genomePosition':[start,end]})\n",
    "    nodesMetadata[nodeID-1][accessionID].setdefault('annotation',{}).update({og: [(0,geneLen-1)]})\n",
    "    nodesMetadata[nodeID-1][accessionID]['annotation'].update({at:[(0,geneLen-1)] for at in atList})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def recordAltChr(nodeID,accessionID,chrID,start,end,nodesMetadata):\n",
    "    nodesMetadata[nodeID-1][accessionID].setdefault('altChrGenPos',[]).append({'chr':chrID, 'genomePosition':[start,end]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def addLink(links,prevPathSegment,name,forward):\n",
    "    '''\n",
    "    `links`: mutable\n",
    "    `prevPathSegment`: mutable\n",
    "    '''\n",
    "    if prevPathSegment is not None:\n",
    "        links[prevPathSegment].add(f'{name}\\t{\"+\" if forward else \"-\"}')\n",
    "    return f'{name}\\t{\"+\" if forward else \"-\"}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def generatePathsLinks(genesAll,sequenceID,accessionID,\n",
    "                       sequences,OGList,segmentIDs,\n",
    "                       nodesMetadata,\n",
    "                       segmentIDToNumDict,links,usCounter,\n",
    "                       doUS=True,segmentData=None,gfaFile=None):\n",
    "    '''\n",
    "    `gfaFile`: file handle to write segments to GFA file\n",
    "    `OGList`: mutable\n",
    "    `links`: mutable\n",
    "    `usCounter`: mutable\n",
    "    \n",
    "    '''\n",
    "    genes = genesAll.loc[genesAll.sequenceID == sequenceID]\n",
    "    path = []\n",
    "    cigar = []\n",
    "    prevEnd = 0\n",
    "    prevPathSegment = None\n",
    "    curSeqID = ''\n",
    "    for generow,gene in genes.iterrows():\n",
    "        og = gene.orthogroup\n",
    "        \n",
    "        geneSeqID = gene.sequenceID\n",
    "        if curSeqID != geneSeqID:\n",
    "            curSeqID = geneSeqID\n",
    "        \n",
    "        atStr = gene.AT_str\n",
    "        if len(atStr[0])>0:\n",
    "            atList = atStr\n",
    "        else:\n",
    "            atList = []\n",
    "        \n",
    "        geneChr = gene.chromosome\n",
    "        geneStart = gene.start\n",
    "        geneEnd = gene.end\n",
    "        geneForward = gene.forward\n",
    "        \n",
    "        if sequences is not None:\n",
    "            if geneForward:\n",
    "                geneSeq = sequences[geneSeqID][geneStart-1:geneEnd]\n",
    "            else:\n",
    "                geneSeq = inverseSequence(sequences[geneSeqID][geneStart-1:geneEnd])\n",
    "        else:\n",
    "            geneSeq = ''\n",
    "        \n",
    "        if doUS:\n",
    "        \n",
    "            if sequences is not None:\n",
    "                usSeq = sequences[geneSeqID][prevEnd-1:geneStart]\n",
    "            else:\n",
    "                usSeq = ''\n",
    "\n",
    "            if len(usSeq)>0:\n",
    "                isUS = True\n",
    "                us = f'US{usCounter:07d}'\n",
    "                usCounter += 1\n",
    "            else:\n",
    "                isUS = False\n",
    "\n",
    "            if isUS:\n",
    "                usID = recordSegment(us,segmentIDs,segmentIDToNumDict,usSeq,gfaFile=gfaFile,segmentData=segmentData)\n",
    "                recordAnnotation(usID,accessionID,geneSeqID,geneChr,prevEnd,geneStart-1,us,[],usSeq,nodesMetadata)\n",
    "        \n",
    "        if og not in OGList:\n",
    "            ogID = recordSegment(og,segmentIDs,segmentIDToNumDict,geneSeq,gfaFile=gfaFile,segmentData=segmentData)\n",
    "            OGList.append(og)\n",
    "        else:\n",
    "            ogID = segmentIDs.index(og)+1\n",
    "        \n",
    "        recordAnnotation(ogID,accessionID,geneSeqID,geneChr,geneStart,geneEnd,og,atList,geneSeq,nodesMetadata)\n",
    "        altPos = genesAll.loc[(genesAll.orthogroup == og) & (genesAll.sequenceID != sequenceID)]\n",
    "        for altrow,altChrOG in altPos.iterrows():\n",
    "            recordAltChr(ogID,accessionID,altChrOG.chromosome,altChrOG.start,altChrOG.end,nodesMetadata)\n",
    "        \n",
    "        pathAdd = [f'{ogID}{\"+\" if geneForward else \"-\"}']\n",
    "        if doUS and isUS:\n",
    "            pathAdd.insert(0,f'{usID}+')\n",
    "            \n",
    "        path.extend(pathAdd)\n",
    "        \n",
    "        if len(cigar)>0 and doUS and isUS:\n",
    "            cigar.extend(['0M','0M']) # with previous block and between two current blocks\n",
    "        else:\n",
    "            cigar.append('0M') # only between current blocks or between previous and current gene\n",
    "                               # without unrelated sequence (intergenic) block.\n",
    "        \n",
    "        if doUS and isUS:\n",
    "            prevPathSegment = addLink(links,prevPathSegment,usID,True)\n",
    "            links[prevPathSegment] = set()\n",
    "        \n",
    "        prevPathSegment = addLink(links,prevPathSegment,ogID,geneForward)\n",
    "        if prevPathSegment not in links:\n",
    "            links[prevPathSegment] = set()\n",
    "        \n",
    "        prevEnd = geneEnd+1\n",
    "    \n",
    "    if doUS:\n",
    "        \n",
    "        if sequences is not None:\n",
    "            usSeq = sequences[curSeqID][prevEnd-1:]\n",
    "        else:\n",
    "            usSeq = ''\n",
    "\n",
    "        if len(usSeq)>0:\n",
    "            us = f'US{usCounter:07d}'\n",
    "            usID = recordSegment(us,segmentIDs,segmentIDToNumDict,usSeq,gfaFile=gfaFile,segmentData=segmentData)\n",
    "            recordAnnotation(usID,accessionID,geneSeqID,prevEnd,len(sequences[curSeqID]),us,[],nodesMetadata)\n",
    "            usCounter += 1\n",
    "            path.append(f'{usID}+')\n",
    "            cigar.append('0M')\n",
    "            prevPathSegment = addLink(links,prevPathSegment,usID,True)\n",
    "\n",
    "    return path,cigar,usCounter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def writeLinks(gfaFile,links,doCigars=True):\n",
    "    for linkLeft,linksRight in links.items():\n",
    "        for linkRight in linksRight:\n",
    "            if doCigars:\n",
    "                gfaFile.write(f'L\\t{linkLeft}\\t{linkRight}\\t0M\\n')\n",
    "            else:\n",
    "                gfaFile.write(f'L\\t{linkLeft}\\t{linkRight}\\t*\\n')\n",
    "                \n",
    "def writePath(gfaFile,AccessionID,path,cigar,doCigars):   \n",
    "    if doCigars:\n",
    "        cigarString = \",\".join(cigar)\n",
    "    else:\n",
    "        cigarString = \"*\"\n",
    "        \n",
    "    gfaFile.write(f'P\\t{AccessionID}\\t{\",\".join(path)}\\t{cigarString}\\n')\n",
    "    \n",
    "def writeSegmentIDs(path,segmentIDs):\n",
    "    with open(path,'w') as jsf:\n",
    "        json.dump(segmentIDs,jsf)\n",
    "        \n",
    "def readSegmentIDs(path):\n",
    "    with open(path,'r') as jsf:\n",
    "        return json.load(jsf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
