{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prime-sapphire",
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "artistic-given",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "engaged-promise",
   "metadata": {},
   "source": [
    "# Graph module\n",
    "\n",
    "> Provides main class and helpers classes and functions to handle pangenome graph and related data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "religious-prairie",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted 05_export.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *\n",
    "from nbdev.export import notebook2script\n",
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "constitutional-amateur",
   "metadata": {},
   "outputs": [],
   "source": [
    "#exporti\n",
    "\n",
    "import os\n",
    "import glob\n",
    "import json\n",
    "import time\n",
    "import itertools\n",
    "import warnings\n",
    "\n",
    "import pdb\n",
    "\n",
    "from copy import deepcopy\n",
    "\n",
    "# import networkx as nx\n",
    "\n",
    "# import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from pangraph_constructor.utils import pathConvert,NpEncoder,adjustZoomLevels\n",
    "from pangraph_constructor.graph import GenomeGraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deluxe-thinking",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "decimal-telephone",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "componentTemplate = {\n",
    "    \"first_bin\": 1,\n",
    "    \"last_bin\": 1,\n",
    "    \"firstCol\": 1,\n",
    "    \"lastCol\": 1,\n",
    "    \"occupants\": [],\n",
    "    \"matrix\": [],\n",
    "    \"larrivals\": [],\n",
    "    \"rarrivals\": [],\n",
    "    \"ldepartures\": [],\n",
    "    \"rdepartures\": [],\n",
    "    \"x\": 0\n",
    "}\n",
    "\n",
    "chunkTemplate = {\n",
    "    \"json_version\":19,\n",
    "    \"first_bin\":1,\n",
    "    \"last_bin\": 1,\n",
    "    \"includes_connectors\": True,\n",
    "    \"components\": []\n",
    "}\n",
    "\n",
    "rootStructTemplate = {\n",
    "    \"json_version\": 19,\n",
    "    \"pangenome_length\": 0,#pangenomeLength, #total number of bp in pangenome\n",
    "    \"includes_connectors\": True,\n",
    "    \"zoom_levels\": {}\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "excited-friday",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def exportToPantograph(graph=None,inputPath=None,GenomeGraphParams={},outputPath=None,outputName=None,outputSuffix=None,isSeq=True,\n",
    "                       listOfExports=['schematise','genomeToPangenome','annotationToGenome'],\n",
    "                       zoomLevels=[1],maxLengthComponent=100,maxLengthChunk=20,invertionThreshold=0.5,\n",
    "                       debug=False,returnDebugData=False):\n",
    "    \n",
    "    if graph is None:\n",
    "        if inputPath is not None:\n",
    "            print('Loading Genome')\n",
    "            graph = GenomeGraph(gfaPath=inputPath,isGFASeq=isSeq,**GenomeGraphParams)\n",
    "        else:\n",
    "            raise ValueError(\"Either graph or inputpath to GFA file should be provided\")\n",
    "    \n",
    "    if outputPath is None and outputName is None:\n",
    "        if inputPath is not None:\n",
    "            if outputSuffix is not None:\n",
    "                outputPath,outputName = pathConvert(inputPath,suffix=outputSuffix)\n",
    "            else:\n",
    "                outputPath,outputName = pathConvert(inputPath)\n",
    "                \n",
    "            \n",
    "        else:\n",
    "            raise ValueError(\"If inputPath is not given, then outputPath and outputName should be provided.\")\n",
    "    else:\n",
    "        if outputSuffix is not None:\n",
    "            outputName = outputName + outputSuffix\n",
    "    print(f'Recording Pantograph data to {outputPath}{os.path.sep}{outputName}')\n",
    "\n",
    "    numNodes = len(graph.nodes)\n",
    "    numNodesDigits = np.int(np.ceil(np.log10(numNodes)))\n",
    "    fromLinks = {}\n",
    "    toLinks = {}\n",
    "\n",
    "    startTime = time.time()\n",
    "\n",
    "    rootStruct = deepcopy(rootStructTemplate)\n",
    "    rootStruct[\"pathNames\"] = graph.accessions\n",
    "    \n",
    "    if returnDebugData:\n",
    "        zoomComponentLengths = {}\n",
    "        zoomNodeToComponent = {}\n",
    "\n",
    "        zoomComponentToNodes = {}\n",
    "        zoomComponents = {} #temporary structure for testing, later each zoomlevel should be just saved as soon as it is processed.\n",
    "        zoomCompNucleotides = {}\n",
    "    \n",
    "    nodeLengths = calcNodeLengths(graph)\n",
    "    \n",
    "    pathLengths,pathNodeArray,pathNodeLengths,pathDirArray,pathNodeLengthsCum = initialPathAnalysis(graph,nodeLengths)\n",
    "    \n",
    "    for zoomLevel in zoomLevels:\n",
    "        \n",
    "        zoomTime = time.time()\n",
    "        \n",
    "        os.makedirs(f'{outputPath}{os.path.sep}{outputName}{os.path.sep}{zoomLevel}',exist_ok=True)\n",
    "\n",
    "        print('\\n===========================')\n",
    "        print(f'Zoom level {zoomLevel}')\n",
    "        print('===========================')\n",
    "        zoom_level_struct = {}\n",
    "        zoom_level_struct[\"files\"] = []\n",
    "        \n",
    "        \n",
    "        \n",
    "        if returnDebugData:\n",
    "            nodeToComponent = zoomNodeToComponent.setdefault(zoomLevel,[])\n",
    "            componentToNode = zoomComponentToNodes.setdefault(zoomLevel,[])\n",
    "            componentLengths = zoomComponentLengths.setdefault(zoomLevel,[])\n",
    "            components = zoomComponents.setdefault(zoomLevel,[])\n",
    "            componentNucleotides = zoomCompNucleotides.setdefault(zoomLevel,[])\n",
    "        else:\n",
    "            nodeToComponent = [] #zoomNodeToComponent.setdefault(zoomLevel,[])\n",
    "            componentToNode = [] #zoomComponentToNodes.setdefault(zoomLevel,[])\n",
    "            componentLengths = [] #zoomComponentLengths.setdefault(zoomLevel,[])\n",
    "            components = [] #zoomComponents.setdefault(zoomLevel,[])\n",
    "            componentNucleotides = [] #zoomCompNucleotides.setdefault(zoomLevel,[])\n",
    "        \n",
    "        component = deepcopy(componentTemplate)\n",
    "\n",
    "        occupants = set()\n",
    "        nucleotides = ''\n",
    "        matrix = {}\n",
    "        \n",
    "        \n",
    "        nodeLinks = []\n",
    "        \n",
    "        nBins = 0\n",
    "        previousInv = {}\n",
    "        nCols = 0\n",
    "        binLength = 0\n",
    "        occupancy = {}\n",
    "        inversion = {}\n",
    "        pos = {}\n",
    "        nodesInComp = set()\n",
    "        breakComponentWhenBinEnds = False\n",
    "        breakCompBeforeBin = False\n",
    "        binOpen = False\n",
    "        breakComponent = False\n",
    "        forceBreak = False\n",
    "\n",
    "        for nodeIdx in range(1,numNodes+1):\n",
    "            if debug:    \n",
    "                print(f'Processing node {nodeIdx:0{numNodesDigits}}/{numNodes:0{numNodesDigits}}')\n",
    "            else:\n",
    "                print(f'\\rProcessing node {nodeIdx:0{numNodesDigits}}/{numNodes:0{numNodesDigits}}',end='')\n",
    "\n",
    "            nodePathsIdx,nodeSeqInPath = np.where(pathNodeArray==nodeIdx)\n",
    "            nodeLen = nodeLengths[nodeIdx-1]\n",
    "\n",
    "            uniqueNodePathsIDx,pathNodeCount = np.unique(nodePathsIdx,return_counts=True)\n",
    "            \n",
    "            incomingFarLink = False\n",
    "            pathStart = False\n",
    "            pathEnd = False\n",
    "            breakCompBeforeBin = False\n",
    "            for j,pathID in enumerate(uniqueNodePathsIDx):\n",
    "                occupants.add(pathID)\n",
    "\n",
    "                pOcc = occupancy.setdefault(pathID,0) \n",
    "                pOcc += pathNodeCount[j]*nodeLengths[nodeIdx-1]\n",
    "                occupancy[pathID] = pOcc # Occupancy\n",
    "                \n",
    "                localInv = 0\n",
    "                for nodeNumInPath in nodeSeqInPath[np.where(nodePathsIdx==pathID)[0]]:\n",
    "                    localInv += int(pathDirArray[pathID,nodeNumInPath])\n",
    "\n",
    "                pInv = inversion.get(pathID,0) \n",
    "                pInv += localInv*nodeLengths[nodeIdx-1]/pathNodeCount[j]\n",
    "                inversion[pathID] = pInv\n",
    "                if (pInv-invertionThreshold)*(previousInv.setdefault(pathID,pInv)-invertionThreshold)<0:\n",
    "                    breakCompBeforeBin = True\n",
    "                previousInv[pathID] = pInv\n",
    "\n",
    "                nodePositions = np.where(nodePathsIdx==pathID)[0]\n",
    "                nodeInversionInPath = np.sum([pathDirArray[pathID,nodeSeqInPath[nodePos]] for nodePos in nodePositions])/len(nodePositions)\n",
    "                for nodePos in nodePositions:\n",
    "                    nodePositionInPath = nodeSeqInPath[nodePos]\n",
    "                    if debug:\n",
    "                        print(f'Node {nodeIdx}, Path {pathID} inversion {nodeInversionInPath}')\n",
    "                    if nodeInversionInPath<=invertionThreshold:\n",
    "\n",
    "                        if nodePositionInPath>0 and nodePositionInPath<pathLengths[pathID]-1:\n",
    "                            if debug:\n",
    "                                print(f\"IncomingLinks: {pathNodeArray[pathID,nodePositionInPath-1]} -> {pathNodeArray[pathID,nodePositionInPath]}\")\n",
    "                                print(f'Existing known links: {fromLinks.get(pathNodeArray[pathID,nodePositionInPath-1],{})}')\n",
    "\n",
    "                            previousNode = pathNodeArray[pathID,nodePositionInPath-1]\n",
    "                            nextNode = pathNodeArray[pathID,nodePositionInPath]\n",
    "                            if previousNode+1!=nextNode:\n",
    "                                if (previousNode<nextNode and nextNode in fromLinks.get(previousNode,{})) or (previousNode>nextNode):\n",
    "                                    incomingFarLink = True\n",
    "                                    break\n",
    "                        elif nodePositionInPath==0:\n",
    "                            pathStart = True\n",
    "                    else:\n",
    "\n",
    "                        if nodePositionInPath>0 and nodePositionInPath<pathLengths[pathID]-1:\n",
    "                            if debug:\n",
    "                                print(f\"IncomingLinks: {pathNodeArray[pathID,nodePositionInPath]} -> {pathNodeArray[pathID,nodePositionInPath+1]}\")\n",
    "                                print(f'Existing known links: {toLinks.get(pathNodeArray[pathID,nodePositionInPath+1],{})}')\n",
    "\n",
    "                            previousNode = pathNodeArray[pathID,nodePositionInPath]\n",
    "                            nextNode = pathNodeArray[pathID,nodePositionInPath+1]\n",
    "                            if previousNode-1!=nextNode:\n",
    "                                if (previousNode>nextNode and nextNode in toLinks.get(nextNode,{})) or (previousNode<nextNode):\n",
    "                                    incomingFarLink = True\n",
    "                                    break\n",
    "                        elif nodePositionInPath==pathLengths[pathID]-1:\n",
    "                            pathEnd = True\n",
    "#                 if incomingFarLink or pathStart or pathEnd:\n",
    "#                     break\n",
    "\n",
    "            if nodeIdx>1 and (pathStart or pathEnd or incomingFarLink or breakCompBeforeBin) and len(matrix)>0:\n",
    "                if debug:\n",
    "                    if incomingFarLink:\n",
    "                        print(f'Node {nodeIdx}: Component broken before node {nodeIdx} due to incoming far link.')\n",
    "                    else:\n",
    "                        print(f'Node {nodeIdx}: Component broken before node {nodeIdx} due to start or end of a path.')\n",
    "                component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = \\\n",
    "                    finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)\n",
    "                componentToNode.append(list(nodesInComp))\n",
    "                nodesInComp = set([nodeIdx])\n",
    "                breakComponent = False\n",
    "                forceBreak = False\n",
    "                breakCompBeforeBin = False\n",
    "            \n",
    "            nodeToComponent.append([len(components)])\n",
    "            nodesInComp.add(nodeIdx)\n",
    "\n",
    "            if zoomLevel<nodeLen:\n",
    "                for posInd,i in enumerate(range(0,nodeLen,zoomLevel)):\n",
    "                    \n",
    "#                     breakCompBeforeBin = False\n",
    "#                     for j,pathID in enumerate(uniqueNodePathsIDx):\n",
    "#                         occupants.add(pathID)\n",
    "\n",
    "#                         pOcc = occupancy.setdefault(pathID,0) \n",
    "#                         pOcc += pathNodeCount[j]*binLength\n",
    "#                         occupancy[pathID] = pOcc # Occupancy\n",
    "\n",
    "#                         posPath = pos.setdefault(pathID,[])\n",
    "#                         localInv = 0\n",
    "#                         for nodeNumInPath in nodeSeqInPath[np.where(nodePathsIdx==pathID)[0]]:\n",
    "#                             localInv += int(pathDirArray[pathID,nodeNumInPath])\n",
    "#                             offset = np.min([i+zoomLevel,nodeLen])\n",
    "#                             nodeStart = pathNodeLengthsCum[pathID,nodeNumInPath]-nodeLen+1\n",
    "#                             posPath.append([nodeStart+i,nodeStart+offset-1])\n",
    "\n",
    "\n",
    "#                         pInv = inversion.get(pathID,0) \n",
    "#                         pInv += localInv*binLength/pathNodeCount[j]\n",
    "#                         inversion[pathID] = pInv\n",
    "#                         if (pInv-invertionThreshold)*(previousInv.setdefault(pathID,pInv)-invertionThreshold)<0:\n",
    "#                             breakCompBeforeBin = True\n",
    "#                         previousInv[pathID] = pInv\n",
    "\n",
    "#                     if breakCompBeforeBin and len(matrix)>0:\n",
    "#                         nodeToComponent[-1].remove(len(components))\n",
    "#                         nodesInComp.remove(nodeIdx)\n",
    "\n",
    "#                         component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = \\\n",
    "#                         finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)                    \n",
    "\n",
    "#                         nodeToComponent[-1].append(len(components))\n",
    "#                         componentToNode.append(list(nodesInComp))\n",
    "#                         nodesInComp = set([nodeIdx])\n",
    "#                         breakCompBeforeBin = False\n",
    "#                         breakComponent = False\n",
    "#                         forceBreak = False\n",
    "                    \n",
    "                    if zoomLevel==1:\n",
    "                        if isSeq:\n",
    "                            nucleotides += graph.nodesData[nodeIdx-1][i:i+zoomLevel]\n",
    "                        else:\n",
    "                            nucleotides += graph.nodes[nodeIdx-1][i:i+zoomLevel]\n",
    "\n",
    "                    if (i+zoomLevel>nodeLen):\n",
    "                        binLength = nodeLen-i\n",
    "                    else:\n",
    "                        binLength = zoomLevel\n",
    "\n",
    "                    for j,pathID in enumerate(uniqueNodePathsIDx):\n",
    "                        posPath = pos.setdefault(pathID,[])\n",
    "                        for nodeNumInPath in nodeSeqInPath[np.where(nodePathsIdx==pathID)[0]]:\n",
    "                            offset = np.min([i+zoomLevel,nodeLen])\n",
    "                            nodeStart = pathNodeLengthsCum[pathID,nodeNumInPath]-nodeLen+1\n",
    "                            posPath.append([nodeStart+i,nodeStart+offset-1])\n",
    "\n",
    "                    matrix,pos,binLength,_,_,nBins,nCols,binOpen = \\\n",
    "                        finaliseBin(matrix,pos,binLength,nodeLengths[nodeIdx-1],occupancy,inversion,nBins,nCols,posInd)\n",
    "\n",
    "                    if nBins+1>maxLengthComponent and (i+zoomLevel<nodeLen) and len(matrix)>0:\n",
    "                        # Because this is intra-node break, all connectors will be handled by rdepartures without separating into inverted or non-inverted. \n",
    "                        # All inter-node links will be handled through links later when chunks are being recorded.\n",
    "                        \n",
    "                        forwardPaths = []\n",
    "                        invertedPaths = []\n",
    "                        for pathID in range(len(graph.paths)):\n",
    "                            if pathID in occupants:\n",
    "                                if matrix[pathID][1][0][1]>invertionThreshold:\n",
    "                                    invertedPaths.append(pathID)\n",
    "                                else:\n",
    "                                    forwardPaths.append(pathID)\n",
    "                            else:\n",
    "                                forwardPaths.append(pathID)\n",
    "                        \n",
    "                        if len(forwardPaths)>0:\n",
    "                            \n",
    "                            component[\"rdepartures\"].append({\n",
    "                                \"upstream\": component[\"first_bin\"]+nBins-1,\n",
    "                                \"downstream\": component[\"first_bin\"]+nBins,\n",
    "                                \"participants\": forwardPaths,\n",
    "                                'otherSideRight': False\n",
    "                            })\n",
    "                        if len(invertedPaths)>0:\n",
    "                            component[\"rarrivals\"].append({\n",
    "                                \"upstream\": component[\"first_bin\"]+nBins,\n",
    "                                \"downstream\": component[\"first_bin\"]+nBins-1,\n",
    "                                \"participants\": invertedPaths,\n",
    "                                'otherSideRight': False\n",
    "                            })\n",
    "\n",
    "                        if debug:\n",
    "                            print(f'Node {nodeIdx}: Component broken inside node {nodeIdx} due to max component length.')\n",
    "\n",
    "                        component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)                    \n",
    "\n",
    "                        nodeToComponent[-1].append(len(components))\n",
    "                        componentToNode.append(list(nodesInComp))\n",
    "                        nodesInComp = set([nodeIdx])\n",
    "                        breakComponent = False\n",
    "                        forceBreak = False\n",
    "                binLength = 0\n",
    "                occupancy = {}\n",
    "                inversion = {}\n",
    "            else:\n",
    "                if nBins==0 and not binOpen:\n",
    "                    if debug:\n",
    "                        print(f'Before processing node {nodeIdx} the nodeLinks was cleared')\n",
    "                    nodeLinks = []\n",
    "                binOpen = True\n",
    "                binLength += nodeLen\n",
    "\n",
    "#                 if isSeq:\n",
    "#                     nucleotides += graph.nodesData[nodeIdx-1]\n",
    "#                 else:\n",
    "#                     nucleotides += graph.nodes[nodeIdx-1]\n",
    "\n",
    "#                 breakCompBeforeBin = False\n",
    "#                 for j,pathID in enumerate(uniqueNodePathsIDx):\n",
    "\n",
    "#                     occupants.add(pathID)\n",
    "\n",
    "#                     pOcc = occupancy.setdefault(pathID,0) \n",
    "#                     pOcc += pathNodeCount[j]*nodeLen\n",
    "#                     occupancy[pathID] = pOcc\n",
    "\n",
    "#                     localInv = 0\n",
    "#                     posPath = pos.setdefault(pathID,[])\n",
    "#                     for nodeNumInPath in nodeSeqInPath[np.where(nodePathsIdx==pathID)[0]]:\n",
    "#                         localInv += int(pathDirArray[pathID,nodeNumInPath])\n",
    "#                         nodeStart = pathNodeLengthsCum[pathID,nodeNumInPath]-nodeLen+1\n",
    "#                         posPath.append([nodeStart,nodeStart+nodeLen-1])\n",
    "#                     pInv = inversion.get(pathID,0) \n",
    "#                     pInv += localInv*nodeLen/pathNodeCount[j]\n",
    "#                     inversion[pathID] = pInv\n",
    "#                     if (pInv-invertionThreshold)*(previousInv.setdefault(pathID,0)-invertionThreshold)<0:\n",
    "#                         breakCompBeforeBin = True\n",
    "#                     previousInv[pathID] = pInv\n",
    "\n",
    "#                 if breakCompBeforeBin and len(matrix)>0:\n",
    "#                     nodeToComponent[-1].remove(len(components))\n",
    "#                     nodesInComp.remove(nodeIdx)\n",
    "\n",
    "#                     component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = \\\n",
    "#                     finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)                    \n",
    "\n",
    "#                     nodeToComponent[-1].append(len(components))\n",
    "#                     componentToNode.append(list(nodesInComp))\n",
    "#                     nodesInComp = set([nodeIdx])\n",
    "#                     breakCompBeforeBin = False\n",
    "#                     breakComponent = False\n",
    "#                     forceBreak = False\n",
    "                \n",
    "                for j,pathID in enumerate(uniqueNodePathsIDx):\n",
    "                    posPath = pos.setdefault(pathID,[])\n",
    "                    for nodeNumInPath in nodeSeqInPath[np.where(nodePathsIdx==pathID)[0]]:\n",
    "                        nodeStart = pathNodeLengthsCum[pathID,nodeNumInPath]-nodeLen+1\n",
    "                        posPath.append([nodeStart,nodeStart+nodeLen-1])\n",
    "\n",
    "                if nodeIdx<len(graph.nodes):\n",
    "                    if binLength+nodeLengths[nodeIdx]>zoomLevel:\n",
    "                        matrix,pos,binLength,occupancy,inversion,nBins,nCols,binOpen = finaliseBin(matrix,pos,binLength,binLength,occupancy,inversion,nBins,nCols,0)\n",
    "                else:\n",
    "                    matrix,pos,binLength,occupancy,inversion,nBins,nCols,binOpen = finaliseBin(matrix,pos,binLength,binLength,occupancy,inversion,nBins,nCols,0)\n",
    "\n",
    "            # marking path ends and breaking component if any exist\n",
    "            pathEnds = np.sort(np.unique([pathIdx for pathIdx,nodeInSeq in zip(nodePathsIdx,nodeSeqInPath) if pathLengths[pathIdx]==nodeInSeq+1]))\n",
    "            if len(pathEnds)>0:\n",
    "                if binOpen:\n",
    "                    component.setdefault(\"ends\",set()).update(pathEnds.tolist())\n",
    "                else:\n",
    "                    if nBins>0:\n",
    "\n",
    "                        component.setdefault(\"ends\",set()).update(pathEnds.tolist())\n",
    "\n",
    "                    else:\n",
    "                        components[-1].setdefault(\"ends\",set()).update(pathEnds.tolist())\n",
    "\n",
    "            pathEndsRelevant = np.sort(np.unique([pathIdx for pathIdx,nodeInSeq in zip(nodePathsIdx,nodeSeqInPath) if pathLengths[pathIdx]==nodeInSeq+1 and (not pathDirArray[pathIdx,nodeInSeq])]))\n",
    "            pathStartsRelevant = np.sort(np.unique([pathIdx for pathIdx,nodeInSeq in zip(nodePathsIdx,nodeSeqInPath) if nodeInSeq==0 and pathDirArray[pathIdx,nodeInSeq]]))\n",
    "            if len(pathEndsRelevant)>0 or len(pathStartsRelevant)>0:\n",
    "                if binOpen:\n",
    "                    if debug:\n",
    "                        print(f'Node {nodeIdx}: Component will be broken due to start or end of a path at node {nodeIdx} when bin is filled.')\n",
    "                    breakComponent = True\n",
    "                    forceBreak = True\n",
    "                elif nBins>0:\n",
    "                    if debug:\n",
    "                        print(f'Node {nodeIdx}: Component broken after node {nodeIdx} due to start or end of a path.')\n",
    "                    component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)\n",
    "                    componentToNode.append(list(nodesInComp))\n",
    "\n",
    "                    nodesInComp = set()\n",
    "                    breakComponent = False\n",
    "                    forceBreak = False\n",
    "\n",
    "            if nodeIdx<len(graph.nodes) and nBins>0:\n",
    "                if nBins+nodeLengths[nodeIdx]/zoomLevel>maxLengthComponent:\n",
    "\n",
    "                    if debug:\n",
    "                        print(f'Node {nodeIdx}: Component broken after node {nodeIdx} because next component will not fit in this component.')\n",
    "                    component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)\n",
    "                    componentToNode.append(list(nodesInComp))\n",
    "\n",
    "                    nodesInComp = set()\n",
    "                    breakComponent = False\n",
    "                    forceBreak = False\n",
    "            elif nBins>0:\n",
    "\n",
    "                if debug:\n",
    "                    print(f'Node {nodeIdx}: Last node in the last component.')\n",
    "                component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)\n",
    "                componentToNode.append(list(nodesInComp))\n",
    "\n",
    "                nodesInComp = set()\n",
    "                breakComponent = False\n",
    "\n",
    "\n",
    "            nodeFromLink = fromLinks.setdefault(nodeIdx,{})\n",
    "\n",
    "            if zoomLevel==1:\n",
    "                for j,pathID in enumerate(uniqueNodePathsIDx):\n",
    "                    nodePositions = np.where(nodePathsIdx==pathID)[0]\n",
    "                    nodeInversionInPath = np.sum([pathDirArray[pathID,nodeSeqInPath[nodePos]] for nodePos in nodePositions])/len(nodePositions)\n",
    "                    for nodePos in nodePositions:\n",
    "                        nodePositionInPath = nodeSeqInPath[nodePos]\n",
    "                        if nodePositionInPath<(pathLengths[pathID]-1):\n",
    "                            nextNode = pathNodeArray[pathID,nodePositionInPath+1]\n",
    "                            nextNodePathsIdx,nextNodeSeqInPath = np.where(pathNodeArray==nextNode)\n",
    "                            nextNodePositions = np.where(nextNodePathsIdx==pathID)[0]\n",
    "                            nextNodeInversionInPath = np.sum([pathDirArray[pathID,nextNodeSeqInPath[nodePos]] for nodePos in nextNodePositions])/len(nextNodePositions)\n",
    "                            if debug:\n",
    "                                print(f'Node {nodeIdx}, next node {nextNode}, path {pathID}, number in path {nodePositionInPath}, length of path {pathLengths[pathID]}')\n",
    "\n",
    "                            if nodeInversionInPath>invertionThreshold and nextNodeInversionInPath>invertionThreshold:\n",
    "                                reverseCond  = (nodeIdx<=nextNode)\n",
    "                                step = -1\n",
    "                            else:\n",
    "                                reverseCond = (nodeIdx>=nextNode)\n",
    "                                step = 1\n",
    "\n",
    "                            if reverseCond:\n",
    "                                if debug:\n",
    "                                    print(f'Reverse link for path {pathID} from node {nodeIdx} to node {nextNode}')\n",
    "\n",
    "                                nodeFromLink.setdefault(nextNode,[]).append(pathID)\n",
    "                                toLinks.setdefault(nextNode,{}).setdefault(nodeIdx,[]).append(pathID)\n",
    "                                if (nBins>0 or binOpen) and (nodeInversionInPath<=0.5):\n",
    "\n",
    "                                    if debug:\n",
    "                                        print(f'Node {nodeIdx}: Component will be broken after node {nodeIdx} due to backward link to node {nextNode}.')\n",
    "                                    nodeLinks.append(nextNode)\n",
    "                                    breakComponent = True\n",
    "                            else:\n",
    "                                if np.any([node in pathNodeArray[pathID,:] for node in range(nodeIdx+1*step,nextNode,step)]):\n",
    "                                    if debug:\n",
    "                                        print(f'Jump link for path {pathID} from node {nodeIdx} to node {nextNode}')\n",
    "\n",
    "                                    nodeFromLink.setdefault(nextNode,[]).append(pathID)\n",
    "                                    toLinks.setdefault(nextNode,{}).setdefault(nodeIdx,[]).append(pathID)\n",
    "                                    if (nBins>0 or binOpen) and (nodeInversionInPath<=0.5): \n",
    "\n",
    "                                        if debug:\n",
    "                                            print(f'Node {nodeIdx}: Component will be broken after node {nodeIdx} due forward jumping link to node {nextNode}.')\n",
    "                                        nodeLinks.append(nextNode)\n",
    "                                        breakComponent = True\n",
    "                                else:\n",
    "\n",
    "\n",
    "                                    startNode = None\n",
    "                                    endNode = None\n",
    "                                    step = None\n",
    "\n",
    "                                    if nodeInversionInPath>invertionThreshold and nextNodeInversionInPath>invertionThreshold:\n",
    "                                        startNode = nodeIdx\n",
    "                                        endNode = nextNode\n",
    "                                        step = -1\n",
    "                                        if debug:\n",
    "                                            print(f'Intermediate nodes for path {pathID} from node {nodeIdx} to node {nextNode}')\n",
    "\n",
    "                                    elif nodeInversionInPath<=invertionThreshold and nextNodeInversionInPath<=invertionThreshold:\n",
    "                                        startNode = nodeIdx\n",
    "                                        endNode = nextNode\n",
    "                                        step = 1\n",
    "                                        if debug:\n",
    "                                            print(f'Intermediate nodes for path {pathID} from node {nodeIdx} to node {nextNode}')\n",
    "                                    else:\n",
    "                                        if debug:\n",
    "                                            print(f'Jump link for path {pathID} from node {nodeIdx} to node {nextNode}')\n",
    "\n",
    "                                        nodeFromLink.setdefault(nextNode,[]).append(pathID)\n",
    "                                        toLinks.setdefault(nextNode,{}).setdefault(nodeIdx,[]).append(pathID)\n",
    "                                        if nBins>0 or binOpen: \n",
    "\n",
    "                                            if debug:\n",
    "                                                print(f'Node {nodeIdx}: Component will be broken after node {nodeIdx} due forward jumping link to node {nextNode}.')\n",
    "                                            nodeLinks.append(nextNode)\n",
    "                                            breakComponent = True\n",
    "\n",
    "                                    if startNode is not None and endNode is not None and step is not None:\n",
    "                                        for intermediateNodeIdx in range(startNode,endNode,step):\n",
    "                                            if debug:\n",
    "                                                print(f'Adding link from node {intermediateNodeIdx} to node {intermediateNodeIdx+1*step} for path {pathID}')\n",
    "                                            fromLinks.setdefault(intermediateNodeIdx,{}).setdefault(intermediateNodeIdx+1*step,[]).append(pathID)\n",
    "                                            toLinks.setdefault(intermediateNodeIdx+1*step,{}).setdefault(intermediateNodeIdx,[]).append(pathID)\n",
    "                        else:\n",
    "                            if debug:\n",
    "                                print(\"Last node in path\")\n",
    "            else:#if not breakComponent:\n",
    "                if nBins>0 or binOpen:\n",
    "                    breakLink,nnodes = checkLinksForBreak(nodeIdx,nodeFromLink,toLinks.setdefault(nodeIdx,{}),matrix,inversion,binOpen,invertionThreshold,True)\n",
    "                    if breakLink==2 or (not binOpen and breakLink==1):\n",
    "                        if debug:\n",
    "                            print(f'Node {nodeIdx}: Component will be broken after node {nodeIdx} due a link from that node (res code {breakLink}) to nodes {nnodes}.')\n",
    "                        breakComponent = True\n",
    "                        nodeLinks.extend(nnodes)\n",
    "\n",
    "            if breakComponent and not binOpen:\n",
    "                if np.any([node not in nodesInComp for node in nodeLinks]) or len(nodeLinks)==0 or forceBreak:\n",
    "                    \n",
    "                    if debug:\n",
    "                        print(f'Node {nodeIdx}: Component broken after node {nodeIdx} because of previously set flag.')\n",
    "                        print(f'Nodes in component: {nodesInComp}')\n",
    "                        print(f'Nodes linked from component: {nodeLinks}')\n",
    "                    component,components,componentNucleotides,matrix,occupants,nBins,nCols,nucleotides = finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides)\n",
    "                    componentToNode.append(list(nodesInComp))\n",
    "                    nodesInComp = set()\n",
    "                    breakComponent = False\n",
    "                    forceBreak = False\n",
    "                    nodeLinks = []\n",
    "                else:\n",
    "                    if debug:\n",
    "                        print(f'Node {nodeIdx}: Component is not broken as all links are internal, flag reset.')\n",
    "                        print(f'Nodes in component: {nodesInComp}')\n",
    "                        print(f'Nodes linked from component: {nodeLinks}')\n",
    "                    breakComponent = False;\n",
    "\n",
    "        print('\\nNodes processed, recording component to files.')\n",
    "\n",
    "        if zoomLevel==1:\n",
    "            rootStruct[\"pangenome_length\"] = np.sum(componentLengths)\n",
    "\n",
    "        chunkList = rootStruct[\"zoom_levels\"].setdefault(zoomLevel,{\n",
    "            \"last_bin\": components[-1][\"last_bin\"],\n",
    "            \"files\":[]\n",
    "        })\n",
    "        # Recording data to files for zoom level\n",
    "\n",
    "    #     chunkTemplate = {\n",
    "    #     \"json_version\":17,\n",
    "    #     \"bin_width\":1,\n",
    "    #     \"first_bin\":1,\n",
    "    #     \"includes_connectors\": True,\n",
    "    #     \"components\": []\n",
    "    # }    \n",
    "        chunk = deepcopy(chunkTemplate)\n",
    "        chunkNum = 0\n",
    "        prevTotalCols = 0\n",
    "        curCompCols = 0\n",
    "        nucleotides = ''\n",
    "        nBins = 0\n",
    "        numComps = len(components)\n",
    "        numCompsDigits = np.int(np.ceil(np.log10(numComps)))\n",
    "        \n",
    "        for compNum in range(numComps):\n",
    "            if debug:    \n",
    "                print(f'Processing component {compNum+1:0{numCompsDigits}}/{numComps:0{numCompsDigits}}')\n",
    "            else:\n",
    "                print(f'\\rProcessing component {compNum+1:0{numCompsDigits}}/{numComps:0{numCompsDigits}}',end='')\n",
    "                \n",
    "            component = components[compNum]\n",
    "\n",
    "            nodeInComp = componentToNode[compNum]\n",
    "            nodeInComp.sort()\n",
    "\n",
    "            if len(nodeInComp)>1:\n",
    "                component = addLinksToComp(nodeInComp,nodeToComponent,toLinks,fromLinks,component,components)\n",
    "            elif len(nodeInComp)==1:\n",
    "\n",
    "                mainNode = nodeInComp[0]\n",
    "    #             compForNode = nodeToComponent[mainNode-1]\n",
    "                doLeft = False\n",
    "                if compNum==0:\n",
    "                    doLeft = True\n",
    "                elif componentToNode[compNum-1][-1]!=mainNode:\n",
    "                    doLeft = True\n",
    "\n",
    "                doRight = False # Check this conditions as they may be causing issue!\n",
    "                if compNum==(len(components)-1):\n",
    "                    doRight = True\n",
    "                elif componentToNode[compNum+1][0]!=mainNode:\n",
    "                    doRight = True\n",
    "\n",
    "                component = addLinksToComp(nodeInComp,nodeToComponent,toLinks,fromLinks,component,components,doLeft,doRight)\n",
    "\n",
    "            else:\n",
    "                raise ValueError(f\"Component {compNum} does not have any associated nodes!\")\n",
    "            \n",
    "            nBins += component[\"last_bin\"]-component[\"first_bin\"]+1\n",
    "            if len(chunk['components'])>0:\n",
    "                if checkLinks(chunk[\"components\"][-1],component):\n",
    "                    \n",
    "                    newComp = joinComponents(chunk[\"components\"].pop(),component,maxLengthComponent,invertionThreshold)\n",
    "                    if isinstance(newComp,list):\n",
    "                        chunk[\"components\"].append(newComp[0])\n",
    "                        component = newComp[1]\n",
    "                    else:\n",
    "                        component = newComp\n",
    "            \n",
    "            nucleotides += componentNucleotides[compNum]\n",
    "            chunk[\"components\"].append(component)\n",
    "            \n",
    "\n",
    "            # End of chunk\n",
    "            if compNum<len(components)-1:\n",
    "\n",
    "                if len(chunk['components'])>=maxLengthChunk:\n",
    "                #nBins+components[compNum+1][\"last_bin\"]-components[compNum+1][\"first_bin\"]+1>maxLengthChunk:\n",
    "                    rootStruct,chunk,nBins,chunkNum,prevTotalCols,curCompCols,nucleotides = finaliseChunk(\n",
    "                                                                                              rootStruct,\n",
    "                                                                                              zoomLevel,\n",
    "                                                                                              chunk,\n",
    "                                                                                              nucleotides,\n",
    "                                                                                              nBins,\n",
    "                                                                                              chunkNum,\n",
    "                                                                                              curCompCols,\n",
    "                                                                                              prevTotalCols,\n",
    "                                                                                              outputPath,\n",
    "                                                                                              outputName)\n",
    "            else:\n",
    "                rootStruct,chunk,nBins,chunkNum,prevTotalCols,curCompCols,nucleotides = finaliseChunk(\n",
    "                                                                                          rootStruct,\n",
    "                                                                                          zoomLevel,\n",
    "                                                                                          chunk,\n",
    "                                                                                          nucleotides,\n",
    "                                                                                          nBins,\n",
    "                                                                                          chunkNum,\n",
    "                                                                                          curCompCols,\n",
    "                                                                                          prevTotalCols,\n",
    "                                                                                          outputPath,\n",
    "                                                                                          outputName)\n",
    "        if not debug:\n",
    "            print()\n",
    "        print(f'Recording finished. Zoom level time is {time.time() - zoomTime}. Time elapsed {time.time() - startTime}')\n",
    "\n",
    "    with open(f'{outputPath}{os.path.sep}{outputName}{os.path.sep}bin2file.json','w') as f:\n",
    "        json.dump(rootStruct,f,cls=NpEncoder)\n",
    "    \n",
    "    if returnDebugData:\n",
    "        return zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "strange-dividend",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def checkLinks(leftComp,rightComp):\n",
    "    leftRdepCond = np.all([link['upstream']+1==link['downstream'] for link in leftComp['rdepartures']])\n",
    "    leftRarrCond = np.all([link['upstream']-1==link['downstream'] for link in leftComp['rarrivals']])\n",
    "\n",
    "    rightRdepCond = np.all([link['upstream']-1==link['downstream'] for link in rightComp['ldepartures']])\n",
    "    rightRarrCond = np.all([link['upstream']+1==link['downstream'] for link in rightComp['larrivals']])\n",
    "    \n",
    "    return leftRdepCond and leftRarrCond and rightRarrCond and rightRdepCond"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continent-unemployment",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def getMatrixPathElement(matrix,pathID):\n",
    "    res = [el for el in matrix if el[0]==pathID]\n",
    "    if len(res)==1:\n",
    "        return res[0]\n",
    "    elif len(res)>0:\n",
    "        warnings.warn(f\"More than one element for path {pathID} is found!\")\n",
    "    \n",
    "    return None\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "raised-treasury",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def joinComponents(leftComp,rightComp, maxLengthComponent, invertionThreshold=0.5):\n",
    "    '''\n",
    "    If the joining was successful, the function will return a joined component.\n",
    "    \n",
    "    If the joining was not successful and was aborted for one of the following reasons, it will return a list of original components. \n",
    "    The reasons for aborting the joining can be the following:\n",
    "    - In one of the paths the invertion is lower than threshold in one component and higher in the other.\n",
    "    - Left component contains at least one end\n",
    "    - Right component contains at least one start\n",
    "    \n",
    "    The function will not check links for coming or going on the right of the left component and left of the right component. \n",
    "    It will just get left links from left component and right links from right component and assign them to the new component.\n",
    "    '''\n",
    "    \n",
    "    if leftComp['last_bin']-leftComp['first_bin']+1 + rightComp['last_bin']-rightComp['first_bin']+1 > maxLengthComponent:\n",
    "        return [leftComp,rightComp]\n",
    "    \n",
    "    if leftComp.get('ends',False):\n",
    "        # End of a path\n",
    "        return [leftComp,rightComp]\n",
    "    \n",
    "    newComp = {}\n",
    "    newComp['first_bin'] = min(leftComp['first_bin'],rightComp['first_bin'])\n",
    "    newComp['last_bin'] = max(leftComp['last_bin'],rightComp['last_bin'])\n",
    "    newComp['firstCol'] = min(leftComp['firstCol'],rightComp['firstCol'])\n",
    "    newComp['lastCol'] = max(leftComp['lastCol'],rightComp['lastCol'])\n",
    "    \n",
    "    leftCompNumBins = leftComp['last_bin']-leftComp['first_bin']+1\n",
    "    \n",
    "    newComp['occupants'] = list(set(leftComp['occupants']).union(rightComp['occupants']))\n",
    "    \n",
    "    for pathID in newComp['occupants']:\n",
    "        leftPathElement = getMatrixPathElement(leftComp['matrix'],pathID)\n",
    "        rightPathElement = getMatrixPathElement(rightComp['matrix'],pathID)\n",
    "        if leftPathElement is None:\n",
    "            if len([el for el in rightPathElement[2][1] if el[2][0][0]==1 or el[2][-1][0]==1])>0:\n",
    "                # Start of a path\n",
    "                return [leftComp,rightComp]\n",
    "            newComp.setdefault(\"matrix\",[]).append(rightPathElement)\n",
    "            continue\n",
    "        \n",
    "        if rightPathElement is None:\n",
    "            newComp.setdefault(\"matrix\",[]).append(leftPathElement)\n",
    "            continue\n",
    "        \n",
    "        if (leftPathElement[1]>invertionThreshold and rightPathElement[1]<=invertionThreshold) or \\\n",
    "           (leftPathElement[1]<=invertionThreshold and rightPathElement[1]>invertionThreshold):\n",
    "            return [leftComp,rightComp]\n",
    "        \n",
    "        newPathElement = []\n",
    "        newPathElement.append(pathID)\n",
    "        newPathElement.append(leftPathElement[1])\n",
    "        pathMatrix = []\n",
    "        pathMatrix.append(leftPathElement[2][0] + [el+leftCompNumBins for el in rightPathElement[2][0]])\n",
    "        pathMatrix.append(leftPathElement[2][1] + rightPathElement[2][1])\n",
    "        newPathElement.append(pathMatrix)\n",
    "        newComp.setdefault(\"matrix\",[]).append(newPathElement)\n",
    "        \n",
    "    newComp['larrivals'] = leftComp['larrivals']\n",
    "    newComp['ldepartures'] = leftComp['ldepartures']\n",
    "    newComp['rarrivals'] = rightComp['rarrivals']\n",
    "    newComp['rdepartures'] = rightComp['rdepartures']\n",
    "    \n",
    "    return newComp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "peaceful-montgomery",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def calcNodeLengths(graph):\n",
    "    print('Calculating nodes length...')\n",
    "    \n",
    "    numNodes = len(graph.nodes)\n",
    "    numNodesDigits = np.int(np.ceil(np.log10(numNodes)))\n",
    "    nodeLengths = [0]*numNodes\n",
    "    \n",
    "    for nodeIdx in range(numNodes):\n",
    "        print(f'\\rProcessing node {nodeIdx+1:0{numNodesDigits}}/{numNodes:0{numNodesDigits}}',end='')\n",
    "        if graph.nodesData[nodeIdx]=='':\n",
    "            nodeLengths[nodeIdx] = len(graph.nodes[nodeIdx])\n",
    "        else:\n",
    "            nodeLengths[nodeIdx] = len(graph.nodesData[nodeIdx])\n",
    "    \n",
    "    print('\\nFinished calculating nodes lengths')\n",
    "    return nodeLengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "harmful-chinese",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def initialPathAnalysis(graph,nodeLengths):\n",
    "    print('Preprocessing paths...')\n",
    "    \n",
    "    numPaths = len(graph.paths)\n",
    "    numPathsZeros = np.int(np.ceil(np.log10(numPaths)))\n",
    "    \n",
    "    maxLengthPath = len(max(graph.paths,key=lambda arr: len(arr)))\n",
    "\n",
    "    pathLengths = []\n",
    "    pathNodeArray = np.zeros((len(graph.paths),maxLengthPath),dtype = np.int)\n",
    "    pathNodeLengths = np.zeros((len(graph.paths),maxLengthPath),dtype = np.int)\n",
    "    pathDirArray = np.zeros((len(graph.paths),maxLengthPath),dtype = np.bool) # True - \"+\", False - \"-\" or padding for shorter paths.\n",
    "\n",
    "    for i,path in enumerate(graph.paths):\n",
    "        print(f'\\rProcessing path {i+1:0{numPathsZeros}}/{numPaths:0{numPathsZeros}}',end='')\n",
    "        pathLengths.append(len(path))\n",
    "\n",
    "        pathNodeArray[i,:pathLengths[-1]] = [int(node[:-1]) for node in path]\n",
    "        pathDirArray[i,:pathLengths[-1]] = [node[-1]=='-' for node in path]\n",
    "        pathNodeLengths[i,:pathLengths[-1]] = [nodeLengths[node-1] for node in pathNodeArray[i,:pathLengths[-1]]]\n",
    "\n",
    "    pathNodeLengthsCum = np.cumsum(pathNodeLengths,axis=1)\n",
    "    print('\\nFinished preprocessing paths')\n",
    "    return pathLengths,pathNodeArray,pathNodeLengths,pathDirArray,pathNodeLengthsCum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hidden-officer",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def finaliseBin(matrix,pos,binLength,occInvAdj,occupancy,inversion,nBins,nCols,posInd):\n",
    "    for pathID in pos.keys():\n",
    "        matrixPath = matrix.setdefault(pathID,[[],[]])\n",
    "        posPath = pos.get(pathID)\n",
    "        posArray = np.array(posPath)\n",
    "#         print(f'[finaliseBin] original pos {posArray}')\n",
    "        posArray = posArray[np.argsort(posArray[:,0]),:]\n",
    "        posIntersect = (posArray[1:,1]-(posArray[:-1,0]-1))*\\\n",
    "                        (posArray[:-1,1]-(posArray[1:,0]-1))\n",
    "#         print(f'[finaliseBin] pos intersections {posIntersect}')\n",
    "        newPos = [[posArray[0,0]]]\n",
    "        candidates = [posArray[0,1]]\n",
    "        for jointNum in range(len(posIntersect)):\n",
    "            if posIntersect[jointNum]>=0:\n",
    "                candidates.extend(posArray[jointNum+1,:].tolist())\n",
    "            else:\n",
    "                newPos[-1].append(np.max(candidates))\n",
    "                newPos.append([posArray[jointNum+1,0]])\n",
    "                candidates = [posArray[jointNum+1,1]]\n",
    "            \n",
    "        newPos[-1].append(np.max(candidates))# !!!!        \n",
    "        \n",
    "#         print(f'[finaliseBin] joined pos {newPos}')\n",
    "        \n",
    "        inversionRate = inversion[pathID]/occInvAdj\n",
    "    \n",
    "        matrixPathRecord = [0,0,[]]\n",
    "        matrixPath[0].append(nBins)\n",
    "        matrixPathRecord[0] = occupancy[pathID]/occInvAdj # Occupancy\n",
    "        matrixPathRecord[1] = inversionRate # Inversion rate\n",
    "        matrixPathRecord[2] = newPos\n",
    "        if inversionRate>0.5:\n",
    "            matrixPath[1].insert(posInd,matrixPathRecord)\n",
    "        else:\n",
    "            matrixPath[1].append(matrixPathRecord)\n",
    "        \n",
    "    return matrix,{},0,{},{},nBins+1,nCols+binLength,False #matrix,pos,binLength,occupancy,inversion,nBins,nCols,binOpen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "robust-rendering",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def adjustMatrixPathArray(matrixPathArray,isInverted,nBins):\n",
    "    if isInverted:\n",
    "        _tempArray = [nBins-binNum-1 for binNum in matrixPathArray[0]]\n",
    "        _tempArray.sort()\n",
    "        matrixPathArray[0] = _tempArray\n",
    "    return matrixPathArray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "impressed-finnish",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def finaliseComponent(component,components,componentNucleotides,matrix,occupants,nBins,nCols,componentLengths,nucleotides,invertionThreshold=0.5):\n",
    "    componentLengths.append(nBins)\n",
    "    component[\"matrix\"].extend([[pathID,\n",
    "                                 int(matrixPathArray[1][0][1]>invertionThreshold),\n",
    "                                 adjustMatrixPathArray(matrixPathArray,int(matrixPathArray[1][0][1]>invertionThreshold),nBins)] \\\n",
    "                                for pathID,matrixPathArray in matrix.items()])\n",
    "    component[\"occupants\"] = sorted(list(occupants))\n",
    "    component['last_bin'] = component['first_bin'] + nBins - 1\n",
    "    component['lastCol'] = component['firstCol'] + nCols - 1\n",
    "    componentNucleotides.append(nucleotides)\n",
    "    firstBin = component['last_bin'] + 1\n",
    "    firstCol = component['lastCol'] + 1\n",
    "    components.append(component)\n",
    "    component = deepcopy(componentTemplate)\n",
    "    component['first_bin'] = firstBin\n",
    "    component['firstCol'] = firstCol\n",
    "    return component,components,componentNucleotides,{},set(),0,0,''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "appointed-terrorist",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def checkLinksForBreak(nodeIdx,nodeFromLinks,nodeToLinks,matrix,inversion,binOpen,inversionThreshold,isCurrent):\n",
    "    res = 0\n",
    "    nodes = set()\n",
    "    if isCurrent:\n",
    "        for node,pathList in nodeFromLinks.items():\n",
    "            if nodeIdx==node:\n",
    "                res = max(1,res)\n",
    "                nodes.add(node)\n",
    "            else:\n",
    "                for pathID in pathList:\n",
    "                    if binOpen:\n",
    "                        invLevel = inversion.get(pathID)\n",
    "                    else:\n",
    "                        invLevel = matrix.get(pathID)\n",
    "                        if invLevel is not None:\n",
    "                            invLevel = invLevel[1][0][1]\n",
    "                    if invLevel is not None:\n",
    "                        if invLevel<=inversionThreshold:\n",
    "                            # Forward path\n",
    "                            if nodeIdx+1!=node:\n",
    "                                res = max(2,res)\n",
    "                                nodes.add(node)\n",
    "        for node,pathList in nodeToLinks.items():\n",
    "            if nodeIdx==node:\n",
    "                res = max(1,res)\n",
    "                nodes.add(node)\n",
    "            else:\n",
    "                for pathID in pathList:\n",
    "                    if binOpen:\n",
    "                        invLevel = inversion.get(pathID)\n",
    "                    else:\n",
    "                        invLevel = matrix.get(pathID)\n",
    "                        if invLevel is not None:\n",
    "                            invLevel = invLevel[1][0][1]\n",
    "                    if invLevel is not None:\n",
    "                        if invLevel>inversionThreshold:\n",
    "                            # Inverted path\n",
    "                            if nodeIdx+1!=node:\n",
    "                                res = max(2,res)\n",
    "                                nodes.add(node)\n",
    "            \n",
    "    else:\n",
    "        # Do not use isCurrent=False !!!\n",
    "        for node,pathList in nodeFromLinks.items():\n",
    "            if nodeIdx==node:\n",
    "                res = max(1,res)\n",
    "                nodes.add(node)\n",
    "            else:\n",
    "                for pathID in pathList:\n",
    "                    if binOpen:\n",
    "                        invLevel = inversion.get(pathID)\n",
    "                    else:\n",
    "                        invLevel = matrix.get(pathID)\n",
    "                        if invLevel is not None:\n",
    "                            invLevel = invLevel[1][0][1]\n",
    "                    if invLevel is not None:\n",
    "                        if invLevel>inversionThreshold:\n",
    "                            # Inverted path\n",
    "                            if nodeIdx-1!=node:\n",
    "                                res = max(2,res)\n",
    "                                nodes.add(node)\n",
    "        for node,pathList in nodeToLinks.items():\n",
    "            if nodeIdx==node:\n",
    "                res = max(1,res)\n",
    "                nodes.add(node)\n",
    "            else:\n",
    "                for pathID in pathList:\n",
    "                    if binOpen:\n",
    "                        invLevel = inversion.get(pathID)\n",
    "                    else:\n",
    "                        invLevel = matrix.get(pathID)\n",
    "                        if invLevel is not None:\n",
    "                            invLevel = invLevel[1][0][1]\n",
    "                    if invLevel is not None:\n",
    "                        if invLevel<=inversionThreshold:\n",
    "                            # Inverted path\n",
    "                            if nodeIdx+1!=node:\n",
    "                                res = max(2,res)\n",
    "                                nodes.add(node)\n",
    "        \n",
    "    return res,list(nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "transparent-representative",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def finaliseChunk(rootStruct,zoomLevel,chunk,nucleotides,nBins,chunkNum,curCompCols,prevTotalCols,outputPath,outputName):\n",
    "    endChunkBin = chunk['components'][-1]['last_bin']\n",
    "    chunk['last_bin'] = endChunkBin\n",
    "    \n",
    "    localPath = f'{outputPath}{os.path.sep}{outputName}{os.path.sep}{zoomLevel}{os.path.sep}'\n",
    "    \n",
    "    fileName = f'chunk{chunkNum}_zoom{zoomLevel}.schematic.json'\n",
    "    \n",
    "    with open(f'{localPath}{fileName}','w') as f:\n",
    "        json.dump(chunk,f,cls=NpEncoder)\n",
    "    \n",
    "    rootStruct['zoom_levels'][zoomLevel]['files'].append({\n",
    "        'file': fileName,\n",
    "        'first_bin':chunk['first_bin'],\n",
    "        'last_bin':chunk['last_bin'],\n",
    "#         'x':prevTotalCols\n",
    "    })\n",
    "    \n",
    "    if zoomLevel==1:\n",
    "        fastaName = f'seq_chunk{chunkNum}_zoom{zoomLevel}.fa'\n",
    "        rootStruct['zoom_levels'][zoomLevel]['files'][-1]['fasta'] = fastaName\n",
    "        \n",
    "        with open(f'{localPath}{fastaName}','w') as f:\n",
    "            f.write(f'>first_bin:{chunk[\"first_bin\"]} last_bin:{chunk[\"last_bin\"]}\\n')\n",
    "            f.write(nucleotides)\n",
    "        \n",
    "    chunk = deepcopy(chunkTemplate)\n",
    "    chunk['first_bin'] = endChunkBin + 1\n",
    "    return rootStruct,chunk,0,chunkNum + 1,prevTotalCols+curCompCols,0,'' #rootStruct,chunk,nBins,chunkNum,prevTotalCols,curCompCols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "checked-consultation",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def addLinksToComp(nodeInComp,nodeToComponent,toLinks,fromLinks,component,components,doLeft=True,doRight=True):\n",
    "    compLinks = fillLinks(nodeInComp,nodeToComponent,toLinks,fromLinks,component,components,doLeft,doRight)\n",
    "    if \"lArrivals\" in compLinks:\n",
    "        for (upstream,downstream,otherSide),participants in compLinks[\"lArrivals\"].items():\n",
    "            if otherSide=='fr':\n",
    "                fromRight = True\n",
    "            else:\n",
    "                fromRight = False\n",
    "            component['larrivals'].append({\n",
    "                    'upstream': upstream,\n",
    "                    'downstream': downstream,\n",
    "                    'otherSideRight': fromRight,\n",
    "                    'participants': participants\n",
    "                })\n",
    "    \n",
    "    if 'rArrivals' in compLinks:\n",
    "        for (upstream,downstream,otherSide),participants in compLinks[\"rArrivals\"].items():\n",
    "            if otherSide == 'fr':\n",
    "                fromRight = True\n",
    "            else:\n",
    "                fromRight = False\n",
    "            component['rarrivals'].append({\n",
    "                    'upstream': upstream,\n",
    "                    'downstream': downstream,\n",
    "                    'otherSideRight': fromRight,\n",
    "                    'participants': participants\n",
    "                })\n",
    "            \n",
    "    if 'rDepartures' in compLinks:\n",
    "        for (upstream,downstream,otherSide),participants in compLinks[\"rDepartures\"].items():\n",
    "            if otherSide == 'tl':\n",
    "                toRight = False\n",
    "            else:\n",
    "                toRight = True\n",
    "            component['rdepartures'].append({\n",
    "                'upstream': upstream,\n",
    "                'downstream': downstream,\n",
    "                'otherSideRight': toRight,\n",
    "                'participants': participants\n",
    "            })\n",
    "    \n",
    "    if 'lDepartures' in compLinks:\n",
    "        for (upstream,downstream,otherSide),participants in compLinks[\"lDepartures\"].items():\n",
    "            if otherSide == 'tl':\n",
    "                toRight = False\n",
    "            else:\n",
    "                toRight = True\n",
    "            component['ldepartures'].append({\n",
    "                'upstream': upstream,\n",
    "                'downstream': downstream,\n",
    "                'otherSideRight': toRight,\n",
    "                'participants': participants\n",
    "            })\n",
    "        \n",
    "    return component"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "flying-darwin",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def splitforwardInversedNodeComp(pathList,component,isInverse):\n",
    "#     if component['firstCol']==460:\n",
    "#         pdb.set_trace()\n",
    "    \n",
    "    forward = []\n",
    "    inversed = []\n",
    "    \n",
    "    for pathID in pathList:\n",
    "            try:\n",
    "                if component[\"matrix\"][component[\"occupants\"].index(pathID)][1]>0:\n",
    "                    inversed.append(pathID)\n",
    "                else:\n",
    "                    forward.append(pathID)\n",
    "            except (IndexError,ValueError):\n",
    "                # If it is artificial pass link.\n",
    "                if isInverse:\n",
    "                    inversed.append(pathID)\n",
    "                else:\n",
    "                    forward.append(pathID)\n",
    "                \n",
    "    return forward,inversed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "split-ranking",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "def fillLinks(nodeInComp,nodeToComponent,toLinks,fromLinks,component,components,doLeft=True,doRight=True):\n",
    "    compLinks = {}\n",
    "    \n",
    "    \n",
    "    \n",
    "#     if component['first_bin']==2 and len(nodeInComp)==2:\n",
    "#         pdb.set_trace()\n",
    "    \n",
    "    for node in nodeInComp:\n",
    "        \n",
    "        # Processing all external arrival links\n",
    "        nodeToLink = toLinks.get(node,{})\n",
    "        for fromNode in nodeToLink.keys():\n",
    "            intermediateCondition = (node<fromNode)\n",
    "            \n",
    "            la,ra = splitforwardInversedNodeComp(nodeToLink[fromNode],component,intermediateCondition)\n",
    "            \n",
    "            fromFirstComp = components[nodeToComponent[fromNode-1][0]]\n",
    "            fromLastComp = components[nodeToComponent[fromNode-1][-1]]\n",
    "            \n",
    "            #left arrivals\n",
    "            if len(la)>0 and doLeft:\n",
    "                frd,fld = splitforwardInversedNodeComp(la,fromFirstComp,intermediateCondition)\n",
    "                \n",
    "                complArrivals = compLinks.setdefault('lArrivals',{})\n",
    "                # from right departure\n",
    "                if len(frd)>0 and (fromNode not in nodeInComp) and (fromLastComp['last_bin']+1!=component['first_bin']):\n",
    "                    complArrivals.setdefault((fromLastComp['last_bin'],component['first_bin'],'fr'),set()).update(frd)\n",
    "                #from left departure\n",
    "                if len(fld)>0 and (fromNode not in nodeInComp):\n",
    "                    complArrivals.setdefault((fromFirstComp['first_bin'],component['first_bin'],'fl'),set()).update(fld)\n",
    "            \n",
    "            #right arrivals\n",
    "            if len(ra)>0 and doRight:\n",
    "                frd,fld = splitforwardInversedNodeComp(ra,fromFirstComp,intermediateCondition)\n",
    "                \n",
    "                comprArrivals = compLinks.setdefault('rArrivals',{})\n",
    "                #from right departures\n",
    "                if len(frd)>0 and (fromNode not in nodeInComp):\n",
    "                    comprArrivals.setdefault((fromLastComp['last_bin'],component['last_bin'],'fr'),set()).update(frd)\n",
    "                #from left departures\n",
    "                if len(fld)>0 and (fromNode not in nodeInComp):\n",
    "                    comprArrivals.setdefault((fromFirstComp['first_bin'],component['last_bin'],'fl'),set()).update(fld)\n",
    "            \n",
    "        # Processing all external departure links\n",
    "        nodeFromLink = fromLinks.get(node,{})\n",
    "        for toNode in nodeFromLink.keys():\n",
    "            intermediateCondition = (node>toNode)\n",
    "            \n",
    "            rd,ld = splitforwardInversedNodeComp(nodeFromLink[toNode],component,intermediateCondition)\n",
    "            \n",
    "            toFirstComp = components[nodeToComponent[toNode-1][0]]\n",
    "            toLastComp = components[nodeToComponent[toNode-1][-1]]\n",
    "            \n",
    "            #right departures\n",
    "            if len(rd)>0 and doRight: # Check if doRight is set incorrectly for our case (121->122 at level 4)\n",
    "                tla,tra = splitforwardInversedNodeComp(rd,toFirstComp,intermediateCondition)\n",
    "                \n",
    "                comprDepartures = compLinks.setdefault('rDepartures',{})\n",
    "                #to left arrivals\n",
    "                if len(tla)>0:\n",
    "                    if toNode not in nodeInComp:\n",
    "                         comprDepartures.setdefault((component['last_bin'],toFirstComp['first_bin'],'tl'),set()).update(tla)\n",
    "                    elif node==nodeInComp[-1] and toNode==nodeInComp[0]:\n",
    "                        # Check that repeated element is actually contiguous inside otherwise it is not a loop and should not be visible.\n",
    "                        tla_update = list(set(tla).intersection(*[set(fromLinks.get(nodeInComp[i],{}).get(nodeInComp[i+1],[])) for i in range(len(nodeInComp)-1)]))\n",
    "                        \n",
    "                        if len(tla_update)>0:\n",
    "                            #Loop link, can only happen when with left arrival and right departure (or vice versa).\n",
    "                            complArrivals = compLinks.setdefault('lArrivals',{})\n",
    "\n",
    "                            complArrivals.setdefault((component['last_bin'],component['first_bin'],'fr'),set()).update(tla)\n",
    "                            comprDepartures.setdefault((component['last_bin'],component['first_bin'],'tl'),set()).update(tla)\n",
    "                \n",
    "                #to right arrivals\n",
    "                if len(tra)>0: # Most probably the problem is here! Check it!\n",
    "                    if toNode not in nodeInComp:\n",
    "                        comprDepartures.setdefault((component['last_bin'],toLastComp['last_bin'],'tr'),set()).update(tra)\n",
    "            \n",
    "            #left departures\n",
    "            if len(ld)>0 and doLeft:\n",
    "                tla,tra = splitforwardInversedNodeComp(ld,toFirstComp,intermediateCondition)\n",
    "                \n",
    "                complDepartures = compLinks.setdefault('lDepartures',{})\n",
    "                #to left arrivals\n",
    "                if len(tla)>0:\n",
    "                    if toNode not in nodeInComp:\n",
    "                        complDepartures.setdefault((component['first_bin'],toFirstComp['first_bin'],'tl'),set()).update(tla)\n",
    "                \n",
    "                if len(tra)>0:\n",
    "                    if toNode not in nodeInComp:\n",
    "                        if component['first_bin']-1!=toFirstComp['last_bin']:\n",
    "                            complDepartures.setdefault((component['first_bin'],toLastComp['last_bin'],'tr'),set()).update(tra)\n",
    "                    elif node==nodeInComp[0] and toNode==nodeInComp[-1]:\n",
    "                        # Check that repeated element is actually contiguous inside otherwise it is not a loop and should not be visible.\n",
    "                        tla_update = list(set(tra).intersection(*[set(fromLinks.get(nodeInComp[i],{}).get(nodeInComp[i-1],[])) for i in range(len(nodeInComp)-1,0,-1)]))\n",
    "                        \n",
    "                        if len(tla_update)>0:\n",
    "                            #Loop link, can only happen when with right arrival and left departure (or vice versa).\n",
    "                            comprArrivals = compLinks.setdefault('rArrivals',{})\n",
    "\n",
    "                            comprArrivals.setdefault((component['first_bin'],component['last_bin'],'fl'),set()).update(tra)\n",
    "                            complDepartures.setdefault((component['first_bin'],component['last_bin'],'tr'),set()).update(tra)\n",
    "                        \n",
    "    return compLinks"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
