{
 "cells": [
  {
   "cell_type": "raw",
   "id": "01c0da9b-f099-489e-b39e-5b44e028aa16",
   "metadata": {},
   "source": [
    "---\n",
    "description: all functions to run command line interface\n",
    "output-file: scripts.html\n",
    "title: Script module\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6ba7bb3-3150-4e0c-b5da-2a7e3b9b71c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a486763-588c-4bad-8d91-c7865ee3f51b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| include: false\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c251ef1f-aaeb-4d0d-9b5a-79db8c3ebc8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| include: false\n",
    "from nbdev import nbdev_export\n",
    "nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dff11c4-4671-474f-a241-b56de288a6a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| exporti\n",
    "\n",
    "import argparse\n",
    "import yaml\n",
    "import os\n",
    "import time\n",
    "\n",
    "from pangraph_constructor.graph import GenomeGraph\n",
    "from pangraph_constructor.export import exportProject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9894dc6-de83-4ae4-a11e-a53d534b80d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def pantograph_script():\n",
    "    parser = argparse.ArgumentParser(prog='pantograph',\n",
    "                                     description='This is command line tool giving access to the library called pangraph_constructor for manipulating and \\\n",
    "                                     visualision pre-processing of genome graphs.')\n",
    "    \n",
    "    subparsers = parser.add_subparsers(help='Multiple commands available')\n",
    "    \n",
    "    export_parser = subparsers.add_parser('export-vis',help='Export a set of genome graphs (project) to a project directory structure to be visualised in \\\n",
    "                                                             Pantorgaph visualisation component.')\n",
    "    \n",
    "    export_parser.add_argument('file',help='YAML configuration file for exporting process.')\n",
    "    export_parser.add_argument('-g','--generate',action='store_true',help='Generate model commented configuration file at the given path.')\n",
    "    \n",
    "    export_parser.set_defaults(func=export_project_script)\n",
    "    \n",
    "    ann2graph_parser = subparsers.add_parser('annotation2graph',help='Generate gene graph (as opposed to nucleotide graph) from a set of annotations for multiple \\\n",
    "                                                                    accessions and multiple molecules/chromosomes.')\n",
    "    ann2graph_parser.add_argument('file',help='YAML configuration file for graph generation from annotation.')\n",
    "    ann2graph_parser.add_argument('-g','--generate',action='store_true',help='Generate model commented configuration file at the given path.')\n",
    "    \n",
    "    ann2graph_parser.set_defaults(func=ann2graph_script)\n",
    "    \n",
    "    paths2graph_parser = subparsers.add_parser('paths2graph', help='Generate graph from path file(s).')\n",
    "    \n",
    "    paths2graph_parser.add_argument('file',help='YAML configuration file for graph generation from paths.')\n",
    "    paths2graph_parser.add_argument('-g','--generate',action='store_true',help='Generate model commented configuration file at the given path.')\n",
    "    \n",
    "    paths2graph_parser.set_defaults(func=paths2graph_script)\n",
    "    \n",
    "    sortgraph_parser = subparsers.add_parser('sort-graph',help='Perform graph sorting to get preferred order of nodes for linearised visualisation.')\n",
    "    \n",
    "    sortgraph_parser.add_argument('--quiet','-q',action='store_true', help='Suppress most of output. False (i.e. verbose) is not set.')\n",
    "    \n",
    "    sortgraph_parser.add_argument('--isseq','-s',action='store_true', help='Does this graph contains nucleotide sequences. False is not set.')\n",
    "    \n",
    "    sortgraph_parser.add_argument('--output','-o', action='store', help='File path where to save sorted graph. If not set, the input will be overwritten.')\n",
    "    \n",
    "    sortgraph_parser.add_argument('input', help='Relative (to current directory) or absolute path to the GFA file with the graph to be sorted.')\n",
    "    \n",
    "    sortgraph_parser.set_defaults(func=sort_graph_script)\n",
    "    \n",
    "    args = parser.parse_args()\n",
    "        \n",
    "    if 'func' in vars(args):\n",
    "        args.func(args)\n",
    "    else:\n",
    "        parser.print_help()\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bccc0f0-f57c-47b2-b190-c1f7c2fdb635",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def paths2graph_script(args):\n",
    "    \n",
    "    if 'file' not in vars(args):\n",
    "        raise AttributeError('Positional argument with path to setting YAML file has to be provided.')\n",
    "        \n",
    "    if args.generate:\n",
    "        defaultYAML = \"\"\"\\\n",
    "#####################################################################################\n",
    "# This is example YAML setting file generated for pantograph paths2graph tool.      #\n",
    "# Paths are stubs and should be substituted!!!                                      #\n",
    "#                                                                                   #\n",
    "# There are two versions of path files:                                             #\n",
    "# - v1: Each chromosome/molecule is a separate file, each row of the file has       #\n",
    "#       format as follows                                                           #\n",
    "#       <accession ID>: <node1><dir1>,<node2><dir2>,...                             #\n",
    "#       <dirX> is a directionality of each node and can be '+' or '-'.              #\n",
    "# - v2: One path file contains all chromosomes/molecules, each row of the file      #\n",
    "#       has the following format:                                                   #\n",
    "#       <accession ID>-<chromosome number>: <node1><dir1>,<node2><dir2>,...         #\n",
    "#       Here <chromosome number> can be any id, which will be added to 'Chr' in     #\n",
    "#       all names and IDs.\n",
    "# There is a possibility not to have directionality. In this case, parameter \n",
    "# `directional` should be set to false.\n",
    "#####################################################################################\n",
    "\n",
    "# `pathfileDir` common path to all path files (like path to your project files)\n",
    "pathfileDir: /path/to/pathfiles\n",
    "\n",
    "# `pathFiles` is a dictionary of chromosomes as keys and values with paths (relative to `pathFileDir`) to path files (for v1 path files), or\n",
    "# `pathFile` is a string containing a single path (relative to `pathFileDir`) to path file (for v2 path file).\n",
    "# Only one options should be used! If both options are used, pathFile option will be ignorred!\n",
    "# If you need to process a single path file for a single molecule (v1), then use pathFiles and provide \n",
    "# a list of one file.\n",
    "# pathFiles:\n",
    "#     Chr1: path/to/pathfile1.txt\n",
    "#     Chr2: path/to/pathfile2.txt\n",
    "pathFile: path/to/pathfile.txt\n",
    "\n",
    "# `outputpath` is an absolute path (directory) where to save the resulting graphs\n",
    "outputpath: /path/to/graphs\n",
    "\n",
    "# `outputBasename` is a base for creating graph file names in the following format:\n",
    "# - for v1 <outputBasename>_<chromosome name>.gfa\n",
    "#   here <chromosome name is taken from keys for `pathFiles`\n",
    "# - for v2 <outputBasename>_Chr<chromosome number>.gfa\n",
    "#   here <chromosome number> is taken from the file itself (see file structure above).\n",
    "outputBasename: 'path_graph'\n",
    "\n",
    "# The following parameters are optional:\n",
    "\n",
    "# `sort` defines how the paths (accessions should be sorted). There are three options:\n",
    "# 1) `sort` can be true or false. If it is false, then no sorting happens and accessions will be sorted in the order\n",
    "#    as they appear in pathFile(s). If they appear differently for different chromosomes, then order of accessions\n",
    "#    in graphs will be different. If it is true, then they will sorted in alphabethical order.\n",
    "# 2) `sort` is a string, it should contain one accession name (e.g. reference accession) and it will be placed on top\n",
    "#    with the rest of the accessions sorted in alphabetical order.\n",
    "# 3) `sort` is list of accession names. In this case, the provided order will be used for ordering paths. If some \n",
    "#    accessions will be missing from the list (even if they present in path file(s)), they will be omitted from graphs.\n",
    "#\n",
    "# If `sort` is not provided, default is true.\n",
    "# sort: false\n",
    "# sort: 'reference'\n",
    "# sort:\n",
    "#     - accession3\n",
    "#     - accession2\n",
    "#     - accession1\n",
    "\n",
    "# `directional` is a boolean (true or false) and indicate whether the tool should look for directionality\n",
    "# indication in the file next to node (true) or it is not there and positive directionality should be assumed for all nodes (false).\n",
    "# Default is true.\n",
    "# directional: false\n",
    "\"\"\"\n",
    "\n",
    "        with open(args.file,'w') as f:\n",
    "            f.write(defaultYAML)\n",
    "        return\n",
    "    \n",
    "    with open(args.file,'r') as f:\n",
    "        params = yaml.load(f,Loader=yaml.Loader)\n",
    "    \n",
    "    # Change code to use params read from YAML file.\n",
    "    \n",
    "    pathfileDir = params['pathfileDir'].rstrip('/').rstrip('\\\\')\n",
    "    \n",
    "    directional = params.get('directional', True)\n",
    "    sort_acc = params.get('sort', True)\n",
    "    \n",
    "    outputpath = params[\"outputpath\"].rstrip(\"/\").rstrip(\"\\\\\")\n",
    "    \n",
    "    outputbase = f'{outputpath}{os.path.sep}{params[outputBasename]}'\n",
    "    \n",
    "    if pathFiles in params:\n",
    "    \n",
    "        # For v1\n",
    "        for seqName,pathfile in params['pathFiles'].items(): \n",
    "\n",
    "            paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{pathfile}',directional,sort_acc)\n",
    "\n",
    "            # paths = {}\n",
    "            # for accession in sorted(list(_paths.keys())):\n",
    "            #     paths[accession] = _paths[accession]\n",
    "            # del _paths\n",
    "\n",
    "            coregraph = GenomeGraph(pathsDict=paths)\n",
    "\n",
    "            # On undirected coregraph sorting is not optimal! Check sorting!!!\n",
    "\n",
    "            coregraph.treeSort()\n",
    "\n",
    "            if len(coregraph.nodes)!=len(coregraph.order):\n",
    "                print('Sorting failed and not all nodes were sorted. Saving unsorted graph')\n",
    "                output = f'{outputbase}_{seqName}_unordered.gfa'\n",
    "                coregraph.order = list(range(1,len(coregraph.nodes)+1))\n",
    "                coregraph.toGFA(output,doSeq=False)\n",
    "            else:\n",
    "                coreGFApath = f'{outputbase}_{seqName}.gfa'\n",
    "                coregraph.toGFA(coreGFApath,False)\n",
    "                \n",
    "    elif pathFile in params:\n",
    "        \n",
    "        # For v2\n",
    "        pathfileDir = '../../1001G/coreGraph/new_Nov2022/'\n",
    "\n",
    "        pathsfile = f'paths.txt'\n",
    "\n",
    "        paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{params[\"pathFile\"]}',directional,sort_acc,True)\n",
    "\n",
    "        for seqNum in paths.keys():#range(1,6): \n",
    "            # paths = {}\n",
    "            # for accession in sorted(list(_paths.keys())):\n",
    "            #     paths[accession] = _paths[accession]\n",
    "            # del _paths\n",
    "\n",
    "            coregraph = GenomeGraph(pathsDict=paths[seqNum])\n",
    "\n",
    "            # On undirected coregraph sorting is not optimal! Check sorting!!!\n",
    "\n",
    "            coregraph.treeSort()\n",
    "\n",
    "            if len(coregraph.nodes)!=len(coregraph.order):\n",
    "                print('Sorting failed and not all nodes were sorted. Saving unsorted graph')\n",
    "                output = f'{outputbase}_Chr{seqNum}_unordered.gfa'\n",
    "                coregraph.order = list(range(1,len(coregraph.nodes)+1))\n",
    "                coregraph.toGFA(output,doSeq=False)\n",
    "            else:\n",
    "                coreGFApath = f'{outputbase}_Chr{seqNum}.gfa'\n",
    "                coregraph.toGFA(coreGFApath,False)\n",
    "                \n",
    "    else:\n",
    "        raise AttributeError('Either `pathFiles` or `pathFile` should be provided in setting file.')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d252a837-860b-44ef-9129-6a1ef0e26aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def ann2graph_script(args):\n",
    "    if 'file' not in vars(args):\n",
    "        raise AttributeError('Positional argument with path to setting YAML file has to be provided.')\n",
    "        \n",
    "    if args.generate:\n",
    "        defaultYAML = \"\"\"\\\n",
    "#####################################################################################\n",
    "# This is example YAML setting file generated for pantograph annotation2graph tool. #\n",
    "# Paths are stubs and should be substituted!!!                                      #\n",
    "#####################################################################################\n",
    "\n",
    "# `graphpath` is full path including filename where the graph should be saved. If extension is not provided, gfa will be added.\n",
    "# If `chromosomes` are provided, then the actual files will get chromosomes names at the end of file name (before extension), e.g.\n",
    "# graphpath=/path/to/graph.gfa and chromosomes contains 'chr1', then final file will be /path/to/graph_chr1.gfa\n",
    "graphpath: /path/to/graph.gfa\n",
    "\n",
    "# `annotationFilesPaths` common path to all files (like path to your project files)\n",
    "annotationFilesPaths: /path/to/project\n",
    "\n",
    "# `annotationFiles` is a list of paths (relative to `annotationFilesPaths`) to annotation (gff3) files, \n",
    "# This is required parameter!\n",
    "annotationFiles:\n",
    "    - 'path/to/file1.gff'\n",
    "    - 'path/to/file2.gff'\n",
    "\n",
    "# The following parameters are optional:\n",
    "\n",
    "# `sequenceFilesDict` is a dictionary, where keys are IDs of accessions used in annotation files and value\n",
    "# is a path to FASTA file (relative to `annotationFilesPaths`). Assumption is that FASTA sequence names\n",
    "# are the same as GFF3 sequence names.\n",
    "# sequenceFilesDict:\n",
    "#     accession1: path/to/fileA.fasta\n",
    "#     accession2: path/to/fileB.fasta\n",
    "\n",
    "# `pangenomeFiles` is a list of GFF3 file paths (relative to `annotationFilesPaths`) for the same intervals as in `annotationFiles`, \n",
    "# ID fields in GFF2 metadata should be the same.\n",
    "# pangenomeFiles:\n",
    "#     - 'path/to/file1.gff'\n",
    "#     - 'path/to/file2.gff'\n",
    "\n",
    "\n",
    "# `accOrder` is a list which gives an order of accessions in the graph. If not given, accessions will be sorted in \n",
    "# alphabetical order.\n",
    "accOrder:\n",
    "    - accession2\n",
    "    - accession1\n",
    "\n",
    "# `chromosomes` a list of molecules/chromosomes, if provided, the tool will create only one graph for each chromosome, otherwise, it will create one graph for \n",
    "# all chromosomes (exporting of such graphs to Pantograph Visualisation is not currently implemented).\n",
    "# chromosomes: \n",
    "#    - 'Chr1'\n",
    "#    - 'Chr2'\n",
    "#    - 'Chr3'\n",
    "\n",
    "# `doUS` (default: False) if set, tool will add unrelated sequence blocks between annotated genes/blocks. Not recommended.\n",
    "# doUS: true\n",
    "\n",
    "# `refAnnotationFile` is a path (relative to `annotationFilesPaths`) to gff3 file with reference notation for gene names. \n",
    "# In main annotations reference gene names should be identified either in \"gene\" records under \"AT\" key (prioritised), \n",
    "# or in \"mRNA\" records under \"Name\" key (fallback). If `transmapFile` is provided, it will be used to find corresponding \n",
    "# similarity IDs and reference genes IDs.\n",
    "# refAnnotationFile: path/to/reference.gff\n",
    "\n",
    "# `refAccession` is an Accession ID for reference annotation (should be provided if `refAnnotationFile` if provided).\n",
    "refAccession: 'reference_accession'\n",
    "\n",
    "# `refSequenceFile` is a FASTA path (relative to `annotationFilesPaths`), where sequences of each block/gene for reference.\n",
    "# refSequenceFile: path/to/reference.fasta\n",
    "\n",
    "# `transmapFile` is a tab delimited file with column \"Orthogroup\", which contains similarity ID for genes and a column with name\n",
    "# given by `transmapFileRefCol` which contains reference annotation gene names. There should be either both or none provided.\n",
    "# transmapFile: path/to/transmap.map\n",
    "# transmapFileRefCol: 'reference'\n",
    "\"\"\"\n",
    "\n",
    "        with open(args.file,'w') as f:\n",
    "            f.write(defaultYAML)\n",
    "        return\n",
    "    \n",
    "    with open(args.file,'r') as f:\n",
    "        params = yaml.load(f,Loader=yaml.Loader)\n",
    "    \n",
    "    commonpath = params.pop('annotationFilesPaths').rstrip('/').rstrip('\\\\')\n",
    "    \n",
    "    graphpath = params.pop('graphpath')\n",
    "    gpath,gext = os.path.splitext(graphpath)\n",
    "    if gext=='':\n",
    "        gext = 'gfa'\n",
    "    \n",
    "    params['annotationFiles'] = [f'{commonpath}{os.path.sep}{p}' for p in params['annotationFiles']]\n",
    "    \n",
    "    if 'sequenceFilesDict' in params:\n",
    "        params['sequenceFilesDict'] = {k:f'{commonpath}{os.path.sep}{v}' for k,v in params['sequenceFilesDict']}\n",
    "    \n",
    "    if 'pangenomeFiles' in params:\n",
    "        params['pangenomeFiles'] = [f'{commonpath}{os.path.sep}{p}' for p in params['pangenomeFiles']]\n",
    "        \n",
    "    if 'refAnnotationFile' in params:\n",
    "        params['refAnnotationFile'] = f'{commonpath}{os.path.sep}{params[\"refAnnotationFile\"]}'\n",
    "    \n",
    "    if 'refSequenceFile' in params:\n",
    "        params['refSequenceFile'] = f'{commonpath}{os.path.sep}{params[\"refSequenceFile\"]}'\n",
    "    \n",
    "    if 'transmapFile' in params:\n",
    "        params['transmapFile'] = f'{commonpath}{os.path.sep}{param[\"transmapFile\"]}'\n",
    "    \n",
    "    \n",
    "    \n",
    "    if 'chromosomes' in params:\n",
    "        chrs = params.pop('chromosomes')\n",
    "        \n",
    "        for chromosome in chrs:\n",
    "            print(f'\\nProcessing {chromosome}\\n============')\n",
    "\n",
    "            curtst = time.time()\n",
    "            \n",
    "            graph = GenomeGraph(**params, chromosome = chromosome)\n",
    "\n",
    "            print(f'Generating graph for {chromosome} took {time.time() - curtst} seconds')\n",
    "            \n",
    "            \n",
    "            \n",
    "            curtst = time.time()\n",
    "            graph.treeSort()\n",
    "            print(f'Sorting graph for {chromosome} took {time.time() - curtst} seconds')\n",
    "            if len(graph.nodes)!=len(graph.order):\n",
    "                print('Sorting failed and not all nodes were sorted. Saving unsorted graph')\n",
    "                output = f'{gpath}_{chromosome}_unsorted.{gext}'\n",
    "                graph.order = list(range(1,len(graph.nodes)+1))\n",
    "                graph.toGFA(output,doSeq=False)\n",
    "            else:\n",
    "                output = f'{gpath}_{chromosome}.{gext}'\n",
    "                graph.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=False)\n",
    "    else:\n",
    "        print(f'\\nProcessing single graph for all chromosomes\\n============')\n",
    "\n",
    "        curtst = time.time()\n",
    "        \n",
    "        graph = GenomeGraph(**params)\n",
    "\n",
    "        print(f'Generating graph for all chromosomes took {time.time() - curtst:.0f} seconds')\n",
    "        \n",
    "        curtst = time.time()\n",
    "        graph.treeSort()\n",
    "        print(f'Sorting graph for all chromosomes took {time.time() - curtst:.0} seconds')\n",
    "        \n",
    "        if len(graph.nodes)!=len(graph.order):\n",
    "            print('Sorting failed and not all nodes were sorted. Saving unsorted graph')\n",
    "            output = f'{gpath}_unordered.{gext}'\n",
    "            graph.order = list(range(1,len(graph.nodes)+1))\n",
    "            graph.toGFA(output,doSeq=False)\n",
    "        else:\n",
    "            output = f'{gpath}.{gext}'\n",
    "            graph.toGFA(output,doSeq=False)\n",
    "    # change function call here.\n",
    "    #exportProject(**params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d036cfa1-d03c-4a79-b30c-983e249eff18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def sort_graph_script(args):\n",
    "    inputfile = args.input\n",
    "    \n",
    "    vargs = vars(args)\n",
    "    isseq = args.get('isseq',False)\n",
    "    \n",
    "    outputfile = vargs.get('output', inputfile)\n",
    "    \n",
    "    quiet = vargs.get('quiet',False)\n",
    "    \n",
    "    graph_new = GenomeGraph(gfaPath=inputfile,isGFASeq=isseq,quiet=quiet)\n",
    "\n",
    "    graph_new.treeSort()\n",
    "\n",
    "    assert len(graph_new.nodes)==len(graph_new.order)\n",
    "\n",
    "    graph_new.toGFA(outputfile,doSeq=isseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "211f1edb-9f8e-4f76-9f29-bd6f12c0afd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| export\n",
    "def export_project_script(args):\n",
    "    if 'file' not in vars(args):\n",
    "        raise AttributeError('Positional argument with path to setting YAML file has to be provided.')\n",
    "    \n",
    "    print(args)\n",
    "    \n",
    "    if args.generate:\n",
    "        defaultYAML = \"\"\"\\\n",
    "###############################################################################\n",
    "# This is example YAML setting file generated for pantograph export-vis tool. #\n",
    "# Paths are stubs and should be substituted!!!                                #\n",
    "###############################################################################\n",
    "\n",
    "# projectID - this is a project id and will be the name of directory that holds all data to visualise the project.\n",
    "#             It is recommended not to have special symbols and spaces in this name\n",
    "# REQUIRED\n",
    "projectID: pangenome_project\n",
    "\n",
    "# projectName - Human readable short name describing the project. It should not be more than 50 symbols.\n",
    "# REQUIRED\n",
    "projectName: \"All chromosomes for 27 accessions of species X\"\n",
    "\n",
    "# caseDict - a dictionary of each individual graph within project. Usually one file per chromosome, although, in some cases \n",
    "#            there can be only one graph for all chromosomes or other cases.\n",
    "#            Graph names can contain path relative to `pathToGraphs` if they are in different directories.\n",
    "# REQUIRED\n",
    "caseDict:\n",
    "    Chr1: pangenome_graph_chromosome1.gfa\n",
    "    Chr2: pangenome_graph_chromosome2.gfa\n",
    "\n",
    "# pathToIndex - along with project data structure, exporting adds information to the index.json file which is\n",
    "#               read by Pantograph visualisation component in the first instance. If it does not exist at the given path\n",
    "#               it will be created at the given path.\n",
    "# REQUIRED\n",
    "pathToIndex: /path/to/pantograph/visualisation/data/root/\n",
    "\n",
    "# pathToGraphs - path where graph files mentioned in `caseDict` can be found.\n",
    "# REQUIRED\n",
    "pathToGraphs: /path/to/all/graph/files/\n",
    "\n",
    "# When graphs are exported some metadata (annotations, genome and pangenome positions) can be recorded to Redis DB\n",
    "# If Redis DB is not available or recording of metadata is not needed, then parameter redisHost should be omitted.\n",
    "# Otherwise, if Redis DB is available and metadata should be recorded, then redisHost should be set to the hostname (or IP address)\n",
    "# of the Redis DB server\n",
    "# redisHost: 'redis.example.com'\n",
    "# By default, this tool will use port 6379 for Redis DB server and DB ID 0 (zero). If any of these should be changed, then set relevant variables as follows:\n",
    "# redisPort: 6379\n",
    "# redisDB: 0\n",
    "\n",
    "# Suffix is used for the naming of the output directories for each graph (molecule or chromosome), e.g. if project is called 'pangenome_project', \n",
    "# case is 'chr1' and suffix is '_new', then the output directory for this graph visualisation data structure will be called 'pangenome_project_chr1_new'\n",
    "# If not set, then no suffix will be added.\n",
    "# suffix: ''\n",
    "\n",
    "# maxLengthComponent defined how large (in visual columns) can the component be. If continuity allows for longer components, it will be broken\n",
    "# into components of maximum `maxLengthComponent` columns. If not set, default of 100 is used.\n",
    "# maxLengthComponent: 100\n",
    "\n",
    "# maxLengthChunk defined how many components will be recorded into a single file in the visualisation data structure. Smaller chunk files allows to\n",
    "# load data into memory more precisely, but if a lot of small components should be loaded, larger chunks make it quicker.\n",
    "# If not set, default of 6 will be used.\n",
    "# maxLengthChunk: 6,\n",
    "\n",
    "# Each cell has calculated inversion ratio (see user manual for details). It will be shown as inverted (shades of red) if its inversion ratio\n",
    "# is more than `inversionThreshold` set\n",
    "# If not set, default is 0.5\n",
    "# inversionThreshold: 0.5,\n",
    "\n",
    "# If the graph is not nucleotide graph (e.g. gene graph generated from paths or annotations), `isSeq` should be set to False.\n",
    "# Default is True\n",
    "# isSeq: False,\n",
    "\n",
    "# Which zoom levels should be generated. The number indicate how many maximum nucleotides (or genes in case of gene graph) can be displayed as\n",
    "# single column in the visualised matrix. Zoom level of 1 is added automatically if it is not in the list. Each previous zoom level should be \n",
    "# a factor or the next one, i.e. next zoom level should be obtained by multiplying previous one by an integer more than 1.\n",
    "# E.g. zoom levels [1,2,4,8] is correct, but zoom levels [1,2,4,6] is incorrect because 4 is not a factor of 6.\n",
    "# Defaults is [1], meaning to generate only lowest zoom level (single nucleotide/gene resolution).\n",
    "# zoomLevels: [1,2,4]\n",
    "\n",
    "# If `fillZoomLevel` set to True, then on top of the highest zoom level set in `zoomLevels`, the function will keep adding more zoom levels\n",
    "# by multiplying it by 2 until there is no non-linear links to remove left.\n",
    "# Default is True\n",
    "# fillZoomLevel: False\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        with open(args.file,'w') as f:\n",
    "            f.write(defaultYAML)\n",
    "        return\n",
    "    \n",
    "    with open(args.file,'r') as f:\n",
    "        params = yaml.load(f,Loader=yaml.Loader)\n",
    "    exportProject(**params)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ccfc81d-3ff2-46d8-92f5-91f0bbd4301b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
