{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "description: This package provides all functionality from graph construction (currently,\n",
    "  from annotation, in the future, from raw assembled sequences) to graph processing\n",
    "  (sorting, grouping, adjusting).\n",
    "output-file: index.html\n",
    "title: pantograph_constructor\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#| include: false\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "%load_ext line_profiler\n",
    "%load_ext memory_profiler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### code profiling help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://jakevdp.github.io/PythonDataScienceHandbook/01.07-timing-and-profiling.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`%time`: Time the execution of a single statement  \n",
    "`%timeit`: Time repeated execution of a single statement for more accuracy  \n",
    "`%prun`: Run code with the profiler  \n",
    "`%lprun`: Run code with the line-by-line profiler  \n",
    "`%memit`: Measure the memory use of a single statement  \n",
    "`%mprun`: Run code with the line-by-line memory profiler  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Install"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enter the directory of the library and enter:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`pip install .`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from nbdev import nbdev_export\n",
    "nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from redis import Redis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cProfile\n",
    "import pstats\n",
    "from pstats import SortKey"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pdb\n",
    "\n",
    "import os\n",
    "import glob\n",
    "import re\n",
    "import time\n",
    "import pdb\n",
    "from random import shuffle\n",
    "from copy import deepcopy\n",
    "import joblib\n",
    "\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "\n",
    "from pangraph_constructor.graph import GenomeGraph\n",
    "from pangraph_constructor.synteny import generateOrder,readTransMap\n",
    "from pangraph_constructor.tree import TremauxTree\n",
    "from pangraph_constructor.utils import pathFileToPathDict,resetDB\n",
    "from pangraph_constructor.utils import iset_add,iset_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resetDB()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "redisConn = Redis(host='redis',port=6379)\n",
    "\n",
    "redisConn.flushall()\n",
    "\n",
    "redisConn.close()\n",
    "\n",
    "del redisConn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pangraph_constructor.utils import adjustZoomLevels,pathConvert\n",
    "from pangraph_constructor.exportDev import exportToPantograph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter('ignore')\n",
    "warnings.filterwarnings('always',category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating from annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "datadir = '../../1001G/annotations/freeze2.1'\n",
    "gfadir = '../../1001G/pantograph/data'"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "datadir = '.'\n",
    "gfadir = '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotationFiles = sorted(glob.glob(f'{datadir}{os.path.sep}*.gff'))\n",
    "sequenceFiles = sorted(glob.glob(f'{datadir}{os.path.sep}sequences{os.path.sep}*.fasta'))\n",
    "transMapFile = f'{datadir}{os.path.sep}TransMap{os.path.sep}TransMap.map'\n",
    "refAnnotationFile = f'{datadir}{os.path.sep}outgroups/araport.gff'\n",
    "refSequenceFile = f'{datadir}{os.path.sep}outgroups/araport.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('../../1001G/annotations/freeze2.1/outgroups/araport.gff',\n",
       " '../../1001G/annotations/freeze2.1/outgroups/araport.fasta',\n",
       " ['../../1001G/annotations/freeze2.1/10002.gff',\n",
       "  '../../1001G/annotations/freeze2.1/10015.gff',\n",
       "  '../../1001G/annotations/freeze2.1/10024.gff',\n",
       "  '../../1001G/annotations/freeze2.1/1741.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22001.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22002.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22003.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22004.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22005.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22006.gff',\n",
       "  '../../1001G/annotations/freeze2.1/22007.gff',\n",
       "  '../../1001G/annotations/freeze2.1/6024.gff',\n",
       "  '../../1001G/annotations/freeze2.1/6069.gff',\n",
       "  '../../1001G/annotations/freeze2.1/6124.gff',\n",
       "  '../../1001G/annotations/freeze2.1/6244.gff',\n",
       "  '../../1001G/annotations/freeze2.1/6909.gff',\n",
       "  '../../1001G/annotations/freeze2.1/6966.gff',\n",
       "  '../../1001G/annotations/freeze2.1/8236.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9075.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9537.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9543.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9638.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9728.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9764.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9888.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9905.gff',\n",
       "  '../../1001G/annotations/freeze2.1/9981.gff'],\n",
       " ['../../1001G/annotations/freeze2.1/sequences/10002.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/10015.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/10024.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/1741.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22001.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22002.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22003.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22004.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22005.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22006.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/22007.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/6024.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/6069.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/6124.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/6244.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/6909.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/6966.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/8236.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9075.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9537.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9543.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9638.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9728.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9764.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9888.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9905.fasta',\n",
       "  '../../1001G/annotations/freeze2.1/sequences/9981.fasta'],\n",
       " '../../1001G/annotations/freeze2.1/TransMap/TransMap.map')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refAnnotationFile,refSequenceFile,annotationFiles,sequenceFiles,transMapFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ATmap = readTransMap(transMapFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 1,\n",
       " 2,\n",
       " 3,\n",
       " 4,\n",
       " 5,\n",
       " 6,\n",
       " 7,\n",
       " 8,\n",
       " 9,\n",
       " 10,\n",
       " 11,\n",
       " 12,\n",
       " 13,\n",
       " 14,\n",
       " 15,\n",
       " 16,\n",
       " 17,\n",
       " 18,\n",
       " 19,\n",
       " 20,\n",
       " 21,\n",
       " 22,\n",
       " 23,\n",
       " 24,\n",
       " 25,\n",
       " 26]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fileOrder = generateOrder(annotationFiles,priorityAccession=None)\n",
    "fileOrder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GenomeGraph(annotationFiles=annotationFiles,\n",
    "#                         sequenceFiles=None,\n",
    "#                         fileOrder=fileOrder,\n",
    "#                         doUS=doUS,\n",
    "#                         seqSuffix=seqSuffix,\n",
    "#                         refAnnotationFile=refAnnotationFile,\n",
    "#                         refAccession='TAIR10',\n",
    "#                         transMap=ATmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def __init__(self,gfaPath=None,doOverlapCleaning=True,\n",
    "#                  paths=None,\n",
    "#                  nodes=None,nodesData=None,links=None,\n",
    "#                  pathsDict=None,\n",
    "#                  sequenceFiles=None,annotationFiles=None,\n",
    "#                  doBack=False,**kwargs)\n",
    "self._graphFromAnnotation(annotationFiles,sequenceFiles,**kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotationFiles,fileOrder,doUS,seqSuffix,ATMap,refAnnotationFile,refAccession#'TAIR10'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# _graphFromAnnotation(self,annotationFiles,sequenceFiles=None,**kwargs):\n",
    "self.nodeNameToID = {}\n",
    "\n",
    "fileOrder = kwargs.get('fileOrder',list(range(len(annotationFiles))))\n",
    "\n",
    "doUS = kwargs.get('doUS',False)\n",
    "self.usCounter = 0\n",
    "self.OGList = []\n",
    "links = self._linksDictToSet(self.forwardLinks)\n",
    "\n",
    "if 'refAnnotationFile' in kwargs:\n",
    "    links = self._processAnnotation(kwargs['refAnnotationFile'], links,\n",
    "                                    ATmap=kwargs.get('transMap', None),\n",
    "                                    seqFile=kwargs.get('refSequenceFile',None),\n",
    "                                    seqSuffix=kwargs.get('seqSuffix', None),\n",
    "                                    doUS=doUS,\n",
    "                                    isRef=True,\n",
    "                                    accID=kwargs.get('refAccession',None))\n",
    "\n",
    "for fileNum in fileOrder:\n",
    "    if sequenceFiles is not None:\n",
    "        seqFile = sequenceFiles[fileNum]\n",
    "    else:\n",
    "        seqFile = None\n",
    "     # annotationFile\n",
    "     # transMap\n",
    "    links = self._processAnnotation(annotationFiles[fileNum], links,\n",
    "                                    ATmap=kwargs.get('transMap', None),\n",
    "                                    seqFile=seqFile,\n",
    "                                    seqSuffix=kwargs.get('seqSuffix', None),\n",
    "                                    doUS=doUS)\n",
    "\n",
    "\n",
    "self.forwardLinks = self._linksSetToDict(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# _processAnnotation(self, annotationFile, links, ATmap=None, seqFile=None, seqSuffix=None, doUS=False, isRef=False, accID=None):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../../1001G/annotations/freeze2.1/10024.gff'"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotationFiles[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pangraph_constructor.synteny import processAccession,generatePathsLinks\n",
    "from skbio import read as skbio_read\n",
    "from skbio.metadata import IntervalMetadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ann = skbio_read(annotationFiles[0], format='gff3', into=IntervalMetadata, seq_id='10002_Chr1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Interval(interval_metadata=<140264681975920>, bounds=[(252677, 255684)], fuzzy=[(False, False)], metadata={'source': 'EVM', 'type': 'gene', 'score': '.', 'strand': '+', 'ID': 'evm.TU.10002_Chr1.68', 'Name': 'EVM%20prediction%2010002_Chr1.68', 'OG': 'OG0001856', 'AT': 'AT1G01690,AT1G01695'})]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = ann.query(bounds=((252670,252680),),metadata={'type':'gene'})\n",
    "list(q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "accessionID, genes, sequences = \\\n",
    "        processAccession(annotationFiles[2],\n",
    "                         None,\n",
    "                         ATmap=ATmap,\n",
    "                         isRef=False,\n",
    "                         accID=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>orthogroup</th>\n",
       "      <th>sequenceID</th>\n",
       "      <th>accessionID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>forward</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>AT_str</th>\n",
       "      <th>geneSeq</th>\n",
       "      <th>overlapGenes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13382</th>\n",
       "      <td>42</td>\n",
       "      <td>OG0000490</td>\n",
       "      <td>10024_Chr1</td>\n",
       "      <td>10024</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>153796</td>\n",
       "      <td>155238</td>\n",
       "      <td>[AT1G01390, AT1G01420, AT4G01070]</td>\n",
       "      <td></td>\n",
       "      <td>[evm.TU.10024_Chr1.42]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13385</th>\n",
       "      <td>45</td>\n",
       "      <td>OG0000490</td>\n",
       "      <td>10024_Chr1</td>\n",
       "      <td>10024</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>False</td>\n",
       "      <td>160034</td>\n",
       "      <td>161479</td>\n",
       "      <td>[AT1G01390, AT1G01420, AT4G01070]</td>\n",
       "      <td></td>\n",
       "      <td>[evm.TU.10024_Chr1.45]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>123</td>\n",
       "      <td>OG0000490</td>\n",
       "      <td>10024_Chr4</td>\n",
       "      <td>10024</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>False</td>\n",
       "      <td>456556</td>\n",
       "      <td>457998</td>\n",
       "      <td>[AT1G01390, AT1G01420, AT4G01070]</td>\n",
       "      <td></td>\n",
       "      <td>[evm.TU.10024_Chr4.123]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      geneID orthogroup  sequenceID accessionID chromosome  forward   start  \\\n",
       "13382     42  OG0000490  10024_Chr1       10024       Chr1    False  153796   \n",
       "13385     45  OG0000490  10024_Chr1       10024       Chr1    False  160034   \n",
       "130      123  OG0000490  10024_Chr4       10024       Chr4    False  456556   \n",
       "\n",
       "          end                             AT_str geneSeq  \\\n",
       "13382  155238  [AT1G01390, AT1G01420, AT4G01070]           \n",
       "13385  161479  [AT1G01390, AT1G01420, AT4G01070]           \n",
       "130    457998  [AT1G01390, AT1G01420, AT4G01070]           \n",
       "\n",
       "                  overlapGenes  \n",
       "13382   [evm.TU.10024_Chr1.42]  \n",
       "13385   [evm.TU.10024_Chr1.45]  \n",
       "130    [evm.TU.10024_Chr4.123]  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.loc[genes.orthogroup=='OG0000490']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>orthogroup</th>\n",
       "      <th>sequenceID</th>\n",
       "      <th>accessionID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>forward</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>AT_str</th>\n",
       "      <th>geneSeq</th>\n",
       "      <th>overlapGenes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>123</td>\n",
       "      <td>OG0000490</td>\n",
       "      <td>10024_Chr4</td>\n",
       "      <td>10024</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>False</td>\n",
       "      <td>456556</td>\n",
       "      <td>457998</td>\n",
       "      <td>[AT1G01390, AT1G01420, AT4G01070]</td>\n",
       "      <td></td>\n",
       "      <td>[evm.TU.10024_Chr4.123]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    geneID orthogroup  sequenceID accessionID chromosome  forward   start  \\\n",
       "130    123  OG0000490  10024_Chr4       10024       Chr4    False  456556   \n",
       "\n",
       "        end                             AT_str geneSeq  \\\n",
       "130  457998  [AT1G01390, AT1G01420, AT4G01070]           \n",
       "\n",
       "                overlapGenes  \n",
       "130  [evm.TU.10024_Chr4.123]  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.loc[(genes.orthogroup=='OG0000490') & (genes.sequenceID!='10024_Chr1')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if seqSuffix is None:\n",
    "        seqList = genes.sequenceID.unique().tolist()\n",
    "        seqList.sort()\n",
    "    else:\n",
    "        seqList = [f'{accessionID}{seqSuffix}']\n",
    "\n",
    "    path = []\n",
    "    for seqID in seqList:\n",
    "\n",
    "        p, cigar, usCounter = generatePathsLinks(genes.loc[genes.sequenceID == seqID], seqID, accessionID, sequences, self.OGList,\n",
    "                                                 self.nodes, self.nodesMetadata, self.nodeNameToID, links,\n",
    "                                                 self.usCounter, doUS=doUS, segmentData=self.nodesData)\n",
    "        path = path + p\n",
    "    if isRef:\n",
    "        self.paths.insert(0,path)\n",
    "        self.accessions.insert(0,accessionID)\n",
    "    else:\n",
    "        self.paths.append(path)\n",
    "        self.accessions.append(accessionID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing Chr1\n",
      "============\n",
      "Calculating nodes length...\n",
      "Processing node 7455/7455\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Node 7455 inverted\n",
      "Generating graph for Chr1 took 2385.88067817688 seconds\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 7454/7455\n",
      "Sorting graph for Chr1 took 138.0292296409607 seconds\n",
      "\n",
      "Processing Chr2\n",
      "============\n",
      "Calculating nodes length...\n",
      "Processing node 4877/4877\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Node 4873 inverted\n",
      "Generating graph for Chr2 took 1497.942242860794 seconds\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 4876/4877\n",
      "Sorting graph for Chr2 took 21.346941709518433 seconds\n",
      "\n",
      "Processing Chr3\n",
      "============\n",
      "Calculating nodes length...\n",
      "Processing node 8442/8442\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Node 8442 inverted\n",
      "Generating graph for Chr3 took 1920.5248539447784 seconds\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 8441/8442\n",
      "Sorting graph for Chr3 took 46.73442721366882 seconds\n",
      "\n",
      "Processing Chr4\n",
      "============\n",
      "Calculating nodes length...\n",
      "Processing node 4620/4620\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Node 4618 inverted\n",
      "Generating graph for Chr4 took 1449.75550699234 seconds\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 4619/4620\n",
      "Sorting graph for Chr4 took 26.44510269165039 seconds\n",
      "\n",
      "Processing Chr5\n",
      "============\n",
      "Calculating nodes length...\n",
      "Processing node 7052/7052\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Node 7050 inverted\n",
      "Generating graph for Chr5 took 2051.643649816513 seconds\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 7051/7052\n",
      "Sorting graph for Chr5 took 44.098389625549316 seconds\n"
     ]
    }
   ],
   "source": [
    "doUS = False\n",
    "for chrnum in range(1,6):\n",
    "    seqID = f'Chr{chrnum}'\n",
    "    gfaFilename = f'AT_{seqID}_OGOnly_2.1.gfa'\n",
    "    seqSuffix = f'_{seqID}'\n",
    "\n",
    "    print(f'\\nProcessing {seqID}\\n============')\n",
    "\n",
    "    curtst = time.time()\n",
    "    graph = GenomeGraph(annotationFiles=annotationFiles,\n",
    "                        sequenceFiles=None,\n",
    "                        fileOrder=fileOrder,\n",
    "                        doUS=doUS,\n",
    "                        seqSuffix=seqSuffix,\n",
    "                        refAnnotationFile=refAnnotationFile,\n",
    "                        refAccession='TAIR10',\n",
    "                        transMap=ATmap)\n",
    "    print(f'Generating graph for {seqID} took {time.time() - curtst} seconds')\n",
    "    \n",
    "    curtst = time.time()\n",
    "    graph.treeSort()\n",
    "    print(f'Sorting graph for {seqID} took {time.time() - curtst} seconds')\n",
    "    assert len(graph.nodes)==len(graph.order)\n",
    "    \n",
    "    \n",
    "    graph.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=False)\n",
    "    # graph.toGFA(f'{gfadir}{os.path.sep}unordered_{gfaFilename}',doSeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "curT = time.localtime()\n",
    "message = f\"Generating gene graphs for all chromosome finished at \\\n",
    "            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\"\n",
    "!ntfy send \"{message}\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "graph = GenomeGraph(gfaPath=f'{gfadir}{os.path.sep}unordered_{gfaFilename}',isGFASeq=False)\n",
    "curtst = time.time()\n",
    "graph.treeSort()\n",
    "print(f'Sorting graph took {time.time() - curtst} seconds')\n",
    "\n",
    "assert len(graph.nodes)==len(graph.order)\n",
    "\n",
    "graph.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "doUS = False\n",
    "\n",
    "for seqID in ['Chr1','Chr2','Chr3','Chr4','Chr5']:\n",
    "    \n",
    "    print(f'\\nProcessing {seqID}\\n============')\n",
    "    \n",
    "    gfaFilename = f'AT_{seqID}_OGOnly_2.1.gfa'\n",
    "    seqSuffix = f'_{seqID}'\n",
    "    \n",
    "#     curtst = time.time()\n",
    "#     graph = GenomeGraph(annotationFiles=annotationFiles,\n",
    "#                         sequenceFiles=None,\n",
    "#                         fileOrder=fileOrder,\n",
    "#                         doUS=doUS,\n",
    "#                         seqSuffix=seqSuffix,\n",
    "#                         refAnnotationFile=refAnnotationFile,\n",
    "#                         refAccession='TAIR10',\n",
    "#                         transMap=ATmap)\n",
    "#     print(f'Generating graph took {time.time() - curtst} seconds')\n",
    "    \n",
    "#     graph.toGFA(f'{gfadir}{os.path.sep}unordered_{gfaFilename}',doSeq=False)\n",
    "    \n",
    "    graph = GenomeGraph(gfaPath=f'{gfadir}{os.path.sep}unordered_{gfaFilename}',isGFASeq=False)\n",
    "    \n",
    "    curtst = time.time()\n",
    "    graph.treeSort()\n",
    "    print(f'Generating graph took {time.time() - curtst} seconds')\n",
    "    assert len(graph.nodes)==len(graph.order)\n",
    "    \n",
    "    \n",
    "    graph.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading Pathfile to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted 05_export.ipynb.\n",
      "Converted 05_exportDev.ipynb.\n",
      "Converted deBruijnGraphProcessing.ipynb.\n",
      "Converted dev.ipynb.\n",
      "Converted graphTesting.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating nodes length...\n",
      "Processing node 35/35\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 34/35\n",
      "Calculating nodes length...\n",
      "Processing node 19/19\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 18/19\n",
      "Calculating nodes length...\n",
      "Processing node 26/26\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 25/26\n",
      "Calculating nodes length...\n",
      "Processing node 15/15\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 14/15\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 6/7\n"
     ]
    }
   ],
   "source": [
    "# For v1\n",
    "pathfileDir = '../../1001G/coreGraph'\n",
    "\n",
    "for seqNum in range(1,6): \n",
    "\n",
    "    pathsfile = f'paths{seqNum:d}.txt'\n",
    "\n",
    "    _paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{pathsfile}',True,'TAIR10')\n",
    "\n",
    "    paths = {}\n",
    "    for accession in sorted(list(_paths.keys())):\n",
    "        paths[accession] = _paths[accession]\n",
    "    del _paths\n",
    "\n",
    "    coregraph = GenomeGraph(pathsDict=paths)\n",
    "\n",
    "    # On undirected coregraph sorting is not optimal! Check sorting!!!\n",
    "\n",
    "    coregraph.treeSort()\n",
    "\n",
    "    assert len(coregraph.nodes)==len(coregraph.order)\n",
    "\n",
    "    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_Chr{seqNum:d}.gfa'\n",
    "\n",
    "    coregraph.toGFA(coreGFApath,False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating nodes length...\n",
      "Processing node 33/33\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 32/33\n",
      "Calculating nodes length...\n",
      "Processing node 19/19\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 18/19\n",
      "Calculating nodes length...\n",
      "Processing node 24/24\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 23/24\n",
      "Calculating nodes length...\n",
      "Processing node 14/14\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 13/14\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Constructing Tremaux tree\n",
      "Preprocessing tree 100%\n",
      "Done!\n",
      "Getting root nodes\n",
      "Start Loop...\n",
      "Nodes in order: 6/7\n"
     ]
    }
   ],
   "source": [
    "# For v2\n",
    "pathfileDir = '../../1001G/coreGraph/new_Nov2022/'\n",
    "\n",
    "pathsfile = f'paths.txt'\n",
    "\n",
    "paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{pathsfile}',True,'TAIR10',True)\n",
    "\n",
    "for seqNum in paths.keys():#range(1,6): \n",
    "    # paths = {}\n",
    "    # for accession in sorted(list(_paths.keys())):\n",
    "    #     paths[accession] = _paths[accession]\n",
    "    # del _paths\n",
    "\n",
    "    coregraph = GenomeGraph(pathsDict=paths[seqNum])\n",
    "\n",
    "    # On undirected coregraph sorting is not optimal! Check sorting!!!\n",
    "\n",
    "    coregraph.treeSort()\n",
    "\n",
    "    assert len(coregraph.nodes)==len(coregraph.order)\n",
    "\n",
    "    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_v2_Chr{seqNum}.gfa'\n",
    "\n",
    "    coregraph.toGFA(coreGFApath,False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "curT = time.localtime()\n",
    "message = f\"Generating chain graphs for all chromosome finished at \\\n",
    "            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\"\n",
    "!ntfy send \"{message}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading test sorting graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/pantograph/data'\n",
    "pathsfile = 'testCollapse_path.txt'\n",
    "\n",
    "paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{pathsfile}',True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "paths = {}\n",
    "for accession in sorted(list(_paths.keys())):\n",
    "    paths[accession] = _paths[accession]\n",
    "del _paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph = GenomeGraph(pathsDict=paths,nodeNameLength=4)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "On undirected coregraph sorting is not optimal! Check sorting!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph.tremauxTree.draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(coregraph.nodes),len(coregraph.order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph.nodes,coregraph.order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[coregraph.nodes[i-1] for i in coregraph.order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coreGFApath = f'{pathfileDir}{os.path.sep}testCollapse.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph.toGFA(coreGFApath,False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading graph from GFA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Large graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f'{gfadir}{os.path.sep}{gfaFilename}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new = GenomeGraph(gfaPath=f'{gfadir}{os.path.sep}{gfaFilename}',isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(graph_new.nodes)==len(graph_new.order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new.order),len(graph_new.nodes),len(graph_new.tremauxTree),len(graph_new.tremauxTree.originalGraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfaFilename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfadir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ntfy send \"Sorting gene graph finished!\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Large graph with unrelated sequences and real sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new2 = GenomeGraph(gfaPath='./test_long_graph_full_seq.gfa',isGFASeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new2.accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new2.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new2.order),len(graph_new2.nodes),len(graph_new2.tremauxTree),len(graph_new2.tremauxTree.originalGraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfaFilename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfadir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfaFilename = 'AT_Chr1_OGOnly_Seq.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new2.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Christian's test graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new = GenomeGraph(gfaPath='../../1001G/graphSorting/graphs/sixRef.panGenome.Chroms.pggb.Chr2.5mb.gfa',isGFASeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new.accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cProfile.run('graph_new.treeSort()','treesort_profile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = pstats.Stats('treesort_profile')\n",
    "p.strip_dirs().sort_stats(SortKey.CUMULATIVE).print_stats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new.order),len(graph_new.nodes),len(graph_new.tremauxTree),len(graph_new.tremauxTree.originalGraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfaFilename='testChristian_new.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfadir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shuffle(graph_new.order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfaFilename='testChristian_shuffled.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new = GenomeGraph(gfaPath='../../1001G/pantograph/data/testChristian_shuffled.gfa',isGFASeq=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Christian's full graph (Chr1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new = GenomeGraph(gfaPath='../../1001G/chrisGraph/chr1.wfmash.n20.a90.s10000.p1,19,39,3,81,1.seqwish.sort.smooth.sort.gfa',\n",
    "                        isGFASeq=True,\n",
    "                       doOverlapCleaning=False,\n",
    "                       accessionsToRemove=['Consensus'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new.accessions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new.order),len(graph_new.nodes),len(graph_new.tremauxTree),len(graph_new.tremauxTree.originalGraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(graph_new.nodes)==len(graph_new.order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph_new.order),len(graph_new.nodes),len(graph_new.tremauxTree),len(graph_new.tremauxTree.originalGraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfaFilename='Christian_chr1.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gfadir='../../1001G/chrisGraph/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph_new.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ntfy send \"Sorting nucleotide graph finished!\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Small simple graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2 = GenomeGraph(gfaPath='./shorttest.gfa',isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2.forwardLinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2.tremauxTree.draw_original()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2.tremauxTree.draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2.order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph2.toGFA('../../1001G/pantograph/data/shorttest1.gfa',doSeq=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Small graph with complex loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph3 = GenomeGraph(gfaPath='./shorttest2.gfa',isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph3.paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph3.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb.pm()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph3.tremauxTree.draw_original()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph3.tremauxTree.draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(graph3.order),len(graph3.nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph3.toGFA('../../1001G/pantograph/data/shorttest2.gfa',doSeq=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Small graph with very complex relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4 = GenomeGraph(gfaPath='./shorttest3.gfa',isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4.paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4.tremauxTree.draw_original()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4.tremauxTree.draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4.order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph4.toGFA('../../1001G/pantograph/data/shorttest3.gfa',doSeq=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Small graph with self-loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5 = GenomeGraph(gfaPath='./shorttest_loop.gfa',isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5.paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5.treeSort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5.tremauxTree.draw_original()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5.tremauxTree.draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5.order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph5.toGFA('../../1001G/pantograph/data/shorttest4.gfa',doSeq=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting to Front-end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exporting presentation graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../Meetings/1001G+_20220518/'\n",
    "coreGFApath = f'{pathfileDir}{os.path.sep}paths_presentation.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading graph from ../../Meetings/1001G+_20220518//paths_presentation.gfa\n",
      "Loading segment 7/7\n",
      "Loading segments finished.\n",
      "Loading link 12/12\n",
      "Loading links finished\n",
      "Loading path 5/5\n",
      "Loading paths finished. 5 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 5/5\n",
      "Finished preprocessing paths\n",
      "\n"
     ]
    }
   ],
   "source": [
    "genome = GenomeGraph(coreGFApath,isGFASeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATGCC', 'GCACGTTC', 'ATTTGCCCAA', 'AG', 'CTGAA', 'ATCCCA', 'GCATTCA']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome.nodesData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading graph from ../../Meetings/1001G+_20220518//paths_presentation.gfa\n",
      "Loading segment 7/7\n",
      "Loading segments finished.\n",
      "Loading link 12/12\n",
      "Loading links finished\n",
      "Loading path 5/5\n",
      "Loading paths finished. 5 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 5/5\n",
      "Finished preprocessing paths\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# changing annotation\n",
    "genome = GenomeGraph(coreGFApath,isGFASeq=True)\n",
    "\n",
    "for nodeID,node in enumerate(genome.nodesAnnotation):\n",
    "    for seqName,seqDict in node.items():\n",
    "        for annText in seqDict.keys():\n",
    "            genome.nodesAnnotation[nodeID][seqName][annText] = [(0,len(genome.nodesData[nodeID])-1)]\n",
    "\n",
    "genome.toGFA(coreGFApath,doSeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{}, {}, {}, {}, {}, {}, {}]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# genome = GenomeGraph(coreGFApath,isGFASeq=True)\n",
    "genome.nodesAnnotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 6\n",
    "inversionThreshold = 0.5\n",
    "isSeq = True\n",
    "zoomLevels = [1,2,4,8,16]#,8,16,32]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('../../Meetings/1001G+_20220518', 'paths_presentation_new', [1, 2, 4, 8, 16])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "outputPath,outputName,zoomLevels"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "print(f'Opening Redis connection for db {dbid}')\n",
    "redisConn = Redis(host='redis',port = 6379,db=dbid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "redisConn=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted 05_export.ipynb.\n",
      "Converted 05_exportDev.ipynb.\n",
      "Converted deBruijnGraphProcessing.ipynb.\n",
      "Converted dev.ipynb.\n",
      "Converted graphTesting.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Genome\n",
      "Loading graph from ../../Meetings/1001G+_20220518//paths_presentation.gfa\n",
      "Found nodeNames file ../../Meetings/1001G+_20220518/nodeNames_paths_presentation.json, loading names.\n",
      "Found node annotation file ../../Meetings/1001G+_20220518/annotation_paths_presentation.dat, loading associations.\n",
      "Loading segment 7/7\n",
      "Loading segments finished.\n",
      "Loading link 12/12\n",
      "Loading links finished\n",
      "Loading path 5/5\n",
      "Loading paths finished. 5 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 5/5\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../Meetings/1001G+_20220518/paths_presentation_new\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 5/5\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 2/2\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 5/5\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 5/5\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 7/7\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 7/7\n",
      "Processing component links 6/6\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 6/6\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <2 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 2\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 2 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <4 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 4\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 4 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <8 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 8\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 8 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <16 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 16\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 16 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <32 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 32\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 32 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <64 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 64\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 64 finished.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#%%capture output\n",
    "startTime = time.time()\n",
    "# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \\\n",
    "# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, \n",
    "#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \\\n",
    "exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, \n",
    "                                outputPath=outputPath, outputName=outputName,\n",
    "                                isSeq=isSeq,\n",
    "                                redisConn=redisConn,\n",
    "                                zoomLevels=zoomLevels,\n",
    "                                fillZoomLevels = True,\n",
    "                                maxLengthComponent=maxLengthComponent, \n",
    "                                maxLengthChunk=maxLengthChunk, \n",
    "                                inversionThreshold=inversionThreshold)\n",
    "#                                 returnDebugData=True)\n",
    "runTime = time.time() - startTime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/data/YandexDisk/Kew/src/graphConstruction/pangraph_constructor/exportDev.py\u001b[0m(3902)\u001b[0;36mexportToPantograph\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m   3900 \u001b[0;31m    \u001b[0;32mif\u001b[0m \u001b[0mfillZoomLevels\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m   3901 \u001b[0;31m        \u001b[0mmaxLinkLength\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlinkLengths\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m-> 3902 \u001b[0;31m        \u001b[0mmaxRearrangementLength\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblockEdges\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m   3903 \u001b[0;31m        \u001b[0mmaxBlock\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmaxLinkLength\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmaxRearrangementLength\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m   3904 \u001b[0;31m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  blockEdges\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  linkLengths\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{25: {(3, 1), (4, 5), (1, 4)}, 36: {(6, 1), (5, 6)}}\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  q\n"
     ]
    }
   ],
   "source": [
    "import pdb\n",
    "pdb.pm()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing removable elements identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pangraph_constructor.exportDev import getRemovableStructures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import IPython.display as ipd\n",
    "import joblib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists('./tests/breakIdentify.dat'):\n",
    "    correctResults = joblib.load('./tests/breakIdentify.dat')\n",
    "else:\n",
    "    correctResults = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '../../1001G/GraphCollapsing/TestGraphs'\n",
    "filePrefix = 'test'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test\n",
    "\n",
    "for filename,resDict in correctResults.items():\n",
    "    print(f'\\n####### Testing on case {filename} ########')\n",
    "    graph = GenomeGraph(f'{path}/{filename}',isGFASeq=False)\n",
    "    linksLengths, pairedLinks, blockEdges, _ = getRemovableStructures(graph=graph)\n",
    "    assert linksLengths==resDict['linksLengths']\n",
    "    assert pairedLinks==resDict['pairedLinks']\n",
    "    assert blockEdges==resDict['blockEdges']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "caseNum = 17\n",
    "filename = f'{filePrefix}{caseNum:02d}.gfa'\n",
    "\n",
    "\n",
    "print('############')\n",
    "print(f'Graph from file {filename}')\n",
    "coreGFApath = f'{path}/{filename}'\n",
    "graph = GenomeGraph(coreGFApath,isGFASeq=False)\n",
    "print('Graph Paths:')\n",
    "ipd.display(graph.paths)\n",
    "linkLengths, pairedLinks, blockEdges, _ = getRemovableStructures(graph=graph)\n",
    "\n",
    "print('Link-Lengths associations:')\n",
    "ipd.display(linkLengths)\n",
    "\n",
    "print('PairedLinks:')\n",
    "ipd.display(pairedLinks)\n",
    "\n",
    "print('Rearrangemenet block edges:')\n",
    "ipd.display(blockEdges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "correctResults[filename]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "correctResults[filename] = {'linksLengths':linkLengths,'pairedLinks':pairedLinks,'blockEdges':blockEdges}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "joblib.dump(correctResults,'./tests/breakIdentify.dat')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exporting test collapse graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/pantograph/data'\n",
    "coreGFApath = f'{pathfileDir}{os.path.sep}testCollapse.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 6\n",
    "inversionThreshold = 0.5\n",
    "isSeq = False\n",
    "zoomLevels = [1,2,4]#,8,16,32]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "outputPath,outputName,zoomLevels"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "print(f'Opening Redis connection for db {dbid}')\n",
    "redisConn = Redis(host='redis',port = 6379,db=dbid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "redisConn=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%capture output\n",
    "startTime = time.time()\n",
    "# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \\\n",
    "[zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, \n",
    "                invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \\\n",
    "exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, \n",
    "                                outputPath=outputPath, outputName=outputName,\n",
    "                                isSeq=isSeq,\n",
    "                                redisConn=redisConn,\n",
    "                                zoomLevels=zoomLevels,\n",
    "                                fillZoomLevels = True,\n",
    "                                maxLengthComponent=maxLengthComponent, \n",
    "                                maxLengthChunk=maxLengthChunk, \n",
    "                                inversionThreshold=inversionThreshold,\n",
    "                                returnDebugData=True)\n",
    "runTime = time.time() - startTime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'Executed in {runTime} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ntfy send \"Exporting test collapse graph finished. Overall time = {runTime} seconds\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exporting coregraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/coreGraph'\n",
    "coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_Chr1.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 6\n",
    "inversionThreshold = 0.5\n",
    "isSeq = False\n",
    "zoomLevels = [1,2,4,8,16]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('../../1001G/coreGraph', 'coregraph_Chr1_new', [1, 2, 4, 8, 16])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "outputPath,outputName,zoomLevels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "# print(f'Opening Redis connection for db {dbid}')\n",
    "redisConn = Redis(host='redis',port = 6379,db=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted 05_export.ipynb.\n",
      "Converted 05_exportDev.ipynb.\n",
      "Converted deBruijnGraphProcessing.ipynb.\n",
      "Converted dev.ipynb.\n",
      "Converted graphTesting.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.\n",
      "Loading segment 35/35\n",
      "Loading segments finished.\n",
      "Loading link 72/72\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 35/35\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/coregraph_Chr1_new\n",
      "Calculating nodes length...\n",
      "Processing node 35/35\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 16/16\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 34/34\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 34/34\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 35/35\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 35/35\n",
      "Processing component links 35/35\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 35/35\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <2 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 2\n",
      "===========================\n",
      "Processing component 35/35\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 8/8\n",
      "Recording zoom level 2 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <4 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 4\n",
      "===========================\n",
      "Processing component 8/8\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 5/5\n",
      "Recording zoom level 4 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <8 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 8\n",
      "===========================\n",
      "Processing component 5/5\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 8 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <16 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 16\n",
      "===========================\n",
      "Processing component 1/1\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 16 finished.\n",
      "27.11550521850586\n"
     ]
    }
   ],
   "source": [
    "#%%capture output\n",
    "startTime = time.time()\n",
    "# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \\\n",
    "exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, \n",
    "                                outputPath=outputPath, outputName=outputName,\n",
    "                                isSeq=isSeq,\n",
    "                                redisConn=redisConn,\n",
    "                                zoomLevels=zoomLevels, \n",
    "                                maxLengthComponent=maxLengthComponent, \n",
    "                                maxLengthChunk=maxLengthChunk, \n",
    "                                inversionThreshold=inversionThreshold)\n",
    "#                                 returnDebugata=True)\n",
    "runTime = time.time() - startTime\n",
    "print(runTime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.11) or chardet (3.0.4) doesn't match a supported version!\n",
      "  warnings.warn(\"urllib3 ({}) or chardet ({}) doesn't match a supported \"\n"
     ]
    }
   ],
   "source": [
    "!ntfy send \"Exporting coregraph finished. Overall time = {runTime} seconds\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "❗️❗️❗️ TODO: Next test API. WHen works, change front end to get jsons without annotation and then to load annotation from API!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.\n",
      "Loading segment 35/35\n",
      "Loading segments finished.\n",
      "Loading link 72/72\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 35/35\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n"
     ]
    }
   ],
   "source": [
    "graph = GenomeGraph(coreGFApath,isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pangraph_constructor.utils import iset_get,iset_score,iset_add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "redisConn = Redis(host='redis',port=6379,db=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'a_0': (1.0, 3.0), 'b_0': (1.0, 5.0)}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iset_get(redisConn,'test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['23']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iset_score(redisConn,'coregraph_Chr1_new.1.10024.Pos',22)\n",
    "# coregraph_Chr1_new/10024/1/22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['3370',\n",
       " 'AT1G15940',\n",
       " 'AT1G15950',\n",
       " 'AT1G16000',\n",
       " 'AT1G16010',\n",
       " 'AT1G16022',\n",
       " 'AT1G62830',\n",
       " 'AT1G62840',\n",
       " 'OG0001502',\n",
       " 'OG0004404',\n",
       " 'OG0007340',\n",
       " 'OG0007341',\n",
       " 'OG0007343',\n",
       " 'OG0008435',\n",
       " 'OG0015400',\n",
       " 'OG0017985',\n",
       " 'OG0022841',\n",
       " 'OG0024046',\n",
       " 'OG0024055',\n",
       " 'OG0025924',\n",
       " 'OG0027925',\n",
       " 'OG0030893']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iset_score(redisConn,'coregraph_Chr1_new.9543.Gene',50,50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'3370_0': (48.0, 51.0)}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iset_get(redisConn,'coregraph_Chr1_new.9543.Gene','3370_0')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "redisConn=None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exporting coregraph with genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/coreGraph'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 6\n",
    "inversionThreshold = 0.5\n",
    "isSeq = True\n",
    "zoomLevels = [1,3,9]# + [9*2**i for i in range(9)]\n",
    "# zoomLevels = [2**i for i in range(12)]#  [1,2,4,8,16,32,9*16,9*32,9*128,9*256]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "# outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "# outputPath,outputName,zoomLevels"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "print(f'Opening Redis connection for db {dbid}')\n",
    "redisConn = Redis(host='redis',port = 6379,db=dbid)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "redisConn=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%capture output\n",
    "for chrNum in range(1,6):\n",
    "    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_genes_Chr{chrNum}.gfa'\n",
    "    \n",
    "    outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "    \n",
    "    dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "    print(f'Opening Redis connection for db {dbid}')\n",
    "    redisConn = Redis(host='redis',port = 6379,db=dbid)\n",
    "    \n",
    "    startTime = time.time()\n",
    "    # [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \\\n",
    "    exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, \n",
    "                                    outputPath=outputPath, outputName=outputName,\n",
    "                                    isSeq=isSeq,\n",
    "                                    redisConn=redisConn,\n",
    "                                    zoomLevels=zoomLevels,\n",
    "                                    fillZoomLevels = True,\n",
    "                                    maxLengthComponent=maxLengthComponent, \n",
    "                                    maxLengthChunk=maxLengthChunk, \n",
    "                                    inversionThreshold=inversionThreshold)\n",
    "    #                                 returnDebugata=True)\n",
    "    runTime = time.time() - startTime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "curT = time.localtime(time.time()+3600)\n",
    "message = f\"Exporting core graph with genes for all chromosomes finished at \\\n",
    "            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\\n\"\n",
    "!ntfy send \"{message}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zoomLevels = [1,2,4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "# outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "# outputPath,outputName,zoomLevels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%capture output2\n",
    "for chrNum in range(1,6):\n",
    "    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_Chr{chrNum}.gfa'\n",
    "    \n",
    "    outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "    \n",
    "    dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "    print(f'Opening Redis connection for db {dbid}')\n",
    "    redisConn = Redis(host='redis',port = 6379,db=dbid)\n",
    "    \n",
    "    startTime = time.time()\n",
    "    # [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \\\n",
    "    exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, \n",
    "                                    outputPath=outputPath, outputName=outputName,\n",
    "                                    isSeq=isSeq,\n",
    "                                    redisConn=redisConn,\n",
    "                                    zoomLevels=zoomLevels,\n",
    "                                    fillZoomLevels = True,\n",
    "                                    maxLengthComponent=maxLengthComponent, \n",
    "                                    maxLengthChunk=maxLengthChunk, \n",
    "                                    inversionThreshold=inversionThreshold)\n",
    "    #                                 returnDebugata=True)\n",
    "    runTime = time.time() - startTime\n",
    "    \n",
    "    print(f'Exporting core graph for Chr{chrNum} took {runTime} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "curT = time.localtime(time.time()+3600)\n",
    "message = f\"Exporting core graph with genes for all chromosomes finished at \\\n",
    " {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\\n\"\n",
    "!ntfy send \"{message}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ntfy send \"Exporting coregraph with genes finished. Overall time = {runTime} seconds\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exporting gene graphs for all chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/pantograph/data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 6\n",
    "inversionThreshold = 0.5\n",
    "isSeq = False\n",
    "zoomLevels = [1]#,2,4]# + [9*2**i for i in range(12)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbdev_export()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing case AT_Chr1_OGOnly_2.1_new\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 7455/7455\n",
      "Loading segments finished.\n",
      "Loading link 14001/14001\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7455/7455\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1_new\n",
      "Calculating nodes length...\n",
      "Processing node 7455/7455\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Identifying rearrangement blocks\n",
      "Processing node 7455/7455\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 7455/7455\n",
      "Processing component links 3869/3869\n",
      "Recording component 3869/3869\n",
      "Recording zoom level 1 finished.\n",
      "Exporting gene graph for Chr1 took 481.7105052471161 seconds\n",
      "\n",
      "Processing case AT_Chr2_OGOnly_2.1_new\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 4877/4877\n",
      "Loading segments finished.\n",
      "Loading link 10302/10302\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 4877/4877\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1_new\n",
      "Calculating nodes length...\n",
      "Processing node 4877/4877\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Identifying rearrangement blocks\n",
      "Processing node 4877/4877\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 4877/4877\n",
      "Processing component links 2534/2534\n",
      "Recording component 2534/2534\n",
      "Recording zoom level 1 finished.\n",
      "Exporting gene graph for Chr2 took 221.0419373512268 seconds\n",
      "\n",
      "Processing case AT_Chr3_OGOnly_2.1_new\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 8442/8442\n",
      "Loading segments finished.\n",
      "Loading link 15985/15985\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 8442/8442\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1_new\n",
      "Calculating nodes length...\n",
      "Processing node 8442/8442\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Identifying rearrangement blocks\n",
      "Processing node 8442/8442\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 8442/8442\n",
      "Processing component links 3383/3383\n",
      "Recording component 3383/3383\n",
      "Recording zoom level 1 finished.\n",
      "Exporting gene graph for Chr3 took 385.5069224834442 seconds\n",
      "\n",
      "Processing case AT_Chr4_OGOnly_2.1_new\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 4620/4620\n",
      "Loading segments finished.\n",
      "Loading link 9980/9980\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 4620/4620\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1_new\n",
      "Calculating nodes length...\n",
      "Processing node 4620/4620\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Identifying rearrangement blocks\n",
      "Processing node 4620/4620\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 4620/4620\n",
      "Processing component links 2526/2526\n",
      "Recording component 2526/2526\n",
      "Recording zoom level 1 finished.\n",
      "Exporting gene graph for Chr4 took 180.05753779411316 seconds\n",
      "\n",
      "Processing case AT_Chr5_OGOnly_2.1_new\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 7052/7052\n",
      "Loading segments finished.\n",
      "Loading link 12737/12737\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7052/7052\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1_new\n",
      "Calculating nodes length...\n",
      "Processing node 7052/7052\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Identifying rearrangement blocks\n",
      "Processing node 7052/7052\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 7052/7052\n",
      "Processing component links 3274/3274\n",
      "Recording component 3274/3274\n",
      "Recording zoom level 1 finished.\n",
      "Exporting gene graph for Chr5 took 379.5864608287811 seconds\n"
     ]
    }
   ],
   "source": [
    "for seqID in ['Chr1','Chr2','Chr3','Chr4','Chr5']:#'Chr1','Chr2',\n",
    "\n",
    "    coreGFApath = f'{pathfileDir}{os.path.sep}AT_{seqID}_OGOnly_2.1.gfa'\n",
    "\n",
    "    outputPath,outputName = pathConvert(coreGFApath,suffix='_new')\n",
    "    \n",
    "    print()\n",
    "    print(f'Processing case {outputName}')\n",
    "    # dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "    # print(f'Opening Redis connection for db {dbid}')\n",
    "    redisConn = Redis(host='redis',port = 6379,db=0)\n",
    "\n",
    "    #%%capture output\n",
    "    startTime = time.time()\n",
    "    # initialLinkLengths, initialPairedLinks, initialInterconnectedLinks, initialBlockEdges, \\\n",
    "    # zoomNodeToComponent,zoomComponentToNodes,zoomComponents,\\\n",
    "    # zoomFromComponentLinks, zoomToComponentLinks, zoomLinkLengths, zoomPairedLinks, zoomInterconnectedLinks, \\\n",
    "    # zoomOldToNewRemoval, zoomNewToOldRemoval, \\\n",
    "    # zoomLinkLengthsRemoval, zoomPairedLinksRemoval, zoomInterconnectedLinksRemoval, zoomBlockEdgesRemoval, \\\n",
    "    # zoomFromComponentLinksRemoval, zoomToComponentLinksRemoval, \\\n",
    "    # graph = \\\n",
    "    exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, \n",
    "                                    outputPath=outputPath, outputName=outputName,\n",
    "                                    isSeq=isSeq,\n",
    "                                    redisConn=redisConn,\n",
    "                                    zoomLevels=zoomLevels, \n",
    "                                    fillZoomLevels = False,\n",
    "                                    maxLengthComponent=maxLengthComponent, \n",
    "                                    maxLengthChunk=maxLengthChunk, \n",
    "                                    inversionThreshold=inversionThreshold)#,\n",
    "                                    # returnDebugData=True)\n",
    "    runTime = time.time() - startTime\n",
    "    \n",
    "    print(f'Exporting gene graph for {seqID} took {runTime} seconds')\n",
    "    \n",
    "    redisConn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: Failed to send notification using default\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/ntfy/backends/default.py\", line 20, in notify\n",
      "    module.notify(title=title, message=message, **kwargs)\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/ntfy/backends/linux.py\", line 32, in notify\n",
      "    bus = dbus.SessionBus()\n",
      "  File \"/usr/lib/python3/dist-packages/dbus/_dbus.py\", line 212, in __new__\n",
      "    return Bus.__new__(cls, Bus.TYPE_SESSION, private=private,\n",
      "  File \"/usr/lib/python3/dist-packages/dbus/_dbus.py\", line 102, in __new__\n",
      "    bus = BusConnection.__new__(subclass, bus_type, mainloop=mainloop)\n",
      "  File \"/usr/lib/python3/dist-packages/dbus/bus.py\", line 124, in __new__\n",
      "    bus = cls._new_for_bus(address_or_type, mainloop=mainloop)\n",
      "dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NotSupported: Unable to autolaunch a dbus-daemon without a $DISPLAY for X11\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/ntfy/__init__.py\", line 53, in notify\n",
      "    notify_ret = notifier.notify(\n",
      "  File \"/usr/local/lib/python3.8/dist-packages/ntfy/backends/default.py\", line 22, in notify\n",
      "    raise DefaultNotifierError(e, module)\n",
      "ntfy.backends.default.DefaultNotifierError: (DBusException('Unable to autolaunch a dbus-daemon without a $DISPLAY for X11'), <module 'ntfy.backends.linux' from '/usr/local/lib/python3.8/dist-packages/ntfy/backends/linux.py'>)\n"
     ]
    }
   ],
   "source": [
    "curT = time.localtime()\n",
    "message = f\"Exporting gene graph for all chromosomes finished at \\\n",
    "            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\"\n",
    "!ntfy send \"{message}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 7455/7455\n",
      "Loading segments finished.\n",
      "Loading link 14001/14001\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7455/7455\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n"
     ]
    }
   ],
   "source": [
    "graph = GenomeGraph(gfaPath='../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa', isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TAIR10': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148013, 149848]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [154367, 156178]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '10002': {'genPos': [{'chr': 'Chr1', 'genomePosition': [150525, 151967]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '10015': {'genPos': [{'chr': 'Chr1', 'genomePosition': [149434, 150876]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [155667, 157112]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '10024': {'genPos': [{'chr': 'Chr1', 'genomePosition': [153796, 155238]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [160034, 161479]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '1741': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148244, 149686]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [154491, 155936]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22001': {'genPos': [{'chr': 'Chr1', 'genomePosition': [151677, 153086]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [157924, 159369]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22002': {'genPos': [{'chr': 'Chr1', 'genomePosition': [152325, 153767]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22003': {'genPos': [{'chr': 'Chr1', 'genomePosition': [154749, 156191]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [160982, 162427]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22004': {'genPos': [{'chr': 'Chr1', 'genomePosition': [152208, 153617]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22005': {'genPos': [{'chr': 'Chr1', 'genomePosition': [149704, 151146]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [155491, 156936]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22006': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148430, 149839]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [154217, 155662]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '22007': {'genPos': [{'chr': 'Chr1', 'genomePosition': [149667, 151109]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '6024': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148707, 150149]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [154925, 156370]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '6069': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148675, 150117]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [154919, 156364]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '6124': {'genPos': [{'chr': 'Chr1', 'genomePosition': [151870, 153279]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '6244': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148529, 149938]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [154773, 156218]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '6909': {'genPos': [{'chr': 'Chr1', 'genomePosition': [151163, 152605]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [157410, 158855]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '6966': {'genPos': [{'chr': 'Chr1', 'genomePosition': [152818, 154227]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [159042, 160487]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '8236': {'genPos': [{'chr': 'Chr1', 'genomePosition': [150738, 152147]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [156985, 158430]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9075': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148586, 150028]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9537': {'genPos': [{'chr': 'Chr1', 'genomePosition': [148112, 149521]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9543': {'genPos': [{'chr': 'Chr1', 'genomePosition': [149923, 151332]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9638': {'genPos': [{'chr': 'Chr1', 'genomePosition': [153515, 154957]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [159764, 161209]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9728': {'genPos': [{'chr': 'Chr1', 'genomePosition': [150904, 152346]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [157151, 158596]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9764': {'genPos': [{'chr': 'Chr1', 'genomePosition': [149045, 150487]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9888': {'genPos': [{'chr': 'Chr1', 'genomePosition': [154457, 155866]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9905': {'genPos': [{'chr': 'Chr1', 'genomePosition': [153937, 155346]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}},\n",
       " '9981': {'genPos': [{'chr': 'Chr1', 'genomePosition': [151436, 152878]},\n",
       "   {'chr': 'Chr1', 'genomePosition': [157683, 159128]}],\n",
       "  'annotation': {'OG0000490': [(0, 0)],\n",
       "   'AT1G01390': [(0, 0)],\n",
       "   'AT1G01420': [(0, 0)],\n",
       "   'AT4G01070': [(0, 0)]}}}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph.nodesMetadata[37]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import joblib"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "joblib.dump(initialLinkLengths,'dumps/initialLinkLengths.dump')\n",
    "joblib.dump(initialPairedLinks,'dumps/initialPairedLinks.dump')\n",
    "joblib.dump(initialInterconnectedLinks,'dumps/initialInterconnectedLinks.dump')\n",
    "joblib.dump(initialBlockEdges,'dumps/initialBlockEdges.dump')\n",
    "joblib.dump(zoomNodeToComponent,'dumps/zoomNodeToComponent.dump')\n",
    "joblib.dump(zoomComponentToNodes,'dumps/zoomComponentToNodes.dump')\n",
    "joblib.dump(zoomComponents,'dumps/zoomComponents.dump')\n",
    "joblib.dump(zoomFromComponentLinks,'dumps/zoomFromComponentLinks.dump')\n",
    "joblib.dump(zoomToComponentLinks,'dumps/zoomToComponentLinks.dump')\n",
    "joblib.dump(zoomLinkLengths,'dumps/zoomLinkLengths.dump')\n",
    "joblib.dump(zoomPairedLinks,'dumps/zoomPairedLinks.dump')\n",
    "joblib.dump(zoomInterconnectedLinks,'dumps/zoomInterconnectedLinks.dump')\n",
    "joblib.dump(zoomOldToNewRemoval,'dumps/zoomOldToNewRemoval.dump')\n",
    "joblib.dump(zoomNewToOldRemoval,'dumps/zoomNewToOldRemoval.dump')\n",
    "joblib.dump(zoomLinkLengthsRemoval,'dumps/zoomLinkLengthsRemoval.dump')\n",
    "joblib.dump(zoomPairedLinksRemoval,'dumps/zoomPairedLinksRemoval.dump')\n",
    "joblib.dump(zoomInterconnectedLinksRemoval,'dumps/zoomInterconnectedLinksRemoval.dump')\n",
    "joblib.dump(zoomBlockEdgesRemoval,'dumps/zoomBlockEdgesRemoval.dump')\n",
    "joblib.dump(zoomFromComponentLinksRemoval,'dumps/zoomFromComponentLinksRemoval.dump')\n",
    "joblib.dump(zoomToComponentLinksRemoval,'dumps/zoomToComponentLinksRemoval.dump')\n",
    "joblib.dump(graph,'dumps/graph.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "initialLinkLengths = joblib.load('dumps/initialLinkLengths.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "initialPairedLinks = joblib.load('dumps/initialPairedLinks.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "initialInterconnectedLinks = joblib.load('dumps/initialInterconnectedLinks.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "initialBlockEdges = joblib.load('dumps/initialBlockEdges.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomNodeToComponent = joblib.load('dumps/zoomNodeToComponent.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomComponentToNodes = joblib.load('dumps/zoomComponentToNodes.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomComponents = joblib.load('dumps/zoomComponents.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomFromComponentLinks = joblib.load('dumps/zoomFromComponentLinks.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomToComponentLinks = joblib.load('dumps/zoomToComponentLinks.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomLinkLengths = joblib.load('dumps/zoomLinkLengths.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomPairedLinks = joblib.load('dumps/zoomPairedLinks.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomInterconnectedLinks = joblib.load('dumps/zoomInterconnectedLinks.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomOldToNewRemoval = joblib.load('dumps/zoomOldToNewRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomNewToOldRemoval = joblib.load('dumps/zoomNewToOldRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomLinkLengthsRemoval = joblib.load('dumps/zoomLinkLengthsRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomPairedLinksRemoval = joblib.load('dumps/zoomPairedLinksRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomInterconnectedLinksRemoval = joblib.load('dumps/zoomInterconnectedLinksRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomBlockEdgesRemoval = joblib.load('dumps/zoomBlockEdgesRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomFromComponentLinksRemoval = joblib.load('dumps/zoomFromComponentLinksRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zoomToComponentLinksRemoval = joblib.load('dumps/zoomToComponentLinksRemoval.dump')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "graph = joblib.load('dumps/graph.dump')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each node processing time increase significantly with overall number of nodes. This is wrong and should be investigated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding nucleotide data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from copy import deepcopy\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/coreGraph'\n",
    "coreGFApath = f'{pathfileDir}{os.path.sep}coreGraph_f2.1_Ref_v04.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph = GenomeGraph(gfaPath=coreGFApath,isGFASeq=False)\n",
    "coregraph_genes = deepcopy(coregraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fullGraphPath = '../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa'\n",
    "fullgraph = GenomeGraph(gfaPath=fullGraphPath,isGFASeq=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chainToGenesFile = 'chain2gene_f2.1_Ref_v04.txt'\n",
    "maxChainLength = len(coregraph.nodes[0])\n",
    "chainToListDict = {}\n",
    "with open(f'{pathfileDir}{os.path.sep}{chainToGenesFile}') as f:\n",
    "    for line in f:\n",
    "        chainName, geneList = line.split(':')\n",
    "        geneList = geneList.lstrip().rstrip().split(',')\n",
    "        chainToListDict[chainName.zfill(maxChainLength)] = geneList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for nodeIdx,nodeName in enumerate(coregraph.nodes):\n",
    "    print(f'\\nNode {nodeIdx+1}/{len(coregraph.nodes)}',end='')\n",
    "    geneList = chainToListDict.get(nodeName.zfill(maxChainLength), [f'ch{nodeName.zfill(7)}'])\n",
    "    geneIds = []\n",
    "    if geneList[0][:2]!='ch':\n",
    "        geneIds = [int(gene.rstrip('+'))-1 for gene in geneList]\n",
    "        geneList = [fullgraph.nodes[geneid] for geneid in geneIds]\n",
    "    coregraph_genes.nodesData[nodeIdx] = ''.join(geneList)\n",
    "    \n",
    "    for accession, chainDict in coregraph.nodesAnnotation[nodeIdx].items():\n",
    "        interval = chainDict[nodeName]\n",
    "        geneCumLengths = np.hstack((0, np.cumsum([len(gene) for gene in geneList])))\n",
    "        \n",
    "        coregraph.nodesAnnotation[nodeIdx][accession].\\\n",
    "            update({geneAnnotation:interval \\\n",
    "                    for geneid in geneIds \\\n",
    "                        for geneAnnotation in fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:None}).keys()})\n",
    "\n",
    "        coregraph_genes.nodesAnnotation[nodeIdx][accession].\\\n",
    "            update({geneAnnotation:[(geneCumLengths[i], geneCumLengths[i+1]-1)] \\\n",
    "                            for i,geneid in enumerate(geneIds) \\\n",
    "                                for geneAnnotation in fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:None}).keys()})\n",
    "\n",
    "        coregraph_genes.nodesAnnotation[nodeIdx][accession].\\\n",
    "            update({nodeName:[(geneCumLengths[0], geneCumLengths[-1]-1)]})\n",
    "print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coregraph.toGFA(f'{pathfileDir}{os.path.sep}coregraph_f2.1_Ref_v04.gfa',doSeq=False)\n",
    "coregraph_genes.toGFA(f'{pathfileDir}{os.path.sep}coregraph_genes_f2.1_Ref_v04.gfa',doSeq=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathfileDir = '../../1001G/coreGraph'\n",
    "coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_genes_f2.1_Ref_v04.gfa'\n",
    "# coreGFApath = f'{pathfileDir}{os.path.sep}coreGraph.gfa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zoomLevels = [1,10,20,100,500,1000,5000,10000,50000,100000,500000,1000000]\n",
    "zoomLevels = [1,3,9,45,90,450,900,4500,9000]\n",
    "# zoomLevels = [1,3,9,18]\n",
    "# zoomLevels = [4,8,16]\n",
    "\n",
    "isSeq = True\n",
    "\n",
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 16\n",
    "invertionThreshold = 0.5\n",
    "# inputPath = '../../1001G/pantograph/data/shorttest2.gfa'\n",
    "# inputPath = '../../1001G/pantograph/data/AT_Chr1_OGOnly.gfa'\n",
    "# inputPath = '../../1001G/chrisGraph/chr1.wfmash.n20.a90.s10000.p1,19,39,3,81,1.seqwish.sort.smooth.sort.gfa'\n",
    "# inputPath = '../../1001G/pantograph/data/shorttest_seq.gfa'\n",
    "inputPath = coreGFApath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "outputPath,outputName = pathConvert(inputPath,suffix='_new')\n",
    "outputPath,outputName,zoomLevels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "print(f'Opening Redis connection for db {dbid}')\n",
    "redisConn = Redis(host='redis',port = 6379,db=dbid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides = \\\n",
    "exportToPantograph(inputPath=inputPath,\n",
    "                   outputName=outputName,\n",
    "                   outputPath=outputPath,\n",
    "                   isSeq=isSeq,\n",
    "                   redisConn=redisConn,\n",
    "                   GenomeGraphParams={'accessionsToRemove':['Consensus']},\n",
    "                   zoomLevels=zoomLevels,\n",
    "                   maxLengthChunk=maxLengthChunk,\n",
    "                   maxLengthComponent=maxLengthComponent,\n",
    "                   invertionThreshold=invertionThreshold,)\n",
    "#                                              debug=True,returnDebugData=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ntfy send \"Pantograph data generation for coregraph finished.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding gene data mass processing several chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_init.ipynb.\n",
      "Converted 01_graph.ipynb.\n",
      "Converted 02_tree.ipynb.\n",
      "Converted 03_synteny.ipynb.\n",
      "Converted 04_utils.ipynb.\n",
      "Converted 05_export.ipynb.\n",
      "Converted 05_exportDev.ipynb.\n",
      "Converted deBruijnGraphProcessing.ipynb.\n",
      "Converted dev.ipynb.\n",
      "Converted graphTesting.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkNodeLengthsFile(GFAPath):\n",
    "    directory = os.path.dirname(GFAPath)\n",
    "    filebase = os.path.splitext(os.path.basename(GFAPath))[0]\n",
    "    nodeLenPath = f'{directory}{os.path.sep}nodeLengths_{filebase}.dat'\n",
    "    \n",
    "    if os.path.exists(nodeLenPath):\n",
    "        return joblib.load(nodeLenPath)\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "doCreateCoreGenes = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr1.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr1.dat, loading associations.\n",
      "Loading segment 33/33\n",
      "Loading segments finished.\n",
      "Loading link 68/68\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 33/33\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 7455/7455\n",
      "Loading segments finished.\n",
      "Loading link 14001/14001\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7455/7455\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Node 33/33\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr1.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr1.dat, loading associations.\n",
      "Loading segment 33/33\n",
      "Loading segments finished.\n",
      "Loading link 68/68\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 33/33\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr1_new\n",
      "Calculating nodes length...\n",
      "Processing node 33/33\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 15/15\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 32/32\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 32/32\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 33/33\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 33/33\n",
      "Processing component links 33/33\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 33/33\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 33/33\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 4/4\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 4/4\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 9 finished.\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr1.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_genes_Chr1.dat, loading associations.\n",
      "Loading segment 33/33\n",
      "Loading segments finished.\n",
      "Loading link 68/68\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 33/33\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr1_new\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 15/15\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 32/32\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 32/32\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 33/33\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 33/33\n",
      "Processing component links 57/57\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 57/57\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 57/57\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 32/32\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 32/32\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 22/22\n",
      "Recording zoom level 9 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <18 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 18\n",
      "===========================\n",
      "Processing component 22/22\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 19/19\n",
      "Recording zoom level 18 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <36 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 36\n",
      "===========================\n",
      "Processing component 19/19\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 12/12\n",
      "Recording zoom level 36 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <72 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 72\n",
      "===========================\n",
      "Processing component 12/12\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 2/2\n",
      "Recording zoom level 72 finished.\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr2.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr2.dat, loading associations.\n",
      "Loading segment 19/19\n",
      "Loading segments finished.\n",
      "Loading link 38/38\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 19/19\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 4877/4877\n",
      "Loading segments finished.\n",
      "Loading link 10302/10302\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 4877/4877\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Node 19/19\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr2.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr2.dat, loading associations.\n",
      "Loading segment 19/19\n",
      "Loading segments finished.\n",
      "Loading link 38/38\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 19/19\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr2_new\n",
      "Calculating nodes length...\n",
      "Processing node 19/19\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 9/9\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 18/18\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 18/18\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 19/19\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 19/19\n",
      "Processing component links 19/19\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 19/19\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 19/19\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 3/3\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 3/3\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 9 finished.\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr2.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_genes_Chr2.dat, loading associations.\n",
      "Loading segment 19/19\n",
      "Loading segments finished.\n",
      "Loading link 38/38\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 19/19\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr2_new\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 9/9\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 18/18\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 18/18\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 19/19\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 19/19\n",
      "Processing component links 34/34\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 34/34\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 34/34\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 21/21\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 21/21\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 12/12\n",
      "Recording zoom level 9 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <18 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 18\n",
      "===========================\n",
      "Processing component 12/12\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 7/7\n",
      "Recording zoom level 18 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <36 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 36\n",
      "===========================\n",
      "Processing component 7/7\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 36 finished.\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr3.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr3.dat, loading associations.\n",
      "Loading segment 24/24\n",
      "Loading segments finished.\n",
      "Loading link 51/51\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 24/24\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 8442/8442\n",
      "Loading segments finished.\n",
      "Loading link 15985/15985\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 8442/8442\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Node 24/24\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr3.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr3.dat, loading associations.\n",
      "Loading segment 24/24\n",
      "Loading segments finished.\n",
      "Loading link 51/51\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 24/24\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr3_new\n",
      "Calculating nodes length...\n",
      "Processing node 24/24\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 13/13\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 26/26\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 26/26\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 24/24\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 24/24\n",
      "Processing component links 24/24\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 24/24\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 24/24\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 3/3\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 3/3\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 9 finished.\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr3.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_genes_Chr3.dat, loading associations.\n",
      "Loading segment 24/24\n",
      "Loading segments finished.\n",
      "Loading link 51/51\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 24/24\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr3_new\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 13/13\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 26/26\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 26/26\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 24/24\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 24/24\n",
      "Processing component links 43/43\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 43/43\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 43/43\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 24/24\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 24/24\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 18/18\n",
      "Recording zoom level 9 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <18 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 18\n",
      "===========================\n",
      "Processing component 18/18\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 11/11\n",
      "Recording zoom level 18 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <36 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 36\n",
      "===========================\n",
      "Processing component 11/11\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 5/5\n",
      "Recording zoom level 36 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <72 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 72\n",
      "===========================\n",
      "Processing component 5/5\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 4/4\n",
      "Recording zoom level 72 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <144 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 144\n",
      "===========================\n",
      "Processing component 4/4\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 4/4\n",
      "Recording zoom level 144 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <288 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 288\n",
      "===========================\n",
      "Processing component 4/4\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 4/4\n",
      "Recording zoom level 288 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <576 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 576\n",
      "===========================\n",
      "Processing component 4/4\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 2/2\n",
      "Recording zoom level 576 finished.\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr4.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr4.dat, loading associations.\n",
      "Loading segment 14/14\n",
      "Loading segments finished.\n",
      "Loading link 27/27\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 14/14\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 4620/4620\n",
      "Loading segments finished.\n",
      "Loading link 9980/9980\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 4620/4620\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Node 14/14\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr4.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr4.dat, loading associations.\n",
      "Loading segment 14/14\n",
      "Loading segments finished.\n",
      "Loading link 27/27\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 14/14\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr4_new\n",
      "Calculating nodes length...\n",
      "Processing node 14/14\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 6/6\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 13/13\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 13/13\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 14/14\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 14/14\n",
      "Processing component links 14/14\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 14/14\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 14/14\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 1/1\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 9 finished.\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr4.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_genes_Chr4.dat, loading associations.\n",
      "Loading segment 14/14\n",
      "Loading segments finished.\n",
      "Loading link 27/27\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 14/14\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr4_new\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 6/6\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 13/13\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 13/13\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 14/14\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 14/14\n",
      "Processing component links 29/29\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 29/29\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 29/29\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 17/17\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 17/17\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 11/11\n",
      "Recording zoom level 9 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <18 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 18\n",
      "===========================\n",
      "Processing component 11/11\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 9/9\n",
      "Recording zoom level 18 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <36 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 36\n",
      "===========================\n",
      "Processing component 9/9\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 6/6\n",
      "Recording zoom level 36 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <72 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 72\n",
      "===========================\n",
      "Processing component 6/6\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 5/5\n",
      "Recording zoom level 72 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <144 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 144\n",
      "===========================\n",
      "Processing component 5/5\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 2/2\n",
      "Recording zoom level 144 finished.\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr5.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr5.dat, loading associations.\n",
      "Loading segment 7/7\n",
      "Loading segments finished.\n",
      "Loading link 12/12\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa\n",
      "Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.\n",
      "Loading segment 7052/7052\n",
      "Loading segments finished.\n",
      "Loading link 12737/12737\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7052/7052\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Node 7/7\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr5.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_Chr5.dat, loading associations.\n",
      "Loading segment 7/7\n",
      "Loading segments finished.\n",
      "Loading link 12/12\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_Chr5_new\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 3/3\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 6/6\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 6/6\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 7/7\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 7/7\n",
      "Processing component links 7/7\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 7/7\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 7/7\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 1/1\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 1/1\n",
      "Recording zoom level 9 finished.\n",
      "Loading Genome\n",
      "Loading graph from ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr5.gfa\n",
      "Found node annotation file ../../1001G/coreGraph/new_Nov2022/annotation_coregraph_v2_genes_Chr5.dat, loading associations.\n",
      "Loading segment 7/7\n",
      "Loading segments finished.\n",
      "Loading link 12/12\n",
      "Loading links finished\n",
      "Loading path 28/28\n",
      "Loading paths finished. 28 paths added, 0 paths ignored.\n",
      "Calculating nodes length...\n",
      "Processing node 7/7\n",
      "Finished calculating nodes lengths\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "\n",
      "Recording Pantograph data to ../../1001G/coreGraph/new_Nov2022/coregraph_v2_genes_Chr5_new\n",
      "Preprocessing paths...\n",
      "Processing path 28/28\n",
      "Finished preprocessing paths\n",
      "Processing path breaks...\n",
      "Postprocessing interconnected links 3/3\n",
      "Preprocessing interconnected links finished.\n",
      "\n",
      "Processing path breaks finished.\n",
      "Converting blocks to block lengths 6/6\n",
      "Conversion finished.\n",
      "Reformating links to block lengths associations 6/6\n",
      "Reformating finished.\n",
      "Identifying rearrangement blocks\n",
      "Processing node 7/7\n",
      "Identifying rearrangement blocks finished.\n",
      "\n",
      "===========================\n",
      "Zoom level 1\n",
      "===========================\n",
      "Processing node 7/7\n",
      "Processing component links 22/22\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "Recording component 22/22\n",
      "Recording zoom level 1 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <3 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 3\n",
      "===========================\n",
      "Processing component 22/22\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 12/12\n",
      "Recording zoom level 3 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <9 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 9\n",
      "===========================\n",
      "Processing component 12/12\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 5/5\n",
      "Recording zoom level 9 finished.\n",
      "Removing links according to collapsible blocks\n",
      "All links associated with collapsibleComponents <18 were removed.     0 components were deleted as isolated.\n",
      "\n",
      "===========================\n",
      "Zoom level 18\n",
      "===========================\n",
      "Processing component 5/5\n",
      "Processing component finished.\n",
      "Converting link to block lengths associations\n",
      "Converting paired links\n",
      "Converting interconnected links\n",
      "Converting rearrangement blocks\n",
      "\n",
      "Recording component 2/2\n",
      "Recording zoom level 18 finished.\n"
     ]
    }
   ],
   "source": [
    "pathfileDir = '../../1001G/coreGraph/new_Nov2022'\n",
    "maxLengthComponent = 100\n",
    "maxLengthChunk = 16\n",
    "invertionThreshold = 0.5\n",
    "zoomLevels = [1,3,9]\n",
    "zoomLevels = adjustZoomLevels(zoomLevels)\n",
    "\n",
    "for seqNum in range(1,6):\n",
    "    if doCreateCoreGenes:\n",
    "        coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_v2_Chr{seqNum:d}.gfa'\n",
    "\n",
    "        coregraph = GenomeGraph(gfaPath=coreGFApath,isGFASeq=False)\n",
    "        coregraph_genes = deepcopy(coregraph)\n",
    "\n",
    "        fullGraphPath = f'../../1001G/pantograph/data/AT_Chr{seqNum:d}_OGOnly_2.1.gfa'\n",
    "        fullgraph = GenomeGraph(gfaPath=fullGraphPath,isGFASeq=False)\n",
    "\n",
    "        chainToGenesFile = f'chain2gene.txt'\n",
    "        # maxChainLength = len(coregraph.nodes[0])\n",
    "        chainToListDict = {}\n",
    "        nodeLengths = []\n",
    "        with open(f'{pathfileDir}{os.path.sep}{chainToGenesFile}') as f:\n",
    "            for line in f:\n",
    "                chainName, geneList = line.split(':')\n",
    "                geneList = geneList.lstrip().rstrip().split(',')\n",
    "                chainToListDict[chainName] = geneList\n",
    "\n",
    "        for nodeIdx,nodeName in enumerate(coregraph.nodes):\n",
    "            print(f'\\rNode {nodeIdx+1}/{len(coregraph.nodes)}',end='')\n",
    "            geneList = chainToListDict.get(nodeName, [])\n",
    "            # geneIds = []\n",
    "            # if geneList[0][:2]!='ch':\n",
    "            geneList = [gene.rstrip('+') for gene in geneList]\n",
    "            geneIds = [fullgraph.nodes.index(genename) for genename in geneList]\n",
    "            # coregraph_genes.nodesData[nodeIdx] = ''.join(geneList)\n",
    "            geneNum = max(1,len(geneList))\n",
    "            nodeLengths.append(geneNum)\n",
    "\n",
    "            for accession, chainDict in coregraph.nodesAnnotation[nodeIdx].items():\n",
    "                interval = chainDict[nodeName]\n",
    "                # geneCumLengths = np.hstack((0, np.cumsum([len(gene) for gene in geneList])))\n",
    "                \n",
    "                coregraph.nodesAnnotation[nodeIdx][accession].pop(nodeName,None)\n",
    "                \n",
    "                coregraph.nodesAnnotation[nodeIdx][accession].\\\n",
    "                    update({geneAnnotation:interval \\\n",
    "                            for i,geneid in enumerate(geneIds) \\\n",
    "                                for geneAnnotation in fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:None}).keys()})\n",
    "\n",
    "                coregraph_genes.nodesAnnotation[nodeIdx][accession].\\\n",
    "                    update({geneAnnotation:[(i,i)] \\\n",
    "                                    for i,geneid in enumerate(geneIds) \\\n",
    "                                        for geneAnnotation in fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:None}).keys()})\n",
    "\n",
    "                # coregraph_genes.nodesAnnotation[nodeIdx][accession].\\\n",
    "                #     update({nodeName:[(0, geneNum-1)]})\n",
    "        print('')\n",
    "\n",
    "    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_v2_Chr{seqNum:d}.gfa'\n",
    "    coreGeneGFApath = f'{pathfileDir}{os.path.sep}coregraph_v2_genes_Chr{seqNum:d}.gfa'\n",
    "    \n",
    "    if doCreateCoreGenes:\n",
    "        coreGeneNodeLengthsPath = f'{pathfileDir}{os.path.sep}nodeLengths_coregraph_v2_genes_Chr{seqNum:d}.dat'\n",
    "\n",
    "        coregraph.toGFA(coreGFApath,doSeq=False)\n",
    "        coregraph_genes.toGFA(coreGeneGFApath,doSeq=True)\n",
    "        joblib.dump(nodeLengths,coreGeneNodeLengthsPath)\n",
    "\n",
    "    #Exporting chain graph with annotation only\n",
    "    isSeq = False\n",
    "    inputPath = coreGFApath\n",
    "    \n",
    "    outputPath,outputName = pathConvert(inputPath,suffix='_new')\n",
    "\n",
    "    # dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "    # print(f'Opening Redis connection for db {dbid}')\n",
    "    redisConn = Redis(host='redis',port = 6379,db=0)\n",
    "\n",
    "    exportToPantograph(inputPath=inputPath,\n",
    "                       outputName=outputName,\n",
    "                       outputPath=outputPath,\n",
    "                       isSeq=isSeq,\n",
    "                       redisConn=redisConn,\n",
    "                       GenomeGraphParams={'accessionsToRemove':['Consensus']},\n",
    "                       zoomLevels=zoomLevels,\n",
    "                       fillZoomLevels=True,\n",
    "                       maxLengthChunk=maxLengthChunk,\n",
    "                       maxLengthComponent=maxLengthComponent,\n",
    "                       inversionThreshold=inversionThreshold,)\n",
    "\n",
    "    #Exporting chain graph with genes\n",
    "    isSeq = False\n",
    "    inputPath = coreGeneGFApath\n",
    "    \n",
    "    outputPath,outputName = pathConvert(inputPath,suffix='_new')\n",
    "\n",
    "    # dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)\n",
    "    # print(f'Opening Redis connection for db {dbid}')\n",
    "    redisConn = Redis(host='redis',port = 6379,db=0)\n",
    "\n",
    "    exportToPantograph(inputPath=inputPath,\n",
    "                       outputName=outputName,\n",
    "                       outputPath=outputPath,\n",
    "                       isSeq=isSeq,\n",
    "                       nodeLengths=checkNodeLengthsFile(inputPath),\n",
    "                       redisConn=redisConn,\n",
    "                       GenomeGraphParams={'accessionsToRemove':['Consensus']},\n",
    "                       zoomLevels=zoomLevels,\n",
    "                       fillZoomLevels=True,\n",
    "                       maxLengthChunk=maxLengthChunk,\n",
    "                       maxLengthComponent=maxLengthComponent,\n",
    "                       inversionThreshold=inversionThreshold,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.11) or chardet (3.0.4) doesn't match a supported version!\n",
      "  warnings.warn(\"urllib3 ({}) or chardet ({}) doesn't match a supported \"\n"
     ]
    }
   ],
   "source": [
    "curT = time.localtime()\n",
    "message = f\"Adding genes and exporting of chain graphs for all chromosome finished at \\\n",
    "            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\"\n",
    "!ntfy send \"{message}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "red = Redis(host='redis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pangraph_constructor.utils import iset_get"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[b'AT_Chr1_OGOnly_2.1_new.4.6124.PosStart',\n",
       " b'coregraph_v2_Chr1_new.3.9905.PosStart',\n",
       " b'coregraph_v2_Chr5_new.1.9905.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.22002.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.1.TAIR10.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.6124.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.9728.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.9.9537.PosStart',\n",
       " b'coregraph_Chr1_new.6244.GeneStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.6124.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.10002.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.3.22005.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.6124.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.9537.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.36.9888.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.1.22004.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.10024.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1024.10015.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.22007.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.9543.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9.TAIR10.PosStart',\n",
       " b'coregraph_v2_Chr5_new.6244.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.22004.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.6244.PosStart',\n",
       " b'coregraph_genes_Chr1_new.18.22007.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.18.9905.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.10002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.22002.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.8236.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.3.9905.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9981.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.128.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.9075.PosStart',\n",
       " b'coregraph_genes_Chr1_new.22001.GeneStart',\n",
       " b'coregraph_v2_Chr4_new.3.9728.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4096.8236.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.1.8236.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.16.22002.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.9981.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.22002.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.3.10015.PosStart',\n",
       " b'coregraph_v2_Chr4_new.6069.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.22001.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4096.9888.PosStart',\n",
       " b'coregraph_v2_Chr3_new.9.9764.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.9.22006.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.9543.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.3.6244.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.9728.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.22004.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1.6966.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.8236.PosStart',\n",
       " b'coregraph_v2_Chr1_new.1.9905.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.9728.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.6966.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1024.10002.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.22004.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.9075.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.1.1741.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.10002.PosStart',\n",
       " b'coregraph_Chr1_new.3.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.3.10024.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.9981.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.6966.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.22001.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.6069.GeneStart',\n",
       " b'coregraph_v2_genes_Chr1_new.36.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.1.1741.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.9.22007.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.6244.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.22003.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.6069.GeneEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.1741.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.18.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.18.1741.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.22003.PosStart',\n",
       " b'coregraph_v2_Chr3_new.9.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.9.1741.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.9905.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.72.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.9.10024.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.1741.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.22005.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.1.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.22004.GeneStart',\n",
       " b'coregraph_v2_genes_Chr3_new.288.9728.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.3.22006.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.6069.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.9981.PosStart',\n",
       " b'coregraph_genes_Chr1_new.1.22007.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.3.6244.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.9888.PosStart',\n",
       " b'coregraph_Chr1_new.1.6024.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9.22007.PosStart',\n",
       " b'coregraph_v2_Chr1_new.9.10024.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.128.22004.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.6909.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.1741.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1024.22006.PosStart',\n",
       " b'coregraph_v2_Chr5_new.3.6069.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.6966.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.3.9888.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.9075.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.9981.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.6909.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.22004.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.18.8236.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9.9888.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.36.9075.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.6909.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.22005.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.6024.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.1741.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.8236.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.18.6244.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.6966.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.22005.PosEnd',\n",
       " b'coregraph_Chr1_new.9.10024.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.1.22005.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.22006.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.22007.GeneEnd',\n",
       " b'coregraph_v2_Chr3_new.1.22005.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.6966.PosStart',\n",
       " b'coregraph_v2_Chr3_new.9.6124.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.TAIR10.GeneEnd',\n",
       " b'coregraph_v2_Chr2_new.3.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.9905.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.22001.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.22001.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.22006.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.10002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.TAIR10.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.1.22005.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.9537.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.6966.GeneEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.6909.PosStart',\n",
       " b'coregraph_v2_Chr5_new.1.1741.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9.9905.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.6069.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.22007.GeneStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.9905.PosStart',\n",
       " b'coregraph_v2_Chr4_new.1.22006.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.1.6024.PosStart',\n",
       " b'coregraph_genes_Chr1_new.9.9888.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.6024.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.9888.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.10024.GeneStart',\n",
       " b'coregraph_v2_Chr1_new.9.1741.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.18.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9981.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.9764.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.9075.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.18.9764.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.8236.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.22004.PosStart',\n",
       " b'coregraph_genes_Chr1_new.9537.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1024.6024.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.18.9888.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.36.9537.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9543.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.6244.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.22003.PosEnd',\n",
       " b'coregraph_Chr1_new.1.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.18.22001.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.10002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.1.9905.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.22002.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.9075.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4096.22001.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4096.TAIR10.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.22006.GeneStart',\n",
       " b'coregraph_v2_Chr1_new.9728.GeneEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.9537.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.9075.PosEnd',\n",
       " b'coregraph_Chr1_new.8.1741.PosStart',\n",
       " b'coregraph_v2_Chr3_new.1.6966.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.9543.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.1.9543.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.6124.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.9981.GeneStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.8236.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.9075.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.22006.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.6244.GeneStart',\n",
       " b'coregraph_genes_Chr1_new.1.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.18.10002.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.22001.GeneEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.22002.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.9981.PosStart',\n",
       " b'coregraph_v2_Chr3_new.9.22007.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.10024.PosStart',\n",
       " b'coregraph_Chr1_new.8.6124.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.9.22002.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.8236.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.144.9543.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.6024.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.9537.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.36.22002.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.9888.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.6124.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.9537.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.8236.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.TAIR10.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.3.1741.PosStart',\n",
       " b'coregraph_genes_Chr1_new.144.9905.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.1.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.1.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.9728.GeneEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.9905.PosEnd',\n",
       " b'coregraph_Chr1_new.1.8236.PosStart',\n",
       " b'coregraph_Chr1_new.8.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.3.10002.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.1.22004.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.9537.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.1741.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.22005.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.3.6024.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.3.22005.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.22004.GeneEnd',\n",
       " b'coregraph_v2_Chr5_new.3.22002.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.10015.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.9543.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.6966.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.10024.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.1.22002.PosStart',\n",
       " b'coregraph_v2_Chr1_new.1.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.144.9543.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.3.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.10002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.144.9075.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.6069.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.3.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.1.9543.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.22004.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.36.6244.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.6966.PosStart',\n",
       " b'coregraph_v2_Chr5_new.3.6124.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.1.9905.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.10002.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.TAIR10.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.9.9905.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.576.9764.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.22004.PosStart',\n",
       " b'coregraph_v2_Chr3_new.TAIR10.GeneStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.TAIR10.PosEnd',\n",
       " b'coregraph_Chr1_new.16.6909.PosEnd',\n",
       " b'coregraph_Chr1_new.1.6966.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.22006.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.1741.PosStart',\n",
       " b'coregraph_v2_Chr3_new.1.9543.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.36.9764.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.9905.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.144.22006.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.6244.GeneStart',\n",
       " b'coregraph_v2_genes_Chr3_new.288.9638.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.9543.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9.9764.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.22001.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.6069.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.18.TAIR10.PosStart',\n",
       " b'coregraph_v2_Chr1_new.3.9764.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.6124.GeneStart',\n",
       " b'coregraph_genes_Chr1_new.9.9728.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.10015.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.9728.GeneStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.9981.PosEnd',\n",
       " b'coregraph_Chr1_new.8.9905.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.8236.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.9981.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.6069.GeneStart',\n",
       " b'coregraph_v2_Chr3_new.3.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.1741.GeneStart',\n",
       " b'coregraph_v2_Chr2_new.9.9075.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.3.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.72.TAIR10.PosEnd',\n",
       " b'coregraph_Chr1_new.9.9638.PosStart',\n",
       " b'coregraph_genes_Chr1_new.3.9905.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.10002.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.1.9888.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.6124.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.9981.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.6244.GeneStart',\n",
       " b'coregraph_v2_genes_Chr2_new.36.9728.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.22003.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.18.9728.PosStart',\n",
       " b'coregraph_v2_Chr3_new.8236.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4096.6124.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1741.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.1.22004.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.TAIR10.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.22001.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.1.1741.PosStart',\n",
       " b'coregraph_v2_Chr1_new.22007.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.22001.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.9888.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.3.22002.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.22005.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.9905.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.9981.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.9888.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.22005.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9764.GeneStart',\n",
       " b'coregraph_v2_genes_Chr1_new.72.22001.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.1.8236.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.6124.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.9543.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9537.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.TAIR10.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.3.9905.PosEnd',\n",
       " b'coregraph_Chr1_new.1741.GeneStart',\n",
       " b'coregraph_genes_Chr1_new.9.9537.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.9905.PosStart',\n",
       " b'coregraph_Chr1_new.9638.GeneEnd',\n",
       " b'coregraph_Chr1_new.1.6966.PosEnd',\n",
       " b'coregraph_Chr1_new.16.22006.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.9543.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4096.9728.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.3.22004.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.10015.PosStart',\n",
       " b'coregraph_Chr1_new.16.6966.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.18.22007.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.6244.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.6124.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.6966.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.9075.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9.9075.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.3.22003.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.22002.PosStart',\n",
       " b'coregraph_v2_Chr4_new.1.9537.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.9.9075.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.3.6244.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.9.6124.PosStart',\n",
       " b'coregraph_Chr1_new.9.22006.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.1.9981.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1024.1741.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.6244.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.22002.GeneStart',\n",
       " b'coregraph_Chr1_new.9.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.1741.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.9.9075.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.TAIR10.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.6966.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.128.10015.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.128.9543.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.22007.GeneEnd',\n",
       " b'coregraph_v2_Chr3_new.3.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.9638.PosStart',\n",
       " b'coregraph_v2_Chr4_new.1.9638.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.22002.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9888.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.22006.GeneStart',\n",
       " b'coregraph_v2_genes_Chr2_new.9.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.1.22004.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.6124.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.22007.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.9728.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.1.8236.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.10024.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.1741.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.144.9543.PosStart',\n",
       " b'coregraph_genes_Chr1_new.9.10002.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.6024.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.9075.PosStart',\n",
       " b'coregraph_v2_Chr2_new.3.9537.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9.6244.PosEnd',\n",
       " b'coregraph_Chr1_new.22007.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.3.6124.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.9638.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.18.9537.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.22002.GeneEnd',\n",
       " b'coregraph_v2_Chr5_new.22006.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.6069.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.9.22006.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.22002.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.22007.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.22005.GeneEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.TAIR10.PosStart',\n",
       " b'coregraph_v2_Chr4_new.9.6909.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.22002.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9.6024.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.6966.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.6909.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.3.22002.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.22006.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.6909.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.9537.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.9543.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.9.1741.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.1.9075.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.1.10002.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.8236.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.3.10002.PosStart',\n",
       " b'coregraph_v2_Chr4_new.6024.GeneStart',\n",
       " b'coregraph_v2_genes_Chr5_new.10015.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.9981.PosStart',\n",
       " b'coregraph_Chr1_new.9.9905.PosStart',\n",
       " b'coregraph_Chr1_new.16.9905.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9537.GeneEnd',\n",
       " b'coregraph_v2_Chr3_new.1.22007.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.1.22005.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.1.9764.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.10015.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.9075.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1.9764.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.128.9728.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.6069.PosStart',\n",
       " b'coregraph_Chr1_new.3.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.36.6124.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.144.6244.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.1.6909.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.9537.PosStart',\n",
       " b'coregraph_Chr1_new.9.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1.9981.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.3.10015.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.36.22005.PosStart',\n",
       " b'coregraph_Chr1_new.9.6069.PosEnd',\n",
       " b'coregraph_Chr1_new.2.6966.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.22004.PosStart',\n",
       " b'coregraph_v2_Chr2_new.22005.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.22004.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.6069.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.1.22005.PosStart',\n",
       " b'coregraph_Chr1_new.8.22002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.10015.PosEnd',\n",
       " b'coregraph_Chr1_new.9.6244.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.36.22006.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.6244.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.144.6069.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.1741.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.10002.PosStart',\n",
       " b'coregraph_v2_Chr3_new.3.22004.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.1.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8236.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.22006.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.1741.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.9.6244.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.6069.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.6909.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9537.GeneStart',\n",
       " b'coregraph_v2_genes_Chr1_new.72.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.22007.GeneEnd',\n",
       " b'coregraph_v2_Chr3_new.1.8236.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.9905.PosStart',\n",
       " b'coregraph_v2_Chr1_new.9543.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1024.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.3.6244.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.36.22007.PosStart',\n",
       " b'coregraph_v2_Chr5_new.3.9537.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.9.9905.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.3.22004.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.144.22002.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.36.22003.PosStart',\n",
       " b'coregraph_genes_Chr1_new.9.9075.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.10002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.36.9728.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9764.PosStart',\n",
       " b'coregraph_v2_Chr2_new.1.8236.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.36.9543.PosStart',\n",
       " b'coregraph_genes_Chr1_new.36.10015.PosStart',\n",
       " b'coregraph_genes_Chr1_new.72.6909.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.10024.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9888.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.9728.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8192.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.72.6024.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.9537.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8236.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.22005.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.22002.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.9.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.1.9075.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.9.22004.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.9537.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.9905.PosStart',\n",
       " b'coregraph_v2_Chr1_new.1.TAIR10.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.18.10024.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.6124.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.9075.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.9.9543.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.576.22003.PosStart',\n",
       " b'coregraph_Chr1_new.8.9537.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.9.9728.PosStart',\n",
       " b'coregraph_v2_Chr4_new.9.6966.PosStart',\n",
       " b'coregraph_v2_Chr3_new.9.9537.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1024.6124.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.9.22005.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.22002.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.64.9981.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.6024.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.22002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.36.9728.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.10002.GeneEnd',\n",
       " b'coregraph_v2_Chr1_new.3.22003.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.9981.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.9.10015.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.6124.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.10002.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.3.22001.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.1.9888.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.22002.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.1.6124.PosStart',\n",
       " b'coregraph_v2_Chr4_new.3.22004.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.22001.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.9.22002.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.9537.GeneStart',\n",
       " b'coregraph_v2_Chr4_new.9.9543.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.6966.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.TAIR10.GeneEnd',\n",
       " b'coregraph_v2_Chr2_new.22002.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.128.8236.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9.9764.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.18.22005.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.6966.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.22001.PosStart',\n",
       " b'coregraph_genes_Chr1_new.1.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.9.9764.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.10024.PosStart',\n",
       " b'coregraph_v2_Chr3_new.1.10015.PosStart',\n",
       " b'coregraph_v2_Chr1_new.3.22001.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.36.22004.PosStart',\n",
       " b'coregraph_v2_Chr2_new.1.9888.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.36.6124.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.10002.PosStart',\n",
       " b'coregraph_genes_Chr1_new.6909.GeneStart',\n",
       " b'coregraph_genes_Chr1_new.9.8236.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.3.6909.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9981.GeneEnd',\n",
       " b'coregraph_v2_Chr5_new.9.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.288.22001.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.8236.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.6966.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.TAIR10.PosStart',\n",
       " b'coregraph_v2_Chr1_new.6124.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.3.9764.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4096.6966.PosStart',\n",
       " b'coregraph_v2_Chr4_new.9728.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.1.9075.PosStart',\n",
       " b'coregraph_v2_Chr3_new.1.6244.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.22006.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.144.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.6909.PosStart',\n",
       " b'coregraph_Chr1_new.2.6966.PosStart',\n",
       " b'coregraph_genes_Chr1_new.144.1741.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.1.9728.PosStart',\n",
       " b'coregraph_Chr1_new.2.6909.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.1741.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.9638.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.22005.PosStart',\n",
       " b'coregraph_Chr1_new.3.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.1741.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.9543.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.6909.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.6909.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.36.6024.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.22007.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.TAIR10.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.9.10002.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.6909.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.9905.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.TAIR10.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.6124.GeneEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.64.9543.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.6244.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.36.9728.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1024.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.72.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9764.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.8236.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.9075.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.22001.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.16.9905.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.9075.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.9543.PosStart',\n",
       " b'coregraph_genes_Chr1_new.72.9888.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.22003.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.6069.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.9543.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.22002.GeneStart',\n",
       " b'coregraph_Chr1_new.4.22005.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.3.9075.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.TAIR10.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.1.22006.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.22005.PosStart',\n",
       " b'coregraph_genes_Chr1_new.144.9075.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.1.9728.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.6124.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.3.6024.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.22003.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.1.9075.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.9981.PosStart',\n",
       " b'coregraph_v2_Chr4_new.3.9764.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.1.22003.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.22006.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.3.9888.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.18.22004.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.9905.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.9638.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.9075.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.1.6069.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.6966.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.22006.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.72.22005.PosStart',\n",
       " b'coregraph_v2_Chr4_new.9075.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.22003.GeneEnd',\n",
       " b'coregraph_v2_Chr5_new.6024.GeneEnd',\n",
       " b'coregraph_v2_Chr4_new.9.10002.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1024.9543.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.10002.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.6909.PosStart',\n",
       " b'coregraph_Chr1_new.16.TAIR10.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.9905.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.144.22001.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.9764.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1024.10015.PosStart',\n",
       " b'coregraph_v2_Chr3_new.3.22001.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.18.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.10015.GeneStart',\n",
       " b'coregraph_v2_Chr1_new.9.6124.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.1.10015.PosStart',\n",
       " b'coregraph_genes_Chr1_new.3.9981.PosEnd',\n",
       " b'coregraph_Chr1_new.4.6909.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.8236.PosEnd',\n",
       " b'coregraph_Chr1_new.9.6124.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.10015.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.3.TAIR10.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.8236.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.16.9764.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.9728.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.3.9537.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.3.6966.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9.TAIR10.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.8236.GeneStart',\n",
       " b'coregraph_v2_genes_Chr2_new.1.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.3.6024.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.22007.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.9905.PosStart',\n",
       " b'coregraph_v2_Chr2_new.1.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.6966.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.9905.PosStart',\n",
       " b'coregraph_v2_Chr1_new.1.6909.PosStart',\n",
       " b'coregraph_v2_Chr3_new.9.9075.PosEnd',\n",
       " b'coregraph_Chr1_new.8.10015.PosEnd',\n",
       " b'coregraph_Chr1_new.16.22005.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.72.TAIR10.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.22005.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.1.10024.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4096.22007.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.9764.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.576.1741.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.18.9981.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.9543.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.36.22007.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.22007.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.3.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.18.9543.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.22002.PosStart',\n",
       " b'coregraph_Chr1_new.1.1741.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.TAIR10.PosStart',\n",
       " b'coregraph_genes_Chr1_new.3.9905.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.3.6909.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.1741.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.72.9543.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.9.22004.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4096.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.6124.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.22003.PosStart',\n",
       " b'coregraph_v2_Chr1_new.1.6244.PosStart',\n",
       " b'coregraph_v2_Chr4_new.9.6244.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.22005.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.22004.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.22005.PosStart',\n",
       " b'coregraph_genes_Chr1_new.144.9728.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.1.6909.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.22001.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.9.9764.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.1741.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.36.10002.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9537.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.18.TAIR10.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4096.1741.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.9764.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.3.22004.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.6909.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.3.9981.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.9764.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.1.6024.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.22003.GeneEnd',\n",
       " b'coregraph_Chr1_new.2.22002.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9.22001.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.6909.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.TAIR10.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.22002.PosStart',\n",
       " b'coregraph_v2_Chr4_new.6909.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.72.1741.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9.8236.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.1.9905.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.36.9905.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.9981.GeneEnd',\n",
       " b'coregraph_v2_Chr2_new.1.22006.PosEnd',\n",
       " b'coregraph_Chr1_new.1.6244.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.3.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.9981.PosStart',\n",
       " b'coregraph_genes_Chr1_new.9.TAIR10.PosStart',\n",
       " b'coregraph_v2_Chr5_new.1.9537.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.22006.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.18.6909.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.10002.PosStart',\n",
       " b'coregraph_v2_Chr1_new.3.9905.PosEnd',\n",
       " b'coregraph_Chr1_new.16.9764.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.72.22007.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.9075.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.9728.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.22004.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.16.9537.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.6909.GeneStart',\n",
       " b'coregraph_v2_Chr1_new.3.9075.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1024.1741.PosStart',\n",
       " b'coregraph_Chr1_new.9.9764.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.1.10002.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.6024.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.8236.PosStart',\n",
       " b'coregraph_Chr1_new.9.8236.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2.22006.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.22004.PosEnd',\n",
       " b'coregraph_Chr1_new.9.9075.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9.9543.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.22001.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8.9764.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.18.9728.PosStart',\n",
       " b'coregraph_v2_Chr3_new.10015.GeneStart',\n",
       " b'coregraph_v2_genes_Chr1_new.18.6069.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.1.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.9.6244.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.22003.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.6024.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.10002.GeneEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.6124.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.22001.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.9981.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.10002.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.6124.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9537.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.9905.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.9075.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9.TAIR10.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.6966.PosStart',\n",
       " b'coregraph_Chr1_new.1.9075.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.9905.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.6966.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.576.8236.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.3.1741.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.22004.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.22006.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.1.6244.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.22005.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.9638.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.6909.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.6909.GeneStart',\n",
       " b'coregraph_v2_Chr4_new.9.9638.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.6024.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.512.22001.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.3.10015.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.6244.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.10024.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.22005.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.9888.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.32.6909.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.6909.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.144.TAIR10.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.9981.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.22006.PosEnd',\n",
       " b'coregraph_Chr1_new.3.9728.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.3.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.9.9888.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.9543.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.6909.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.1.10024.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.9.22001.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.22007.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4096.9905.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.22007.PosStart',\n",
       " b'coregraph_v2_Chr3_new.3.10024.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.22002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.18.8236.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.9543.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1.9537.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.22002.PosStart',\n",
       " b'coregraph_v2_Chr3_new.3.22002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.72.10015.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.8236.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.22005.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.16.22004.PosStart',\n",
       " b'coregraph_Chr1_new.9.6069.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.9.9888.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.22005.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.36.9638.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.9.22007.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.1024.22004.PosEnd',\n",
       " b'coregraph_v2_genes_Chr1_new.3.10002.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.9.9764.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.18.9537.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.1.22002.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.22006.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.TAIR10.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9.6069.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.9.22001.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.1741.PosEnd',\n",
       " b'coregraph_v2_Chr4_new.1.10002.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.22007.PosStart',\n",
       " b'coregraph_v2_genes_Chr1_new.9.9728.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9.22005.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.1.6024.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4096.9075.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.9638.PosStart',\n",
       " b'coregraph_v2_Chr2_new.9888.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.9.8236.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.TAIR10.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.9075.GeneEnd',\n",
       " b'coregraph_Chr1_new.3.10002.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.9543.GeneEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1024.9888.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.18.9728.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.9.9638.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.10015.GeneEnd',\n",
       " b'coregraph_v2_Chr4_new.6124.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.32.9543.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.9888.GeneStart',\n",
       " b'coregraph_v2_genes_Chr4_new.9638.GeneStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1.22006.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.6244.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.22003.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.6909.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.3.22003.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.9075.GeneStart',\n",
       " b'coregraph_v2_genes_Chr4_new.18.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.22001.GeneStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.4.22007.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.1.9075.PosStart',\n",
       " b'coregraph_v2_Chr1_new.9764.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.72.10015.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.1.6909.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.22007.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.6244.PosEnd',\n",
       " b'coregraph_v2_Chr2_new.9.9905.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.9.22004.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.10024.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.22004.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.8236.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.6124.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.10015.PosEnd',\n",
       " b'coregraph_Chr1_new.3.6124.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.3.10024.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.9905.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.1741.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.10015.PosStart',\n",
       " b'coregraph_Chr1_new.8.22007.PosStart',\n",
       " b'coregraph_Chr1_new.3.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.36.9981.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.22002.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.9.TAIR10.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.16.10024.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.128.9888.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9888.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.22002.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.256.6966.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.9.6024.PosStart',\n",
       " b'coregraph_v2_Chr4_new.3.10015.PosStart',\n",
       " b'coregraph_Chr1_new.16.6966.PosEnd',\n",
       " b'coregraph_v2_Chr3_new.3.6069.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.6069.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.22006.GeneStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.2048.22002.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.9764.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.64.22003.PosStart',\n",
       " b'coregraph_v2_genes_Chr2_new.36.9888.PosStart',\n",
       " b'coregraph_v2_Chr5_new.9.6244.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.64.10002.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.TAIR10.GeneStart',\n",
       " b'coregraph_v2_Chr5_new.1.22001.PosStart',\n",
       " b'coregraph_Chr1_new.8.10002.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.3.9543.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.1.22007.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8192.6124.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.9728.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.9888.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.512.9888.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.18.6024.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.22003.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.18.22006.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.4.6024.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.TAIR10.GeneEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.4.9764.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.6069.GeneEnd',\n",
       " b'coregraph_genes_Chr1_new.22002.GeneEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.1.22001.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.144.6124.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.1.9728.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.6124.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.8.22005.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.256.9905.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.9888.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.36.10015.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9638.PosStart',\n",
       " b'coregraph_Chr1_new.16.8236.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.9.22006.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9075.PosEnd',\n",
       " b'coregraph_genes_Chr1_new.72.TAIR10.PosStart',\n",
       " b'coregraph_Chr1_new.16.9764.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.9981.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.512.9075.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.9638.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.1.22003.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.18.6124.PosStart',\n",
       " b'coregraph_genes_Chr1_new.36.6244.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.9075.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.6966.PosStart',\n",
       " b'coregraph_v2_Chr1_new.9543.GeneEnd',\n",
       " b'coregraph_Chr1_new.4.9728.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2.9888.PosStart',\n",
       " b'AT_Chr2_OGOnly_2.1_new.64.22007.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.32.22001.PosStart',\n",
       " b'coregraph_Chr1_new.4.6244.PosStart',\n",
       " b'coregraph_genes_Chr1_new.18.9905.PosStart',\n",
       " b'coregraph_v2_Chr1_new.22005.GeneStart',\n",
       " b'coregraph_v2_Chr5_new.9.10002.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9764.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.22005.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.18.22006.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.2048.9981.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.3.9075.PosStart',\n",
       " b'coregraph_v2_Chr3_new.1.6244.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.9.6909.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2048.10024.PosEnd',\n",
       " b'coregraph_v2_genes_Chr4_new.1.1741.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.9.TAIR10.PosEnd',\n",
       " b'coregraph_v2_genes_Chr3_new.288.6244.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.256.9543.PosStart',\n",
       " b'coregraph_Chr1_new.3.9728.PosEnd',\n",
       " b'coregraph_v2_Chr1_new.1.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr3_new.18.22006.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.36.6024.PosStart',\n",
       " b'coregraph_genes_Chr1_new.1.9537.PosEnd',\n",
       " b'coregraph_v2_Chr5_new.9.8236.PosEnd',\n",
       " b'coregraph_v2_genes_Chr5_new.1.8236.PosStart',\n",
       " b'coregraph_v2_genes_Chr4_new.1.9764.PosStart',\n",
       " b'coregraph_v2_Chr4_new.3.9075.PosStart',\n",
       " b'AT_Chr1_OGOnly_2.1_new.8.8236.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.TAIR10.GeneStart',\n",
       " b'coregraph_v2_genes_Chr3_new.576.22005.PosEnd',\n",
       " b'coregraph_v2_genes_Chr2_new.1.6244.PosEnd',\n",
       " b'AT_Chr3_OGOnly_2.1_new.2.22001.PosEnd',\n",
       " b'AT_Chr1_OGOnly_2.1_new.128.9728.PosEnd',\n",
       " b'AT_Chr2_OGOnly_2.1_new.8192.10015.PosStart',\n",
       " b'coregraph_v2_genes_Chr5_new.3.6966.PosStart',\n",
       " b'AT_Chr3_OGOnly_2.1_new.1.22001.PosStart',\n",
       " ...]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "red.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'OG0008080_0': (1777.0, 1777.0)}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iset_get(red,'coregraph_v2_genes_Chr1_new.10002.Gene','OG0008080_0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AT3G06020_0': (0.0, 0.0),\n",
       " 'OG0010622_0': (0.0, 0.0),\n",
       " 'AT3G06030_0': (1.0, 1.0),\n",
       " 'OG0005087_0': (1.0, 1.0),\n",
       " 'AT3G06035_0': (2.0, 2.0),\n",
       " 'OG0019490_0': (2.0, 2.0),\n",
       " 'AT3G06070_0': (3.0, 3.0),\n",
       " 'OG0010623_0': (3.0, 3.0),\n",
       " 'AT3G06080_0': (4.0, 4.0),\n",
       " 'OG0010624_0': (4.0, 4.0),\n",
       " 'AT3G06100_0': (5.0, 5.0),\n",
       " 'OG0010625_0': (5.0, 5.0),\n",
       " 'AT3G06110_0': (6.0, 6.0),\n",
       " 'OG0010626_0': (6.0, 6.0),\n",
       " 'AT3G06120_0': (7.0, 7.0),\n",
       " 'OG0010627_0': (7.0, 7.0),\n",
       " 'AT3G06130_0': (8.0, 8.0),\n",
       " 'OG0010628_0': (8.0, 8.0),\n",
       " 'AT3G06140_0': (9.0, 9.0),\n",
       " 'OG0010629_0': (9.0, 9.0),\n",
       " 'AT3G06145_0': (10.0, 10.0),\n",
       " 'OG0010630_0': (10.0, 10.0),\n",
       " 'AT3G06150_0': (11.0, 11.0),\n",
       " 'OG0005088_0': (11.0, 11.0),\n",
       " 'AT3G06160_0': (12.0, 12.0),\n",
       " 'OG0005089_0': (12.0, 12.0),\n",
       " 'AT3G06170_0': (13.0, 13.0),\n",
       " 'OG0005090_0': (13.0, 13.0),\n",
       " 'AT3G06180_0': (14.0, 14.0),\n",
       " 'OG0010631_0': (14.0, 14.0),\n",
       " 'AT3G06190_0': (15.0, 15.0),\n",
       " 'OG0010632_0': (15.0, 15.0),\n",
       " 'AT3G06200_0': (16.0, 16.0),\n",
       " 'OG0010633_0': (16.0, 16.0),\n",
       " 'AT3G06210_0': (17.0, 17.0),\n",
       " 'OG0010634_0': (17.0, 17.0),\n",
       " 'AT3G06220_0': (18.0, 18.0),\n",
       " 'OG0010635_0': (18.0, 18.0),\n",
       " 'AT3G06230_0': (19.0, 19.0),\n",
       " 'OG0005091_0': (19.0, 19.0),\n",
       " 'AT3G06240_0': (20.0, 20.0),\n",
       " 'OG0010636_0': (20.0, 20.0),\n",
       " 'AT3G06250_0': (21.0, 21.0),\n",
       " 'OG0010637_0': (21.0, 21.0),\n",
       " 'AT3G06260_0': (22.0, 22.0),\n",
       " 'OG0016380_0': (22.0, 22.0),\n",
       " 'AT3G06270_0': (23.0, 23.0),\n",
       " 'OG0005092_0': (23.0, 23.0),\n",
       " 'AT3G06290_0': (24.0, 24.0),\n",
       " 'OG0003367_0': (24.0, 24.0),\n",
       " 'AT3G06300_0': (25.0, 25.0),\n",
       " 'OG0016381_0': (25.0, 25.0),\n",
       " 'AT3G06310_0': (26.0, 26.0),\n",
       " 'OG0016382_0': (26.0, 26.0),\n",
       " 'AT3G06330_0': (27.0, 27.0),\n",
       " 'OG0010638_0': (27.0, 27.0),\n",
       " 'AT3G06340_0': (28.0, 28.0),\n",
       " 'OG0010639_0': (28.0, 28.0),\n",
       " 'AT3G06350_0': (29.0, 29.0),\n",
       " 'OG0016383_0': (29.0, 29.0),\n",
       " 'AT3G06420_0': (30.0, 30.0),\n",
       " 'OG0010640_0': (30.0, 30.0),\n",
       " 'AT3G06430_0': (31.0, 31.0),\n",
       " 'OG0010641_0': (31.0, 31.0),\n",
       " 'AT3G06460_0': (32.0, 32.0),\n",
       " 'OG0003369_0': (32.0, 32.0),\n",
       " 'AT3G06470_0': (33.0, 33.0),\n",
       " 'OG0005094_0': (33.0, 33.0),\n",
       " 'AT3G06480_0': (34.0, 34.0),\n",
       " 'OG0010642_0': (34.0, 34.0),\n",
       " 'AT3G06483_0': (35.0, 35.0),\n",
       " 'OG0005095_0': (35.0, 35.0),\n",
       " 'AT3G06490_0': (36.0, 36.0),\n",
       " 'OG0010643_0': (36.0, 36.0),\n",
       " 'AT3G06500_0': (37.0, 37.0),\n",
       " 'OG0003370_0': (37.0, 37.0),\n",
       " 'AT3G06510_0': (38.0, 38.0),\n",
       " 'OG0016385_0': (38.0, 38.0),\n",
       " 'AT3G06520_0': (39.0, 39.0),\n",
       " 'OG0016386_0': (39.0, 39.0),\n",
       " 'AT3G06530_0': (40.0, 40.0),\n",
       " 'OG0003371_0': (40.0, 40.0),\n",
       " 'AT3G06540_0': (41.0, 41.0),\n",
       " 'OG0010644_0': (41.0, 41.0),\n",
       " 'AT3G06550_0': (42.0, 42.0),\n",
       " 'OG0002398_0': (42.0, 42.0),\n",
       " 'AT3G06560_0': (43.0, 43.0),\n",
       " 'OG0002709_0': (43.0, 43.0),\n",
       " 'AT3G06570_0': (44.0, 44.0),\n",
       " 'OG0019491_0': (44.0, 44.0),\n",
       " 'AT3G06580_0': (45.0, 45.0),\n",
       " 'OG0010645_0': (45.0, 45.0),\n",
       " 'AT3G06590_0': (46.0, 46.0),\n",
       " 'OG0010646_0': (46.0, 46.0),\n",
       " 'AT3G06600_0': (47.0, 47.0),\n",
       " 'OG0010647_0': (47.0, 47.0),\n",
       " 'AT3G06610_0': (48.0, 48.0),\n",
       " 'OG0010648_0': (48.0, 48.0),\n",
       " 'AT3G06620_0': (49.0, 49.0),\n",
       " 'OG0005096_0': (49.0, 49.0),\n",
       " 'AT3G06630_0': (50.0, 50.0),\n",
       " 'OG0019492_0': (50.0, 50.0),\n",
       " 'AT3G06640_0': (51.0, 51.0),\n",
       " 'OG0019493_0': (51.0, 51.0),\n",
       " 'AT3G66652_0': (52.0, 52.0),\n",
       " 'OG0005097_0': (52.0, 52.0),\n",
       " 'AT3G66654_0': (53.0, 53.0),\n",
       " 'OG0010649_0': (53.0, 53.0),\n",
       " 'AT3G66656_0': (54.0, 54.0),\n",
       " 'OG0019494_0': (54.0, 54.0),\n",
       " 'AT3G06730_0': (55.0, 55.0),\n",
       " 'OG0010652_0': (55.0, 55.0),\n",
       " 'AT3G06740_0': (56.0, 56.0),\n",
       " 'OG0020159_0': (56.0, 56.0),\n",
       " 'AT3G06750_0': (57.0, 57.0),\n",
       " 'OG0010653_0': (57.0, 57.0),\n",
       " 'AT3G06760_0': (58.0, 58.0),\n",
       " 'OG0003372_0': (58.0, 58.0),\n",
       " 'AT3G06770_0': (59.0, 59.0),\n",
       " 'OG0010654_0': (59.0, 59.0),\n",
       " 'AT3G06778_0': (60.0, 60.0),\n",
       " 'OG0010655_0': (60.0, 60.0),\n",
       " 'AT3G06780_0': (61.0, 61.0),\n",
       " 'OG0010656_0': (61.0, 61.0),\n",
       " 'AT3G06790_0': (62.0, 62.0),\n",
       " 'OG0010657_0': (62.0, 62.0),\n",
       " 'AT3G06810_0': (63.0, 63.0),\n",
       " 'OG0010658_0': (63.0, 63.0),\n",
       " 'AT3G06840_0': (64.0, 64.0),\n",
       " 'OG0010659_0': (64.0, 64.0),\n",
       " 'AT3G06850_0': (65.0, 65.0),\n",
       " 'OG0010660_0': (65.0, 65.0),\n",
       " 'AT3G06860_0': (66.0, 66.0),\n",
       " 'OG0010661_0': (66.0, 66.0),\n",
       " 'AT3G06868_0': (67.0, 67.0),\n",
       " 'OG0010662_0': (67.0, 67.0),\n",
       " 'AT3G06880_0': (68.0, 68.0),\n",
       " 'OG0005098_0': (68.0, 68.0),\n",
       " 'AT3G06890_0': (69.0, 69.0),\n",
       " 'OG0016388_0': (69.0, 69.0),\n",
       " 'AT3G06910_0': (70.0, 70.0),\n",
       " 'OG0016389_0': (70.0, 70.0),\n",
       " 'AT3G06920_0': (71.0, 71.0),\n",
       " 'OG0003373_0': (71.0, 71.0),\n",
       " 'AT3G06950_0': (72.0, 72.0),\n",
       " 'OG0005099_0': (72.0, 72.0),\n",
       " 'AT3G06960_0': (73.0, 73.0),\n",
       " 'OG0005100_0': (73.0, 73.0),\n",
       " 'AT3G06970_0': (74.0, 74.0),\n",
       " 'OG0018479_0': (74.0, 74.0),\n",
       " 'AT3G06980_0': (75.0, 75.0),\n",
       " 'OG0005101_0': (75.0, 75.0),\n",
       " 'AT3G07020_0': (76.0, 76.0),\n",
       " 'OG0010664_0': (76.0, 76.0),\n",
       " 'AT3G07030_0': (77.0, 77.0),\n",
       " 'OG0016390_0': (77.0, 77.0),\n",
       " 'AT3G07050_0': (78.0, 78.0),\n",
       " 'OG0010665_0': (78.0, 78.0),\n",
       " 'AT3G07060_0': (79.0, 79.0),\n",
       " 'OG0010666_0': (79.0, 79.0),\n",
       " 'AT3G07080_0': (80.0, 80.0),\n",
       " 'OG0010667_0': (80.0, 80.0),\n",
       " 'AT3G07090_0': (81.0, 81.0),\n",
       " 'OG0010668_0': (81.0, 81.0),\n",
       " 'AT3G07100_0': (82.0, 82.0),\n",
       " 'OG0010669_0': (82.0, 82.0),\n",
       " 'AT3G07120_0': (83.0, 83.0),\n",
       " 'OG0016392_0': (83.0, 83.0),\n",
       " 'AT3G07130_0': (84.0, 84.0),\n",
       " 'OG0010670_0': (84.0, 84.0),\n",
       " 'AT3G07140_0': (85.0, 85.0),\n",
       " 'OG0005102_0': (85.0, 85.0),\n",
       " 'AT3G07150_0': (86.0, 86.0),\n",
       " 'OG0019495_0': (86.0, 86.0),\n",
       " 'AT3G07160_0': (87.0, 87.0),\n",
       " 'OG0003374_0': (87.0, 87.0),\n",
       " 'AT3G07170_0': (88.0, 88.0),\n",
       " 'OG0010671_0': (88.0, 88.0),\n",
       " 'AT3G07180_0': (89.0, 89.0),\n",
       " 'OG0003375_0': (89.0, 89.0),\n",
       " 'AT3G07190_0': (90.0, 90.0),\n",
       " 'OG0018481_0': (90.0, 90.0),\n",
       " 'AT3G07195_0': (91.0, 91.0),\n",
       " 'OG0010672_0': (91.0, 91.0),\n",
       " 'AT3G07200_0': (92.0, 92.0),\n",
       " 'OG0010673_0': (92.0, 92.0),\n",
       " 'AT3G07210_0': (93.0, 93.0),\n",
       " 'OG0010674_0': (93.0, 93.0),\n",
       " 'AT3G07270_0': (94.0, 94.0),\n",
       " 'OG0010675_0': (94.0, 94.0),\n",
       " 'AT3G07273_0': (95.0, 95.0),\n",
       " 'OG0010676_0': (95.0, 95.0),\n",
       " 'AT3G07290_0': (96.0, 96.0),\n",
       " 'OG0016393_0': (96.0, 96.0),\n",
       " 'AT3G07300_0': (97.0, 97.0),\n",
       " 'OG0005103_0': (97.0, 97.0),\n",
       " 'AT3G07340_0': (98.0, 98.0),\n",
       " 'OG0010677_0': (98.0, 98.0),\n",
       " 'AT3G07350_0': (99.0, 99.0),\n",
       " 'OG0005104_0': (99.0, 99.0),\n",
       " 'AT3G07390_0': (100.0, 100.0),\n",
       " 'OG0018483_0': (100.0, 100.0),\n",
       " 'AT3G07400_0': (101.0, 101.0),\n",
       " 'OG0010678_0': (101.0, 101.0),\n",
       " 'AT3G07430_0': (102.0, 102.0),\n",
       " 'OG0005107_0': (102.0, 102.0),\n",
       " 'AT3G07440_0': (103.0, 103.0),\n",
       " 'OG0016396_0': (103.0, 103.0),\n",
       " 'AT3G07480_0': (104.0, 104.0),\n",
       " 'OG0010680_0': (104.0, 104.0),\n",
       " 'AT3G07490_0': (105.0, 105.0),\n",
       " 'OG0018484_0': (105.0, 105.0),\n",
       " 'AT3G07500_0': (106.0, 106.0),\n",
       " 'OG0010681_0': (106.0, 106.0),\n",
       " 'AT3G07530_0': (107.0, 107.0),\n",
       " 'OG0002244_0': (107.0, 107.0),\n",
       " 'AT3G07540_0': (108.0, 108.0),\n",
       " 'OG0010683_0': (108.0, 108.0),\n",
       " 'AT3G07550_0': (109.0, 109.0),\n",
       " 'OG0010684_0': (109.0, 109.0),\n",
       " 'AT3G07560_0': (110.0, 110.0),\n",
       " 'OG0010685_0': (110.0, 110.0),\n",
       " 'AT3G07565_0': (111.0, 111.0),\n",
       " 'OG0002711_0': (111.0, 111.0),\n",
       " 'AT3G07570_0': (112.0, 112.0),\n",
       " 'OG0010686_0': (112.0, 112.0),\n",
       " 'AT3G07580_0': (113.0, 113.0),\n",
       " 'OG0016398_0': (113.0, 113.0),\n",
       " 'AT3G07600_0': (114.0, 114.0),\n",
       " 'OG0010687_0': (114.0, 114.0),\n",
       " 'AT3G07610_0': (115.0, 115.0),\n",
       " 'OG0003376_0': (115.0, 115.0),\n",
       " 'AT3G07620_0': (116.0, 116.0),\n",
       " 'OG0005108_0': (116.0, 116.0),\n",
       " 'AT3G07630_0': (117.0, 117.0),\n",
       " 'OG0010688_0': (117.0, 117.0),\n",
       " 'AT3G07640_0': (118.0, 118.0),\n",
       " 'OG0010689_0': (118.0, 118.0),\n",
       " 'AT3G07650_0': (119.0, 119.0),\n",
       " 'OG0010690_0': (119.0, 119.0),\n",
       " 'AT3G07660_0': (120.0, 120.0),\n",
       " 'OG0010691_0': (120.0, 120.0),\n",
       " 'AT3G07670_0': (121.0, 121.0),\n",
       " 'OG0010692_0': (121.0, 121.0),\n",
       " 'AT3G07680_0': (122.0, 122.0),\n",
       " 'OG0018485_0': (122.0, 122.0),\n",
       " 'AT3G07690_0': (123.0, 123.0),\n",
       " 'OG0010693_0': (123.0, 123.0),\n",
       " 'AT3G07700_0': (124.0, 124.0),\n",
       " 'OG0010694_0': (124.0, 124.0),\n",
       " 'AT3G07740_0': (125.0, 125.0),\n",
       " 'OG0003377_0': (125.0, 125.0),\n",
       " 'AT3G07750_0': (126.0, 126.0),\n",
       " 'OG0010696_0': (126.0, 126.0),\n",
       " 'AT3G07760_0': (127.0, 127.0),\n",
       " 'OG0003378_0': (127.0, 127.0),\n",
       " 'AT3G07770_0': (128.0, 128.0),\n",
       " 'OG0003379_0': (128.0, 128.0),\n",
       " 'AT3G07780_0': (129.0, 129.0),\n",
       " 'OG0018486_0': (129.0, 129.0),\n",
       " 'AT3G07790_0': (130.0, 130.0),\n",
       " 'OG0010697_0': (130.0, 130.0),\n",
       " 'AT3G07800_0': (131.0, 131.0),\n",
       " 'OG0010698_0': (131.0, 131.0),\n",
       " 'AT3G07810_0': (132.0, 132.0),\n",
       " 'OG0010699_0': (132.0, 132.0),\n",
       " 'AT3G07860_0': (133.0, 133.0),\n",
       " 'OG0010700_0': (133.0, 133.0),\n",
       " 'AT3G07870_0': (134.0, 134.0),\n",
       " 'OG0010701_0': (134.0, 134.0),\n",
       " 'AT3G07880_0': (135.0, 135.0),\n",
       " 'OG0016399_0': (135.0, 135.0),\n",
       " 'AT3G07890_0': (136.0, 136.0),\n",
       " 'OG0010702_0': (136.0, 136.0),\n",
       " 'AT3G07900_0': (137.0, 137.0),\n",
       " 'OG0010703_0': (137.0, 137.0),\n",
       " 'AT3G07910_0': (138.0, 138.0),\n",
       " 'OG0020161_0': (138.0, 138.0),\n",
       " 'AT3G07930_0': (139.0, 139.0),\n",
       " 'OG0010704_0': (139.0, 139.0),\n",
       " 'AT3G07940_0': (140.0, 140.0),\n",
       " 'OG0005109_0': (140.0, 140.0),\n",
       " 'AT3G07950_0': (141.0, 141.0),\n",
       " 'OG0018487_0': (141.0, 141.0),\n",
       " 'AT3G07960_0': (142.0, 142.0),\n",
       " 'OG0018488_0': (142.0, 142.0),\n",
       " 'AT3G07970_0': (143.0, 143.0),\n",
       " 'OG0010705_0': (143.0, 143.0),\n",
       " 'AT3G07980_0': (144.0, 144.0),\n",
       " 'OG0016400_0': (144.0, 144.0),\n",
       " 'AT3G07990_0': (145.0, 145.0),\n",
       " 'OG0010706_0': (145.0, 145.0),\n",
       " 'AT3G08000_0': (146.0, 146.0),\n",
       " 'OG0005110_0': (146.0, 146.0),\n",
       " 'AT3G08010_0': (147.0, 147.0),\n",
       " 'OG0010707_0': (147.0, 147.0),\n",
       " 'AT3G08020_0': (148.0, 148.0),\n",
       " 'OG0010708_0': (148.0, 148.0),\n",
       " 'AT3G08030_0': (149.0, 149.0),\n",
       " 'OG0010709_0': (149.0, 149.0),\n",
       " 'AT3G08040_0': (150.0, 150.0),\n",
       " 'OG0010710_0': (150.0, 150.0),\n",
       " 'AT3G08490_0': (151.0, 151.0),\n",
       " 'OG0010711_0': (151.0, 151.0),\n",
       " 'AT3G08500_0': (152.0, 152.0),\n",
       " 'OG0010712_0': (152.0, 152.0),\n",
       " 'AT3G08505_0': (153.0, 153.0),\n",
       " 'OG0010713_0': (153.0, 153.0),\n",
       " 'AT3G08510_0': (154.0, 154.0),\n",
       " 'OG0010714_0': (154.0, 154.0),\n",
       " 'AT3G08550_0': (155.0, 155.0),\n",
       " 'OG0010715_0': (155.0, 155.0),\n",
       " 'AT3G08560_0': (156.0, 156.0),\n",
       " 'OG0010716_0': (156.0, 156.0),\n",
       " 'AT3G08570_0': (157.0, 157.0),\n",
       " 'OG0010717_0': (157.0, 157.0),\n",
       " 'AT3G08580_0': (158.0, 158.0),\n",
       " 'OG0018489_0': (158.0, 158.0),\n",
       " 'AT3G08590_0': (159.0, 159.0),\n",
       " 'OG0010718_0': (159.0, 159.0),\n",
       " 'AT3G08600_0': (160.0, 160.0),\n",
       " 'OG0010719_0': (160.0, 160.0),\n",
       " 'AT3G08610_0': (161.0, 161.0),\n",
       " 'OG0010720_0': (161.0, 161.0),\n",
       " 'AT3G08620_0': (162.0, 162.0),\n",
       " 'OG0010721_0': (162.0, 162.0),\n",
       " 'AT3G08630_0': (163.0, 163.0),\n",
       " 'OG0010722_0': (163.0, 163.0),\n",
       " 'AT3G08640_0': (164.0, 164.0),\n",
       " 'OG0003380_0': (164.0, 164.0),\n",
       " 'AT3G08650_0': (165.0, 165.0),\n",
       " 'OG0005111_0': (165.0, 165.0),\n",
       " 'AT3G08660_0': (166.0, 166.0),\n",
       " 'OG0005112_0': (166.0, 166.0),\n",
       " 'AT3G08670_0': (167.0, 167.0),\n",
       " 'OG0019497_0': (167.0, 167.0),\n",
       " 'AT3G08680_0': (168.0, 168.0),\n",
       " 'OG0005113_0': (168.0, 168.0),\n",
       " 'AT3G08760_0': (169.0, 169.0),\n",
       " 'OG0005114_0': (169.0, 169.0),\n",
       " 'AT3G08770_0': (170.0, 170.0),\n",
       " 'OG0005115_0': (170.0, 170.0),\n",
       " 'AT3G08890_0': (171.0, 171.0),\n",
       " 'OG0010726_0': (171.0, 171.0),\n",
       " 'AT3G08900_0': (172.0, 172.0),\n",
       " 'OG0016403_0': (172.0, 172.0),\n",
       " 'AT3G08910_0': (173.0, 173.0),\n",
       " 'OG0010727_0': (173.0, 173.0),\n",
       " 'AT3G08920_0': (174.0, 174.0),\n",
       " 'OG0019499_0': (174.0, 174.0),\n",
       " 'AT3G08950_0': (175.0, 175.0),\n",
       " 'OG0010728_0': (175.0, 175.0),\n",
       " 'AT3G08960_0': (176.0, 176.0),\n",
       " 'OG0010729_0': (176.0, 176.0),\n",
       " 'AT3G08970_0': (177.0, 177.0),\n",
       " 'OG0010730_0': (177.0, 177.0),\n",
       " 'AT3G08980_0': (178.0, 178.0),\n",
       " 'OG0016405_0': (178.0, 178.0),\n",
       " 'AT3G08990_0': (179.0, 179.0),\n",
       " 'OG0010731_0': (179.0, 179.0),\n",
       " 'AT3G09000_0': (180.0, 180.0),\n",
       " 'OG0018491_0': (180.0, 180.0),\n",
       " 'AT3G09020_0': (181.0, 181.0),\n",
       " 'OG0010732_0': (181.0, 181.0),\n",
       " 'AT3G09030_0': (182.0, 182.0),\n",
       " 'OG0010733_0': (182.0, 182.0),\n",
       " 'AT3G09032_0': (183.0, 183.0),\n",
       " 'OG0020162_0': (183.0, 183.0),\n",
       " 'AT3G09060_0': (184.0, 184.0),\n",
       " 'OG0010735_0': (184.0, 184.0),\n",
       " 'AT3G09070_0': (185.0, 185.0),\n",
       " 'OG0005116_0': (185.0, 185.0),\n",
       " 'AT3G09090_0': (186.0, 186.0),\n",
       " 'OG0010737_0': (186.0, 186.0),\n",
       " 'AT3G09100_0': (187.0, 187.0),\n",
       " 'OG0019500_0': (187.0, 187.0),\n",
       " 'AT3G09110_0': (188.0, 188.0),\n",
       " 'OG0010738_0': (188.0, 188.0),\n",
       " 'AT3G09140_0': (189.0, 189.0),\n",
       " 'OG0010739_0': (189.0, 189.0),\n",
       " 'AT3G09150_0': (190.0, 190.0),\n",
       " 'OG0010740_0': (190.0, 190.0),\n",
       " 'AT3G09210_0': (191.0, 191.0),\n",
       " 'OG0018493_0': (191.0, 191.0),\n",
       " 'AT3G09220_0': (192.0, 192.0),\n",
       " 'OG0010743_0': (192.0, 192.0),\n",
       " 'AT3G09230_0': (193.0, 193.0),\n",
       " 'OG0010744_0': (193.0, 193.0),\n",
       " 'AT3G09240_0': (194.0, 194.0),\n",
       " 'OG0002245_0': (194.0, 194.0),\n",
       " 'AT3G09250_0': (195.0, 195.0),\n",
       " 'OG0010745_0': (195.0, 195.0),\n",
       " 'AT3G09270_0': (196.0, 196.0),\n",
       " 'OG0016407_0': (196.0, 196.0),\n",
       " 'AT3G09280_0': (197.0, 197.0),\n",
       " 'OG0016408_0': (197.0, 197.0),\n",
       " 'AT3G09290_0': (198.0, 198.0),\n",
       " 'OG0016409_0': (198.0, 198.0),\n",
       " 'AT3G09300_0': (199.0, 199.0),\n",
       " 'OG0016410_0': (199.0, 199.0),\n",
       " 'AT3G09310_0': (200.0, 200.0),\n",
       " 'OG0016411_0': (200.0, 200.0),\n",
       " 'AT3G09320_0': (201.0, 201.0),\n",
       " 'OG0005117_0': (201.0, 201.0),\n",
       " 'AT3G09360_0': (202.0, 202.0),\n",
       " 'OG0020672_0': (202.0, 202.0),\n",
       " 'AT3G09370_0': (203.0, 203.0),\n",
       " 'OG0010747_0': (203.0, 203.0),\n",
       " 'AT3G09390_0': (204.0, 204.0),\n",
       " 'OG0018494_0': (204.0, 204.0),\n",
       " 'AT3G09400_0': (205.0, 205.0),\n",
       " 'OG0005118_0': (205.0, 205.0),\n",
       " 'AT3G09540_0': (206.0, 206.0),\n",
       " 'OG0010751_0': (206.0, 206.0),\n",
       " 'AT3G09550_0': (207.0, 207.0),\n",
       " 'OG0005119_0': (207.0, 207.0),\n",
       " 'AT3G09560_0': (208.0, 208.0),\n",
       " 'OG0010752_0': (208.0, 208.0),\n",
       " 'AT3G09600_0': (209.0, 209.0),\n",
       " 'OG0003382_0': (209.0, 209.0),\n",
       " 'AT3G09620_0': (210.0, 210.0),\n",
       " 'OG0005121_0': (210.0, 210.0),\n",
       " 'AT3G09650_0': (211.0, 211.0),\n",
       " 'OG0010753_0': (211.0, 211.0),\n",
       " 'AT3G09660_0': (212.0, 212.0),\n",
       " 'OG0010754_0': (212.0, 212.0),\n",
       " 'AT3G09670_0': (213.0, 213.0),\n",
       " 'OG0010755_0': (213.0, 213.0),\n",
       " 'AT3G09690_0': (214.0, 214.0),\n",
       " 'OG0005122_0': (214.0, 214.0),\n",
       " 'AT3G09700_0': (215.0, 215.0),\n",
       " 'OG0019501_0': (215.0, 215.0),\n",
       " 'AT3G09710_0': (216.0, 216.0),\n",
       " 'OG0018499_0': (216.0, 216.0),\n",
       " 'AT3G09720_0': (217.0, 217.0),\n",
       " 'OG0003383_0': (217.0, 217.0),\n",
       " 'AT3G09730_0': (218.0, 218.0),\n",
       " 'OG0005123_0': (218.0, 218.0),\n",
       " 'AT3G09760_0': (219.0, 219.0),\n",
       " 'OG0010757_0': (219.0, 219.0),\n",
       " 'AT3G09770_0': (220.0, 220.0),\n",
       " 'OG0020164_0': (220.0, 220.0),\n",
       " 'AT3G09780_0': (221.0, 221.0),\n",
       " 'OG0005124_0': (221.0, 221.0),\n",
       " 'AT3G09800_0': (222.0, 222.0),\n",
       " 'OG0005125_0': (222.0, 222.0),\n",
       " 'AT3G09810_0': (223.0, 223.0),\n",
       " 'OG0003384_0': (223.0, 223.0),\n",
       " 'AT3G09850_0': (224.0, 224.0),\n",
       " 'OG0010758_0': (224.0, 224.0),\n",
       " 'AT3G09860_0': (225.0, 225.0),\n",
       " 'OG0010759_0': (225.0, 225.0),\n",
       " 'AT3G09925_0': (226.0, 226.0),\n",
       " 'OG0016416_0': (226.0, 226.0),\n",
       " 'AT3G09930_0': (227.0, 227.0),\n",
       " 'OG0010761_0': (227.0, 227.0),\n",
       " 'AT3G09940_0': (228.0, 228.0),\n",
       " 'OG0010762_0': (228.0, 228.0),\n",
       " 'AT3G09950_0': (229.0, 229.0),\n",
       " 'OG0010763_0': (229.0, 229.0),\n",
       " 'AT3G09960_0': (230.0, 230.0),\n",
       " 'OG0018500_0': (230.0, 230.0),\n",
       " 'AT3G09970_0': (231.0, 231.0),\n",
       " 'OG0010764_0': (231.0, 231.0),\n",
       " 'AT3G09980_0': (232.0, 232.0),\n",
       " 'OG0010765_0': (232.0, 232.0),\n",
       " 'AT3G09990_0': (233.0, 233.0),\n",
       " 'OG0010766_0': (233.0, 233.0),\n",
       " 'AT3G10000_0': (234.0, 234.0),\n",
       " 'OG0010767_0': (234.0, 234.0),\n",
       " 'AT3G10020_0': (235.0, 235.0),\n",
       " 'OG0016417_0': (235.0, 235.0),\n",
       " 'AT3G10030_0': (236.0, 236.0),\n",
       " 'OG0016418_0': (236.0, 236.0),\n",
       " 'AT3G10050_0': (237.0, 237.0),\n",
       " 'OG0010768_0': (237.0, 237.0),\n",
       " 'AT3G10060_0': (238.0, 238.0),\n",
       " 'OG0010769_0': (238.0, 238.0),\n",
       " 'AT3G10070_0': (239.0, 239.0),\n",
       " 'OG0010770_0': (239.0, 239.0),\n",
       " 'AT3G10160_0': (240.0, 240.0),\n",
       " 'OG0010773_0': (240.0, 240.0),\n",
       " 'AT3G10185_0': (241.0, 241.0),\n",
       " 'OG0019504_0': (241.0, 241.0),\n",
       " 'AT3G10180_0': (242.0, 242.0),\n",
       " 'OG0010774_0': (242.0, 242.0),\n",
       " 'AT3G10190_0': (243.0, 243.0),\n",
       " 'OG0010775_0': (243.0, 243.0),\n",
       " 'AT3G10210_0': (244.0, 244.0),\n",
       " 'OG0010776_0': (244.0, 244.0),\n",
       " 'AT3G10220_0': (245.0, 245.0),\n",
       " 'OG0010777_0': (245.0, 245.0),\n",
       " 'AT3G10250_0': (246.0, 246.0),\n",
       " 'OG0005126_0': (246.0, 246.0),\n",
       " 'AT3G10260_0': (247.0, 247.0),\n",
       " 'OG0016419_0': (247.0, 247.0),\n",
       " 'AT3G10270_0': (248.0, 248.0),\n",
       " 'OG0010779_0': (248.0, 248.0),\n",
       " 'AT3G10300_0': (249.0, 249.0),\n",
       " 'OG0010780_0': (249.0, 249.0),\n",
       " 'AT3G10310_0': (250.0, 250.0),\n",
       " 'OG0010781_0': (250.0, 250.0),\n",
       " 'AT3G10320_0': (251.0, 251.0),\n",
       " 'OG0010782_0': (251.0, 251.0),\n",
       " 'AT3G10330_0': (252.0, 252.0),\n",
       " 'OG0010783_0': (252.0, 252.0),\n",
       " 'AT3G10340_0': (253.0, 253.0),\n",
       " 'OG0003385_0': (253.0, 253.0),\n",
       " 'AT3G10350_0': (254.0, 254.0),\n",
       " 'OG0010784_0': (254.0, 254.0),\n",
       " 'AT3G10360_0': (255.0, 255.0),\n",
       " 'OG0010785_0': (255.0, 255.0),\n",
       " 'AT3G10370_0': (256.0, 256.0),\n",
       " 'OG0010786_0': (256.0, 256.0),\n",
       " 'AT3G10380_0': (257.0, 257.0),\n",
       " 'OG0010787_0': (257.0, 257.0),\n",
       " 'AT3G10390_0': (258.0, 258.0),\n",
       " 'OG0010788_0': (258.0, 258.0),\n",
       " 'AT3G10410_0': (259.0, 259.0),\n",
       " 'OG0010791_0': (259.0, 259.0),\n",
       " 'AT3G10420_0': (260.0, 260.0),\n",
       " 'OG0010792_0': (260.0, 260.0),\n",
       " 'AT3G10430_0': (261.0, 261.0),\n",
       " 'OG0020165_0': (261.0, 261.0),\n",
       " 'AT3G10530_0': (262.0, 262.0),\n",
       " 'OG0010796_0': (262.0, 262.0),\n",
       " 'AT3G10540_0': (263.0, 263.0),\n",
       " 'OG0010797_0': (263.0, 263.0),\n",
       " 'AT3G10550_0': (264.0, 264.0),\n",
       " 'OG0005127_0': (264.0, 264.0),\n",
       " 'AT3G10570_0': (265.0, 265.0),\n",
       " 'OG0016421_0': (265.0, 265.0),\n",
       " 'AT3G10572_0': (266.0, 266.0),\n",
       " 'OG0003386_0': (266.0, 266.0),\n",
       " 'AT3G10580_0': (267.0, 267.0),\n",
       " 'OG0005128_0': (267.0, 267.0),\n",
       " 'AT3G10590_0': (268.0, 268.0),\n",
       " 'OG0020166_0': (268.0, 268.0),\n",
       " 'AT3G10595_0': (269.0, 269.0),\n",
       " 'OG0003387_0': (269.0, 269.0),\n",
       " 'AT3G10600_0': (270.0, 270.0),\n",
       " 'OG0010798_0': (270.0, 270.0),\n",
       " 'AT3G10610_0': (271.0, 271.0),\n",
       " 'OG0010799_0': (271.0, 271.0),\n",
       " 'AT3G10620_0': (272.0, 272.0),\n",
       " 'OG0010800_0': (272.0, 272.0),\n",
       " 'AT3G10630_0': (273.0, 273.0),\n",
       " 'OG0005129_0': (273.0, 273.0),\n",
       " 'AT3G10640_0': (274.0, 274.0),\n",
       " 'OG0018504_0': (274.0, 274.0),\n",
       " 'AT3G10650_0': (275.0, 275.0),\n",
       " 'OG0010801_0': (275.0, 275.0),\n",
       " 'AT3G10670_0': (276.0, 276.0),\n",
       " 'OG0005130_0': (276.0, 276.0),\n",
       " 'AT3G10680_0': (277.0, 277.0),\n",
       " 'OG0010802_0': (277.0, 277.0),\n",
       " 'AT3G10690_0': (278.0, 278.0),\n",
       " 'OG0005131_0': (278.0, 278.0),\n",
       " 'AT3G10700_0': (279.0, 279.0),\n",
       " 'OG0010803_0': (279.0, 279.0),\n",
       " 'AT3G10730_0': (280.0, 280.0),\n",
       " 'OG0016422_0': (280.0, 280.0),\n",
       " 'AT3G10740_0': (281.0, 281.0),\n",
       " 'OG0010804_0': (281.0, 281.0),\n",
       " 'AT3G10760_0': (282.0, 282.0),\n",
       " 'OG0010805_0': (282.0, 282.0),\n",
       " 'AT3G10770_0': (283.0, 283.0),\n",
       " 'OG0003388_0': (283.0, 283.0),\n",
       " 'AT3G10780_0': (284.0, 284.0),\n",
       " 'OG0010806_0': (284.0, 284.0),\n",
       " 'AT3G10800_0': (285.0, 285.0),\n",
       " 'OG0005132_0': (285.0, 285.0),\n",
       " 'AT3G10810_0': (286.0, 286.0),\n",
       " 'OG0002399_0': (286.0, 286.0),\n",
       " 'AT3G10840_0': (287.0, 287.0),\n",
       " 'OG0003390_0': (287.0, 287.0),\n",
       " 'AT3G10845_0': (288.0, 288.0),\n",
       " 'OG0020674_0': (288.0, 288.0),\n",
       " 'AT3G10850_0': (289.0, 289.0),\n",
       " 'OG0016424_0': (289.0, 289.0),\n",
       " 'AT3G10870_0': (290.0, 290.0),\n",
       " 'OG0005133_0': (290.0, 290.0),\n",
       " 'AT3G10880_0': (291.0, 291.0),\n",
       " 'OG0005134_0': (291.0, 291.0),\n",
       " 'AT3G10910_0': (292.0, 292.0),\n",
       " 'OG0010807_0': (292.0, 292.0),\n",
       " 'AT3G10915_0': (293.0, 293.0),\n",
       " 'OG0010808_0': (293.0, 293.0),\n",
       " 'AT3G10920_0': (294.0, 294.0),\n",
       " 'OG0005135_0': (294.0, 294.0),\n",
       " 'AT3G10960_0': (295.0, 295.0),\n",
       " 'OG0005136_0': (295.0, 295.0),\n",
       " 'AT3G10970_0': (296.0, 296.0),\n",
       " 'OG0005137_0': (296.0, 296.0),\n",
       " 'AT3G11020_0': (297.0, 297.0),\n",
       " 'OG0010811_0': (297.0, 297.0),\n",
       " 'AT3G11030_0': (298.0, 298.0),\n",
       " 'OG0010812_0': (298.0, 298.0),\n",
       " 'AT3G11040_0': (299.0, 299.0),\n",
       " 'OG0010813_0': (299.0, 299.0),\n",
       " 'AT3G11090_0': (300.0, 300.0),\n",
       " 'OG0010814_0': (300.0, 300.0),\n",
       " 'AT3G11100_0': (301.0, 301.0),\n",
       " 'OG0016429_0': (301.0, 301.0),\n",
       " 'AT3G11110_0': (302.0, 302.0),\n",
       " 'OG0010815_0': (302.0, 302.0),\n",
       " 'AT3G11150_0': (303.0, 303.0),\n",
       " 'OG0010816_0': (303.0, 303.0),\n",
       " 'AT3G11160_0': (304.0, 304.0),\n",
       " 'OG0020169_0': (304.0, 304.0),\n",
       " 'AT3G11170_0': (305.0, 305.0),\n",
       " 'OG0010817_0': (305.0, 305.0),\n",
       " 'AT3G11180_0': (306.0, 306.0),\n",
       " 'OG0010818_0': (306.0, 306.0),\n",
       " 'AT3G11200_0': (307.0, 307.0),\n",
       " 'OG0010819_0': (307.0, 307.0),\n",
       " 'AT3G11210_0': (308.0, 308.0),\n",
       " 'OG0016430_0': (308.0, 308.0),\n",
       " 'AT3G11220_0': (309.0, 309.0),\n",
       " 'OG0010820_0': (309.0, 309.0),\n",
       " 'AT3G11230_0': (310.0, 310.0),\n",
       " 'OG0010821_0': (310.0, 310.0),\n",
       " 'AT3G11240_0': (311.0, 311.0),\n",
       " 'OG0010822_0': (311.0, 311.0),\n",
       " 'AT3G11280_0': (312.0, 312.0),\n",
       " 'OG0010823_0': (312.0, 312.0),\n",
       " 'AT3G11290_0': (313.0, 313.0),\n",
       " 'OG0010824_0': (313.0, 313.0),\n",
       " 'AT3G11397_0': (314.0, 314.0),\n",
       " 'OG0010827_0': (314.0, 314.0),\n",
       " 'AT3G11400_0': (315.0, 315.0),\n",
       " 'OG0002246_0': (315.0, 315.0),\n",
       " 'AT3G11410_0': (316.0, 316.0),\n",
       " 'OG0010828_0': (316.0, 316.0),\n",
       " 'AT3G11420_0': (317.0, 317.0),\n",
       " 'OG0005138_0': (317.0, 317.0),\n",
       " 'AT3G11430_0': (318.0, 318.0),\n",
       " 'OG0016431_0': (318.0, 318.0),\n",
       " 'AT3G11440_0': (319.0, 319.0),\n",
       " 'OG0005139_0': (319.0, 319.0),\n",
       " 'AT3G11450_0': (320.0, 320.0),\n",
       " 'OG0005140_0': (320.0, 320.0),\n",
       " 'AT3G11460_0': (321.0, 321.0),\n",
       " 'OG0005141_0': (321.0, 321.0),\n",
       " 'AT3G11520_0': (322.0, 322.0),\n",
       " 'OG0016432_0': (322.0, 322.0),\n",
       " 'AT3G11530_0': (323.0, 323.0),\n",
       " 'OG0016433_0': (323.0, 323.0),\n",
       " 'AT3G11540_0': (324.0, 324.0),\n",
       " 'OG0005142_0': (324.0, 324.0),\n",
       " 'AT3G11550_0': (325.0, 325.0),\n",
       " 'OG0010829_0': (325.0, 325.0),\n",
       " 'AT3G11580_0': (326.0, 326.0),\n",
       " 'OG0018506_0': (326.0, 326.0),\n",
       " 'AT3G11590_0': (327.0, 327.0),\n",
       " 'OG0005144_0': (327.0, 327.0),\n",
       " 'AT3G11600_0': (328.0, 328.0),\n",
       " 'OG0010830_0': (328.0, 328.0),\n",
       " 'AT3G11620_0': (329.0, 329.0),\n",
       " 'OG0002713_0': (329.0, 329.0),\n",
       " 'AT3G11630_0': (330.0, 330.0),\n",
       " 'OG0003392_0': (330.0, 330.0),\n",
       " 'AT3G11640_0': (331.0, 331.0),\n",
       " 'OG0005145_0': (331.0, 331.0),\n",
       " 'AT3G11650_0': (332.0, 332.0),\n",
       " 'OG0003393_0': (332.0, 332.0),\n",
       " 'AT3G11670_0': (333.0, 333.0),\n",
       " 'OG0003394_0': (333.0, 333.0),\n",
       " 'AT3G11680_0': (334.0, 334.0),\n",
       " 'OG0005146_0': (334.0, 334.0),\n",
       " 'AT3G11690_0': (335.0, 335.0),\n",
       " 'OG0005147_0': (335.0, 335.0),\n",
       " 'AT3G11700_0': (336.0, 336.0),\n",
       " 'OG0010831_0': (336.0, 336.0),\n",
       " 'AT3G11710_0': (337.0, 337.0),\n",
       " 'OG0002714_0': (337.0, 337.0),\n",
       " 'AT3G11720_0': (338.0, 338.0),\n",
       " 'OG0018507_0': (338.0, 338.0),\n",
       " 'AT3G11730_0': (339.0, 339.0),\n",
       " 'OG0018508_0': (339.0, 339.0),\n",
       " 'AT3G11745_0': (340.0, 340.0),\n",
       " 'OG0019506_0': (340.0, 340.0),\n",
       " 'AT3G11750_0': (341.0, 341.0),\n",
       " 'OG0010832_0': (341.0, 341.0),\n",
       " 'AT3G11760_0': (342.0, 342.0),\n",
       " 'OG0010833_0': (342.0, 342.0),\n",
       " 'AT3G11770_0': (343.0, 343.0),\n",
       " 'OG0010834_0': (343.0, 343.0),\n",
       " 'AT3G11773_0': (344.0, 344.0),\n",
       " 'OG0010835_0': (344.0, 344.0),\n",
       " 'AT3G11800_0': (345.0, 345.0),\n",
       " 'OG0010836_0': (345.0, 345.0),\n",
       " 'AT3G11810_0': (346.0, 346.0),\n",
       " 'OG0010837_0': (346.0, 346.0),\n",
       " 'AT3G11820_0': (347.0, 347.0),\n",
       " 'OG0010838_0': (347.0, 347.0),\n",
       " 'AT3G11830_0': (348.0, 348.0),\n",
       " 'OG0010839_0': (348.0, 348.0),\n",
       " 'AT3G11840_0': (349.0, 349.0),\n",
       " 'OG0010840_0': (349.0, 349.0),\n",
       " 'AT3G11850_0': (350.0, 350.0),\n",
       " 'OG0010841_0': (350.0, 350.0),\n",
       " 'AT3G11890_0': (351.0, 351.0),\n",
       " 'OG0010842_0': (351.0, 351.0),\n",
       " 'AT3G11900_0': (352.0, 352.0),\n",
       " 'OG0010843_0': (352.0, 352.0),\n",
       " 'AT3G11920_0': (353.0, 353.0),\n",
       " 'OG0005148_0': (353.0, 353.0),\n",
       " 'AT3G11930_0': (354.0, 354.0),\n",
       " 'OG0005149_0': (354.0, 354.0),\n",
       " 'AT3G11945_0': (355.0, 355.0),\n",
       " 'OG0010844_0': (355.0, 355.0),\n",
       " 'AT3G11950_0': (356.0, 356.0),\n",
       " 'OG0010845_0': (356.0, 356.0),\n",
       " 'AT3G12010_0': (357.0, 357.0),\n",
       " 'OG0010847_0': (357.0, 357.0),\n",
       " 'AT3G12020_0': (358.0, 358.0),\n",
       " 'OG0010848_0': (358.0, 358.0),\n",
       " 'AT3G12040_0': (359.0, 359.0),\n",
       " 'OG0010849_0': (359.0, 359.0),\n",
       " 'AT3G12050_0': (360.0, 360.0),\n",
       " 'OG0010850_0': (360.0, 360.0),\n",
       " 'AT3G12080_0': (361.0, 361.0),\n",
       " 'OG0005152_0': (361.0, 361.0),\n",
       " 'AT3G12090_0': (362.0, 362.0),\n",
       " 'OG0010851_0': (362.0, 362.0),\n",
       " 'AT3G12100_0': (363.0, 363.0),\n",
       " 'OG0005153_0': (363.0, 363.0),\n",
       " 'AT3G12260_0': (364.0, 364.0),\n",
       " 'OG0010858_0': (364.0, 364.0),\n",
       " 'AT3G12270_0': (365.0, 365.0),\n",
       " 'OG0010859_0': (365.0, 365.0),\n",
       " 'AT3G12280_0': (366.0, 366.0),\n",
       " 'OG0002401_0': (366.0, 366.0),\n",
       " 'AT3G12290_0': (367.0, 367.0),\n",
       " 'OG0010860_0': (367.0, 367.0),\n",
       " 'AT3G12300_0': (368.0, 368.0),\n",
       " 'OG0010861_0': (368.0, 368.0),\n",
       " 'AT3G12320_0': (369.0, 369.0),\n",
       " 'OG0010862_0': (369.0, 369.0),\n",
       " 'AT3G12340_0': (370.0, 370.0),\n",
       " 'OG0005154_0': (370.0, 370.0),\n",
       " 'AT3G12345_0': (371.0, 371.0),\n",
       " 'OG0010863_0': (371.0, 371.0),\n",
       " 'AT3G12350_0': (372.0, 372.0),\n",
       " 'OG0010864_0': (372.0, 372.0),\n",
       " 'AT3G12360_0': (373.0, 373.0),\n",
       " 'OG0010865_0': (373.0, 373.0),\n",
       " 'AT3G12370_0': (374.0, 374.0),\n",
       " 'OG0016435_0': (374.0, 374.0),\n",
       " 'AT3G12380_0': (375.0, 375.0),\n",
       " 'OG0003395_0': (375.0, 375.0),\n",
       " 'AT3G12390_0': (376.0, 376.0),\n",
       " 'OG0010866_0': (376.0, 376.0),\n",
       " 'AT3G12400_0': (377.0, 377.0),\n",
       " 'OG0018511_0': (377.0, 377.0),\n",
       " 'AT3G12480_0': (378.0, 378.0),\n",
       " 'OG0010867_0': (378.0, 378.0),\n",
       " 'AT3G12490_0': (379.0, 379.0),\n",
       " 'OG0016436_0': (379.0, 379.0),\n",
       " 'AT3G12500_0': (380.0, 380.0),\n",
       " 'OG0003396_0': (380.0, 380.0),\n",
       " 'AT3G12520_0': (381.0, 381.0),\n",
       " 'OG0003397_0': (381.0, 381.0),\n",
       " 'AT3G12530_0': (382.0, 382.0),\n",
       " 'OG0016437_0': (382.0, 382.0),\n",
       " 'AT3G12540_0': (383.0, 383.0),\n",
       " 'OG0016438_0': (383.0, 383.0),\n",
       " 'AT3G12550_0': (384.0, 384.0),\n",
       " 'OG0016439_0': (384.0, 384.0),\n",
       " 'AT3G12560_0': (385.0, 385.0),\n",
       " 'OG0010868_0': (385.0, 385.0),\n",
       " 'AT3G12570_0': (386.0, 386.0),\n",
       " 'OG0010869_0': (386.0, 386.0),\n",
       " 'AT3G12590_0': (387.0, 387.0),\n",
       " 'OG0010870_0': (387.0, 387.0),\n",
       " 'AT3G12600_0': (388.0, 388.0),\n",
       " 'OG0010871_0': (388.0, 388.0),\n",
       " 'AT3G12610_0': (389.0, 389.0),\n",
       " 'OG0010872_0': (389.0, 389.0),\n",
       " 'AT3G12620_0': (390.0, 390.0),\n",
       " 'OG0010873_0': (390.0, 390.0),\n",
       " 'AT3G12630_0': (391.0, 391.0),\n",
       " 'OG0010874_0': (391.0, 391.0),\n",
       " 'AT3G12640_0': (392.0, 392.0),\n",
       " 'OG0010875_0': (392.0, 392.0),\n",
       " 'AT3G12650_0': (393.0, 393.0),\n",
       " 'OG0010876_0': (393.0, 393.0),\n",
       " 'AT3G12680_0': (394.0, 394.0),\n",
       " 'OG0010879_0': (394.0, 394.0),\n",
       " 'AT3G12685_0': (395.0, 395.0),\n",
       " 'OG0019507_0': (395.0, 395.0),\n",
       " 'AT3G12700_0': (396.0, 396.0),\n",
       " 'OG0005155_0': (396.0, 396.0),\n",
       " 'AT3G12710_0': (397.0, 397.0),\n",
       " 'OG0016440_0': (397.0, 397.0),\n",
       " 'AT3G12720_0': (398.0, 398.0),\n",
       " 'OG0016441_0': (398.0, 398.0),\n",
       " 'AT3G12730_0': (399.0, 399.0),\n",
       " 'OG0010880_0': (399.0, 399.0),\n",
       " 'AT3G12740_0': (400.0, 400.0),\n",
       " 'OG0010881_0': (400.0, 400.0),\n",
       " 'AT3G12750_0': (401.0, 401.0),\n",
       " 'OG0010882_0': (401.0, 401.0),\n",
       " 'AT3G12760_0': (402.0, 402.0),\n",
       " 'OG0010883_0': (402.0, 402.0),\n",
       " 'AT3G12770_0': (403.0, 403.0),\n",
       " 'OG0018512_0': (403.0, 403.0),\n",
       " 'AT3G12775_0': (404.0, 404.0),\n",
       " 'OG0010884_0': (404.0, 404.0),\n",
       " 'AT3G12800_0': (405.0, 405.0),\n",
       " 'OG0016442_0': (405.0, 405.0),\n",
       " 'AT3G12810_0': (406.0, 406.0),\n",
       " 'OG0005156_0': (406.0, 406.0),\n",
       " 'AT3G12820_0': (407.0, 407.0),\n",
       " 'OG0010885_0': (407.0, 407.0),\n",
       " 'AT3G12870_0': (408.0, 408.0),\n",
       " 'OG0010887_0': (408.0, 408.0),\n",
       " 'AT3G12880_0': (409.0, 409.0),\n",
       " 'OG0010888_0': (409.0, 409.0),\n",
       " 'AT3G12890_0': (410.0, 410.0),\n",
       " 'OG0010889_0': (410.0, 410.0),\n",
       " 'AT3G12900_0': (411.0, 411.0),\n",
       " 'OG0016443_0': (411.0, 411.0),\n",
       " 'AT3G12910_0': (412.0, 412.0),\n",
       " 'OG0016444_0': (412.0, 412.0),\n",
       " 'AT3G12920_0': (413.0, 413.0),\n",
       " 'OG0010890_0': (413.0, 413.0),\n",
       " 'AT3G12930_0': (414.0, 414.0),\n",
       " 'OG0010891_0': (414.0, 414.0),\n",
       " 'AT3G12990_0': (415.0, 415.0),\n",
       " 'OG0010894_0': (415.0, 415.0),\n",
       " 'AT3G13000_0': (416.0, 416.0),\n",
       " 'OG0010895_0': (416.0, 416.0),\n",
       " 'AT3G13010_0': (417.0, 417.0),\n",
       " 'OG0018514_0': (417.0, 417.0),\n",
       " 'AT3G13020_0': (418.0, 418.0),\n",
       " 'OG0010896_0': (418.0, 418.0),\n",
       " 'AT3G13030_0': (419.0, 419.0),\n",
       " 'OG0010897_0': (419.0, 419.0),\n",
       " 'AT3G13040_0': (420.0, 420.0),\n",
       " 'OG0005157_0': (420.0, 420.0),\n",
       " 'AT3G13050_0': (421.0, 421.0),\n",
       " 'OG0016446_0': (421.0, 421.0),\n",
       " 'AT3G13060_0': (422.0, 422.0),\n",
       " 'OG0002402_0': (422.0, 422.0),\n",
       " 'AT3G13120_0': (423.0, 423.0),\n",
       " 'OG0010898_0': (423.0, 423.0),\n",
       " 'AT3G13130_0': (424.0, 424.0),\n",
       " 'OG0018515_0': (424.0, 424.0),\n",
       " 'AT3G13140_0': (425.0, 425.0),\n",
       " 'OG0010899_0': (425.0, 425.0),\n",
       " 'AT3G13150_0': (426.0, 426.0),\n",
       " 'OG0016447_0': (426.0, 426.0),\n",
       " 'AT3G13160_0': (427.0, 427.0),\n",
       " 'OG0018516_0': (427.0, 427.0),\n",
       " 'AT3G13170_0': (428.0, 428.0),\n",
       " 'OG0010900_0': (428.0, 428.0),\n",
       " 'AT3G13175_0': (429.0, 429.0),\n",
       " 'OG0016448_0': (429.0, 429.0),\n",
       " 'AT3G13180_0': (430.0, 430.0),\n",
       " 'OG0005158_0': (430.0, 430.0),\n",
       " 'AT3G13190_0': (431.0, 431.0),\n",
       " 'OG0010901_0': (431.0, 431.0),\n",
       " 'AT3G13200_0': (432.0, 432.0),\n",
       " 'OG0018517_0': (432.0, 432.0),\n",
       " 'AT3G13220_0': (433.0, 433.0),\n",
       " 'OG0005159_0': (433.0, 433.0),\n",
       " 'AT3G13222_0': (434.0, 434.0),\n",
       " 'OG0016449_0': (434.0, 434.0),\n",
       " 'AT3G13224_0': (435.0, 435.0),\n",
       " 'OG0010902_0': (435.0, 435.0),\n",
       " 'AT3G13227_0': (436.0, 436.0),\n",
       " 'OG0019509_0': (436.0, 436.0),\n",
       " 'AT3G13225_0': (437.0, 437.0),\n",
       " 'OG0010903_0': (437.0, 437.0),\n",
       " 'AT3G13226_0': (438.0, 438.0),\n",
       " 'OG0010904_0': (438.0, 438.0),\n",
       " 'AT3G13228_0': (439.0, 439.0),\n",
       " 'OG0002717_0': (439.0, 439.0),\n",
       " 'AT3G13229_0': (440.0, 440.0),\n",
       " 'OG0010905_0': (440.0, 440.0),\n",
       " 'AT3G13230_0': (441.0, 441.0),\n",
       " 'OG0003398_0': (441.0, 441.0),\n",
       " 'AT3G13235_0': (442.0, 442.0),\n",
       " 'OG0010906_0': (442.0, 442.0),\n",
       " 'AT3G13280_0': (443.0, 443.0),\n",
       " 'OG0019510_0': (443.0, 443.0),\n",
       " 'AT3G13290_0': (444.0, 444.0),\n",
       " 'OG0010907_0': (444.0, 444.0),\n",
       " 'AT3G13300_0': (445.0, 445.0),\n",
       " 'OG0010908_0': (445.0, 445.0),\n",
       " 'AT3G13310_0': (446.0, 446.0),\n",
       " 'OG0010909_0': (446.0, 446.0),\n",
       " 'AT3G13320_0': (447.0, 447.0),\n",
       " 'OG0005160_0': (447.0, 447.0),\n",
       " 'AT3G13330_0': (448.0, 448.0),\n",
       " 'OG0010910_0': (448.0, 448.0),\n",
       " 'AT3G13340_0': (449.0, 449.0),\n",
       " 'OG0005161_0': (449.0, 449.0),\n",
       " 'AT3G13350_0': (450.0, 450.0),\n",
       " 'OG0018518_0': (450.0, 450.0),\n",
       " 'AT3G13360_0': (451.0, 451.0),\n",
       " 'OG0010911_0': (451.0, 451.0),\n",
       " 'AT3G13410_0': (452.0, 452.0),\n",
       " 'OG0005163_0': (452.0, 452.0),\n",
       " 'AT3G13420_0': (453.0, 453.0),\n",
       " 'OG0010912_0': (453.0, 453.0),\n",
       " 'AT3G13430_0': (454.0, 454.0),\n",
       " 'OG0010913_0': (454.0, 454.0),\n",
       " 'AT3G13450_0': (455.0, 455.0),\n",
       " 'OG0019511_0': (455.0, 455.0),\n",
       " 'AT3G13460_0': (456.0, 456.0),\n",
       " 'OG0010915_0': (456.0, 456.0),\n",
       " 'AT3G13480_0': (457.0, 457.0),\n",
       " 'OG0010916_0': (457.0, 457.0),\n",
       " 'AT3G13490_0': (458.0, 458.0),\n",
       " 'OG0010917_0': (458.0, 458.0),\n",
       " 'AT3G13500_0': (459.0, 459.0),\n",
       " 'OG0019512_0': (459.0, 459.0),\n",
       " 'AT3G13520_0': (460.0, 460.0),\n",
       " 'OG0016451_0': (460.0, 460.0),\n",
       " 'AT3G13530_0': (461.0, 461.0),\n",
       " 'OG0016452_0': (461.0, 461.0),\n",
       " 'AT3G13540_0': (462.0, 462.0),\n",
       " 'OG0005164_0': (462.0, 462.0),\n",
       " 'AT3G13550_0': (463.0, 463.0),\n",
       " 'OG0002403_0': (463.0, 463.0),\n",
       " 'AT3G13560_0': (464.0, 464.0),\n",
       " 'OG0010918_0': (464.0, 464.0),\n",
       " 'AT3G13570_0': (465.0, 465.0),\n",
       " 'OG0003399_0': (465.0, 465.0),\n",
       " 'AT3G13580_0': (466.0, 466.0),\n",
       " 'OG0005165_0': (466.0, 466.0),\n",
       " 'AT3G13600_0': (467.0, 467.0),\n",
       " 'OG0003400_0': (467.0, 467.0),\n",
       " 'AT3G13610_0': (468.0, 468.0),\n",
       " 'OG0003401_0': (468.0, 468.0),\n",
       " 'AT3G13620_0': (469.0, 469.0),\n",
       " 'OG0016453_0': (469.0, 469.0),\n",
       " 'AT3G13690_0': (470.0, 470.0),\n",
       " 'OG0010923_0': (470.0, 470.0),\n",
       " 'AT3G13700_0': (471.0, 471.0),\n",
       " 'OG0005166_0': (471.0, 471.0),\n",
       " 'AT3G13710_0': (472.0, 472.0),\n",
       " 'OG0018520_0': (472.0, 472.0),\n",
       " 'AT3G13720_0': (473.0, 473.0),\n",
       " 'OG0016454_0': (473.0, 473.0),\n",
       " 'AT3G13730_0': (474.0, 474.0),\n",
       " 'OG0010924_0': (474.0, 474.0),\n",
       " 'AT3G13740_0': (475.0, 475.0),\n",
       " 'OG0010925_0': (475.0, 475.0),\n",
       " 'AT3G13770_0': (476.0, 476.0),\n",
       " 'OG0018521_0': (476.0, 476.0),\n",
       " 'AT3G13772_0': (477.0, 477.0),\n",
       " 'OG0010926_0': (477.0, 477.0),\n",
       " 'AT3G13780_0': (478.0, 478.0),\n",
       " 'OG0019513_0': (478.0, 478.0),\n",
       " 'AT3G13782_0': (479.0, 479.0),\n",
       " 'OG0010927_0': (479.0, 479.0),\n",
       " 'AT3G13784_0': (480.0, 480.0),\n",
       " 'OG0010928_0': (480.0, 480.0),\n",
       " 'AT3G13790_0': (481.0, 481.0),\n",
       " 'OG0005167_0': (481.0, 481.0),\n",
       " 'AT3G13800_0': (482.0, 482.0),\n",
       " 'OG0010929_0': (482.0, 482.0),\n",
       " 'AT3G13810_0': (483.0, 483.0),\n",
       " 'OG0010930_0': (483.0, 483.0),\n",
       " 'AT3G13820_0': (484.0, 484.0),\n",
       " 'OG0020679_0': (484.0, 484.0),\n",
       " 'AT3G13845_0': (485.0, 485.0),\n",
       " 'OG0010931_0': (485.0, 485.0),\n",
       " 'AT3G13850_0': (486.0, 486.0),\n",
       " 'OG0010932_0': (486.0, 486.0),\n",
       " 'AT3G13930_0': (487.0, 487.0),\n",
       " 'OG0010933_0': (487.0, 487.0),\n",
       " 'AT3G13940_0': (488.0, 488.0),\n",
       " 'OG0010934_0': (488.0, 488.0),\n",
       " 'AT3G13980_0': (489.0, 489.0),\n",
       " 'OG0010937_0': (489.0, 489.0),\n",
       " 'AT3G13990_0': (490.0, 490.0),\n",
       " 'OG0002086_0': (490.0, 490.0),\n",
       " 'AT3G14010_0': (491.0, 491.0),\n",
       " 'OG0010938_0': (491.0, 491.0),\n",
       " 'AT3G14020_0': (492.0, 492.0),\n",
       " 'OG0016456_0': (492.0, 492.0),\n",
       " 'AT3G14050_0': (493.0, 493.0),\n",
       " 'OG0016457_0': (493.0, 493.0),\n",
       " 'AT3G14060_0': (494.0, 494.0),\n",
       " 'OG0010939_0': (494.0, 494.0),\n",
       " 'AT3G14067_0': (495.0, 495.0),\n",
       " 'OG0018525_0': (495.0, 495.0),\n",
       " 'AT3G14070_0': (496.0, 496.0),\n",
       " 'OG0010940_0': (496.0, 496.0),\n",
       " 'AT3G14075_0': (497.0, 497.0),\n",
       " 'OG0016458_0': (497.0, 497.0),\n",
       " 'AT3G14080_0': (498.0, 498.0),\n",
       " 'OG0010941_0': (498.0, 498.0),\n",
       " 'AT3G14090_0': (499.0, 499.0),\n",
       " 'OG0010942_0': (499.0, 499.0),\n",
       " ...}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iset_get(red,'coregraph_v2_genes_Chr3_new.9981.Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
