pygengraph
================

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

## Install

Enter the directory of the library and enter:

`pip install .`

and for development use

`pip install -e .`

## How to use

``` python
from nbdev import nbdev_export
nbdev_export()
```

``` python
import os
import glob
import re
import time

from pygengraph.graph import GenomeGraph
from pygengraph.utils import pathFileToPathDict
from pygengraph.export import exportProject
```

# If redis database needs to be cleaned.

from redis import Redis

redisConn = Redis(host=‘redis’,port=6379)

redisConn.flushall()

redisConn.close()

del redisConn

``` python
import warnings
warnings.simplefilter('ignore')
warnings.filterwarnings('always',category=RuntimeWarning)
```

# Generating graphs

## Generating from annotation

### Preparing list of files

``` python
refdir = '/path/to/reference/'
annotationdir = '/path/to/annotation'
gfadir = '/path/to/graphs'
```

``` python
annotationFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}*.gff'))
pangenomeFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}*pangen.gff'))
# If you want to include sequences instead of simple notion of genes.
# It should also be converted to sequenceFileDict, see details in documentation for GenomeGraph Class constructor.
# sequenceFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}sequences{os.path.sep}*.fasta'))
refAnnotationFile = f'{refdir}{os.path.sep}reference.gff'
# If you want to include sequences instead of simple notion of genes
# refSequenceFile = f'{refdir}{os.path.sep}reference.fasta'
```

``` python
refdir = '../../1001G/annotations/freeze2.1/outgroups'
annotationdir = '../../1001G/annotations/freeze2.1'
gfadir = '../../1001G/annotations/graphs'
```

``` python
annotationFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}*.gff'))
# pangenomeFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}*pangen.gff'))
# If you want to include sequences instead of simple notion of genes.
# It should also be converted to sequenceFileDict, see details in documentation for GenomeGraph Class constructor.
# sequenceFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}sequences{os.path.sep}*.fasta'))
refAnnotationFile = f'{refdir}{os.path.sep}araport.gff'
# If you want to include sequences instead of simple notion of genes
# refSequenceFile = f'{refdir}{os.path.sep}reference.fasta'
```

``` python
annotationFiles
```

    ['../../1001G/annotations/freeze2.1/10002.gff',
     '../../1001G/annotations/freeze2.1/10015.gff',
     '../../1001G/annotations/freeze2.1/10024.gff',
     '../../1001G/annotations/freeze2.1/1741.gff',
     '../../1001G/annotations/freeze2.1/22001.gff',
     '../../1001G/annotations/freeze2.1/22002.gff',
     '../../1001G/annotations/freeze2.1/22003.gff',
     '../../1001G/annotations/freeze2.1/22004.gff',
     '../../1001G/annotations/freeze2.1/22005.gff',
     '../../1001G/annotations/freeze2.1/22006.gff',
     '../../1001G/annotations/freeze2.1/22007.gff',
     '../../1001G/annotations/freeze2.1/6024.gff',
     '../../1001G/annotations/freeze2.1/6069.gff',
     '../../1001G/annotations/freeze2.1/6124.gff',
     '../../1001G/annotations/freeze2.1/6244.gff',
     '../../1001G/annotations/freeze2.1/6909.gff',
     '../../1001G/annotations/freeze2.1/6966.gff',
     '../../1001G/annotations/freeze2.1/8236.gff',
     '../../1001G/annotations/freeze2.1/9075.gff',
     '../../1001G/annotations/freeze2.1/9537.gff',
     '../../1001G/annotations/freeze2.1/9543.gff',
     '../../1001G/annotations/freeze2.1/9638.gff',
     '../../1001G/annotations/freeze2.1/9728.gff',
     '../../1001G/annotations/freeze2.1/9764.gff',
     '../../1001G/annotations/freeze2.1/9888.gff',
     '../../1001G/annotations/freeze2.1/9905.gff',
     '../../1001G/annotations/freeze2.1/9981.gff']

### Generaton of gene graph

``` python
doUS = False
n = 1
for chrnum in range(1,n+1): # here n is number of chromosomes.
    chromosome = f'Chr{chrnum}'

    print(f'\nProcessing {chromosome}\n============')

    curtst = time.time()
    
    graph = GenomeGraph(annotationFiles = annotationFiles,
                        pangenomeFiles = None,
                        sequenceFilesDict = None,
                        doUS = doUS,
                        chromosome = chromosome,
                        refAnnotationFile=refAnnotationFile,
                        refAccession='TAIR10')
    
    print(f'Generating graph for {chromosome} took {time.time() - curtst} seconds')
    
    curtst = time.time()
    graph.treeSort()
    print(f'Sorting graph for {chromosome} took {time.time() - curtst} seconds')
    if len(graph.nodes)!=len(graph.order):
            print('Sorting failed and not all nodes were sorted. Saving unsorted graph')
            gfaFilename = f'Gene_{chromosome}_simOnly_unordered.gfa'
            graph.order = list(range(1,len(graph.nodes)+1))
    else:
        gfaFilename = f'Gene_{chromosome}_simOnly.gfa'
    
    graph.toGFA(f'{gfadir}{os.path.sep}{gfaFilename}',doSeq=False)
```

``` python
q = [1,2,3,4]
```

``` python
q.remove(3)
```

``` python
q
```

    [1, 2, 4]

## Loading Pathfile to graph

``` python
# For path file v1
pathfileDir = 'examples/gene_graph'

pathsfile = 'paths_genegraph.txt'

paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{pathsfile}', True, True)

graph = GenomeGraph(pathsDict=paths)

graph.treeSort()

if len(graph.nodes)!=len(graph.order):
    print('Sorting failed and not all nodes were sorted. Saving unsorted graph')
    output = 'paths_genegraph_unordered.gfa'
    graph.order = list(range(1,len(graph.nodes)+1))
    graph.toGFA(output,doSeq=False)
else:
    coreGFApath = f'paths_genegraph.gfa'
    coregraph.toGFA(coreGFApath,doSeq=False)
```

``` python
# For v2
# This is example, no v2 file currently available for demonstration.
pathfileDir = '/path/to/file'

pathsfile = f'paths.txt'

paths = pathFileToPathDict(f'{pathfileDir}{os.path.sep}{pathsfile}',True,'reference',True)

for seqNum in paths.keys():

    graph = GenomeGraph(pathsDict=paths[seqNum])

    # On undirected coregraph sorting is not optimal! Check sorting!!!

    graph.treeSort()

    if len(graph.nodes)!=len(graph.order):
        print('Sorting failed and not all nodes were sorted. Saving unsorted graph')
        output = f'{pathfileDir}{os.path.sep}graph_Chr{seqNum}_unordered.gfa'
        graph.order = list(range(1,len(graph.nodes)+1))
        graph.toGFA(output,doSeq=False)
    else:
        coreGFApath = f'{pathfileDir}{os.path.sep}graph_Chr{seqNum}.gfa'
        graph.toGFA(coreGFApath,doSeq=False)
```

# Loading graph from GFA and sorting it

``` python
gfadir = 'examples/nucleotide_graph'

# It is nucleotide graph. If it is not nucleotide graph, then `isSeq` variable should be changed to False.
gfafilename = 'paths_presentation.gfa'
isSeq = True

graph = GenomeGraph(gfaPath=f'{gfadir}{os.path.sep}{gfafilename}',isGFASeq=isSeq)

graph_new.treeSort()

assert len(graph_new.nodes)==len(graph_new.order)

basename,ext = os.path.splitext(gfafilename)

graph_new.toGFA(f'{gfadir}{os.path.sep}{basename}_ordered.{ext}',doSeq=isSeq)
```

### Exporting presentation graph

``` python
pathfileDir = '../../Meetings/1001G+_20220518/'
coreGFApath = f'{pathfileDir}{os.path.sep}paths_presentation.gfa'
```

# changing annotation

genome = GenomeGraph(coreGFApath,isGFASeq=True)

for nodeID,node in enumerate(genome.nodesAnnotation): for
seqName,seqDict in node.items(): for annText in seqDict.keys():
genome.nodesAnnotation\[nodeID\]\[seqName\]\[annText\] =
\[(0,len(genome.nodesData\[nodeID\])-1)\]

genome.toGFA(coreGFApath,doSeq=True)

``` python
maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = True
zoomLevels = [1,2,4,8,16]#,8,16,32]
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
outputPath,outputName,zoomLevels
```

    ('../../Meetings/1001G+_20220518', 'paths_presentation_new', [1, 2, 4, 8, 16])

dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName)
print(f’Opening Redis connection for db {dbid}‘) redisConn =
Redis(host=’redis’,port = 6379,db=dbid)

``` python
redisConn=None
```

``` python
#%%capture output
startTime = time.time()
# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \
# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, 
#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \
exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                outputPath=outputPath, outputName=outputName,
                                isSeq=isSeq,
                                redisConn=redisConn,
                                zoomLevels=zoomLevels,
                                fillZoomLevels = True,
                                maxLengthComponent=maxLengthComponent, 
                                maxLengthChunk=maxLengthChunk, 
                                inversionThreshold=inversionThreshold)
#                                 returnDebugData=True)
runTime = time.time() - startTime
```

    Loading Genome
    Loading graph from ../../Meetings/1001G+_20220518//paths_presentation.gfa
    Found nodeNames file ../../Meetings/1001G+_20220518/nodeNames_paths_presentation.json, loading names.
    Found node annotation file ../../Meetings/1001G+_20220518/annotation_paths_presentation.dat, loading associations.
    Loading segment 7/7
    Loading segments finished.
    Loading link 12/12
    Loading links finished
    Loading path 5/5
    Loading paths finished. 5 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7/7
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 5/5
    Finished preprocessing paths

    Recording Pantograph data to ../../Meetings/1001G+_20220518/paths_presentation_new
    Calculating nodes length...
    Processing node 7/7
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 5/5
    Finished preprocessing paths
    Processing path breaks...
    Postprocessing interconnected links 2/2
    Preprocessing interconnected links finished.

    Processing path breaks finished.
    Converting blocks to block lengths 5/5
    Conversion finished.
    Reformating links to block lengths associations 5/5
    Reformating finished.
    Identifying rearrangement blocks
    Processing node 7/7
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 7/7
    Processing component links 6/6
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks
    Recording component 6/6
    Recording zoom level 1 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <2 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 2
    ===========================
    Processing component 6/6
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 6/6
    Recording zoom level 2 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <4 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 4
    ===========================
    Processing component 6/6
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 6/6
    Recording zoom level 4 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <8 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 8
    ===========================
    Processing component 6/6
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 6/6
    Recording zoom level 8 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <16 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 16
    ===========================
    Processing component 6/6
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 6/6
    Recording zoom level 16 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <32 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 32
    ===========================
    Processing component 6/6
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 6/6
    Recording zoom level 32 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <64 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 64
    ===========================
    Processing component 6/6
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 6/6
    Recording zoom level 64 finished.

### Exporting tutorial gene graph

``` python
pathfileDir = '../../1001G/pantograph'
coreGFApath = f'{pathfileDir}{os.path.sep}genegraph_tutorial.gfa'
```

``` python
graph = GenomeGraph(gfaPath=coreGFApath,isGFASeq=False)
```

    Loading graph from ../../1001G/pantograph/genegraph_tutorial.gfa
    Found node annotation file ../../1001G/pantograph/annotation_genegraph_tutorial.dat, loading associations.
    Loading segment 7/7
    Loading segments finished.
    Loading link 13/13
    Loading links finished
    Loading path 5/5
    Loading paths finished. 5 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7/7
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 5/5
    Finished preprocessing paths

``` python
graph.paths
```

    [['1+', '2+', '3+', '4+'],
     ['1+', '5+', '6+', '7+', '1+'],
     ['4+', '1+', '2+', '3+'],
     ['2+', '3+', '4+', '1+', '5-', '4+', '7+'],
     ['6+', '7-', '1+', '2+', '3+']]

# changing annotation

genome = GenomeGraph(coreGFApath,isGFASeq=True)

for nodeID,node in enumerate(genome.nodesAnnotation): for
seqName,seqDict in node.items(): for annText in seqDict.keys():
genome.nodesAnnotation\[nodeID\]\[seqName\]\[annText\] =
\[(0,len(genome.nodesData\[nodeID\])-1)\]

genome.toGFA(coreGFApath,doSeq=True)

``` python
maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = False
zoomLevels = [1]#,8,16,32]
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
outputPath,outputName = pathConvert(coreGFApath,suffix='')
outputPath,outputName,zoomLevels
```

    ('../../1001G/pantograph', 'genegraph_tutorial', [1])

``` python
# dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)
# print(f'Opening Redis connection for db {dbid}')
redisConn = Redis(host = 'redis',port = 6379, db = 0)
```

redisConn=None

``` python
#%%capture output
startTime = time.time()
# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \
# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, 
#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \
exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                outputPath=outputPath, outputName=outputName,
                                isSeq=isSeq,
                                redisConn=redisConn,
                                zoomLevels=zoomLevels,
                                fillZoomLevels = False,
                                maxLengthComponent=maxLengthComponent, 
                                maxLengthChunk=maxLengthChunk, 
                                inversionThreshold=inversionThreshold)
#                                 returnDebugData=True)
runTime = time.time() - startTime
```

    Loading Genome
    Loading graph from ../../1001G/pantograph/genegraph_tutorial.gfa
    Found node annotation file ../../1001G/pantograph/annotation_genegraph_tutorial.dat, loading associations.
    Loading segment 7/7
    Loading segments finished.
    Loading link 13/13
    Loading links finished
    Loading path 5/5
    Loading paths finished. 5 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7/7
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 5/5
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/genegraph_tutorial
    Calculating nodes length...
    Processing node 7/7
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 5/5
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 7/7
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 7/7
    Processing component links 6/6
    Recording component 6/6
    Recording zoom level 1 finished.

### Testing removable elements identification

``` python
from pangraph_constructor.exportDev import getRemovableStructures
```

``` python
import IPython.display as ipd
import joblib
```

``` python
if os.path.exists('./tests/breakIdentify.dat'):
    correctResults = joblib.load('./tests/breakIdentify.dat')
else:
    correctResults = {}
```

``` python
notebook2script()
```

``` python
path = '../../1001G/GraphCollapsing/TestGraphs'
filePrefix = 'test'
```

``` python
#test

for filename,resDict in correctResults.items():
    print(f'\n####### Testing on case {filename} ########')
    graph = GenomeGraph(f'{path}/{filename}',isGFASeq=False)
    linksLengths, pairedLinks, blockEdges, _ = getRemovableStructures(graph=graph)
    assert linksLengths==resDict['linksLengths']
    assert pairedLinks==resDict['pairedLinks']
    assert blockEdges==resDict['blockEdges']
```

``` python
caseNum = 17
filename = f'{filePrefix}{caseNum:02d}.gfa'


print('############')
print(f'Graph from file {filename}')
coreGFApath = f'{path}/{filename}'
graph = GenomeGraph(coreGFApath,isGFASeq=False)
print('Graph Paths:')
ipd.display(graph.paths)
linkLengths, pairedLinks, blockEdges, _ = getRemovableStructures(graph=graph)

print('Link-Lengths associations:')
ipd.display(linkLengths)

print('PairedLinks:')
ipd.display(pairedLinks)

print('Rearrangemenet block edges:')
ipd.display(blockEdges)
```

``` python
correctResults[filename]
```

``` python
correctResults[filename] = {'linksLengths':linkLengths,'pairedLinks':pairedLinks,'blockEdges':blockEdges}
```

``` python
joblib.dump(correctResults,'./tests/breakIdentify.dat')
```

### Exporting test collapse graph

``` python
pathfileDir = '../../1001G/pantograph/data'
coreGFApath = f'{pathfileDir}{os.path.sep}testCollapse.gfa'
```

``` python
maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = False
zoomLevels = [1,2,4]#,8,16,32]
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
outputPath,outputName,zoomLevels
```

dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName)
print(f’Opening Redis connection for db {dbid}‘) redisConn =
Redis(host=’redis’,port = 6379,db=dbid)

``` python
redisConn=None
```

``` python
#%%capture output
startTime = time.time()
# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \
[zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, 
                invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \
exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                outputPath=outputPath, outputName=outputName,
                                isSeq=isSeq,
                                redisConn=redisConn,
                                zoomLevels=zoomLevels,
                                fillZoomLevels = True,
                                maxLengthComponent=maxLengthComponent, 
                                maxLengthChunk=maxLengthChunk, 
                                inversionThreshold=inversionThreshold,
                                returnDebugData=True)
runTime = time.time() - startTime
```

``` python
print(f'Executed in {runTime} seconds')
```

``` python
!ntfy send "Exporting test collapse graph finished. Overall time = {runTime} seconds"
```

### Exporting coregraph

``` python
pathfileDir = '../../1001G/coreGraph'
coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_Chr1.gfa'
```

``` python
maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = False
zoomLevels = [1,2,4,8,16]
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
outputPath,outputName,zoomLevels
```

    ('../../1001G/coreGraph', 'coregraph_Chr1_new', [1, 2, 4, 8, 16])

``` python
# dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)
# print(f'Opening Redis connection for db {dbid}')
redisConn = Redis(host='redis',port = 6379,db=0)
```

``` python
notebook2script()
```

    Converted 00_init.ipynb.
    Converted 01_graph.ipynb.
    Converted 02_tree.ipynb.
    Converted 03_synteny.ipynb.
    Converted 04_utils.ipynb.
    Converted 05_export.ipynb.
    Converted 05_exportDev.ipynb.
    Converted deBruijnGraphProcessing.ipynb.
    Converted dev.ipynb.
    Converted graphTesting.ipynb.
    Converted index.ipynb.

``` python
#%%capture output
startTime = time.time()
# [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \
exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                outputPath=outputPath, outputName=outputName,
                                isSeq=isSeq,
                                redisConn=redisConn,
                                zoomLevels=zoomLevels, 
                                maxLengthComponent=maxLengthComponent, 
                                maxLengthChunk=maxLengthChunk, 
                                inversionThreshold=inversionThreshold)
#                                 returnDebugata=True)
runTime = time.time() - startTime
print(runTime)
```

    Loading Genome
    Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa
    Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.
    Loading segment 35/35
    Loading segments finished.
    Loading link 72/72
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 35/35
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/coreGraph/coregraph_Chr1_new
    Calculating nodes length...
    Processing node 35/35
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Processing path breaks...
    Postprocessing interconnected links 16/16
    Preprocessing interconnected links finished.

    Processing path breaks finished.
    Converting blocks to block lengths 34/34
    Conversion finished.
    Reformating links to block lengths associations 34/34
    Reformating finished.
    Identifying rearrangement blocks
    Processing node 35/35
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 35/35
    Processing component links 35/35
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks
    Recording component 35/35
    Recording zoom level 1 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <2 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 2
    ===========================
    Processing component 35/35
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 8/8
    Recording zoom level 2 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <4 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 4
    ===========================
    Processing component 8/8
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 5/5
    Recording zoom level 4 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <8 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 8
    ===========================
    Processing component 5/5
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 1/1
    Recording zoom level 8 finished.
    Removing links according to collapsible blocks
    All links associated with collapsibleComponents <16 were removed.     0 components were deleted as isolated.

    ===========================
    Zoom level 16
    ===========================
    Processing component 1/1
    Processing component finished.
    Converting link to block lengths associations
    Converting paired links
    Converting interconnected links
    Converting rearrangement blocks

    Recording component 1/1
    Recording zoom level 16 finished.
    27.11550521850586

``` python
!ntfy send "Exporting coregraph finished. Overall time = {runTime} seconds"
```

    /usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.11) or chardet (3.0.4) doesn't match a supported version!
      warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "

❗️❗️❗️ TODO: Next test API. WHen works, change front end to get jsons
without annotation and then to load annotation from API!

``` python
graph = GenomeGraph(coreGFApath,isGFASeq=False)
```

    Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa
    Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.
    Loading segment 35/35
    Loading segments finished.
    Loading link 72/72
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 35/35
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

``` python
from pangraph_constructor.utils import iset_get,iset_score,iset_add
```

``` python
redisConn = Redis(host='redis',port=6379,db=0)
```

``` python
iset_get(redisConn,'test')
```

    {'a_0': (1.0, 3.0), 'b_0': (1.0, 5.0)}

``` python
iset_score(redisConn,'coregraph_Chr1_new.1.10024.Pos',22)
# coregraph_Chr1_new/10024/1/22
```

    ['23']

``` python
iset_score(redisConn,'coregraph_Chr1_new.9543.Gene',50,50)
```

    ['3370',
     'AT1G15940',
     'AT1G15950',
     'AT1G16000',
     'AT1G16010',
     'AT1G16022',
     'AT1G62830',
     'AT1G62840',
     'OG0001502',
     'OG0004404',
     'OG0007340',
     'OG0007341',
     'OG0007343',
     'OG0008435',
     'OG0015400',
     'OG0017985',
     'OG0022841',
     'OG0024046',
     'OG0024055',
     'OG0025924',
     'OG0027925',
     'OG0030893']

``` python
iset_get(redisConn,'coregraph_Chr1_new.9543.Gene','3370_0')
```

    {'3370_0': (48.0, 51.0)}

redisConn=None

### Exporting coregraph with genes

``` python
notebook2script()
```

``` python
pathfileDir = '../../1001G/coreGraph'
```

``` python
maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = True
zoomLevels = [1,3,9]# + [9*2**i for i in range(9)]
# zoomLevels = [2**i for i in range(12)]#  [1,2,4,8,16,32,9*16,9*32,9*128,9*256]
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
# outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
# outputPath,outputName,zoomLevels
```

dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName)
print(f’Opening Redis connection for db {dbid}‘) redisConn =
Redis(host=’redis’,port = 6379,db=dbid)

redisConn=None

``` python
# %%capture output
for chrNum in range(1,6):
    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_genes_Chr{chrNum}.gfa'
    
    outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
    
    dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)
    print(f'Opening Redis connection for db {dbid}')
    redisConn = Redis(host='redis',port = 6379,db=dbid)
    
    startTime = time.time()
    # [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \
    exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                    outputPath=outputPath, outputName=outputName,
                                    isSeq=isSeq,
                                    redisConn=redisConn,
                                    zoomLevels=zoomLevels,
                                    fillZoomLevels = True,
                                    maxLengthComponent=maxLengthComponent, 
                                    maxLengthChunk=maxLengthChunk, 
                                    inversionThreshold=inversionThreshold)
    #                                 returnDebugata=True)
    runTime = time.time() - startTime
```

``` python
curT = time.localtime(time.time()+3600)
message = f"Exporting core graph with genes for all chromosomes finished at \
            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\n"
!ntfy send "{message}"
```

``` python
notebook2script()
```

``` python
zoomLevels = [1,2,4]
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
# outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
# outputPath,outputName,zoomLevels
```

``` python
#%%capture output2
for chrNum in range(1,6):
    coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_Chr{chrNum}.gfa'
    
    outputPath,outputName = pathConvert(coreGFApath,suffix='_new')
    
    dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)
    print(f'Opening Redis connection for db {dbid}')
    redisConn = Redis(host='redis',port = 6379,db=dbid)
    
    startTime = time.time()
    # [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \
    exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                    outputPath=outputPath, outputName=outputName,
                                    isSeq=isSeq,
                                    redisConn=redisConn,
                                    zoomLevels=zoomLevels,
                                    fillZoomLevels = True,
                                    maxLengthComponent=maxLengthComponent, 
                                    maxLengthChunk=maxLengthChunk, 
                                    inversionThreshold=inversionThreshold)
    #                                 returnDebugata=True)
    runTime = time.time() - startTime
    
    print(f'Exporting core graph for Chr{chrNum} took {runTime} seconds')
```

``` python
curT = time.localtime(time.time()+3600)
message = f"Exporting core graph with genes for all chromosomes finished at \
 {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!\n"
!ntfy send "{message}"
```

``` python
!ntfy send "Exporting coregraph with genes finished. Overall time = {runTime} seconds"
```

### Exporting gene graphs for all chromosomes

``` python
pathfileDir = '../../1001G/pantograph/data'
```

``` python
maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = False
zoomLevels = [1]#,2,4]# + [9*2**i for i in range(12)]
fillZoomLevel = False
```

``` python
nbdev_export()
```

``` python
caseDict = {}
for i in range(1,6):
    GFA = f'AT_Chr{i}_OGOnly_2.1.gfa'
    caseDict[f'Chr{i}'] = GFA
```

    {'Chr1': 'AT_Chr1_OGOnly_2.1.gfa',
     'Chr2': 'AT_Chr2_OGOnly_2.1.gfa',
     'Chr3': 'AT_Chr3_OGOnly_2.1.gfa',
     'Chr4': 'AT_Chr4_OGOnly_2.1.gfa',
     'Chr5': 'AT_Chr5_OGOnly_2.1.gfa'}

``` python
exportProject(projectID = 'AT_OGOnly_2.1', projectName = 'A. thaliana 27 genomes + TAIR10 gene graph 2.1', 
              caseDict = caseDict, pathToIndex = pathfileDir, pathToGraphs = pathfileDir,
              redisHost='redis', redisPort = 6379, redisDB = 0,
              suffix = 'pr',
              maxLengthComponent = maxLengthComponent, maxLengthChunk = maxLengthChunk,
              inversionThreshold = inversionThreshold,
              isSeq = isSeq, zoomLevels = zoomLevels, fillZoomLevel = fillZoomLevel)
```


    Processing case AT_OGOnly_2.1_Chr1pr
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
    Loading segment 7455/7455
    Loading segments finished.
    Loading link 14001/14001
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7455/7455
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr1pr
    Calculating nodes length...
    Processing node 7455/7455
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 7455/7455
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 7455/7455
    Processing component links 3869/3869
    Recording component 3869/3869
    Recording zoom level 1 finished.
    Exporting gene graph for Chr1 took 465.6470103263855 seconds

    Processing case AT_OGOnly_2.1_Chr2pr
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.
    Loading segment 4877/4877
    Loading segments finished.
    Loading link 10302/10302
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 4877/4877
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr2pr
    Calculating nodes length...
    Processing node 4877/4877
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 4877/4877
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 4877/4877
    Processing component links 2534/2534
    Recording component 2534/2534
    Recording zoom level 1 finished.
    Exporting gene graph for Chr2 took 214.97250485420227 seconds

    Processing case AT_OGOnly_2.1_Chr3pr
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.
    Loading segment 8442/8442
    Loading segments finished.
    Loading link 15985/15985
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 8442/8442
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr3pr
    Calculating nodes length...
    Processing node 8442/8442
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 8442/8442
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 8442/8442
    Processing component links 3383/3383
    Recording component 3383/3383
    Recording zoom level 1 finished.
    Exporting gene graph for Chr3 took 430.0764102935791 seconds

    Processing case AT_OGOnly_2.1_Chr4pr
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.
    Loading segment 4620/4620
    Loading segments finished.
    Loading link 9980/9980
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 4620/4620
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr4pr
    Calculating nodes length...
    Processing node 4620/4620
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 4620/4620
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 4620/4620
    Processing component links 2526/2526
    Recording component 2526/2526
    Recording zoom level 1 finished.
    Exporting gene graph for Chr4 took 199.40259623527527 seconds

    Processing case AT_OGOnly_2.1_Chr5pr
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.
    Loading segment 7052/7052
    Loading segments finished.
    Loading link 12737/12737
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7052/7052
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr5pr
    Calculating nodes length...
    Processing node 7052/7052
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 7052/7052
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 7052/7052
    Processing component links 3274/3274
    Recording component 3274/3274
    Recording zoom level 1 finished.
    Exporting gene graph for Chr5 took 348.297509431839 seconds

``` python
for seqID in ['Chr1','Chr2','Chr3','Chr4','Chr5']:#'Chr1','Chr2',

    coreGFApath = f'{pathfileDir}{os.path.sep}AT_{seqID}_OGOnly_2.1.gfa'

    outputPath,outputName = pathConvert(coreGFApath,suffix='_viscol')
    
    print()
    print(f'Processing case {outputName}')
    # dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)
    # print(f'Opening Redis connection for db {dbid}')
    redisConn = Redis(host='redis',port = 6379,db=0)

    #%%capture output
    startTime = time.time()
    # initialLinkLengths, initialPairedLinks, initialInterconnectedLinks, initialBlockEdges, \
    # zoomNodeToComponent,zoomComponentToNodes,zoomComponents,\
    # zoomFromComponentLinks, zoomToComponentLinks, zoomLinkLengths, zoomPairedLinks, zoomInterconnectedLinks, \
    # zoomOldToNewRemoval, zoomNewToOldRemoval, \
    # zoomLinkLengthsRemoval, zoomPairedLinksRemoval, zoomInterconnectedLinksRemoval, zoomBlockEdgesRemoval, \
    # zoomFromComponentLinksRemoval, zoomToComponentLinksRemoval, \
    # graph = \
    exportToPantograph(inputPath=coreGFApath, GenomeGraphParams={}, 
                                    outputPath=outputPath, outputName=outputName,
                                    isSeq=isSeq,
                                    redisConn=redisConn,
                                    zoomLevels=zoomLevels, 
                                    fillZoomLevels = False,
                                    maxLengthComponent=maxLengthComponent, 
                                    maxLengthChunk=maxLengthChunk, 
                                    inversionThreshold=inversionThreshold)#,
                                    # returnDebugData=True)
    runTime = time.time() - startTime
    
    print(f'Exporting gene graph for {seqID} took {runTime} seconds')
    
    redisConn.close()
```


    Processing case AT_Chr1_OGOnly_2.1_viscol
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
    Loading segment 7455/7455
    Loading segments finished.
    Loading link 14001/14001
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7455/7455
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1_viscol
    Calculating nodes length...
    Processing node 7455/7455
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 7455/7455
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 7455/7455
    Processing component links 3869/3869
    Recording component 3869/3869
    Recording zoom level 1 finished.
    Exporting gene graph for Chr1 took 445.7378799915314 seconds

    Processing case AT_Chr2_OGOnly_2.1_viscol
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.
    Loading segment 4877/4877
    Loading segments finished.
    Loading link 10302/10302
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 4877/4877
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1_viscol
    Calculating nodes length...
    Processing node 4877/4877
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 4877/4877
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 4877/4877
    Processing component links 2534/2534
    Recording component 2534/2534
    Recording zoom level 1 finished.
    Exporting gene graph for Chr2 took 193.71776700019836 seconds

    Processing case AT_Chr3_OGOnly_2.1_viscol
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.
    Loading segment 8442/8442
    Loading segments finished.
    Loading link 15985/15985
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 8442/8442
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1_viscol
    Calculating nodes length...
    Processing node 8442/8442
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 8442/8442
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 8442/8442
    Processing component links 3383/3383
    Recording component 3383/3383
    Recording zoom level 1 finished.
    Exporting gene graph for Chr3 took 411.77689695358276 seconds

    Processing case AT_Chr4_OGOnly_2.1_viscol
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.
    Loading segment 4620/4620
    Loading segments finished.
    Loading link 9980/9980
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 4620/4620
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1_viscol
    Calculating nodes length...
    Processing node 4620/4620
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 4620/4620
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 4620/4620
    Processing component links 2526/2526
    Recording component 2526/2526
    Recording zoom level 1 finished.
    Exporting gene graph for Chr4 took 180.36719918251038 seconds

    Processing case AT_Chr5_OGOnly_2.1_viscol
    Loading Genome
    Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.
    Loading segment 7052/7052
    Loading segments finished.
    Loading link 12737/12737
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7052/7052
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

    Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1_viscol
    Calculating nodes length...
    Processing node 7052/7052
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths
    Identifying rearrangement blocks
    Processing node 7052/7052
    Identifying rearrangement blocks finished.

    ===========================
    Zoom level 1
    ===========================
    Processing node 7052/7052
    Processing component links 3274/3274
    Recording component 3274/3274
    Recording zoom level 1 finished.
    Exporting gene graph for Chr5 took 338.8390169143677 seconds

``` python
curT = time.localtime()
message = f"Exporting gene graph for all chromosomes finished at \
            {curT.tm_hour:02d}:{curT.tm_min:02d} on {curT.tm_mday:02d}/{curT.tm_mon:02d}!"
!ntfy send "{message}"
```

``` python
coreGFApath = f'{pathfileDir}{os.path.sep}AT_Chr1_OGOnly_2.1.gfa'
graph = GenomeGraph(gfaPath=coreGFApath, isGFASeq=False)
```

    Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
    Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
    Loading segment 7455/7455
    Loading segments finished.
    Loading link 14001/14001
    Loading links finished
    Loading path 28/28
    Loading paths finished. 28 paths added, 0 paths ignored.
    Calculating nodes length...
    Processing node 7455/7455
    Finished calculating nodes lengths
    Preprocessing paths...
    Processing path 28/28
    Finished preprocessing paths

``` python
graph.nodesMetadata[1361]['22006']
```

    {'genPos': [{'chr': 'Chr1', 'genomePosition': [19020282, 19020893]},
      {'chr': 'Chr1', 'genomePosition': [19022280, 19023284]}],
     'annotation': {'OG0001585': [(0, 0)],
      'AT1G55230': [(0, 0)],
      'AT1G55240': [(0, 0)]}}

import joblib

joblib.dump(initialLinkLengths,‘dumps/initialLinkLengths.dump’)
joblib.dump(initialPairedLinks,‘dumps/initialPairedLinks.dump’)
joblib.dump(initialInterconnectedLinks,‘dumps/initialInterconnectedLinks.dump’)
joblib.dump(initialBlockEdges,‘dumps/initialBlockEdges.dump’)
joblib.dump(zoomNodeToComponent,‘dumps/zoomNodeToComponent.dump’)
joblib.dump(zoomComponentToNodes,‘dumps/zoomComponentToNodes.dump’)
joblib.dump(zoomComponents,‘dumps/zoomComponents.dump’)
joblib.dump(zoomFromComponentLinks,‘dumps/zoomFromComponentLinks.dump’)
joblib.dump(zoomToComponentLinks,‘dumps/zoomToComponentLinks.dump’)
joblib.dump(zoomLinkLengths,‘dumps/zoomLinkLengths.dump’)
joblib.dump(zoomPairedLinks,‘dumps/zoomPairedLinks.dump’)
joblib.dump(zoomInterconnectedLinks,‘dumps/zoomInterconnectedLinks.dump’)
joblib.dump(zoomOldToNewRemoval,‘dumps/zoomOldToNewRemoval.dump’)
joblib.dump(zoomNewToOldRemoval,‘dumps/zoomNewToOldRemoval.dump’)
joblib.dump(zoomLinkLengthsRemoval,‘dumps/zoomLinkLengthsRemoval.dump’)
joblib.dump(zoomPairedLinksRemoval,‘dumps/zoomPairedLinksRemoval.dump’)
joblib.dump(zoomInterconnectedLinksRemoval,‘dumps/zoomInterconnectedLinksRemoval.dump’)
joblib.dump(zoomBlockEdgesRemoval,‘dumps/zoomBlockEdgesRemoval.dump’)
joblib.dump(zoomFromComponentLinksRemoval,‘dumps/zoomFromComponentLinksRemoval.dump’)
joblib.dump(zoomToComponentLinksRemoval,‘dumps/zoomToComponentLinksRemoval.dump’)
joblib.dump(graph,‘dumps/graph.dump’)

initialLinkLengths = joblib.load(‘dumps/initialLinkLengths.dump’)

initialPairedLinks = joblib.load(‘dumps/initialPairedLinks.dump’)

initialInterconnectedLinks =
joblib.load(‘dumps/initialInterconnectedLinks.dump’)

initialBlockEdges = joblib.load(‘dumps/initialBlockEdges.dump’)

zoomNodeToComponent = joblib.load(‘dumps/zoomNodeToComponent.dump’)

zoomComponentToNodes = joblib.load(‘dumps/zoomComponentToNodes.dump’)

zoomComponents = joblib.load(‘dumps/zoomComponents.dump’)

zoomFromComponentLinks =
joblib.load(‘dumps/zoomFromComponentLinks.dump’)

zoomToComponentLinks = joblib.load(‘dumps/zoomToComponentLinks.dump’)

zoomLinkLengths = joblib.load(‘dumps/zoomLinkLengths.dump’)

zoomPairedLinks = joblib.load(‘dumps/zoomPairedLinks.dump’)

zoomInterconnectedLinks =
joblib.load(‘dumps/zoomInterconnectedLinks.dump’)

zoomOldToNewRemoval = joblib.load(‘dumps/zoomOldToNewRemoval.dump’)

zoomNewToOldRemoval = joblib.load(‘dumps/zoomNewToOldRemoval.dump’)

zoomLinkLengthsRemoval =
joblib.load(‘dumps/zoomLinkLengthsRemoval.dump’)

zoomPairedLinksRemoval =
joblib.load(‘dumps/zoomPairedLinksRemoval.dump’)

zoomInterconnectedLinksRemoval =
joblib.load(‘dumps/zoomInterconnectedLinksRemoval.dump’)

zoomBlockEdgesRemoval = joblib.load(‘dumps/zoomBlockEdgesRemoval.dump’)

zoomFromComponentLinksRemoval =
joblib.load(‘dumps/zoomFromComponentLinksRemoval.dump’)

zoomToComponentLinksRemoval =
joblib.load(‘dumps/zoomToComponentLinksRemoval.dump’)

graph = joblib.load(‘dumps/graph.dump’)

### Comments

Each node processing time increase significantly with overall number of
nodes. This is wrong and should be investigated.

``` python
from copy import deepcopy
import numpy as np
```

``` python
pathfileDir = '../../1001G/coreGraph'
coreGFApath = f'{pathfileDir}{os.path.sep}coreGraph_f2.1_Ref_v04.gfa'
```

``` python
coregraph = GenomeGraph(gfaPath=coreGFApath,isGFASeq=False)
coregraph_genes = deepcopy(coregraph)
```

``` python
fullGraphPath = '../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa'
fullgraph = GenomeGraph(gfaPath=fullGraphPath,isGFASeq=False)
```

``` python
chainToGenesFile = 'chain2gene_f2.1_Ref_v04.txt'
maxChainLength = len(coregraph.nodes[0])
chainToListDict = {}
with open(f'{pathfileDir}{os.path.sep}{chainToGenesFile}') as f:
    for line in f:
        chainName, geneList = line.split(':')
        geneList = geneList.lstrip().rstrip().split(',')
        chainToListDict[chainName.zfill(maxChainLength)] = geneList
```

``` python
for nodeIdx,nodeName in enumerate(coregraph.nodes):
    print(f'\nNode {nodeIdx+1}/{len(coregraph.nodes)}',end='')
    geneList = chainToListDict.get(nodeName.zfill(maxChainLength), [f'ch{nodeName.zfill(7)}'])
    geneIds = []
    if geneList[0][:2]!='ch':
        geneIds = [int(gene.rstrip('+'))-1 for gene in geneList]
        geneList = [fullgraph.nodes[geneid] for geneid in geneIds]
    coregraph_genes.nodesData[nodeIdx] = ''.join(geneList)
    
    for accession, chainDict in coregraph.nodesAnnotation[nodeIdx].items():
        interval = chainDict[nodeName]
        geneCumLengths = np.hstack((0, np.cumsum([len(gene) for gene in geneList])))
        
        coregraph.nodesAnnotation[nodeIdx][accession].\
            update({geneAnnotation:interval \
                    for geneid in geneIds \
                        for geneAnnotation in fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:None}).keys()})

        coregraph_genes.nodesAnnotation[nodeIdx][accession].\
            update({geneAnnotation:[(geneCumLengths[i], geneCumLengths[i+1]-1)] \
                            for i,geneid in enumerate(geneIds) \
                                for geneAnnotation in fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:None}).keys()})

        coregraph_genes.nodesAnnotation[nodeIdx][accession].\
            update({nodeName:[(geneCumLengths[0], geneCumLengths[-1]-1)]})
print('')
```

``` python
coregraph.toGFA(f'{pathfileDir}{os.path.sep}coregraph_f2.1_Ref_v04.gfa',doSeq=False)
coregraph_genes.toGFA(f'{pathfileDir}{os.path.sep}coregraph_genes_f2.1_Ref_v04.gfa',doSeq=True)
```

``` python
pathfileDir = '../../1001G/coreGraph'
coreGFApath = f'{pathfileDir}{os.path.sep}coregraph_genes_f2.1_Ref_v04.gfa'
# coreGFApath = f'{pathfileDir}{os.path.sep}coreGraph.gfa'
```

``` python
# zoomLevels = [1,10,20,100,500,1000,5000,10000,50000,100000,500000,1000000]
zoomLevels = [1,3,9,45,90,450,900,4500,9000]
# zoomLevels = [1,3,9,18]
# zoomLevels = [4,8,16]

isSeq = True

maxLengthComponent = 100
maxLengthChunk = 16
invertionThreshold = 0.5
# inputPath = '../../1001G/pantograph/data/shorttest2.gfa'
# inputPath = '../../1001G/pantograph/data/AT_Chr1_OGOnly.gfa'
# inputPath = '../../1001G/chrisGraph/chr1.wfmash.n20.a90.s10000.p1,19,39,3,81,1.seqwish.sort.smooth.sort.gfa'
# inputPath = '../../1001G/pantograph/data/shorttest_seq.gfa'
inputPath = coreGFApath
```

``` python
zoomLevels = adjustZoomLevels(zoomLevels)
outputPath,outputName = pathConvert(inputPath,suffix='_new')
outputPath,outputName,zoomLevels
```

``` python
dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)
print(f'Opening Redis connection for db {dbid}')
redisConn = Redis(host='redis',port = 6379,db=dbid)
```

``` python
# zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides = \
exportToPantograph(inputPath=inputPath,
                   outputName=outputName,
                   outputPath=outputPath,
                   isSeq=isSeq,
                   redisConn=redisConn,
                   GenomeGraphParams={'accessionsToRemove':['Consensus']},
                   zoomLevels=zoomLevels,
                   maxLengthChunk=maxLengthChunk,
                   maxLengthComponent=maxLengthComponent,
                   invertionThreshold=invertionThreshold,)
#                                              debug=True,returnDebugData=True)
```

``` python
!ntfy send "Pantograph data generation for coregraph finished."
```

# Exporting to Pantograph visualisation

``` python
projectID = 'paths_genegraph'
projectName = 'Example gene graph'
pathToGraphs = 'examples/gene_graph'
caseDict = {'Main': 'paths_genegraph.gfa'}
pathToIndex = 'examples/Visdata'

# This is if you run it in Docker compose together with active Redis image, which is named "redis".
# If you have separate redis server, enter full address here.
# If you do not want to add any annotation, `redisHost` should be None.
redisHost = 'redis'
redisPort = 6379
redisDB = 0

suffix = ''

maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = False
zoomLevels = [1]
fillZoomLevel = True

exportProject(projectID, projectName, caseDict, pathToIndex, pathToGraphs,
              redisHost = redisHost, redisPort = redisPort, redisDB = redisDB,
              suffix = suffix,
              maxLengthComponent = maxLengthComponent, maxLengthChunk = maxLengthChunk,
              inversionThreshold = inversionThreshold,
              isSeq = isSeq,
              zoomLevels = zoomLevels, fillZoomLevel = fillZoomLevel):
```

``` python
projectID = 'tutorial_graph'
projectName = 'Example nucleotide graph'
pathToGraphs = 'examples/nucleotide_graph'
caseDict = {'Main': 'paths_presentation.gfa'}
pathToIndex = 'examples/Visdata'

# This is if you run it in Docker compose together with active Redis image, which is named "redis".
# If you have separate redis server, enter full address here.
# If you do not want to add any annotation, `redisHost` should be None.
redisHost = 'redis'
redisPort = 6379
redisDB = 0

suffix = ''

maxLengthComponent = 100
maxLengthChunk = 6
inversionThreshold = 0.5
isSeq = 
zoomLevels = [1]
fillZoomLevel = True

exportProject(projectID, projectName, caseDict, pathToIndex, pathToGraphs,
              redisHost = redisHost, redisPort = redisPort, redisDB = redisDB,
              suffix = suffix,
              maxLengthComponent = maxLengthComponent, maxLengthChunk = maxLengthChunk,
              inversionThreshold = inversionThreshold,
              isSeq = isSeq,
              zoomLevels = zoomLevels, fillZoomLevel = fillZoomLevel):
```
