<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).">

<title>pygengraph</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="pygengraph">
<meta property="og:description" content="This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).">
<meta property="og:site-name" content="pygengraph">
<meta name="twitter:title" content="pygengraph">
<meta name="twitter:description" content="This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">pygengraph</span>
    </a>
  </div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">pygengraph</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link active">pygengraph</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./init.html" class="sidebar-item-text sidebar-link">Synteny module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./graph.html" class="sidebar-item-text sidebar-link">Graph module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tree.html" class="sidebar-item-text sidebar-link">Tree module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./synteny.html" class="sidebar-item-text sidebar-link">Synteny module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">Utils module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./export.html" class="sidebar-item-text sidebar-link">Export module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scripts.html" class="sidebar-item-text sidebar-link">Script module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./debruijngraphprocessing.html" class="sidebar-item-text sidebar-link">Processing GFA produced by CuttleFish compacted de Bruijn graph constructor</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dev.html" class="sidebar-item-text sidebar-link">Development notebook</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#install" id="toc-install" class="nav-link active" data-scroll-target="#install">Install</a></li>
  <li><a href="#how-to-use" id="toc-how-to-use" class="nav-link" data-scroll-target="#how-to-use">How to use</a></li>
  <li><a href="#if-redis-database-needs-to-be-cleaned." id="toc-if-redis-database-needs-to-be-cleaned." class="nav-link" data-scroll-target="#if-redis-database-needs-to-be-cleaned.">If redis database needs to be cleaned.</a></li>
  <li><a href="#generating-graphs" id="toc-generating-graphs" class="nav-link" data-scroll-target="#generating-graphs">Generating graphs</a>
  <ul class="collapse">
  <li><a href="#generating-from-annotation" id="toc-generating-from-annotation" class="nav-link" data-scroll-target="#generating-from-annotation">Generating from annotation</a>
  <ul class="collapse">
  <li><a href="#preparing-list-of-files" id="toc-preparing-list-of-files" class="nav-link" data-scroll-target="#preparing-list-of-files">Preparing list of files</a></li>
  <li><a href="#generaton-of-gene-graph" id="toc-generaton-of-gene-graph" class="nav-link" data-scroll-target="#generaton-of-gene-graph">Generaton of gene graph</a></li>
  </ul></li>
  <li><a href="#loading-pathfile-to-graph" id="toc-loading-pathfile-to-graph" class="nav-link" data-scroll-target="#loading-pathfile-to-graph">Loading Pathfile to graph</a></li>
  </ul></li>
  <li><a href="#loading-graph-from-gfa-and-sorting-it" id="toc-loading-graph-from-gfa-and-sorting-it" class="nav-link" data-scroll-target="#loading-graph-from-gfa-and-sorting-it">Loading graph from GFA and sorting it</a>
  <ul class="collapse">
  <li><a href="#exporting-presentation-graph" id="toc-exporting-presentation-graph" class="nav-link" data-scroll-target="#exporting-presentation-graph">Exporting presentation graph</a></li>
  </ul></li>
  <li><a href="#changing-annotation" id="toc-changing-annotation" class="nav-link" data-scroll-target="#changing-annotation">changing annotation</a>
  <ul class="collapse">
  <li><a href="#exporting-tutorial-gene-graph" id="toc-exporting-tutorial-gene-graph" class="nav-link" data-scroll-target="#exporting-tutorial-gene-graph">Exporting tutorial gene graph</a></li>
  </ul></li>
  <li><a href="#changing-annotation-1" id="toc-changing-annotation-1" class="nav-link" data-scroll-target="#changing-annotation-1">changing annotation</a>
  <ul class="collapse">
  <li><a href="#testing-removable-elements-identification" id="toc-testing-removable-elements-identification" class="nav-link" data-scroll-target="#testing-removable-elements-identification">Testing removable elements identification</a></li>
  <li><a href="#exporting-test-collapse-graph" id="toc-exporting-test-collapse-graph" class="nav-link" data-scroll-target="#exporting-test-collapse-graph">Exporting test collapse graph</a></li>
  <li><a href="#exporting-coregraph" id="toc-exporting-coregraph" class="nav-link" data-scroll-target="#exporting-coregraph">Exporting coregraph</a></li>
  <li><a href="#exporting-coregraph-with-genes" id="toc-exporting-coregraph-with-genes" class="nav-link" data-scroll-target="#exporting-coregraph-with-genes">Exporting coregraph with genes</a></li>
  <li><a href="#exporting-gene-graphs-for-all-chromosomes" id="toc-exporting-gene-graphs-for-all-chromosomes" class="nav-link" data-scroll-target="#exporting-gene-graphs-for-all-chromosomes">Exporting gene graphs for all chromosomes</a></li>
  <li><a href="#comments" id="toc-comments" class="nav-link" data-scroll-target="#comments">Comments</a></li>
  </ul></li>
  <li><a href="#exporting-to-pantograph-visualisation" id="toc-exporting-to-pantograph-visualisation" class="nav-link" data-scroll-target="#exporting-to-pantograph-visualisation">Exporting to Pantograph visualisation</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-git"></i></div><div class="action-links"><p><a href="https://gitlab.com/pigrenok/pygengraph/tree/master/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">pygengraph</h1>
</div>

<div>
  <div class="description">
    This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="install" class="level2">
<h2 class="anchored" data-anchor-id="install">Install</h2>
<p>Enter the directory of the library and enter:</p>
<p><code>pip install .</code></p>
<p>and for development use</p>
<p><code>pip install -e .</code></p>
</section>
<section id="how-to-use" class="level2">
<h2 class="anchored" data-anchor-id="how-to-use">How to use</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> nbdev <span class="im">import</span> nbdev_export</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>nbdev_export()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pygengraph.graph <span class="im">import</span> GenomeGraph</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pygengraph.utils <span class="im">import</span> pathFileToPathDict</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pygengraph.export <span class="im">import</span> exportProject</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="if-redis-database-needs-to-be-cleaned." class="level1">
<h1>If redis database needs to be cleaned.</h1>
<p>from redis import Redis</p>
<p>redisConn = Redis(host=‘redis’,port=6379)</p>
<p>redisConn.flushall()</p>
<p>redisConn.close()</p>
<p>del redisConn</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(<span class="st">'ignore'</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">'always'</span>,category<span class="op">=</span><span class="pp">RuntimeWarning</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="generating-graphs" class="level1">
<h1>Generating graphs</h1>
<section id="generating-from-annotation" class="level2">
<h2 class="anchored" data-anchor-id="generating-from-annotation">Generating from annotation</h2>
<section id="preparing-list-of-files" class="level3">
<h3 class="anchored" data-anchor-id="preparing-list-of-files">Preparing list of files</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>refdir <span class="op">=</span> <span class="st">'/path/to/reference/'</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>annotationdir <span class="op">=</span> <span class="st">'/path/to/annotation'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>gfadir <span class="op">=</span> <span class="st">'/path/to/graphs'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>annotationFiles <span class="op">=</span> <span class="bu">sorted</span>(glob.glob(<span class="ss">f'</span><span class="sc">{</span>annotationdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">*.gff'</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pangenomeFiles <span class="op">=</span> <span class="bu">sorted</span>(glob.glob(<span class="ss">f'</span><span class="sc">{</span>annotationdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">*pangen.gff'</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes.</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It should also be converted to sequenceFileDict, see details in documentation for GenomeGraph Class constructor.</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sequenceFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}sequences{os.path.sep}*.fasta'))</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>refAnnotationFile <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>refdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">reference.gff'</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># refSequenceFile = f'{refdir}{os.path.sep}reference.fasta'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>refdir <span class="op">=</span> <span class="st">'../../1001G/annotations/freeze2.1/outgroups'</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>annotationdir <span class="op">=</span> <span class="st">'../../1001G/annotations/freeze2.1'</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>gfadir <span class="op">=</span> <span class="st">'../../1001G/annotations/graphs'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>annotationFiles <span class="op">=</span> <span class="bu">sorted</span>(glob.glob(<span class="ss">f'</span><span class="sc">{</span>annotationdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">*.gff'</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pangenomeFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}*pangen.gff'))</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes.</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It should also be converted to sequenceFileDict, see details in documentation for GenomeGraph Class constructor.</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sequenceFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}sequences{os.path.sep}*.fasta'))</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>refAnnotationFile <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>refdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">araport.gff'</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># refSequenceFile = f'{refdir}{os.path.sep}reference.fasta'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>annotationFiles</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['../../1001G/annotations/freeze2.1/10002.gff',
 '../../1001G/annotations/freeze2.1/10015.gff',
 '../../1001G/annotations/freeze2.1/10024.gff',
 '../../1001G/annotations/freeze2.1/1741.gff',
 '../../1001G/annotations/freeze2.1/22001.gff',
 '../../1001G/annotations/freeze2.1/22002.gff',
 '../../1001G/annotations/freeze2.1/22003.gff',
 '../../1001G/annotations/freeze2.1/22004.gff',
 '../../1001G/annotations/freeze2.1/22005.gff',
 '../../1001G/annotations/freeze2.1/22006.gff',
 '../../1001G/annotations/freeze2.1/22007.gff',
 '../../1001G/annotations/freeze2.1/6024.gff',
 '../../1001G/annotations/freeze2.1/6069.gff',
 '../../1001G/annotations/freeze2.1/6124.gff',
 '../../1001G/annotations/freeze2.1/6244.gff',
 '../../1001G/annotations/freeze2.1/6909.gff',
 '../../1001G/annotations/freeze2.1/6966.gff',
 '../../1001G/annotations/freeze2.1/8236.gff',
 '../../1001G/annotations/freeze2.1/9075.gff',
 '../../1001G/annotations/freeze2.1/9537.gff',
 '../../1001G/annotations/freeze2.1/9543.gff',
 '../../1001G/annotations/freeze2.1/9638.gff',
 '../../1001G/annotations/freeze2.1/9728.gff',
 '../../1001G/annotations/freeze2.1/9764.gff',
 '../../1001G/annotations/freeze2.1/9888.gff',
 '../../1001G/annotations/freeze2.1/9905.gff',
 '../../1001G/annotations/freeze2.1/9981.gff']</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>refAnnotationFile</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>'../../1001G/annotations/freeze2.1/outgroups/araport.gff'</code></pre>
</div>
</div>
</section>
<section id="generaton-of-gene-graph" class="level3">
<h3 class="anchored" data-anchor-id="generaton-of-gene-graph">Generaton of gene graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>nbdev_export()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>doUS <span class="op">=</span> <span class="va">False</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>n <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrnum <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,n<span class="op">+</span><span class="dv">1</span>): <span class="co"># here n is number of chromosomes.</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    chromosome <span class="op">=</span> <span class="ss">f'Chr</span><span class="sc">{</span>chrnum<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="ch">\n</span><span class="ss">Processing </span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ch">\n</span><span class="ss">============'</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    curtst <span class="op">=</span> time.time()</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    graph <span class="op">=</span> GenomeGraph(annotationFiles <span class="op">=</span> annotationFiles,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>                        pangenomeFiles <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>                        sequenceFilesDict <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>                        doUS <span class="op">=</span> doUS,</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>                        chromosome <span class="op">=</span> chromosome,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>                        refAnnotationFile<span class="op">=</span>refAnnotationFile,</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>                        refAccession<span class="op">=</span><span class="st">'TAIR10'</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Generating graph for </span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>time<span class="sc">.</span>time() <span class="op">-</span> curtst<span class="sc">}</span><span class="ss"> seconds'</span>)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    curtst <span class="op">=</span> time.time()</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    graph.treeSort()</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Sorting graph for </span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>time<span class="sc">.</span>time() <span class="op">-</span> curtst<span class="sc">}</span><span class="ss"> seconds'</span>)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(graph.nodes)<span class="op">!=</span><span class="bu">len</span>(graph.order):</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">'Sorting failed and not all nodes were sorted. Saving unsorted graph'</span>)</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>            gfaFilename <span class="op">=</span> <span class="ss">f'Gene_</span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss">_simOnly_unordered.gfa'</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>            graph.order <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(graph.nodes)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>        gfaFilename <span class="op">=</span> <span class="ss">f'Gene_</span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss">_simOnly.gfa'</span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>    graph.toGFA(<span class="ss">f'</span><span class="sc">{</span>gfadir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>gfaFilename<span class="sc">}</span><span class="ss">'</span>,doSeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing Chr1
============
Processing file ../../1001G/annotations/freeze2.1/10002.gff...
Annotation file ../../1001G/annotations/freeze2.1/10002.gff read.
Processing file ../../1001G/annotations/freeze2.1/10002.gff finished in 28.162004947662354 seconds.
Processing file ../../1001G/annotations/freeze2.1/10015.gff...
Annotation file ../../1001G/annotations/freeze2.1/10015.gff read.
Processing file ../../1001G/annotations/freeze2.1/10015.gff finished in 23.948190689086914 seconds.
Processing file ../../1001G/annotations/freeze2.1/10024.gff...
Annotation file ../../1001G/annotations/freeze2.1/10024.gff read.
Processing file ../../1001G/annotations/freeze2.1/10024.gff finished in 28.291413068771362 seconds.
Processing file ../../1001G/annotations/freeze2.1/1741.gff...
Annotation file ../../1001G/annotations/freeze2.1/1741.gff read.
Processing file ../../1001G/annotations/freeze2.1/1741.gff finished in 23.79175639152527 seconds.
Processing file ../../1001G/annotations/freeze2.1/22001.gff...
Annotation file ../../1001G/annotations/freeze2.1/22001.gff read.
Processing file ../../1001G/annotations/freeze2.1/22001.gff finished in 24.502596139907837 seconds.
Processing file ../../1001G/annotations/freeze2.1/22002.gff...
Annotation file ../../1001G/annotations/freeze2.1/22002.gff read.
Processing file ../../1001G/annotations/freeze2.1/22002.gff finished in 27.929000854492188 seconds.
Processing file ../../1001G/annotations/freeze2.1/22003.gff...
Annotation file ../../1001G/annotations/freeze2.1/22003.gff read.
Processing file ../../1001G/annotations/freeze2.1/22003.gff finished in 25.13773822784424 seconds.
Processing file ../../1001G/annotations/freeze2.1/22004.gff...
Annotation file ../../1001G/annotations/freeze2.1/22004.gff read.
Processing file ../../1001G/annotations/freeze2.1/22004.gff finished in 27.87197232246399 seconds.
Processing file ../../1001G/annotations/freeze2.1/22005.gff...
Annotation file ../../1001G/annotations/freeze2.1/22005.gff read.
Processing file ../../1001G/annotations/freeze2.1/22005.gff finished in 27.37407636642456 seconds.
Processing file ../../1001G/annotations/freeze2.1/22006.gff...
Annotation file ../../1001G/annotations/freeze2.1/22006.gff read.
Processing file ../../1001G/annotations/freeze2.1/22006.gff finished in 28.385839223861694 seconds.
Processing file ../../1001G/annotations/freeze2.1/22007.gff...
Annotation file ../../1001G/annotations/freeze2.1/22007.gff read.
Processing file ../../1001G/annotations/freeze2.1/22007.gff finished in 27.108904600143433 seconds.
Processing file ../../1001G/annotations/freeze2.1/6024.gff...
Annotation file ../../1001G/annotations/freeze2.1/6024.gff read.
Processing file ../../1001G/annotations/freeze2.1/6024.gff finished in 23.66122317314148 seconds.
Processing file ../../1001G/annotations/freeze2.1/6069.gff...
Annotation file ../../1001G/annotations/freeze2.1/6069.gff read.
Processing file ../../1001G/annotations/freeze2.1/6069.gff finished in 27.961503267288208 seconds.
Processing file ../../1001G/annotations/freeze2.1/6124.gff...
Annotation file ../../1001G/annotations/freeze2.1/6124.gff read.
Processing file ../../1001G/annotations/freeze2.1/6124.gff finished in 25.66038990020752 seconds.
Processing file ../../1001G/annotations/freeze2.1/6244.gff...
Annotation file ../../1001G/annotations/freeze2.1/6244.gff read.
Processing file ../../1001G/annotations/freeze2.1/6244.gff finished in 28.785216569900513 seconds.
Processing file ../../1001G/annotations/freeze2.1/6909.gff...
Annotation file ../../1001G/annotations/freeze2.1/6909.gff read.
Processing file ../../1001G/annotations/freeze2.1/6909.gff finished in 26.238420486450195 seconds.
Processing file ../../1001G/annotations/freeze2.1/6966.gff...
Annotation file ../../1001G/annotations/freeze2.1/6966.gff read.
Processing file ../../1001G/annotations/freeze2.1/6966.gff finished in 26.252767086029053 seconds.
Processing file ../../1001G/annotations/freeze2.1/8236.gff...
Annotation file ../../1001G/annotations/freeze2.1/8236.gff read.
Processing file ../../1001G/annotations/freeze2.1/8236.gff finished in 30.00630521774292 seconds.
Processing file ../../1001G/annotations/freeze2.1/9075.gff...
Annotation file ../../1001G/annotations/freeze2.1/9075.gff read.
Processing file ../../1001G/annotations/freeze2.1/9075.gff finished in 25.003291130065918 seconds.
Processing file ../../1001G/annotations/freeze2.1/9537.gff...
Annotation file ../../1001G/annotations/freeze2.1/9537.gff read.
Processing file ../../1001G/annotations/freeze2.1/9537.gff finished in 29.329904794692993 seconds.
Processing file ../../1001G/annotations/freeze2.1/9543.gff...
Annotation file ../../1001G/annotations/freeze2.1/9543.gff read.
Processing file ../../1001G/annotations/freeze2.1/9543.gff finished in 25.647650480270386 seconds.
Processing file ../../1001G/annotations/freeze2.1/9638.gff...
Annotation file ../../1001G/annotations/freeze2.1/9638.gff read.
Processing file ../../1001G/annotations/freeze2.1/9638.gff finished in 31.790740966796875 seconds.
Processing file ../../1001G/annotations/freeze2.1/9728.gff...
Annotation file ../../1001G/annotations/freeze2.1/9728.gff read.
Processing file ../../1001G/annotations/freeze2.1/9728.gff finished in 26.191680669784546 seconds.
Processing file ../../1001G/annotations/freeze2.1/9764.gff...
Annotation file ../../1001G/annotations/freeze2.1/9764.gff read.
Processing file ../../1001G/annotations/freeze2.1/9764.gff finished in 25.5216326713562 seconds.
Processing file ../../1001G/annotations/freeze2.1/9888.gff...
Annotation file ../../1001G/annotations/freeze2.1/9888.gff read.
Processing file ../../1001G/annotations/freeze2.1/9888.gff finished in 27.076696395874023 seconds.
Processing file ../../1001G/annotations/freeze2.1/9905.gff...
Annotation file ../../1001G/annotations/freeze2.1/9905.gff read.
Processing file ../../1001G/annotations/freeze2.1/9905.gff finished in 27.86582350730896 seconds.
Processing file ../../1001G/annotations/freeze2.1/9981.gff...
Annotation file ../../1001G/annotations/freeze2.1/9981.gff read.
Processing file ../../1001G/annotations/freeze2.1/9981.gff finished in 26.4543399810791 seconds.
Converting gene tables into pandas DataFrames.
Reading annotation files finished in 728.7494006156921 seconds
Converting gene table into graph...

Processing gene table finished in 0.005377531051635742
Converting gene table into graph...

Processing gene table finished in 0.0049724578857421875
Converting gene table into graph...

Processing gene table finished in 0.004971981048583984
Converting gene table into graph...

Processing gene table finished in 0.004911899566650391
Converting gene table into graph...

Processing gene table finished in 0.0049860477447509766
Converting gene table into graph...

Processing gene table finished in 0.005093097686767578
Converting gene table into graph...

Processing gene table finished in 0.005003690719604492
Converting gene table into graph...

Processing gene table finished in 0.00496673583984375
Converting gene table into graph...

Processing gene table finished in 0.0050051212310791016
Converting gene table into graph...

Processing gene table finished in 0.004990339279174805
Converting gene table into graph...

Processing gene table finished in 0.0053675174713134766
Converting gene table into graph...

Processing gene table finished in 0.0049898624420166016
Converting gene table into graph...

Processing gene table finished in 0.004930019378662109
Converting gene table into graph...

Processing gene table finished in 0.0050830841064453125
Converting gene table into graph...

Processing gene table finished in 0.005053043365478516
Converting gene table into graph...

Processing gene table finished in 0.005028247833251953
Converting gene table into graph...

Processing gene table finished in 0.004963874816894531
Converting gene table into graph...

Processing gene table finished in 0.005001068115234375
Converting gene table into graph...

Processing gene table finished in 0.004966020584106445
Converting gene table into graph...

Processing gene table finished in 0.00488591194152832
Converting gene table into graph...

Processing gene table finished in 0.00497126579284668
Converting gene table into graph...

Processing gene table finished in 0.004948854446411133
Converting gene table into graph...

Processing gene table finished in 0.005037784576416016
Converting gene table into graph...

Processing gene table finished in 0.004958152770996094
Converting gene table into graph...

Processing gene table finished in 0.004926204681396484
Converting gene table into graph...

Processing gene table finished in 0.005283832550048828
Converting gene table into graph...

Processing gene table finished in 0.0049571990966796875
Processing file ../../1001G/annotations/freeze2.1/outgroups/araport.gff...
Annotation file ../../1001G/annotations/freeze2.1/outgroups/araport.gff read.
Processing file ../../1001G/annotations/freeze2.1/outgroups/araport.gff finished in 60.026296615600586 seconds.
Converting gene tables into pandas DataFrames.
Reading annotation files finished in 60.15704894065857 seconds
Converting gene table into graph...

Processing gene table finished in 0.00645756721496582</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>TypeError: list indices must be integers or slices, not NoneType</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pdb</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>pdb.pm()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt; /data/YandexDisk/Kew/src/graphConstruction/pangraph_constructor/graph.py(592)_pathCount()
    590                 self.inPath[pathNode] -= 1
    591             previousNode = pathNode
--&gt; 592         self.outPath[previousNode] -= 1
    593 
    594     # We are not interested in pathIDs in *Unique blocks, so,
</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  l</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    587                                         .add(pathID)
    588             else:
    589                 self.pathStarts[pathNode] += 1
    590                 self.inPath[pathNode] -= 1
    591             previousNode = pathNode
--&gt; 592         self.outPath[previousNode] -= 1
    593 
    594     # We are not interested in pathIDs in *Unique blocks, so,
    595     # converting them back to standard format.
    596     self.nodePassUnique = [len(self.nodePassUnique.get(key,[])) for key in range(len(self.nodes))]
    597     self.inPathUnique = [len(self.inPathUnique.get(key,[])) for key in range(len(self.nodes))]
</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  previousNod</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>*** NameError: name 'previousNod' is not defined</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  previousNode
ipdb&gt;  l 587</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    582             if previousNode is not None:
    583                 self.edgePaths.setdefault((previousNode+1,pathNode+1),0)
    584                 self.edgePaths[(previousNode+1,pathNode+1)] += 1
    585 
    586                 self.edgePathsUnique.setdefault((previousNode+1,pathNode+1),set())\
    587                                         .add(pathID)
    588             else:
    589                 self.pathStarts[pathNode] += 1
    590                 self.inPath[pathNode] -= 1
    591             previousNode = pathNode
--&gt; 592         self.outPath[previousNode] -= 1
</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  l 582</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    577             self.nodeStrandPaths[pathNode][pathStrand] += 1
    578             self.outPath[pathNode] += 1
    579             self.outPathUnique.setdefault(pathNode,set()).add(pathID)
    580             self.inPath[pathNode] += 1
    581             self.inPathUnique.setdefault(pathNode,set()).add(pathID)
    582             if previousNode is not None:
    583                 self.edgePaths.setdefault((previousNode+1,pathNode+1),0)
    584                 self.edgePaths[(previousNode+1,pathNode+1)] += 1
    585 
    586                 self.edgePathsUnique.setdefault((previousNode+1,pathNode+1),set())\
    587                                         .add(pathID)
</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  l 577</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    572         for nodeStrand in path:
    573             pathNode = int(nodeStrand[:-1])-1
    574             pathStrand = int(nodeStrand[-1]=='-') # 0 if "+", 1 if "-"
    575             self.nodePass[pathNode] += 1
    576             self.nodePassUnique.setdefault(pathNode,set()).add(pathID)
    577             self.nodeStrandPaths[pathNode][pathStrand] += 1
    578             self.outPath[pathNode] += 1
    579             self.outPathUnique.setdefault(pathNode,set()).add(pathID)
    580             self.inPath[pathNode] += 1
    581             self.inPathUnique.setdefault(pathNode,set()).add(pathID)
    582             if previousNode is not None:
</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  l 573</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    568 
    569     for pathID,path in enumerate(self.paths):
    570 
    571         previousNode = None
    572         for nodeStrand in path:
    573             pathNode = int(nodeStrand[:-1])-1
    574             pathStrand = int(nodeStrand[-1]=='-') # 0 if "+", 1 if "-"
    575             self.nodePass[pathNode] += 1
    576             self.nodePassUnique.setdefault(pathNode,set()).add(pathID)
    577             self.nodeStrandPaths[pathNode][pathStrand] += 1
    578             self.outPath[pathNode] += 1
</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  nodeStrand</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>*** NameError: name 'nodeStrand' is not defined</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  pathNode</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>*** NameError: name 'pathNode' is not defined</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  path</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[]</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  self.paths</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[[], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []]</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  len(self.paths)</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>28</code></pre>
</div>
<div class="cell-output cell-output-stdin">
<pre><code>ipdb&gt;  q</code></pre>
</div>
</div>
</section>
</section>
<section id="loading-pathfile-to-graph" class="level2">
<h2 class="anchored" data-anchor-id="loading-pathfile-to-graph">Loading Pathfile to graph</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For path file v1</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'examples/gene_graph'</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>pathsfile <span class="op">=</span> <span class="st">'paths_genegraph.txt'</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>paths <span class="op">=</span> pathFileToPathDict(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>pathsfile<span class="sc">}</span><span class="ss">'</span>, <span class="va">True</span>, <span class="va">True</span>)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(pathsDict<span class="op">=</span>paths)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>graph.treeSort()</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">len</span>(graph.nodes)<span class="op">!=</span><span class="bu">len</span>(graph.order):</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'Sorting failed and not all nodes were sorted. Saving unsorted graph'</span>)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>    output <span class="op">=</span> <span class="st">'paths_genegraph_unordered.gfa'</span></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>    graph.order <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(graph.nodes)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>    graph.toGFA(output,doSeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'paths_genegraph.gfa'</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>    coregraph.toGFA(coreGFApath,doSeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For v2</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is example, no v2 file currently available for demonstration.</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'/path/to/file'</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>pathsfile <span class="op">=</span> <span class="ss">f'paths.txt'</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>paths <span class="op">=</span> pathFileToPathDict(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>pathsfile<span class="sc">}</span><span class="ss">'</span>,<span class="va">True</span>,<span class="st">'reference'</span>,<span class="va">True</span>)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seqNum <span class="kw">in</span> paths.keys():</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    graph <span class="op">=</span> GenomeGraph(pathsDict<span class="op">=</span>paths[seqNum])</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># On undirected coregraph sorting is not optimal! Check sorting!!!</span></span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>    graph.treeSort()</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(graph.nodes)<span class="op">!=</span><span class="bu">len</span>(graph.order):</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">'Sorting failed and not all nodes were sorted. Saving unsorted graph'</span>)</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>        output <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">graph_Chr</span><span class="sc">{</span>seqNum<span class="sc">}</span><span class="ss">_unordered.gfa'</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>        graph.order <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(graph.nodes)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>        graph.toGFA(output,doSeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>        coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">graph_Chr</span><span class="sc">{</span>seqNum<span class="sc">}</span><span class="ss">.gfa'</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>        graph.toGFA(coreGFApath,doSeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="loading-graph-from-gfa-and-sorting-it" class="level1">
<h1>Loading graph from GFA and sorting it</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>gfadir <span class="op">=</span> <span class="st">'examples/nucleotide_graph'</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># It is nucleotide graph. If it is not nucleotide graph, then `isSeq` variable should be changed to False.</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>gfafilename <span class="op">=</span> <span class="st">'paths_presentation.gfa'</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>gfadir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>gfafilename<span class="sc">}</span><span class="ss">'</span>,isGFASeq<span class="op">=</span>isSeq)</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>graph_new.treeSort()</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">len</span>(graph_new.nodes)<span class="op">==</span><span class="bu">len</span>(graph_new.order)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>basename,ext <span class="op">=</span> os.path.splitext(gfafilename)</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>graph_new.toGFA(<span class="ss">f'</span><span class="sc">{</span>gfadir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>basename<span class="sc">}</span><span class="ss">_ordered.</span><span class="sc">{</span>ext<span class="sc">}</span><span class="ss">'</span>,doSeq<span class="op">=</span>isSeq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exporting-presentation-graph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-presentation-graph">Exporting presentation graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../Meetings/1001G+_20220518/'</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">paths_presentation.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="changing-annotation" class="level1">
<h1>changing annotation</h1>
<p>genome = GenomeGraph(coreGFApath,isGFASeq=True)</p>
<p>for nodeID,node in enumerate(genome.nodesAnnotation): for seqName,seqDict in node.items(): for annText in seqDict.keys(): genome.nodesAnnotation[nodeID][seqName][annText] = [(0,len(genome.nodesData[nodeID])-1)]</p>
<p>genome.toGFA(coreGFApath,doSeq=True)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">16</span>]<span class="co">#,8,16,32]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>('../../Meetings/1001G+_20220518', 'paths_presentation_new', [1, 2, 4, 8, 16])</code></pre>
</div>
</div>
<p>dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName) print(f’Opening Redis connection for db {dbid}‘) redisConn = Redis(host=’redis’,port = 6379,db=dbid)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>redisConn<span class="op">=</span><span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, </span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>                                fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="co">#                                 returnDebugData=True)</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Genome
Loading graph from ../../Meetings/1001G+_20220518//paths_presentation.gfa
Found nodeNames file ../../Meetings/1001G+_20220518/nodeNames_paths_presentation.json, loading names.
Found node annotation file ../../Meetings/1001G+_20220518/annotation_paths_presentation.dat, loading associations.
Loading segment 7/7
Loading segments finished.
Loading link 12/12
Loading links finished
Loading path 5/5
Loading paths finished. 5 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths

Recording Pantograph data to ../../Meetings/1001G+_20220518/paths_presentation_new
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths
Processing path breaks...
Postprocessing interconnected links 2/2
Preprocessing interconnected links finished.

Processing path breaks finished.
Converting blocks to block lengths 5/5
Conversion finished.
Reformating links to block lengths associations 5/5
Reformating finished.
Identifying rearrangement blocks
Processing node 7/7
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7/7
Processing component links 6/6
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks
Recording component 6/6
Recording zoom level 1 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;2 were removed.     0 components were deleted as isolated.

===========================
Zoom level 2
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 2 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;4 were removed.     0 components were deleted as isolated.

===========================
Zoom level 4
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 4 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;8 were removed.     0 components were deleted as isolated.

===========================
Zoom level 8
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 8 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;16 were removed.     0 components were deleted as isolated.

===========================
Zoom level 16
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 16 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;32 were removed.     0 components were deleted as isolated.

===========================
Zoom level 32
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 32 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;64 were removed.     0 components were deleted as isolated.

===========================
Zoom level 64
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 64 finished.</code></pre>
</div>
</div>
<section id="exporting-tutorial-gene-graph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-tutorial-gene-graph">Exporting tutorial gene graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/pantograph'</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">genegraph_tutorial.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading graph from ../../1001G/pantograph/genegraph_tutorial.gfa
Found node annotation file ../../1001G/pantograph/annotation_genegraph_tutorial.dat, loading associations.
Loading segment 7/7
Loading segments finished.
Loading link 13/13
Loading links finished
Loading path 5/5
Loading paths finished. 5 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>graph.paths</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[['1+', '2+', '3+', '4+'],
 ['1+', '5+', '6+', '7+', '1+'],
 ['4+', '1+', '2+', '3+'],
 ['2+', '3+', '4+', '1+', '5-', '4+', '7+'],
 ['6+', '7-', '1+', '2+', '3+']]</code></pre>
</div>
</div>
</section>
</section>
<section id="changing-annotation-1" class="level1">
<h1>changing annotation</h1>
<p>genome = GenomeGraph(coreGFApath,isGFASeq=True)</p>
<p>for nodeID,node in enumerate(genome.nodesAnnotation): for seqName,seqDict in node.items(): for annText in seqDict.keys(): genome.nodesAnnotation[nodeID][seqName][annText] = [(0,len(genome.nodesData[nodeID])-1)]</p>
<p>genome.toGFA(coreGFApath,doSeq=True)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]<span class="co">#,8,16,32]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>('../../1001G/pantograph', 'genegraph_tutorial', [1])</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="co"># print(f'Opening Redis connection for db {dbid}')</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host <span class="op">=</span> <span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>, db <span class="op">=</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>redisConn=None</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, </span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>                                fillZoomLevels <span class="op">=</span> <span class="va">False</span>,</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a><span class="co">#                                 returnDebugData=True)</span></span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Genome
Loading graph from ../../1001G/pantograph/genegraph_tutorial.gfa
Found node annotation file ../../1001G/pantograph/annotation_genegraph_tutorial.dat, loading associations.
Loading segment 7/7
Loading segments finished.
Loading link 13/13
Loading links finished
Loading path 5/5
Loading paths finished. 5 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/genegraph_tutorial
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7/7
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7/7
Processing component links 6/6
Recording component 6/6
Recording zoom level 1 finished.</code></pre>
</div>
</div>
<section id="testing-removable-elements-identification" class="level3">
<h3 class="anchored" data-anchor-id="testing-removable-elements-identification">Testing removable elements identification</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pangraph_constructor.exportDev <span class="im">import</span> getRemovableStructures</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> IPython.display <span class="im">as</span> ipd</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> joblib</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.path.exists(<span class="st">'./tests/breakIdentify.dat'</span>):</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>    correctResults <span class="op">=</span> joblib.load(<span class="st">'./tests/breakIdentify.dat'</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    correctResults <span class="op">=</span> {}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> <span class="st">'../../1001G/GraphCollapsing/TestGraphs'</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>filePrefix <span class="op">=</span> <span class="st">'test'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co">#test</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> filename,resDict <span class="kw">in</span> correctResults.items():</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="ch">\n</span><span class="ss">####### Testing on case </span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss"> ########'</span>)</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    graph <span class="op">=</span> GenomeGraph(<span class="ss">f'</span><span class="sc">{</span>path<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss">'</span>,isGFASeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    linksLengths, pairedLinks, blockEdges, _ <span class="op">=</span> getRemovableStructures(graph<span class="op">=</span>graph)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> linksLengths<span class="op">==</span>resDict[<span class="st">'linksLengths'</span>]</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> pairedLinks<span class="op">==</span>resDict[<span class="st">'pairedLinks'</span>]</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> blockEdges<span class="op">==</span>resDict[<span class="st">'blockEdges'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>caseNum <span class="op">=</span> <span class="dv">17</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>filePrefix<span class="sc">}{</span>caseNum<span class="sc">:02d}</span><span class="ss">.gfa'</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'############'</span>)</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Graph from file </span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>path<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Graph Paths:'</span>)</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>ipd.display(graph.paths)</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>linkLengths, pairedLinks, blockEdges, _ <span class="op">=</span> getRemovableStructures(graph<span class="op">=</span>graph)</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Link-Lengths associations:'</span>)</span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>ipd.display(linkLengths)</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'PairedLinks:'</span>)</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>ipd.display(pairedLinks)</span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Rearrangemenet block edges:'</span>)</span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a>ipd.display(blockEdges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>correctResults[filename]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>correctResults[filename] <span class="op">=</span> {<span class="st">'linksLengths'</span>:linkLengths,<span class="st">'pairedLinks'</span>:pairedLinks,<span class="st">'blockEdges'</span>:blockEdges}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>joblib.dump(correctResults,<span class="st">'./tests/breakIdentify.dat'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-test-collapse-graph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-test-collapse-graph">Exporting test collapse graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/pantograph/data'</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">testCollapse.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>]<span class="co">#,8,16,32]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName) print(f’Opening Redis connection for db {dbid}‘) redisConn = Redis(host=’redis’,port = 6379,db=dbid)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>redisConn<span class="op">=</span><span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>[zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, </span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>                invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] <span class="op">=</span> <span class="op">\</span></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>                                fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold,</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a>                                returnDebugData<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Executed in </span><span class="sc">{</span>runTime<span class="sc">}</span><span class="ss"> seconds'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Exporting test collapse graph finished. Overall time = </span><span class="sc">{runTime}</span><span class="st"> seconds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-coregraph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-coregraph">Exporting coregraph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_Chr1.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">16</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>('../../1001G/coreGraph', 'coregraph_Chr1_new', [1, 2, 4, 8, 16])</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="co"># print(f'Opening Redis connection for db {dbid}')</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Converted 00_init.ipynb.
Converted 01_graph.ipynb.
Converted 02_tree.ipynb.
Converted 03_synteny.ipynb.
Converted 04_utils.ipynb.
Converted 05_export.ipynb.
Converted 05_exportDev.ipynb.
Converted deBruijnGraphProcessing.ipynb.
Converted dev.ipynb.
Converted graphTesting.ipynb.
Converted index.ipynb.</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels, </span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a><span class="co">#                                 returnDebugata=True)</span></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(runTime)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Genome
Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa
Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.
Loading segment 35/35
Loading segments finished.
Loading link 72/72
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 35/35
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/coreGraph/coregraph_Chr1_new
Calculating nodes length...
Processing node 35/35
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Processing path breaks...
Postprocessing interconnected links 16/16
Preprocessing interconnected links finished.

Processing path breaks finished.
Converting blocks to block lengths 34/34
Conversion finished.
Reformating links to block lengths associations 34/34
Reformating finished.
Identifying rearrangement blocks
Processing node 35/35
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 35/35
Processing component links 35/35
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks
Recording component 35/35
Recording zoom level 1 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;2 were removed.     0 components were deleted as isolated.

===========================
Zoom level 2
===========================
Processing component 35/35
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 8/8
Recording zoom level 2 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;4 were removed.     0 components were deleted as isolated.

===========================
Zoom level 4
===========================
Processing component 8/8
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 5/5
Recording zoom level 4 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;8 were removed.     0 components were deleted as isolated.

===========================
Zoom level 8
===========================
Processing component 5/5
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 1/1
Recording zoom level 8 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;16 were removed.     0 components were deleted as isolated.

===========================
Zoom level 16
===========================
Processing component 1/1
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 1/1
Recording zoom level 16 finished.
27.11550521850586</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Exporting coregraph finished. Overall time = </span><span class="sc">{runTime}</span><span class="st"> seconds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.11) or chardet (3.0.4) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "</code></pre>
</div>
</div>
<p>❗️❗️❗️ TODO: Next test API. WHen works, change front end to get jsons without annotation and then to load annotation from API!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa
Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.
Loading segment 35/35
Loading segments finished.
Loading link 72/72
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 35/35
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pangraph_constructor.utils <span class="im">import</span> iset_get,iset_score,iset_add</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port<span class="op">=</span><span class="dv">6379</span>,db<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>iset_get(redisConn,<span class="st">'test'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'a_0': (1.0, 3.0), 'b_0': (1.0, 5.0)}</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>iset_score(redisConn,<span class="st">'coregraph_Chr1_new.1.10024.Pos'</span>,<span class="dv">22</span>)</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="co"># coregraph_Chr1_new/10024/1/22</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['23']</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>iset_score(redisConn,<span class="st">'coregraph_Chr1_new.9543.Gene'</span>,<span class="dv">50</span>,<span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['3370',
 'AT1G15940',
 'AT1G15950',
 'AT1G16000',
 'AT1G16010',
 'AT1G16022',
 'AT1G62830',
 'AT1G62840',
 'OG0001502',
 'OG0004404',
 'OG0007340',
 'OG0007341',
 'OG0007343',
 'OG0008435',
 'OG0015400',
 'OG0017985',
 'OG0022841',
 'OG0024046',
 'OG0024055',
 'OG0025924',
 'OG0027925',
 'OG0030893']</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>iset_get(redisConn,<span class="st">'coregraph_Chr1_new.9543.Gene'</span>,<span class="st">'3370_0'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'3370_0': (48.0, 51.0)}</code></pre>
</div>
</div>
<p>redisConn=None</p>
</section>
<section id="exporting-coregraph-with-genes" class="level3">
<h3 class="anchored" data-anchor-id="exporting-coregraph-with-genes">Exporting coregraph with genes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">9</span>]<span class="co"># + [9*2**i for i in range(9)]</span></span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [2**i for i in range(12)]#  [1,2,4,8,16,32,9*16,9*32,9*128,9*256]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName = pathConvert(coreGFApath,suffix='_new')</span></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName,zoomLevels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName) print(f’Opening Redis connection for db {dbid}‘) redisConn = Redis(host=’redis’,port = 6379,db=dbid)</p>
<p>redisConn=None</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># %%capture output</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrNum <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">6</span>):</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_genes_Chr</span><span class="sc">{</span>chrNum<span class="sc">}</span><span class="ss">.gfa'</span></span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>    outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>    dbid <span class="op">=</span> getDBID(<span class="st">'../pantograph_API/data/caseToDBID.dict'</span>,outputName)</span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Opening Redis connection for db </span><span class="sc">{</span>dbid<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb106-9"><a href="#cb106-9" aria-hidden="true" tabindex="-1"></a>    redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span>dbid)</span>
<span id="cb106-10"><a href="#cb106-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb106-11"><a href="#cb106-11" aria-hidden="true" tabindex="-1"></a>    startTime <span class="op">=</span> time.time()</span>
<span id="cb106-12"><a href="#cb106-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb106-13"><a href="#cb106-13" aria-hidden="true" tabindex="-1"></a>    exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb106-14"><a href="#cb106-14" aria-hidden="true" tabindex="-1"></a>                                    outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb106-15"><a href="#cb106-15" aria-hidden="true" tabindex="-1"></a>                                    isSeq<span class="op">=</span>isSeq,</span>
<span id="cb106-16"><a href="#cb106-16" aria-hidden="true" tabindex="-1"></a>                                    redisConn<span class="op">=</span>redisConn,</span>
<span id="cb106-17"><a href="#cb106-17" aria-hidden="true" tabindex="-1"></a>                                    zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb106-18"><a href="#cb106-18" aria-hidden="true" tabindex="-1"></a>                                    fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb106-19"><a href="#cb106-19" aria-hidden="true" tabindex="-1"></a>                                    maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb106-20"><a href="#cb106-20" aria-hidden="true" tabindex="-1"></a>                                    maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb106-21"><a href="#cb106-21" aria-hidden="true" tabindex="-1"></a>                                    inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb106-22"><a href="#cb106-22" aria-hidden="true" tabindex="-1"></a>    <span class="co">#                                 returnDebugata=True)</span></span>
<span id="cb106-23"><a href="#cb106-23" aria-hidden="true" tabindex="-1"></a>    runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>curT <span class="op">=</span> time.localtime(time.time()<span class="op">+</span><span class="dv">3600</span>)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>message <span class="op">=</span> <span class="ss">f"Exporting core graph with genes for all chromosomes finished at </span><span class="ch">\</span></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="ss">            </span><span class="sc">{</span>curT<span class="sc">.</span>tm_hour<span class="sc">:02d}</span><span class="ss">:</span><span class="sc">{</span>curT<span class="sc">.</span>tm_min<span class="sc">:02d}</span><span class="ss"> on </span><span class="sc">{</span>curT<span class="sc">.</span>tm_mday<span class="sc">:02d}</span><span class="ss">/</span><span class="sc">{</span>curT<span class="sc">.</span>tm_mon<span class="sc">:02d}</span><span class="ss">!</span><span class="ch">\n</span><span class="ss">"</span></span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"</span><span class="sc">{message}</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName = pathConvert(coreGFApath,suffix='_new')</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName,zoomLevels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output2</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrNum <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">6</span>):</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_Chr</span><span class="sc">{</span>chrNum<span class="sc">}</span><span class="ss">.gfa'</span></span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>    outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a>    dbid <span class="op">=</span> getDBID(<span class="st">'../pantograph_API/data/caseToDBID.dict'</span>,outputName)</span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Opening Redis connection for db </span><span class="sc">{</span>dbid<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb111-9"><a href="#cb111-9" aria-hidden="true" tabindex="-1"></a>    redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span>dbid)</span>
<span id="cb111-10"><a href="#cb111-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb111-11"><a href="#cb111-11" aria-hidden="true" tabindex="-1"></a>    startTime <span class="op">=</span> time.time()</span>
<span id="cb111-12"><a href="#cb111-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb111-13"><a href="#cb111-13" aria-hidden="true" tabindex="-1"></a>    exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb111-14"><a href="#cb111-14" aria-hidden="true" tabindex="-1"></a>                                    outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb111-15"><a href="#cb111-15" aria-hidden="true" tabindex="-1"></a>                                    isSeq<span class="op">=</span>isSeq,</span>
<span id="cb111-16"><a href="#cb111-16" aria-hidden="true" tabindex="-1"></a>                                    redisConn<span class="op">=</span>redisConn,</span>
<span id="cb111-17"><a href="#cb111-17" aria-hidden="true" tabindex="-1"></a>                                    zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb111-18"><a href="#cb111-18" aria-hidden="true" tabindex="-1"></a>                                    fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb111-19"><a href="#cb111-19" aria-hidden="true" tabindex="-1"></a>                                    maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb111-20"><a href="#cb111-20" aria-hidden="true" tabindex="-1"></a>                                    maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb111-21"><a href="#cb111-21" aria-hidden="true" tabindex="-1"></a>                                    inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb111-22"><a href="#cb111-22" aria-hidden="true" tabindex="-1"></a>    <span class="co">#                                 returnDebugata=True)</span></span>
<span id="cb111-23"><a href="#cb111-23" aria-hidden="true" tabindex="-1"></a>    runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span>
<span id="cb111-24"><a href="#cb111-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb111-25"><a href="#cb111-25" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Exporting core graph for Chr</span><span class="sc">{</span>chrNum<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>runTime<span class="sc">}</span><span class="ss"> seconds'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>curT <span class="op">=</span> time.localtime(time.time()<span class="op">+</span><span class="dv">3600</span>)</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>message <span class="op">=</span> <span class="ss">f"Exporting core graph with genes for all chromosomes finished at </span><span class="ch">\</span></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="ss"> </span><span class="sc">{</span>curT<span class="sc">.</span>tm_hour<span class="sc">:02d}</span><span class="ss">:</span><span class="sc">{</span>curT<span class="sc">.</span>tm_min<span class="sc">:02d}</span><span class="ss"> on </span><span class="sc">{</span>curT<span class="sc">.</span>tm_mday<span class="sc">:02d}</span><span class="ss">/</span><span class="sc">{</span>curT<span class="sc">.</span>tm_mon<span class="sc">:02d}</span><span class="ss">!</span><span class="ch">\n</span><span class="ss">"</span></span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"</span><span class="sc">{message}</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Exporting coregraph with genes finished. Overall time = </span><span class="sc">{runTime}</span><span class="st"> seconds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-gene-graphs-for-all-chromosomes" class="level3">
<h3 class="anchored" data-anchor-id="exporting-gene-graphs-for-all-chromosomes">Exporting gene graphs for all chromosomes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/pantograph/data'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]<span class="co">#,2,4]# + [9*2**i for i in range(12)]</span></span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a>fillZoomLevel <span class="op">=</span> <span class="va">False</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>nbdev_export()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>caseDict <span class="op">=</span> {}</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">6</span>):</span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>    GFA <span class="op">=</span> <span class="ss">f'AT_Chr</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">_OGOnly_2.1.gfa'</span></span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a>    caseDict[<span class="ss">f'Chr</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">'</span>] <span class="op">=</span> GFA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'Chr1': 'AT_Chr1_OGOnly_2.1.gfa',
 'Chr2': 'AT_Chr2_OGOnly_2.1.gfa',
 'Chr3': 'AT_Chr3_OGOnly_2.1.gfa',
 'Chr4': 'AT_Chr4_OGOnly_2.1.gfa',
 'Chr5': 'AT_Chr5_OGOnly_2.1.gfa'}</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>exportProject(projectID <span class="op">=</span> <span class="st">'AT_OGOnly_2.1'</span>, projectName <span class="op">=</span> <span class="st">'A. thaliana 27 genomes + TAIR10 gene graph 2.1'</span>, </span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>              caseDict <span class="op">=</span> caseDict, pathToIndex <span class="op">=</span> pathfileDir, pathToGraphs <span class="op">=</span> pathfileDir,</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>              redisHost<span class="op">=</span><span class="st">'redis'</span>, redisPort <span class="op">=</span> <span class="dv">6379</span>, redisDB <span class="op">=</span> <span class="dv">0</span>,</span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>              suffix <span class="op">=</span> <span class="st">'pr'</span>,</span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>              maxLengthComponent <span class="op">=</span> maxLengthComponent, maxLengthChunk <span class="op">=</span> maxLengthChunk,</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>              inversionThreshold <span class="op">=</span> inversionThreshold,</span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>              isSeq <span class="op">=</span> isSeq, zoomLevels <span class="op">=</span> zoomLevels, fillZoomLevel <span class="op">=</span> fillZoomLevel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing case AT_OGOnly_2.1_Chr1pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
Loading segment 7455/7455
Loading segments finished.
Loading link 14001/14001
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr1pr
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7455/7455
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7455/7455
Processing component links 3869/3869
Recording component 3869/3869
Recording zoom level 1 finished.
Exporting gene graph for Chr1 took 465.6470103263855 seconds

Processing case AT_OGOnly_2.1_Chr2pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.
Loading segment 4877/4877
Loading segments finished.
Loading link 10302/10302
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr2pr
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4877/4877
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4877/4877
Processing component links 2534/2534
Recording component 2534/2534
Recording zoom level 1 finished.
Exporting gene graph for Chr2 took 214.97250485420227 seconds

Processing case AT_OGOnly_2.1_Chr3pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.
Loading segment 8442/8442
Loading segments finished.
Loading link 15985/15985
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr3pr
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 8442/8442
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 8442/8442
Processing component links 3383/3383
Recording component 3383/3383
Recording zoom level 1 finished.
Exporting gene graph for Chr3 took 430.0764102935791 seconds

Processing case AT_OGOnly_2.1_Chr4pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.
Loading segment 4620/4620
Loading segments finished.
Loading link 9980/9980
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr4pr
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4620/4620
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4620/4620
Processing component links 2526/2526
Recording component 2526/2526
Recording zoom level 1 finished.
Exporting gene graph for Chr4 took 199.40259623527527 seconds

Processing case AT_OGOnly_2.1_Chr5pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.
Loading segment 7052/7052
Loading segments finished.
Loading link 12737/12737
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr5pr
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7052/7052
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7052/7052
Processing component links 3274/3274
Recording component 3274/3274
Recording zoom level 1 finished.
Exporting gene graph for Chr5 took 348.297509431839 seconds</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seqID <span class="kw">in</span> [<span class="st">'Chr1'</span>,<span class="st">'Chr2'</span>,<span class="st">'Chr3'</span>,<span class="st">'Chr4'</span>,<span class="st">'Chr5'</span>]:<span class="co">#'Chr1','Chr2',</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">AT_</span><span class="sc">{</span>seqID<span class="sc">}</span><span class="ss">_OGOnly_2.1.gfa'</span></span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>    outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_viscol'</span>)</span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>()</span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Processing case </span><span class="sc">{</span>outputName<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)</span></span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print(f'Opening Redis connection for db {dbid}')</span></span>
<span id="cb121-11"><a href="#cb121-11" aria-hidden="true" tabindex="-1"></a>    redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb121-12"><a href="#cb121-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-13"><a href="#cb121-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">#%%capture output</span></span>
<span id="cb121-14"><a href="#cb121-14" aria-hidden="true" tabindex="-1"></a>    startTime <span class="op">=</span> time.time()</span>
<span id="cb121-15"><a href="#cb121-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># initialLinkLengths, initialPairedLinks, initialInterconnectedLinks, initialBlockEdges, \</span></span>
<span id="cb121-16"><a href="#cb121-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomNodeToComponent,zoomComponentToNodes,zoomComponents,\</span></span>
<span id="cb121-17"><a href="#cb121-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomFromComponentLinks, zoomToComponentLinks, zoomLinkLengths, zoomPairedLinks, zoomInterconnectedLinks, \</span></span>
<span id="cb121-18"><a href="#cb121-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomOldToNewRemoval, zoomNewToOldRemoval, \</span></span>
<span id="cb121-19"><a href="#cb121-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomLinkLengthsRemoval, zoomPairedLinksRemoval, zoomInterconnectedLinksRemoval, zoomBlockEdgesRemoval, \</span></span>
<span id="cb121-20"><a href="#cb121-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomFromComponentLinksRemoval, zoomToComponentLinksRemoval, \</span></span>
<span id="cb121-21"><a href="#cb121-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># graph = \</span></span>
<span id="cb121-22"><a href="#cb121-22" aria-hidden="true" tabindex="-1"></a>    exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb121-23"><a href="#cb121-23" aria-hidden="true" tabindex="-1"></a>                                    outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb121-24"><a href="#cb121-24" aria-hidden="true" tabindex="-1"></a>                                    isSeq<span class="op">=</span>isSeq,</span>
<span id="cb121-25"><a href="#cb121-25" aria-hidden="true" tabindex="-1"></a>                                    redisConn<span class="op">=</span>redisConn,</span>
<span id="cb121-26"><a href="#cb121-26" aria-hidden="true" tabindex="-1"></a>                                    zoomLevels<span class="op">=</span>zoomLevels, </span>
<span id="cb121-27"><a href="#cb121-27" aria-hidden="true" tabindex="-1"></a>                                    fillZoomLevels <span class="op">=</span> <span class="va">False</span>,</span>
<span id="cb121-28"><a href="#cb121-28" aria-hidden="true" tabindex="-1"></a>                                    maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb121-29"><a href="#cb121-29" aria-hidden="true" tabindex="-1"></a>                                    maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb121-30"><a href="#cb121-30" aria-hidden="true" tabindex="-1"></a>                                    inversionThreshold<span class="op">=</span>inversionThreshold)<span class="co">#,</span></span>
<span id="cb121-31"><a href="#cb121-31" aria-hidden="true" tabindex="-1"></a>                                    <span class="co"># returnDebugData=True)</span></span>
<span id="cb121-32"><a href="#cb121-32" aria-hidden="true" tabindex="-1"></a>    runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span>
<span id="cb121-33"><a href="#cb121-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb121-34"><a href="#cb121-34" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Exporting gene graph for </span><span class="sc">{</span>seqID<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>runTime<span class="sc">}</span><span class="ss"> seconds'</span>)</span>
<span id="cb121-35"><a href="#cb121-35" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb121-36"><a href="#cb121-36" aria-hidden="true" tabindex="-1"></a>    redisConn.close()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing case AT_Chr1_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
Loading segment 7455/7455
Loading segments finished.
Loading link 14001/14001
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7455/7455
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7455/7455
Processing component links 3869/3869
Recording component 3869/3869
Recording zoom level 1 finished.
Exporting gene graph for Chr1 took 445.7378799915314 seconds

Processing case AT_Chr2_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.
Loading segment 4877/4877
Loading segments finished.
Loading link 10302/10302
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4877/4877
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4877/4877
Processing component links 2534/2534
Recording component 2534/2534
Recording zoom level 1 finished.
Exporting gene graph for Chr2 took 193.71776700019836 seconds

Processing case AT_Chr3_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.
Loading segment 8442/8442
Loading segments finished.
Loading link 15985/15985
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 8442/8442
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 8442/8442
Processing component links 3383/3383
Recording component 3383/3383
Recording zoom level 1 finished.
Exporting gene graph for Chr3 took 411.77689695358276 seconds

Processing case AT_Chr4_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.
Loading segment 4620/4620
Loading segments finished.
Loading link 9980/9980
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4620/4620
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4620/4620
Processing component links 2526/2526
Recording component 2526/2526
Recording zoom level 1 finished.
Exporting gene graph for Chr4 took 180.36719918251038 seconds

Processing case AT_Chr5_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.
Loading segment 7052/7052
Loading segments finished.
Loading link 12737/12737
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7052/7052
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7052/7052
Processing component links 3274/3274
Recording component 3274/3274
Recording zoom level 1 finished.
Exporting gene graph for Chr5 took 338.8390169143677 seconds</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>curT <span class="op">=</span> time.localtime()</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>message <span class="op">=</span> <span class="ss">f"Exporting gene graph for all chromosomes finished at </span><span class="ch">\</span></span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a><span class="ss">            </span><span class="sc">{</span>curT<span class="sc">.</span>tm_hour<span class="sc">:02d}</span><span class="ss">:</span><span class="sc">{</span>curT<span class="sc">.</span>tm_min<span class="sc">:02d}</span><span class="ss"> on </span><span class="sc">{</span>curT<span class="sc">.</span>tm_mday<span class="sc">:02d}</span><span class="ss">/</span><span class="sc">{</span>curT<span class="sc">.</span>tm_mon<span class="sc">:02d}</span><span class="ss">!"</span></span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"</span><span class="sc">{message}</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">AT_Chr1_OGOnly_2.1.gfa'</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>coreGFApath, isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
Loading segment 7455/7455
Loading segments finished.
Loading link 14001/14001
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>graph.nodesMetadata[<span class="dv">1361</span>][<span class="st">'22006'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'genPos': [{'chr': 'Chr1', 'genomePosition': [19020282, 19020893]},
  {'chr': 'Chr1', 'genomePosition': [19022280, 19023284]}],
 'annotation': {'OG0001585': [(0, 0)],
  'AT1G55230': [(0, 0)],
  'AT1G55240': [(0, 0)]}}</code></pre>
</div>
</div>
<p>import joblib</p>
<p>joblib.dump(initialLinkLengths,‘dumps/initialLinkLengths.dump’) joblib.dump(initialPairedLinks,‘dumps/initialPairedLinks.dump’) joblib.dump(initialInterconnectedLinks,‘dumps/initialInterconnectedLinks.dump’) joblib.dump(initialBlockEdges,‘dumps/initialBlockEdges.dump’) joblib.dump(zoomNodeToComponent,‘dumps/zoomNodeToComponent.dump’) joblib.dump(zoomComponentToNodes,‘dumps/zoomComponentToNodes.dump’) joblib.dump(zoomComponents,‘dumps/zoomComponents.dump’) joblib.dump(zoomFromComponentLinks,‘dumps/zoomFromComponentLinks.dump’) joblib.dump(zoomToComponentLinks,‘dumps/zoomToComponentLinks.dump’) joblib.dump(zoomLinkLengths,‘dumps/zoomLinkLengths.dump’) joblib.dump(zoomPairedLinks,‘dumps/zoomPairedLinks.dump’) joblib.dump(zoomInterconnectedLinks,‘dumps/zoomInterconnectedLinks.dump’) joblib.dump(zoomOldToNewRemoval,‘dumps/zoomOldToNewRemoval.dump’) joblib.dump(zoomNewToOldRemoval,‘dumps/zoomNewToOldRemoval.dump’) joblib.dump(zoomLinkLengthsRemoval,‘dumps/zoomLinkLengthsRemoval.dump’) joblib.dump(zoomPairedLinksRemoval,‘dumps/zoomPairedLinksRemoval.dump’) joblib.dump(zoomInterconnectedLinksRemoval,‘dumps/zoomInterconnectedLinksRemoval.dump’) joblib.dump(zoomBlockEdgesRemoval,‘dumps/zoomBlockEdgesRemoval.dump’) joblib.dump(zoomFromComponentLinksRemoval,‘dumps/zoomFromComponentLinksRemoval.dump’) joblib.dump(zoomToComponentLinksRemoval,‘dumps/zoomToComponentLinksRemoval.dump’) joblib.dump(graph,‘dumps/graph.dump’)</p>
<p>initialLinkLengths = joblib.load(‘dumps/initialLinkLengths.dump’)</p>
<p>initialPairedLinks = joblib.load(‘dumps/initialPairedLinks.dump’)</p>
<p>initialInterconnectedLinks = joblib.load(‘dumps/initialInterconnectedLinks.dump’)</p>
<p>initialBlockEdges = joblib.load(‘dumps/initialBlockEdges.dump’)</p>
<p>zoomNodeToComponent = joblib.load(‘dumps/zoomNodeToComponent.dump’)</p>
<p>zoomComponentToNodes = joblib.load(‘dumps/zoomComponentToNodes.dump’)</p>
<p>zoomComponents = joblib.load(‘dumps/zoomComponents.dump’)</p>
<p>zoomFromComponentLinks = joblib.load(‘dumps/zoomFromComponentLinks.dump’)</p>
<p>zoomToComponentLinks = joblib.load(‘dumps/zoomToComponentLinks.dump’)</p>
<p>zoomLinkLengths = joblib.load(‘dumps/zoomLinkLengths.dump’)</p>
<p>zoomPairedLinks = joblib.load(‘dumps/zoomPairedLinks.dump’)</p>
<p>zoomInterconnectedLinks = joblib.load(‘dumps/zoomInterconnectedLinks.dump’)</p>
<p>zoomOldToNewRemoval = joblib.load(‘dumps/zoomOldToNewRemoval.dump’)</p>
<p>zoomNewToOldRemoval = joblib.load(‘dumps/zoomNewToOldRemoval.dump’)</p>
<p>zoomLinkLengthsRemoval = joblib.load(‘dumps/zoomLinkLengthsRemoval.dump’)</p>
<p>zoomPairedLinksRemoval = joblib.load(‘dumps/zoomPairedLinksRemoval.dump’)</p>
<p>zoomInterconnectedLinksRemoval = joblib.load(‘dumps/zoomInterconnectedLinksRemoval.dump’)</p>
<p>zoomBlockEdgesRemoval = joblib.load(‘dumps/zoomBlockEdgesRemoval.dump’)</p>
<p>zoomFromComponentLinksRemoval = joblib.load(‘dumps/zoomFromComponentLinksRemoval.dump’)</p>
<p>zoomToComponentLinksRemoval = joblib.load(‘dumps/zoomToComponentLinksRemoval.dump’)</p>
<p>graph = joblib.load(‘dumps/graph.dump’)</p>
</section>
<section id="comments" class="level3">
<h3 class="anchored" data-anchor-id="comments">Comments</h3>
<p>Each node processing time increase significantly with overall number of nodes. This is wrong and should be investigated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> copy <span class="im">import</span> deepcopy</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coreGraph_f2.1_Ref_v04.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>coregraph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>coregraph_genes <span class="op">=</span> deepcopy(coregraph)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>fullGraphPath <span class="op">=</span> <span class="st">'../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa'</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>fullgraph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>fullGraphPath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>chainToGenesFile <span class="op">=</span> <span class="st">'chain2gene_f2.1_Ref_v04.txt'</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>maxChainLength <span class="op">=</span> <span class="bu">len</span>(coregraph.nodes[<span class="dv">0</span>])</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>chainToListDict <span class="op">=</span> {}</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>chainToGenesFile<span class="sc">}</span><span class="ss">'</span>) <span class="im">as</span> f:</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> f:</span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a>        chainName, geneList <span class="op">=</span> line.split(<span class="st">':'</span>)</span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a>        geneList <span class="op">=</span> geneList.lstrip().rstrip().split(<span class="st">','</span>)</span>
<span id="cb132-8"><a href="#cb132-8" aria-hidden="true" tabindex="-1"></a>        chainToListDict[chainName.zfill(maxChainLength)] <span class="op">=</span> geneList</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> nodeIdx,nodeName <span class="kw">in</span> <span class="bu">enumerate</span>(coregraph.nodes):</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="ch">\n</span><span class="ss">Node </span><span class="sc">{</span>nodeIdx<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">/</span><span class="sc">{</span><span class="bu">len</span>(coregraph.nodes)<span class="sc">}</span><span class="ss">'</span>,end<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>    geneList <span class="op">=</span> chainToListDict.get(nodeName.zfill(maxChainLength), [<span class="ss">f'ch</span><span class="sc">{</span>nodeName<span class="sc">.</span>zfill(<span class="dv">7</span>)<span class="sc">}</span><span class="ss">'</span>])</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>    geneIds <span class="op">=</span> []</span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> geneList[<span class="dv">0</span>][:<span class="dv">2</span>]<span class="op">!=</span><span class="st">'ch'</span>:</span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>        geneIds <span class="op">=</span> [<span class="bu">int</span>(gene.rstrip(<span class="st">'+'</span>))<span class="op">-</span><span class="dv">1</span> <span class="cf">for</span> gene <span class="kw">in</span> geneList]</span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>        geneList <span class="op">=</span> [fullgraph.nodes[geneid] <span class="cf">for</span> geneid <span class="kw">in</span> geneIds]</span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a>    coregraph_genes.nodesData[nodeIdx] <span class="op">=</span> <span class="st">''</span>.join(geneList)</span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> accession, chainDict <span class="kw">in</span> coregraph.nodesAnnotation[nodeIdx].items():</span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a>        interval <span class="op">=</span> chainDict[nodeName]</span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a>        geneCumLengths <span class="op">=</span> np.hstack((<span class="dv">0</span>, np.cumsum([<span class="bu">len</span>(gene) <span class="cf">for</span> gene <span class="kw">in</span> geneList])))</span>
<span id="cb133-13"><a href="#cb133-13" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb133-14"><a href="#cb133-14" aria-hidden="true" tabindex="-1"></a>        coregraph.nodesAnnotation[nodeIdx][accession].<span class="op">\</span></span>
<span id="cb133-15"><a href="#cb133-15" aria-hidden="true" tabindex="-1"></a>            update({geneAnnotation:interval <span class="op">\</span></span>
<span id="cb133-16"><a href="#cb133-16" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> geneid <span class="kw">in</span> geneIds <span class="op">\</span></span>
<span id="cb133-17"><a href="#cb133-17" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">for</span> geneAnnotation <span class="kw">in</span> fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:<span class="va">None</span>}).keys()})</span>
<span id="cb133-18"><a href="#cb133-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-19"><a href="#cb133-19" aria-hidden="true" tabindex="-1"></a>        coregraph_genes.nodesAnnotation[nodeIdx][accession].<span class="op">\</span></span>
<span id="cb133-20"><a href="#cb133-20" aria-hidden="true" tabindex="-1"></a>            update({geneAnnotation:[(geneCumLengths[i], geneCumLengths[i<span class="op">+</span><span class="dv">1</span>]<span class="op">-</span><span class="dv">1</span>)] <span class="op">\</span></span>
<span id="cb133-21"><a href="#cb133-21" aria-hidden="true" tabindex="-1"></a>                            <span class="cf">for</span> i,geneid <span class="kw">in</span> <span class="bu">enumerate</span>(geneIds) <span class="op">\</span></span>
<span id="cb133-22"><a href="#cb133-22" aria-hidden="true" tabindex="-1"></a>                                <span class="cf">for</span> geneAnnotation <span class="kw">in</span> fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:<span class="va">None</span>}).keys()})</span>
<span id="cb133-23"><a href="#cb133-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-24"><a href="#cb133-24" aria-hidden="true" tabindex="-1"></a>        coregraph_genes.nodesAnnotation[nodeIdx][accession].<span class="op">\</span></span>
<span id="cb133-25"><a href="#cb133-25" aria-hidden="true" tabindex="-1"></a>            update({nodeName:[(geneCumLengths[<span class="dv">0</span>], geneCumLengths[<span class="op">-</span><span class="dv">1</span>]<span class="op">-</span><span class="dv">1</span>)]})</span>
<span id="cb133-26"><a href="#cb133-26" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">''</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>coregraph.toGFA(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_f2.1_Ref_v04.gfa'</span>,doSeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>coregraph_genes.toGFA(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_genes_f2.1_Ref_v04.gfa'</span>,doSeq<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_genes_f2.1_Ref_v04.gfa'</span></span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a><span class="co"># coreGFApath = f'{pathfileDir}{os.path.sep}coreGraph.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [1,10,20,100,500,1000,5000,10000,50000,100000,500000,1000000]</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">9</span>,<span class="dv">45</span>,<span class="dv">90</span>,<span class="dv">450</span>,<span class="dv">900</span>,<span class="dv">4500</span>,<span class="dv">9000</span>]</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [1,3,9,18]</span></span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [4,8,16]</span></span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">16</span></span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a>invertionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/pantograph/data/shorttest2.gfa'</span></span>
<span id="cb136-12"><a href="#cb136-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/pantograph/data/AT_Chr1_OGOnly.gfa'</span></span>
<span id="cb136-13"><a href="#cb136-13" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/chrisGraph/chr1.wfmash.n20.a90.s10000.p1,19,39,3,81,1.seqwish.sort.smooth.sort.gfa'</span></span>
<span id="cb136-14"><a href="#cb136-14" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/pantograph/data/shorttest_seq.gfa'</span></span>
<span id="cb136-15"><a href="#cb136-15" aria-hidden="true" tabindex="-1"></a>inputPath <span class="op">=</span> coreGFApath</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(inputPath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>dbid <span class="op">=</span> getDBID(<span class="st">'../pantograph_API/data/caseToDBID.dict'</span>,outputName)</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Opening Redis connection for db </span><span class="sc">{</span>dbid<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span>dbid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides = \</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>inputPath,</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>                   outputName<span class="op">=</span>outputName,</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>                   outputPath<span class="op">=</span>outputPath,</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>                   isSeq<span class="op">=</span>isSeq,</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>                   redisConn<span class="op">=</span>redisConn,</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>                   GenomeGraphParams<span class="op">=</span>{<span class="st">'accessionsToRemove'</span>:[<span class="st">'Consensus'</span>]},</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>                   zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>                   maxLengthChunk<span class="op">=</span>maxLengthChunk,</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>                   maxLengthComponent<span class="op">=</span>maxLengthComponent,</span>
<span id="cb139-11"><a href="#cb139-11" aria-hidden="true" tabindex="-1"></a>                   invertionThreshold<span class="op">=</span>invertionThreshold,)</span>
<span id="cb139-12"><a href="#cb139-12" aria-hidden="true" tabindex="-1"></a><span class="co">#                                              debug=True,returnDebugData=True)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Pantograph data generation for coregraph finished."</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exporting-to-pantograph-visualisation" class="level1">
<h1>Exporting to Pantograph visualisation</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>projectID <span class="op">=</span> <span class="st">'paths_genegraph'</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>projectName <span class="op">=</span> <span class="st">'Example gene graph'</span></span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>pathToGraphs <span class="op">=</span> <span class="st">'examples/gene_graph'</span></span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>caseDict <span class="op">=</span> {<span class="st">'Main'</span>: <span class="st">'paths_genegraph.gfa'</span>}</span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a>pathToIndex <span class="op">=</span> <span class="st">'examples/Visdata'</span></span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This is if you run it in Docker compose together with active Redis image, which is named "redis".</span></span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a><span class="co"># If you have separate redis server, enter full address here.</span></span>
<span id="cb141-9"><a href="#cb141-9" aria-hidden="true" tabindex="-1"></a><span class="co"># If you do not want to add any annotation, `redisHost` should be None.</span></span>
<span id="cb141-10"><a href="#cb141-10" aria-hidden="true" tabindex="-1"></a>redisHost <span class="op">=</span> <span class="st">'redis'</span></span>
<span id="cb141-11"><a href="#cb141-11" aria-hidden="true" tabindex="-1"></a>redisPort <span class="op">=</span> <span class="dv">6379</span></span>
<span id="cb141-12"><a href="#cb141-12" aria-hidden="true" tabindex="-1"></a>redisDB <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb141-13"><a href="#cb141-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-14"><a href="#cb141-14" aria-hidden="true" tabindex="-1"></a>suffix <span class="op">=</span> <span class="st">''</span></span>
<span id="cb141-15"><a href="#cb141-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-16"><a href="#cb141-16" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb141-17"><a href="#cb141-17" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb141-18"><a href="#cb141-18" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb141-19"><a href="#cb141-19" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb141-20"><a href="#cb141-20" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]</span>
<span id="cb141-21"><a href="#cb141-21" aria-hidden="true" tabindex="-1"></a>fillZoomLevel <span class="op">=</span> <span class="va">True</span></span>
<span id="cb141-22"><a href="#cb141-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-23"><a href="#cb141-23" aria-hidden="true" tabindex="-1"></a>exportProject(projectID, projectName, caseDict, pathToIndex, pathToGraphs,</span>
<span id="cb141-24"><a href="#cb141-24" aria-hidden="true" tabindex="-1"></a>              redisHost <span class="op">=</span> redisHost, redisPort <span class="op">=</span> redisPort, redisDB <span class="op">=</span> redisDB,</span>
<span id="cb141-25"><a href="#cb141-25" aria-hidden="true" tabindex="-1"></a>              suffix <span class="op">=</span> suffix,</span>
<span id="cb141-26"><a href="#cb141-26" aria-hidden="true" tabindex="-1"></a>              maxLengthComponent <span class="op">=</span> maxLengthComponent, maxLengthChunk <span class="op">=</span> maxLengthChunk,</span>
<span id="cb141-27"><a href="#cb141-27" aria-hidden="true" tabindex="-1"></a>              inversionThreshold <span class="op">=</span> inversionThreshold,</span>
<span id="cb141-28"><a href="#cb141-28" aria-hidden="true" tabindex="-1"></a>              isSeq <span class="op">=</span> isSeq,</span>
<span id="cb141-29"><a href="#cb141-29" aria-hidden="true" tabindex="-1"></a>              zoomLevels <span class="op">=</span> zoomLevels, fillZoomLevel <span class="op">=</span> fillZoomLevel):</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>projectID <span class="op">=</span> <span class="st">'tutorial_graph'</span></span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>projectName <span class="op">=</span> <span class="st">'Example nucleotide graph'</span></span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>pathToGraphs <span class="op">=</span> <span class="st">'examples/nucleotide_graph'</span></span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>caseDict <span class="op">=</span> {<span class="st">'Main'</span>: <span class="st">'paths_presentation.gfa'</span>}</span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>pathToIndex <span class="op">=</span> <span class="st">'examples/Visdata'</span></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This is if you run it in Docker compose together with active Redis image, which is named "redis".</span></span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a><span class="co"># If you have separate redis server, enter full address here.</span></span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a><span class="co"># If you do not want to add any annotation, `redisHost` should be None.</span></span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>redisHost <span class="op">=</span> <span class="st">'redis'</span></span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a>redisPort <span class="op">=</span> <span class="dv">6379</span></span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>redisDB <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a>suffix <span class="op">=</span> <span class="st">''</span></span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb142-18"><a href="#cb142-18" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb142-19"><a href="#cb142-19" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> </span>
<span id="cb142-20"><a href="#cb142-20" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]</span>
<span id="cb142-21"><a href="#cb142-21" aria-hidden="true" tabindex="-1"></a>fillZoomLevel <span class="op">=</span> <span class="va">True</span></span>
<span id="cb142-22"><a href="#cb142-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-23"><a href="#cb142-23" aria-hidden="true" tabindex="-1"></a>exportProject(projectID, projectName, caseDict, pathToIndex, pathToGraphs,</span>
<span id="cb142-24"><a href="#cb142-24" aria-hidden="true" tabindex="-1"></a>              redisHost <span class="op">=</span> redisHost, redisPort <span class="op">=</span> redisPort, redisDB <span class="op">=</span> redisDB,</span>
<span id="cb142-25"><a href="#cb142-25" aria-hidden="true" tabindex="-1"></a>              suffix <span class="op">=</span> suffix,</span>
<span id="cb142-26"><a href="#cb142-26" aria-hidden="true" tabindex="-1"></a>              maxLengthComponent <span class="op">=</span> maxLengthComponent, maxLengthChunk <span class="op">=</span> maxLengthChunk,</span>
<span id="cb142-27"><a href="#cb142-27" aria-hidden="true" tabindex="-1"></a>              inversionThreshold <span class="op">=</span> inversionThreshold,</span>
<span id="cb142-28"><a href="#cb142-28" aria-hidden="true" tabindex="-1"></a>              isSeq <span class="op">=</span> isSeq,</span>
<span id="cb142-29"><a href="#cb142-29" aria-hidden="true" tabindex="-1"></a>              zoomLevels <span class="op">=</span> zoomLevels, fillZoomLevel <span class="op">=</span> fillZoomLevel):</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>