<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).">

<title>pygengraph</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="pygengraph">
<meta property="og:description" content="This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).">
<meta property="og:site-name" content="pygengraph">
<meta name="twitter:title" content="pygengraph">
<meta name="twitter:description" content="This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">pygengraph</span>
    </a>
  </div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">pygengraph</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link active">pygengraph</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./init.html" class="sidebar-item-text sidebar-link">Synteny module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./graph.html" class="sidebar-item-text sidebar-link">Graph module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tree.html" class="sidebar-item-text sidebar-link">Tree module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./synteny.html" class="sidebar-item-text sidebar-link">Synteny module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">Utils module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./export.html" class="sidebar-item-text sidebar-link">Export module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scripts.html" class="sidebar-item-text sidebar-link">Script module</a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">docs</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  docs/index.html
  </li>
          <li class="sidebar-item">
  docs/CONTRIBUTING.html
  </li>
          <li class="sidebar-item">
  docs/debruijngraphprocessing.html
  </li>
          <li class="sidebar-item">
  docs/dev.html
  </li>
          <li class="sidebar-item">
  docs/export.html
  </li>
          <li class="sidebar-item">
  docs/graph.html
  </li>
          <li class="sidebar-item">
  docs/init.html
  </li>
          <li class="sidebar-item">
  docs/scripts.html
  </li>
          <li class="sidebar-item">
  docs/synteny.html
  </li>
          <li class="sidebar-item">
  docs/tree.html
  </li>
          <li class="sidebar-item">
  docs/utils.html
  </li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#install" id="toc-install" class="nav-link active" data-scroll-target="#install">Install</a></li>
  <li><a href="#how-to-use" id="toc-how-to-use" class="nav-link" data-scroll-target="#how-to-use">How to use</a></li>
  <li><a href="#if-redis-database-needs-to-be-cleaned." id="toc-if-redis-database-needs-to-be-cleaned." class="nav-link" data-scroll-target="#if-redis-database-needs-to-be-cleaned.">If redis database needs to be cleaned.</a></li>
  <li><a href="#generating-graphs" id="toc-generating-graphs" class="nav-link" data-scroll-target="#generating-graphs">Generating graphs</a>
  <ul class="collapse">
  <li><a href="#generating-from-annotation" id="toc-generating-from-annotation" class="nav-link" data-scroll-target="#generating-from-annotation">Generating from annotation</a>
  <ul class="collapse">
  <li><a href="#preparing-list-of-files" id="toc-preparing-list-of-files" class="nav-link" data-scroll-target="#preparing-list-of-files">Preparing list of files</a></li>
  <li><a href="#generaton-of-gene-graph" id="toc-generaton-of-gene-graph" class="nav-link" data-scroll-target="#generaton-of-gene-graph">Generaton of gene graph</a></li>
  </ul></li>
  <li><a href="#loading-pathfile-to-graph" id="toc-loading-pathfile-to-graph" class="nav-link" data-scroll-target="#loading-pathfile-to-graph">Loading Pathfile to graph</a></li>
  </ul></li>
  <li><a href="#loading-graph-from-gfa-and-sorting-it" id="toc-loading-graph-from-gfa-and-sorting-it" class="nav-link" data-scroll-target="#loading-graph-from-gfa-and-sorting-it">Loading graph from GFA and sorting it</a>
  <ul class="collapse">
  <li><a href="#exporting-presentation-graph" id="toc-exporting-presentation-graph" class="nav-link" data-scroll-target="#exporting-presentation-graph">Exporting presentation graph</a></li>
  </ul></li>
  <li><a href="#changing-annotation" id="toc-changing-annotation" class="nav-link" data-scroll-target="#changing-annotation">changing annotation</a>
  <ul class="collapse">
  <li><a href="#exporting-tutorial-gene-graph" id="toc-exporting-tutorial-gene-graph" class="nav-link" data-scroll-target="#exporting-tutorial-gene-graph">Exporting tutorial gene graph</a></li>
  </ul></li>
  <li><a href="#changing-annotation-1" id="toc-changing-annotation-1" class="nav-link" data-scroll-target="#changing-annotation-1">changing annotation</a>
  <ul class="collapse">
  <li><a href="#testing-removable-elements-identification" id="toc-testing-removable-elements-identification" class="nav-link" data-scroll-target="#testing-removable-elements-identification">Testing removable elements identification</a></li>
  <li><a href="#exporting-test-collapse-graph" id="toc-exporting-test-collapse-graph" class="nav-link" data-scroll-target="#exporting-test-collapse-graph">Exporting test collapse graph</a></li>
  <li><a href="#exporting-coregraph" id="toc-exporting-coregraph" class="nav-link" data-scroll-target="#exporting-coregraph">Exporting coregraph</a></li>
  <li><a href="#exporting-coregraph-with-genes" id="toc-exporting-coregraph-with-genes" class="nav-link" data-scroll-target="#exporting-coregraph-with-genes">Exporting coregraph with genes</a></li>
  <li><a href="#exporting-gene-graphs-for-all-chromosomes" id="toc-exporting-gene-graphs-for-all-chromosomes" class="nav-link" data-scroll-target="#exporting-gene-graphs-for-all-chromosomes">Exporting gene graphs for all chromosomes</a></li>
  <li><a href="#comments" id="toc-comments" class="nav-link" data-scroll-target="#comments">Comments</a></li>
  </ul></li>
  <li><a href="#exporting-to-pantograph-visualisation" id="toc-exporting-to-pantograph-visualisation" class="nav-link" data-scroll-target="#exporting-to-pantograph-visualisation">Exporting to Pantograph visualisation</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-git"></i></div><div class="action-links"><p><a href="https://gitlab.com/pigrenok/pygengraph/tree/master/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">pygengraph</h1>
</div>

<div>
  <div class="description">
    This package provides all functionality from graph construction (currently, from annotation, in the future, from raw assembled sequences) to graph processing (sorting, grouping, adjusting).
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<section id="install" class="level2">
<h2 class="anchored" data-anchor-id="install">Install</h2>
<p>Enter the directory of the library and enter:</p>
<p><code>pip install .</code></p>
<p>and for development use</p>
<p><code>pip install -e .</code></p>
</section>
<section id="how-to-use" class="level2">
<h2 class="anchored" data-anchor-id="how-to-use">How to use</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> nbdev <span class="im">import</span> nbdev_export</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>nbdev_export()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pygengraph.graph <span class="im">import</span> GenomeGraph</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pygengraph.utils <span class="im">import</span> pathFileToPathDict</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pygengraph.export <span class="im">import</span> exportProject</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="if-redis-database-needs-to-be-cleaned." class="level1">
<h1>If redis database needs to be cleaned.</h1>
<p>from redis import Redis</p>
<p>redisConn = Redis(host=‘redis’,port=6379)</p>
<p>redisConn.flushall()</p>
<p>redisConn.close()</p>
<p>del redisConn</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>warnings.simplefilter(<span class="st">'ignore'</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">'always'</span>,category<span class="op">=</span><span class="pp">RuntimeWarning</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="generating-graphs" class="level1">
<h1>Generating graphs</h1>
<section id="generating-from-annotation" class="level2">
<h2 class="anchored" data-anchor-id="generating-from-annotation">Generating from annotation</h2>
<section id="preparing-list-of-files" class="level3">
<h3 class="anchored" data-anchor-id="preparing-list-of-files">Preparing list of files</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>refdir <span class="op">=</span> <span class="st">'/path/to/reference/'</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>annotationdir <span class="op">=</span> <span class="st">'/path/to/annotation'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>gfadir <span class="op">=</span> <span class="st">'/path/to/graphs'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>annotationFiles <span class="op">=</span> <span class="bu">sorted</span>(glob.glob(<span class="ss">f'</span><span class="sc">{</span>annotationdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">*.gff'</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>pangenomeFiles <span class="op">=</span> <span class="bu">sorted</span>(glob.glob(<span class="ss">f'</span><span class="sc">{</span>annotationdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">*pangen.gff'</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes.</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It should also be converted to sequenceFileDict, see details in documentation for GenomeGraph Class constructor.</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sequenceFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}sequences{os.path.sep}*.fasta'))</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>refAnnotationFile <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>refdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">reference.gff'</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># refSequenceFile = f'{refdir}{os.path.sep}reference.fasta'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>refdir <span class="op">=</span> <span class="st">'../../1001G/annotations/freeze2.1/outgroups'</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>annotationdir <span class="op">=</span> <span class="st">'../../1001G/annotations/freeze2.1'</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>gfadir <span class="op">=</span> <span class="st">'../../1001G/annotations/graphs'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>annotationFiles <span class="op">=</span> <span class="bu">sorted</span>(glob.glob(<span class="ss">f'</span><span class="sc">{</span>annotationdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">*.gff'</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># pangenomeFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}*pangen.gff'))</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes.</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># It should also be converted to sequenceFileDict, see details in documentation for GenomeGraph Class constructor.</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sequenceFiles = sorted(glob.glob(f'{annotationdir}{os.path.sep}sequences{os.path.sep}*.fasta'))</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>refAnnotationFile <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>refdir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">araport.gff'</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to include sequences instead of simple notion of genes</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># refSequenceFile = f'{refdir}{os.path.sep}reference.fasta'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>annotationFiles</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['../../1001G/annotations/freeze2.1/10002.gff',
 '../../1001G/annotations/freeze2.1/10015.gff',
 '../../1001G/annotations/freeze2.1/10024.gff',
 '../../1001G/annotations/freeze2.1/1741.gff',
 '../../1001G/annotations/freeze2.1/22001.gff',
 '../../1001G/annotations/freeze2.1/22002.gff',
 '../../1001G/annotations/freeze2.1/22003.gff',
 '../../1001G/annotations/freeze2.1/22004.gff',
 '../../1001G/annotations/freeze2.1/22005.gff',
 '../../1001G/annotations/freeze2.1/22006.gff',
 '../../1001G/annotations/freeze2.1/22007.gff',
 '../../1001G/annotations/freeze2.1/6024.gff',
 '../../1001G/annotations/freeze2.1/6069.gff',
 '../../1001G/annotations/freeze2.1/6124.gff',
 '../../1001G/annotations/freeze2.1/6244.gff',
 '../../1001G/annotations/freeze2.1/6909.gff',
 '../../1001G/annotations/freeze2.1/6966.gff',
 '../../1001G/annotations/freeze2.1/8236.gff',
 '../../1001G/annotations/freeze2.1/9075.gff',
 '../../1001G/annotations/freeze2.1/9537.gff',
 '../../1001G/annotations/freeze2.1/9543.gff',
 '../../1001G/annotations/freeze2.1/9638.gff',
 '../../1001G/annotations/freeze2.1/9728.gff',
 '../../1001G/annotations/freeze2.1/9764.gff',
 '../../1001G/annotations/freeze2.1/9888.gff',
 '../../1001G/annotations/freeze2.1/9905.gff',
 '../../1001G/annotations/freeze2.1/9981.gff']</code></pre>
</div>
</div>
</section>
<section id="generaton-of-gene-graph" class="level3">
<h3 class="anchored" data-anchor-id="generaton-of-gene-graph">Generaton of gene graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>doUS <span class="op">=</span> <span class="va">False</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>n <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrnum <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,n<span class="op">+</span><span class="dv">1</span>): <span class="co"># here n is number of chromosomes.</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    chromosome <span class="op">=</span> <span class="ss">f'Chr</span><span class="sc">{</span>chrnum<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="ch">\n</span><span class="ss">Processing </span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ch">\n</span><span class="ss">============'</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    curtst <span class="op">=</span> time.time()</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    graph <span class="op">=</span> GenomeGraph(annotationFiles <span class="op">=</span> annotationFiles,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                        pangenomeFiles <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                        sequenceFilesDict <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>                        doUS <span class="op">=</span> doUS,</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>                        chromosome <span class="op">=</span> chromosome,</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>                        refAnnotationFile<span class="op">=</span>refAnnotationFile,</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>                        refAccession<span class="op">=</span><span class="st">'TAIR10'</span>)</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Generating graph for </span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>time<span class="sc">.</span>time() <span class="op">-</span> curtst<span class="sc">}</span><span class="ss"> seconds'</span>)</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    curtst <span class="op">=</span> time.time()</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    graph.treeSort()</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Sorting graph for </span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>time<span class="sc">.</span>time() <span class="op">-</span> curtst<span class="sc">}</span><span class="ss"> seconds'</span>)</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(graph.nodes)<span class="op">!=</span><span class="bu">len</span>(graph.order):</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="st">'Sorting failed and not all nodes were sorted. Saving unsorted graph'</span>)</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>            gfaFilename <span class="op">=</span> <span class="ss">f'Gene_</span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss">_simOnly_unordered.gfa'</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>            graph.order <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(graph.nodes)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>        gfaFilename <span class="op">=</span> <span class="ss">f'Gene_</span><span class="sc">{</span>chromosome<span class="sc">}</span><span class="ss">_simOnly.gfa'</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>    graph.toGFA(<span class="ss">f'</span><span class="sc">{</span>gfadir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>gfaFilename<span class="sc">}</span><span class="ss">'</span>,doSeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>q <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>q.remove(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>q</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[1, 2, 4]</code></pre>
</div>
</div>
</section>
</section>
<section id="loading-pathfile-to-graph" class="level2">
<h2 class="anchored" data-anchor-id="loading-pathfile-to-graph">Loading Pathfile to graph</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For path file v1</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'examples/gene_graph'</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>pathsfile <span class="op">=</span> <span class="st">'paths_genegraph.txt'</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>paths <span class="op">=</span> pathFileToPathDict(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>pathsfile<span class="sc">}</span><span class="ss">'</span>, <span class="va">True</span>, <span class="va">True</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(pathsDict<span class="op">=</span>paths)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>graph.treeSort()</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">len</span>(graph.nodes)<span class="op">!=</span><span class="bu">len</span>(graph.order):</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'Sorting failed and not all nodes were sorted. Saving unsorted graph'</span>)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    output <span class="op">=</span> <span class="st">'paths_genegraph_unordered.gfa'</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    graph.order <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(graph.nodes)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    graph.toGFA(output,doSeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'paths_genegraph.gfa'</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>    coregraph.toGFA(coreGFApath,doSeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For v2</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is example, no v2 file currently available for demonstration.</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'/path/to/file'</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>pathsfile <span class="op">=</span> <span class="ss">f'paths.txt'</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>paths <span class="op">=</span> pathFileToPathDict(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>pathsfile<span class="sc">}</span><span class="ss">'</span>,<span class="va">True</span>,<span class="st">'reference'</span>,<span class="va">True</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seqNum <span class="kw">in</span> paths.keys():</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    graph <span class="op">=</span> GenomeGraph(pathsDict<span class="op">=</span>paths[seqNum])</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># On undirected coregraph sorting is not optimal! Check sorting!!!</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    graph.treeSort()</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(graph.nodes)<span class="op">!=</span><span class="bu">len</span>(graph.order):</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">'Sorting failed and not all nodes were sorted. Saving unsorted graph'</span>)</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>        output <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">graph_Chr</span><span class="sc">{</span>seqNum<span class="sc">}</span><span class="ss">_unordered.gfa'</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>        graph.order <span class="op">=</span> <span class="bu">list</span>(<span class="bu">range</span>(<span class="dv">1</span>,<span class="bu">len</span>(graph.nodes)<span class="op">+</span><span class="dv">1</span>))</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>        graph.toGFA(output,doSeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>        coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">graph_Chr</span><span class="sc">{</span>seqNum<span class="sc">}</span><span class="ss">.gfa'</span></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>        graph.toGFA(coreGFApath,doSeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="loading-graph-from-gfa-and-sorting-it" class="level1">
<h1>Loading graph from GFA and sorting it</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>gfadir <span class="op">=</span> <span class="st">'examples/nucleotide_graph'</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># It is nucleotide graph. If it is not nucleotide graph, then `isSeq` variable should be changed to False.</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>gfafilename <span class="op">=</span> <span class="st">'paths_presentation.gfa'</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span><span class="ss">f'</span><span class="sc">{</span>gfadir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>gfafilename<span class="sc">}</span><span class="ss">'</span>,isGFASeq<span class="op">=</span>isSeq)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>graph_new.treeSort()</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="bu">len</span>(graph_new.nodes)<span class="op">==</span><span class="bu">len</span>(graph_new.order)</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>basename,ext <span class="op">=</span> os.path.splitext(gfafilename)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>graph_new.toGFA(<span class="ss">f'</span><span class="sc">{</span>gfadir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>basename<span class="sc">}</span><span class="ss">_ordered.</span><span class="sc">{</span>ext<span class="sc">}</span><span class="ss">'</span>,doSeq<span class="op">=</span>isSeq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="exporting-presentation-graph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-presentation-graph">Exporting presentation graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../Meetings/1001G+_20220518/'</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">paths_presentation.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="changing-annotation" class="level1">
<h1>changing annotation</h1>
<p>genome = GenomeGraph(coreGFApath,isGFASeq=True)</p>
<p>for nodeID,node in enumerate(genome.nodesAnnotation): for seqName,seqDict in node.items(): for annText in seqDict.keys(): genome.nodesAnnotation[nodeID][seqName][annText] = [(0,len(genome.nodesData[nodeID])-1)]</p>
<p>genome.toGFA(coreGFApath,doSeq=True)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">16</span>]<span class="co">#,8,16,32]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>('../../Meetings/1001G+_20220518', 'paths_presentation_new', [1, 2, 4, 8, 16])</code></pre>
</div>
</div>
<p>dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName) print(f’Opening Redis connection for db {dbid}‘) redisConn = Redis(host=’redis’,port = 6379,db=dbid)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>redisConn<span class="op">=</span><span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, </span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>                                fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a><span class="co">#                                 returnDebugData=True)</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Genome
Loading graph from ../../Meetings/1001G+_20220518//paths_presentation.gfa
Found nodeNames file ../../Meetings/1001G+_20220518/nodeNames_paths_presentation.json, loading names.
Found node annotation file ../../Meetings/1001G+_20220518/annotation_paths_presentation.dat, loading associations.
Loading segment 7/7
Loading segments finished.
Loading link 12/12
Loading links finished
Loading path 5/5
Loading paths finished. 5 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths

Recording Pantograph data to ../../Meetings/1001G+_20220518/paths_presentation_new
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths
Processing path breaks...
Postprocessing interconnected links 2/2
Preprocessing interconnected links finished.

Processing path breaks finished.
Converting blocks to block lengths 5/5
Conversion finished.
Reformating links to block lengths associations 5/5
Reformating finished.
Identifying rearrangement blocks
Processing node 7/7
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7/7
Processing component links 6/6
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks
Recording component 6/6
Recording zoom level 1 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;2 were removed.     0 components were deleted as isolated.

===========================
Zoom level 2
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 2 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;4 were removed.     0 components were deleted as isolated.

===========================
Zoom level 4
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 4 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;8 were removed.     0 components were deleted as isolated.

===========================
Zoom level 8
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 8 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;16 were removed.     0 components were deleted as isolated.

===========================
Zoom level 16
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 16 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;32 were removed.     0 components were deleted as isolated.

===========================
Zoom level 32
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 32 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;64 were removed.     0 components were deleted as isolated.

===========================
Zoom level 64
===========================
Processing component 6/6
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 6/6
Recording zoom level 64 finished.</code></pre>
</div>
</div>
<section id="exporting-tutorial-gene-graph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-tutorial-gene-graph">Exporting tutorial gene graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/pantograph'</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">genegraph_tutorial.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading graph from ../../1001G/pantograph/genegraph_tutorial.gfa
Found node annotation file ../../1001G/pantograph/annotation_genegraph_tutorial.dat, loading associations.
Loading segment 7/7
Loading segments finished.
Loading link 13/13
Loading links finished
Loading path 5/5
Loading paths finished. 5 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>graph.paths</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[['1+', '2+', '3+', '4+'],
 ['1+', '5+', '6+', '7+', '1+'],
 ['4+', '1+', '2+', '3+'],
 ['2+', '3+', '4+', '1+', '5-', '4+', '7+'],
 ['6+', '7-', '1+', '2+', '3+']]</code></pre>
</div>
</div>
</section>
</section>
<section id="changing-annotation-1" class="level1">
<h1>changing annotation</h1>
<p>genome = GenomeGraph(coreGFApath,isGFASeq=True)</p>
<p>for nodeID,node in enumerate(genome.nodesAnnotation): for seqName,seqDict in node.items(): for annText in seqDict.keys(): genome.nodesAnnotation[nodeID][seqName][annText] = [(0,len(genome.nodesData[nodeID])-1)]</p>
<p>genome.toGFA(coreGFApath,doSeq=True)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]<span class="co">#,8,16,32]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>('../../1001G/pantograph', 'genegraph_tutorial', [1])</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># print(f'Opening Redis connection for db {dbid}')</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host <span class="op">=</span> <span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>, db <span class="op">=</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>redisConn=None</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, </span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                 invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] = \</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>                                fillZoomLevels <span class="op">=</span> <span class="va">False</span>,</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a><span class="co">#                                 returnDebugData=True)</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Genome
Loading graph from ../../1001G/pantograph/genegraph_tutorial.gfa
Found node annotation file ../../1001G/pantograph/annotation_genegraph_tutorial.dat, loading associations.
Loading segment 7/7
Loading segments finished.
Loading link 13/13
Loading links finished
Loading path 5/5
Loading paths finished. 5 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/genegraph_tutorial
Calculating nodes length...
Processing node 7/7
Finished calculating nodes lengths
Preprocessing paths...
Processing path 5/5
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7/7
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7/7
Processing component links 6/6
Recording component 6/6
Recording zoom level 1 finished.</code></pre>
</div>
</div>
<section id="testing-removable-elements-identification" class="level3">
<h3 class="anchored" data-anchor-id="testing-removable-elements-identification">Testing removable elements identification</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pangraph_constructor.exportDev <span class="im">import</span> getRemovableStructures</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> IPython.display <span class="im">as</span> ipd</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> joblib</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.path.exists(<span class="st">'./tests/breakIdentify.dat'</span>):</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    correctResults <span class="op">=</span> joblib.load(<span class="st">'./tests/breakIdentify.dat'</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    correctResults <span class="op">=</span> {}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> <span class="st">'../../1001G/GraphCollapsing/TestGraphs'</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>filePrefix <span class="op">=</span> <span class="st">'test'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#test</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> filename,resDict <span class="kw">in</span> correctResults.items():</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="ch">\n</span><span class="ss">####### Testing on case </span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss"> ########'</span>)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    graph <span class="op">=</span> GenomeGraph(<span class="ss">f'</span><span class="sc">{</span>path<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss">'</span>,isGFASeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    linksLengths, pairedLinks, blockEdges, _ <span class="op">=</span> getRemovableStructures(graph<span class="op">=</span>graph)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> linksLengths<span class="op">==</span>resDict[<span class="st">'linksLengths'</span>]</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> pairedLinks<span class="op">==</span>resDict[<span class="st">'pairedLinks'</span>]</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> blockEdges<span class="op">==</span>resDict[<span class="st">'blockEdges'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>caseNum <span class="op">=</span> <span class="dv">17</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>filePrefix<span class="sc">}{</span>caseNum<span class="sc">:02d}</span><span class="ss">.gfa'</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'############'</span>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Graph from file </span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>path<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Graph Paths:'</span>)</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>ipd.display(graph.paths)</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>linkLengths, pairedLinks, blockEdges, _ <span class="op">=</span> getRemovableStructures(graph<span class="op">=</span>graph)</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Link-Lengths associations:'</span>)</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>ipd.display(linkLengths)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'PairedLinks:'</span>)</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>ipd.display(pairedLinks)</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Rearrangemenet block edges:'</span>)</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>ipd.display(blockEdges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>correctResults[filename]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>correctResults[filename] <span class="op">=</span> {<span class="st">'linksLengths'</span>:linkLengths,<span class="st">'pairedLinks'</span>:pairedLinks,<span class="st">'blockEdges'</span>:blockEdges}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>joblib.dump(correctResults,<span class="st">'./tests/breakIdentify.dat'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-test-collapse-graph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-test-collapse-graph">Exporting test collapse graph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/pantograph/data'</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">testCollapse.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>]<span class="co">#,8,16,32]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName) print(f’Opening Redis connection for db {dbid}‘) redisConn = Redis(host=’redis’,port = 6379,db=dbid)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>redisConn<span class="op">=</span><span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>[zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,zoomAccStarts,zoomAccEnds, </span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>                invertedStarts,invertedEnds,toComponentLinks,fromComponentLinks,collapsibleBlocks,fromLinks,toLinks,graph,rootStruct] <span class="op">=</span> <span class="op">\</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>                                fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold,</span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>                                returnDebugData<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Executed in </span><span class="sc">{</span>runTime<span class="sc">}</span><span class="ss"> seconds'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Exporting test collapse graph finished. Overall time = </span><span class="sc">{runTime}</span><span class="st"> seconds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-coregraph" class="level3">
<h3 class="anchored" data-anchor-id="exporting-coregraph">Exporting coregraph</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_Chr1.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">8</span>,<span class="dv">16</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>('../../1001G/coreGraph', 'coregraph_Chr1_new', [1, 2, 4, 8, 16])</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="co"># print(f'Opening Redis connection for db {dbid}')</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Converted 00_init.ipynb.
Converted 01_graph.ipynb.
Converted 02_tree.ipynb.
Converted 03_synteny.ipynb.
Converted 04_utils.ipynb.
Converted 05_export.ipynb.
Converted 05_exportDev.ipynb.
Converted deBruijnGraphProcessing.ipynb.
Converted dev.ipynb.
Converted graphTesting.ipynb.
Converted index.ipynb.</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>startTime <span class="op">=</span> time.time()</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>                                outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>                                isSeq<span class="op">=</span>isSeq,</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>                                redisConn<span class="op">=</span>redisConn,</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>                                zoomLevels<span class="op">=</span>zoomLevels, </span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>                                maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>                                maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>                                inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a><span class="co">#                                 returnDebugata=True)</span></span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(runTime)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading Genome
Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa
Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.
Loading segment 35/35
Loading segments finished.
Loading link 72/72
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 35/35
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/coreGraph/coregraph_Chr1_new
Calculating nodes length...
Processing node 35/35
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Processing path breaks...
Postprocessing interconnected links 16/16
Preprocessing interconnected links finished.

Processing path breaks finished.
Converting blocks to block lengths 34/34
Conversion finished.
Reformating links to block lengths associations 34/34
Reformating finished.
Identifying rearrangement blocks
Processing node 35/35
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 35/35
Processing component links 35/35
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks
Recording component 35/35
Recording zoom level 1 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;2 were removed.     0 components were deleted as isolated.

===========================
Zoom level 2
===========================
Processing component 35/35
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 8/8
Recording zoom level 2 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;4 were removed.     0 components were deleted as isolated.

===========================
Zoom level 4
===========================
Processing component 8/8
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 5/5
Recording zoom level 4 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;8 were removed.     0 components were deleted as isolated.

===========================
Zoom level 8
===========================
Processing component 5/5
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 1/1
Recording zoom level 8 finished.
Removing links according to collapsible blocks
All links associated with collapsibleComponents &lt;16 were removed.     0 components were deleted as isolated.

===========================
Zoom level 16
===========================
Processing component 1/1
Processing component finished.
Converting link to block lengths associations
Converting paired links
Converting interconnected links
Converting rearrangement blocks

Recording component 1/1
Recording zoom level 16 finished.
27.11550521850586</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Exporting coregraph finished. Overall time = </span><span class="sc">{runTime}</span><span class="st"> seconds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.11) or chardet (3.0.4) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "</code></pre>
</div>
</div>
<p>❗️❗️❗️ TODO: Next test API. WHen works, change front end to get jsons without annotation and then to load annotation from API!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading graph from ../../1001G/coreGraph/coregraph_Chr1.gfa
Found node annotation file ../../1001G/coreGraph/annotation_coregraph_Chr1.dat, loading associations.
Loading segment 35/35
Loading segments finished.
Loading link 72/72
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 35/35
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pangraph_constructor.utils <span class="im">import</span> iset_get,iset_score,iset_add</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port<span class="op">=</span><span class="dv">6379</span>,db<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>iset_get(redisConn,<span class="st">'test'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'a_0': (1.0, 3.0), 'b_0': (1.0, 5.0)}</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>iset_score(redisConn,<span class="st">'coregraph_Chr1_new.1.10024.Pos'</span>,<span class="dv">22</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="co"># coregraph_Chr1_new/10024/1/22</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['23']</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>iset_score(redisConn,<span class="st">'coregraph_Chr1_new.9543.Gene'</span>,<span class="dv">50</span>,<span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['3370',
 'AT1G15940',
 'AT1G15950',
 'AT1G16000',
 'AT1G16010',
 'AT1G16022',
 'AT1G62830',
 'AT1G62840',
 'OG0001502',
 'OG0004404',
 'OG0007340',
 'OG0007341',
 'OG0007343',
 'OG0008435',
 'OG0015400',
 'OG0017985',
 'OG0022841',
 'OG0024046',
 'OG0024055',
 'OG0025924',
 'OG0027925',
 'OG0030893']</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>iset_get(redisConn,<span class="st">'coregraph_Chr1_new.9543.Gene'</span>,<span class="st">'3370_0'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'3370_0': (48.0, 51.0)}</code></pre>
</div>
</div>
<p>redisConn=None</p>
</section>
<section id="exporting-coregraph-with-genes" class="level3">
<h3 class="anchored" data-anchor-id="exporting-coregraph-with-genes">Exporting coregraph with genes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">9</span>]<span class="co"># + [9*2**i for i in range(9)]</span></span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [2**i for i in range(12)]#  [1,2,4,8,16,32,9*16,9*32,9*128,9*256]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName = pathConvert(coreGFApath,suffix='_new')</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName,zoomLevels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>dbid = getDBID(‘../pantograph_API/data/caseToDBID.dict’,outputName) print(f’Opening Redis connection for db {dbid}‘) redisConn = Redis(host=’redis’,port = 6379,db=dbid)</p>
<p>redisConn=None</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># %%capture output</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrNum <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">6</span>):</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_genes_Chr</span><span class="sc">{</span>chrNum<span class="sc">}</span><span class="ss">.gfa'</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>    outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>    dbid <span class="op">=</span> getDBID(<span class="st">'../pantograph_API/data/caseToDBID.dict'</span>,outputName)</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Opening Redis connection for db </span><span class="sc">{</span>dbid<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>    redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span>dbid)</span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>    startTime <span class="op">=</span> time.time()</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a>    exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>                                    outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a>                                    isSeq<span class="op">=</span>isSeq,</span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a>                                    redisConn<span class="op">=</span>redisConn,</span>
<span id="cb80-17"><a href="#cb80-17" aria-hidden="true" tabindex="-1"></a>                                    zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb80-18"><a href="#cb80-18" aria-hidden="true" tabindex="-1"></a>                                    fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb80-19"><a href="#cb80-19" aria-hidden="true" tabindex="-1"></a>                                    maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb80-20"><a href="#cb80-20" aria-hidden="true" tabindex="-1"></a>                                    maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb80-21"><a href="#cb80-21" aria-hidden="true" tabindex="-1"></a>                                    inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb80-22"><a href="#cb80-22" aria-hidden="true" tabindex="-1"></a>    <span class="co">#                                 returnDebugata=True)</span></span>
<span id="cb80-23"><a href="#cb80-23" aria-hidden="true" tabindex="-1"></a>    runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>curT <span class="op">=</span> time.localtime(time.time()<span class="op">+</span><span class="dv">3600</span>)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>message <span class="op">=</span> <span class="ss">f"Exporting core graph with genes for all chromosomes finished at </span><span class="ch">\</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="ss">            </span><span class="sc">{</span>curT<span class="sc">.</span>tm_hour<span class="sc">:02d}</span><span class="ss">:</span><span class="sc">{</span>curT<span class="sc">.</span>tm_min<span class="sc">:02d}</span><span class="ss"> on </span><span class="sc">{</span>curT<span class="sc">.</span>tm_mday<span class="sc">:02d}</span><span class="ss">/</span><span class="sc">{</span>curT<span class="sc">.</span>tm_mon<span class="sc">:02d}</span><span class="ss">!</span><span class="ch">\n</span><span class="ss">"</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"</span><span class="sc">{message}</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>notebook2script()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName = pathConvert(coreGFApath,suffix='_new')</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="co"># outputPath,outputName,zoomLevels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co">#%%capture output2</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrNum <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">6</span>):</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_Chr</span><span class="sc">{</span>chrNum<span class="sc">}</span><span class="ss">.gfa'</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>    outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>    dbid <span class="op">=</span> getDBID(<span class="st">'../pantograph_API/data/caseToDBID.dict'</span>,outputName)</span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Opening Redis connection for db </span><span class="sc">{</span>dbid<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>    redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span>dbid)</span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a>    startTime <span class="op">=</span> time.time()</span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># [zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides,toComponentLinks,fromComponentLinks,graph] = \</span></span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a>    exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a>                                    outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb85-15"><a href="#cb85-15" aria-hidden="true" tabindex="-1"></a>                                    isSeq<span class="op">=</span>isSeq,</span>
<span id="cb85-16"><a href="#cb85-16" aria-hidden="true" tabindex="-1"></a>                                    redisConn<span class="op">=</span>redisConn,</span>
<span id="cb85-17"><a href="#cb85-17" aria-hidden="true" tabindex="-1"></a>                                    zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb85-18"><a href="#cb85-18" aria-hidden="true" tabindex="-1"></a>                                    fillZoomLevels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb85-19"><a href="#cb85-19" aria-hidden="true" tabindex="-1"></a>                                    maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb85-20"><a href="#cb85-20" aria-hidden="true" tabindex="-1"></a>                                    maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb85-21"><a href="#cb85-21" aria-hidden="true" tabindex="-1"></a>                                    inversionThreshold<span class="op">=</span>inversionThreshold)</span>
<span id="cb85-22"><a href="#cb85-22" aria-hidden="true" tabindex="-1"></a>    <span class="co">#                                 returnDebugata=True)</span></span>
<span id="cb85-23"><a href="#cb85-23" aria-hidden="true" tabindex="-1"></a>    runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span>
<span id="cb85-24"><a href="#cb85-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb85-25"><a href="#cb85-25" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Exporting core graph for Chr</span><span class="sc">{</span>chrNum<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>runTime<span class="sc">}</span><span class="ss"> seconds'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>curT <span class="op">=</span> time.localtime(time.time()<span class="op">+</span><span class="dv">3600</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>message <span class="op">=</span> <span class="ss">f"Exporting core graph with genes for all chromosomes finished at </span><span class="ch">\</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="ss"> </span><span class="sc">{</span>curT<span class="sc">.</span>tm_hour<span class="sc">:02d}</span><span class="ss">:</span><span class="sc">{</span>curT<span class="sc">.</span>tm_min<span class="sc">:02d}</span><span class="ss"> on </span><span class="sc">{</span>curT<span class="sc">.</span>tm_mday<span class="sc">:02d}</span><span class="ss">/</span><span class="sc">{</span>curT<span class="sc">.</span>tm_mon<span class="sc">:02d}</span><span class="ss">!</span><span class="ch">\n</span><span class="ss">"</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"</span><span class="sc">{message}</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Exporting coregraph with genes finished. Overall time = </span><span class="sc">{runTime}</span><span class="st"> seconds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exporting-gene-graphs-for-all-chromosomes" class="level3">
<h3 class="anchored" data-anchor-id="exporting-gene-graphs-for-all-chromosomes">Exporting gene graphs for all chromosomes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/pantograph/data'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]<span class="co">#,2,4]# + [9*2**i for i in range(12)]</span></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>fillZoomLevel <span class="op">=</span> <span class="va">False</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>nbdev_export()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>caseDict <span class="op">=</span> {}</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">1</span>,<span class="dv">6</span>):</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>    GFA <span class="op">=</span> <span class="ss">f'AT_Chr</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">_OGOnly_2.1.gfa'</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>    caseDict[<span class="ss">f'Chr</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">'</span>] <span class="op">=</span> GFA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'Chr1': 'AT_Chr1_OGOnly_2.1.gfa',
 'Chr2': 'AT_Chr2_OGOnly_2.1.gfa',
 'Chr3': 'AT_Chr3_OGOnly_2.1.gfa',
 'Chr4': 'AT_Chr4_OGOnly_2.1.gfa',
 'Chr5': 'AT_Chr5_OGOnly_2.1.gfa'}</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>exportProject(projectID <span class="op">=</span> <span class="st">'AT_OGOnly_2.1'</span>, projectName <span class="op">=</span> <span class="st">'A. thaliana 27 genomes + TAIR10 gene graph 2.1'</span>, </span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>              caseDict <span class="op">=</span> caseDict, pathToIndex <span class="op">=</span> pathfileDir, pathToGraphs <span class="op">=</span> pathfileDir,</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>              redisHost<span class="op">=</span><span class="st">'redis'</span>, redisPort <span class="op">=</span> <span class="dv">6379</span>, redisDB <span class="op">=</span> <span class="dv">0</span>,</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>              suffix <span class="op">=</span> <span class="st">'pr'</span>,</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>              maxLengthComponent <span class="op">=</span> maxLengthComponent, maxLengthChunk <span class="op">=</span> maxLengthChunk,</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>              inversionThreshold <span class="op">=</span> inversionThreshold,</span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>              isSeq <span class="op">=</span> isSeq, zoomLevels <span class="op">=</span> zoomLevels, fillZoomLevel <span class="op">=</span> fillZoomLevel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing case AT_OGOnly_2.1_Chr1pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
Loading segment 7455/7455
Loading segments finished.
Loading link 14001/14001
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr1pr
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7455/7455
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7455/7455
Processing component links 3869/3869
Recording component 3869/3869
Recording zoom level 1 finished.
Exporting gene graph for Chr1 took 465.6470103263855 seconds

Processing case AT_OGOnly_2.1_Chr2pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.
Loading segment 4877/4877
Loading segments finished.
Loading link 10302/10302
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr2pr
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4877/4877
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4877/4877
Processing component links 2534/2534
Recording component 2534/2534
Recording zoom level 1 finished.
Exporting gene graph for Chr2 took 214.97250485420227 seconds

Processing case AT_OGOnly_2.1_Chr3pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.
Loading segment 8442/8442
Loading segments finished.
Loading link 15985/15985
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr3pr
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 8442/8442
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 8442/8442
Processing component links 3383/3383
Recording component 3383/3383
Recording zoom level 1 finished.
Exporting gene graph for Chr3 took 430.0764102935791 seconds

Processing case AT_OGOnly_2.1_Chr4pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.
Loading segment 4620/4620
Loading segments finished.
Loading link 9980/9980
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr4pr
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4620/4620
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4620/4620
Processing component links 2526/2526
Recording component 2526/2526
Recording zoom level 1 finished.
Exporting gene graph for Chr4 took 199.40259623527527 seconds

Processing case AT_OGOnly_2.1_Chr5pr
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.
Loading segment 7052/7052
Loading segments finished.
Loading link 12737/12737
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_OGOnly_2.1/AT_OGOnly_2.1_Chr5pr
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7052/7052
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7052/7052
Processing component links 3274/3274
Recording component 3274/3274
Recording zoom level 1 finished.
Exporting gene graph for Chr5 took 348.297509431839 seconds</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> seqID <span class="kw">in</span> [<span class="st">'Chr1'</span>,<span class="st">'Chr2'</span>,<span class="st">'Chr3'</span>,<span class="st">'Chr4'</span>,<span class="st">'Chr5'</span>]:<span class="co">#'Chr1','Chr2',</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>    coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">AT_</span><span class="sc">{</span>seqID<span class="sc">}</span><span class="ss">_OGOnly_2.1.gfa'</span></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>    outputPath,outputName <span class="op">=</span> pathConvert(coreGFApath,suffix<span class="op">=</span><span class="st">'_viscol'</span>)</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>()</span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Processing case </span><span class="sc">{</span>outputName<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb95-9"><a href="#cb95-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># dbid = getDBID('../pantograph_API/data/caseToDBID.dict',outputName)</span></span>
<span id="cb95-10"><a href="#cb95-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print(f'Opening Redis connection for db {dbid}')</span></span>
<span id="cb95-11"><a href="#cb95-11" aria-hidden="true" tabindex="-1"></a>    redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb95-12"><a href="#cb95-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-13"><a href="#cb95-13" aria-hidden="true" tabindex="-1"></a>    <span class="co">#%%capture output</span></span>
<span id="cb95-14"><a href="#cb95-14" aria-hidden="true" tabindex="-1"></a>    startTime <span class="op">=</span> time.time()</span>
<span id="cb95-15"><a href="#cb95-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># initialLinkLengths, initialPairedLinks, initialInterconnectedLinks, initialBlockEdges, \</span></span>
<span id="cb95-16"><a href="#cb95-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomNodeToComponent,zoomComponentToNodes,zoomComponents,\</span></span>
<span id="cb95-17"><a href="#cb95-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomFromComponentLinks, zoomToComponentLinks, zoomLinkLengths, zoomPairedLinks, zoomInterconnectedLinks, \</span></span>
<span id="cb95-18"><a href="#cb95-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomOldToNewRemoval, zoomNewToOldRemoval, \</span></span>
<span id="cb95-19"><a href="#cb95-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomLinkLengthsRemoval, zoomPairedLinksRemoval, zoomInterconnectedLinksRemoval, zoomBlockEdgesRemoval, \</span></span>
<span id="cb95-20"><a href="#cb95-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># zoomFromComponentLinksRemoval, zoomToComponentLinksRemoval, \</span></span>
<span id="cb95-21"><a href="#cb95-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># graph = \</span></span>
<span id="cb95-22"><a href="#cb95-22" aria-hidden="true" tabindex="-1"></a>    exportToPantograph(inputPath<span class="op">=</span>coreGFApath, GenomeGraphParams<span class="op">=</span>{}, </span>
<span id="cb95-23"><a href="#cb95-23" aria-hidden="true" tabindex="-1"></a>                                    outputPath<span class="op">=</span>outputPath, outputName<span class="op">=</span>outputName,</span>
<span id="cb95-24"><a href="#cb95-24" aria-hidden="true" tabindex="-1"></a>                                    isSeq<span class="op">=</span>isSeq,</span>
<span id="cb95-25"><a href="#cb95-25" aria-hidden="true" tabindex="-1"></a>                                    redisConn<span class="op">=</span>redisConn,</span>
<span id="cb95-26"><a href="#cb95-26" aria-hidden="true" tabindex="-1"></a>                                    zoomLevels<span class="op">=</span>zoomLevels, </span>
<span id="cb95-27"><a href="#cb95-27" aria-hidden="true" tabindex="-1"></a>                                    fillZoomLevels <span class="op">=</span> <span class="va">False</span>,</span>
<span id="cb95-28"><a href="#cb95-28" aria-hidden="true" tabindex="-1"></a>                                    maxLengthComponent<span class="op">=</span>maxLengthComponent, </span>
<span id="cb95-29"><a href="#cb95-29" aria-hidden="true" tabindex="-1"></a>                                    maxLengthChunk<span class="op">=</span>maxLengthChunk, </span>
<span id="cb95-30"><a href="#cb95-30" aria-hidden="true" tabindex="-1"></a>                                    inversionThreshold<span class="op">=</span>inversionThreshold)<span class="co">#,</span></span>
<span id="cb95-31"><a href="#cb95-31" aria-hidden="true" tabindex="-1"></a>                                    <span class="co"># returnDebugData=True)</span></span>
<span id="cb95-32"><a href="#cb95-32" aria-hidden="true" tabindex="-1"></a>    runTime <span class="op">=</span> time.time() <span class="op">-</span> startTime</span>
<span id="cb95-33"><a href="#cb95-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb95-34"><a href="#cb95-34" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Exporting gene graph for </span><span class="sc">{</span>seqID<span class="sc">}</span><span class="ss"> took </span><span class="sc">{</span>runTime<span class="sc">}</span><span class="ss"> seconds'</span>)</span>
<span id="cb95-35"><a href="#cb95-35" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb95-36"><a href="#cb95-36" aria-hidden="true" tabindex="-1"></a>    redisConn.close()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing case AT_Chr1_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
Loading segment 7455/7455
Loading segments finished.
Loading link 14001/14001
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7455/7455
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7455/7455
Processing component links 3869/3869
Recording component 3869/3869
Recording zoom level 1 finished.
Exporting gene graph for Chr1 took 445.7378799915314 seconds

Processing case AT_Chr2_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr2_OGOnly_2.1.dat, loading associations.
Loading segment 4877/4877
Loading segments finished.
Loading link 10302/10302
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr2_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 4877/4877
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4877/4877
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4877/4877
Processing component links 2534/2534
Recording component 2534/2534
Recording zoom level 1 finished.
Exporting gene graph for Chr2 took 193.71776700019836 seconds

Processing case AT_Chr3_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr3_OGOnly_2.1.dat, loading associations.
Loading segment 8442/8442
Loading segments finished.
Loading link 15985/15985
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr3_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 8442/8442
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 8442/8442
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 8442/8442
Processing component links 3383/3383
Recording component 3383/3383
Recording zoom level 1 finished.
Exporting gene graph for Chr3 took 411.77689695358276 seconds

Processing case AT_Chr4_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr4_OGOnly_2.1.dat, loading associations.
Loading segment 4620/4620
Loading segments finished.
Loading link 9980/9980
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr4_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 4620/4620
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 4620/4620
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 4620/4620
Processing component links 2526/2526
Recording component 2526/2526
Recording zoom level 1 finished.
Exporting gene graph for Chr4 took 180.36719918251038 seconds

Processing case AT_Chr5_OGOnly_2.1_viscol
Loading Genome
Loading graph from ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr5_OGOnly_2.1.dat, loading associations.
Loading segment 7052/7052
Loading segments finished.
Loading link 12737/12737
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths

Recording Pantograph data to ../../1001G/pantograph/data/AT_Chr5_OGOnly_2.1_viscol
Calculating nodes length...
Processing node 7052/7052
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
Identifying rearrangement blocks
Processing node 7052/7052
Identifying rearrangement blocks finished.

===========================
Zoom level 1
===========================
Processing node 7052/7052
Processing component links 3274/3274
Recording component 3274/3274
Recording zoom level 1 finished.
Exporting gene graph for Chr5 took 338.8390169143677 seconds</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>curT <span class="op">=</span> time.localtime()</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>message <span class="op">=</span> <span class="ss">f"Exporting gene graph for all chromosomes finished at </span><span class="ch">\</span></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="ss">            </span><span class="sc">{</span>curT<span class="sc">.</span>tm_hour<span class="sc">:02d}</span><span class="ss">:</span><span class="sc">{</span>curT<span class="sc">.</span>tm_min<span class="sc">:02d}</span><span class="ss"> on </span><span class="sc">{</span>curT<span class="sc">.</span>tm_mday<span class="sc">:02d}</span><span class="ss">/</span><span class="sc">{</span>curT<span class="sc">.</span>tm_mon<span class="sc">:02d}</span><span class="ss">!"</span></span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"</span><span class="sc">{message}</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">AT_Chr1_OGOnly_2.1.gfa'</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>graph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>coreGFApath, isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Loading graph from ../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa
Found node annotation file ../../1001G/pantograph/data/annotation_AT_Chr1_OGOnly_2.1.dat, loading associations.
Loading segment 7455/7455
Loading segments finished.
Loading link 14001/14001
Loading links finished
Loading path 28/28
Loading paths finished. 28 paths added, 0 paths ignored.
Calculating nodes length...
Processing node 7455/7455
Finished calculating nodes lengths
Preprocessing paths...
Processing path 28/28
Finished preprocessing paths
</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>graph.nodesMetadata[<span class="dv">1361</span>][<span class="st">'22006'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'genPos': [{'chr': 'Chr1', 'genomePosition': [19020282, 19020893]},
  {'chr': 'Chr1', 'genomePosition': [19022280, 19023284]}],
 'annotation': {'OG0001585': [(0, 0)],
  'AT1G55230': [(0, 0)],
  'AT1G55240': [(0, 0)]}}</code></pre>
</div>
</div>
<p>import joblib</p>
<p>joblib.dump(initialLinkLengths,‘dumps/initialLinkLengths.dump’) joblib.dump(initialPairedLinks,‘dumps/initialPairedLinks.dump’) joblib.dump(initialInterconnectedLinks,‘dumps/initialInterconnectedLinks.dump’) joblib.dump(initialBlockEdges,‘dumps/initialBlockEdges.dump’) joblib.dump(zoomNodeToComponent,‘dumps/zoomNodeToComponent.dump’) joblib.dump(zoomComponentToNodes,‘dumps/zoomComponentToNodes.dump’) joblib.dump(zoomComponents,‘dumps/zoomComponents.dump’) joblib.dump(zoomFromComponentLinks,‘dumps/zoomFromComponentLinks.dump’) joblib.dump(zoomToComponentLinks,‘dumps/zoomToComponentLinks.dump’) joblib.dump(zoomLinkLengths,‘dumps/zoomLinkLengths.dump’) joblib.dump(zoomPairedLinks,‘dumps/zoomPairedLinks.dump’) joblib.dump(zoomInterconnectedLinks,‘dumps/zoomInterconnectedLinks.dump’) joblib.dump(zoomOldToNewRemoval,‘dumps/zoomOldToNewRemoval.dump’) joblib.dump(zoomNewToOldRemoval,‘dumps/zoomNewToOldRemoval.dump’) joblib.dump(zoomLinkLengthsRemoval,‘dumps/zoomLinkLengthsRemoval.dump’) joblib.dump(zoomPairedLinksRemoval,‘dumps/zoomPairedLinksRemoval.dump’) joblib.dump(zoomInterconnectedLinksRemoval,‘dumps/zoomInterconnectedLinksRemoval.dump’) joblib.dump(zoomBlockEdgesRemoval,‘dumps/zoomBlockEdgesRemoval.dump’) joblib.dump(zoomFromComponentLinksRemoval,‘dumps/zoomFromComponentLinksRemoval.dump’) joblib.dump(zoomToComponentLinksRemoval,‘dumps/zoomToComponentLinksRemoval.dump’) joblib.dump(graph,‘dumps/graph.dump’)</p>
<p>initialLinkLengths = joblib.load(‘dumps/initialLinkLengths.dump’)</p>
<p>initialPairedLinks = joblib.load(‘dumps/initialPairedLinks.dump’)</p>
<p>initialInterconnectedLinks = joblib.load(‘dumps/initialInterconnectedLinks.dump’)</p>
<p>initialBlockEdges = joblib.load(‘dumps/initialBlockEdges.dump’)</p>
<p>zoomNodeToComponent = joblib.load(‘dumps/zoomNodeToComponent.dump’)</p>
<p>zoomComponentToNodes = joblib.load(‘dumps/zoomComponentToNodes.dump’)</p>
<p>zoomComponents = joblib.load(‘dumps/zoomComponents.dump’)</p>
<p>zoomFromComponentLinks = joblib.load(‘dumps/zoomFromComponentLinks.dump’)</p>
<p>zoomToComponentLinks = joblib.load(‘dumps/zoomToComponentLinks.dump’)</p>
<p>zoomLinkLengths = joblib.load(‘dumps/zoomLinkLengths.dump’)</p>
<p>zoomPairedLinks = joblib.load(‘dumps/zoomPairedLinks.dump’)</p>
<p>zoomInterconnectedLinks = joblib.load(‘dumps/zoomInterconnectedLinks.dump’)</p>
<p>zoomOldToNewRemoval = joblib.load(‘dumps/zoomOldToNewRemoval.dump’)</p>
<p>zoomNewToOldRemoval = joblib.load(‘dumps/zoomNewToOldRemoval.dump’)</p>
<p>zoomLinkLengthsRemoval = joblib.load(‘dumps/zoomLinkLengthsRemoval.dump’)</p>
<p>zoomPairedLinksRemoval = joblib.load(‘dumps/zoomPairedLinksRemoval.dump’)</p>
<p>zoomInterconnectedLinksRemoval = joblib.load(‘dumps/zoomInterconnectedLinksRemoval.dump’)</p>
<p>zoomBlockEdgesRemoval = joblib.load(‘dumps/zoomBlockEdgesRemoval.dump’)</p>
<p>zoomFromComponentLinksRemoval = joblib.load(‘dumps/zoomFromComponentLinksRemoval.dump’)</p>
<p>zoomToComponentLinksRemoval = joblib.load(‘dumps/zoomToComponentLinksRemoval.dump’)</p>
<p>graph = joblib.load(‘dumps/graph.dump’)</p>
</section>
<section id="comments" class="level3">
<h3 class="anchored" data-anchor-id="comments">Comments</h3>
<p>Each node processing time increase significantly with overall number of nodes. This is wrong and should be investigated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> copy <span class="im">import</span> deepcopy</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coreGraph_f2.1_Ref_v04.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>coregraph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>coreGFApath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>coregraph_genes <span class="op">=</span> deepcopy(coregraph)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>fullGraphPath <span class="op">=</span> <span class="st">'../../1001G/pantograph/data/AT_Chr1_OGOnly_2.1.gfa'</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>fullgraph <span class="op">=</span> GenomeGraph(gfaPath<span class="op">=</span>fullGraphPath,isGFASeq<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>chainToGenesFile <span class="op">=</span> <span class="st">'chain2gene_f2.1_Ref_v04.txt'</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>maxChainLength <span class="op">=</span> <span class="bu">len</span>(coregraph.nodes[<span class="dv">0</span>])</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>chainToListDict <span class="op">=</span> {}</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}{</span>chainToGenesFile<span class="sc">}</span><span class="ss">'</span>) <span class="im">as</span> f:</span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> f:</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>        chainName, geneList <span class="op">=</span> line.split(<span class="st">':'</span>)</span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>        geneList <span class="op">=</span> geneList.lstrip().rstrip().split(<span class="st">','</span>)</span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>        chainToListDict[chainName.zfill(maxChainLength)] <span class="op">=</span> geneList</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> nodeIdx,nodeName <span class="kw">in</span> <span class="bu">enumerate</span>(coregraph.nodes):</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="ch">\n</span><span class="ss">Node </span><span class="sc">{</span>nodeIdx<span class="op">+</span><span class="dv">1</span><span class="sc">}</span><span class="ss">/</span><span class="sc">{</span><span class="bu">len</span>(coregraph.nodes)<span class="sc">}</span><span class="ss">'</span>,end<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>    geneList <span class="op">=</span> chainToListDict.get(nodeName.zfill(maxChainLength), [<span class="ss">f'ch</span><span class="sc">{</span>nodeName<span class="sc">.</span>zfill(<span class="dv">7</span>)<span class="sc">}</span><span class="ss">'</span>])</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>    geneIds <span class="op">=</span> []</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> geneList[<span class="dv">0</span>][:<span class="dv">2</span>]<span class="op">!=</span><span class="st">'ch'</span>:</span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>        geneIds <span class="op">=</span> [<span class="bu">int</span>(gene.rstrip(<span class="st">'+'</span>))<span class="op">-</span><span class="dv">1</span> <span class="cf">for</span> gene <span class="kw">in</span> geneList]</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>        geneList <span class="op">=</span> [fullgraph.nodes[geneid] <span class="cf">for</span> geneid <span class="kw">in</span> geneIds]</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a>    coregraph_genes.nodesData[nodeIdx] <span class="op">=</span> <span class="st">''</span>.join(geneList)</span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb107-10"><a href="#cb107-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> accession, chainDict <span class="kw">in</span> coregraph.nodesAnnotation[nodeIdx].items():</span>
<span id="cb107-11"><a href="#cb107-11" aria-hidden="true" tabindex="-1"></a>        interval <span class="op">=</span> chainDict[nodeName]</span>
<span id="cb107-12"><a href="#cb107-12" aria-hidden="true" tabindex="-1"></a>        geneCumLengths <span class="op">=</span> np.hstack((<span class="dv">0</span>, np.cumsum([<span class="bu">len</span>(gene) <span class="cf">for</span> gene <span class="kw">in</span> geneList])))</span>
<span id="cb107-13"><a href="#cb107-13" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb107-14"><a href="#cb107-14" aria-hidden="true" tabindex="-1"></a>        coregraph.nodesAnnotation[nodeIdx][accession].<span class="op">\</span></span>
<span id="cb107-15"><a href="#cb107-15" aria-hidden="true" tabindex="-1"></a>            update({geneAnnotation:interval <span class="op">\</span></span>
<span id="cb107-16"><a href="#cb107-16" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">for</span> geneid <span class="kw">in</span> geneIds <span class="op">\</span></span>
<span id="cb107-17"><a href="#cb107-17" aria-hidden="true" tabindex="-1"></a>                        <span class="cf">for</span> geneAnnotation <span class="kw">in</span> fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:<span class="va">None</span>}).keys()})</span>
<span id="cb107-18"><a href="#cb107-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-19"><a href="#cb107-19" aria-hidden="true" tabindex="-1"></a>        coregraph_genes.nodesAnnotation[nodeIdx][accession].<span class="op">\</span></span>
<span id="cb107-20"><a href="#cb107-20" aria-hidden="true" tabindex="-1"></a>            update({geneAnnotation:[(geneCumLengths[i], geneCumLengths[i<span class="op">+</span><span class="dv">1</span>]<span class="op">-</span><span class="dv">1</span>)] <span class="op">\</span></span>
<span id="cb107-21"><a href="#cb107-21" aria-hidden="true" tabindex="-1"></a>                            <span class="cf">for</span> i,geneid <span class="kw">in</span> <span class="bu">enumerate</span>(geneIds) <span class="op">\</span></span>
<span id="cb107-22"><a href="#cb107-22" aria-hidden="true" tabindex="-1"></a>                                <span class="cf">for</span> geneAnnotation <span class="kw">in</span> fullgraph.nodesAnnotation[geneid].get(accession,{fullgraph.nodes[geneid]:<span class="va">None</span>}).keys()})</span>
<span id="cb107-23"><a href="#cb107-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-24"><a href="#cb107-24" aria-hidden="true" tabindex="-1"></a>        coregraph_genes.nodesAnnotation[nodeIdx][accession].<span class="op">\</span></span>
<span id="cb107-25"><a href="#cb107-25" aria-hidden="true" tabindex="-1"></a>            update({nodeName:[(geneCumLengths[<span class="dv">0</span>], geneCumLengths[<span class="op">-</span><span class="dv">1</span>]<span class="op">-</span><span class="dv">1</span>)]})</span>
<span id="cb107-26"><a href="#cb107-26" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">''</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>coregraph.toGFA(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_f2.1_Ref_v04.gfa'</span>,doSeq<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>coregraph_genes.toGFA(<span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_genes_f2.1_Ref_v04.gfa'</span>,doSeq<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>pathfileDir <span class="op">=</span> <span class="st">'../../1001G/coreGraph'</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>coreGFApath <span class="op">=</span> <span class="ss">f'</span><span class="sc">{</span>pathfileDir<span class="sc">}{</span>os<span class="sc">.</span>path<span class="sc">.</span>sep<span class="sc">}</span><span class="ss">coregraph_genes_f2.1_Ref_v04.gfa'</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a><span class="co"># coreGFApath = f'{pathfileDir}{os.path.sep}coreGraph.gfa'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [1,10,20,100,500,1000,5000,10000,50000,100000,500000,1000000]</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">9</span>,<span class="dv">45</span>,<span class="dv">90</span>,<span class="dv">450</span>,<span class="dv">900</span>,<span class="dv">4500</span>,<span class="dv">9000</span>]</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [1,3,9,18]</span></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomLevels = [4,8,16]</span></span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">True</span></span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb110-9"><a href="#cb110-9" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">16</span></span>
<span id="cb110-10"><a href="#cb110-10" aria-hidden="true" tabindex="-1"></a>invertionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb110-11"><a href="#cb110-11" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/pantograph/data/shorttest2.gfa'</span></span>
<span id="cb110-12"><a href="#cb110-12" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/pantograph/data/AT_Chr1_OGOnly.gfa'</span></span>
<span id="cb110-13"><a href="#cb110-13" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/chrisGraph/chr1.wfmash.n20.a90.s10000.p1,19,39,3,81,1.seqwish.sort.smooth.sort.gfa'</span></span>
<span id="cb110-14"><a href="#cb110-14" aria-hidden="true" tabindex="-1"></a><span class="co"># inputPath = '../../1001G/pantograph/data/shorttest_seq.gfa'</span></span>
<span id="cb110-15"><a href="#cb110-15" aria-hidden="true" tabindex="-1"></a>inputPath <span class="op">=</span> coreGFApath</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> adjustZoomLevels(zoomLevels)</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>outputPath,outputName <span class="op">=</span> pathConvert(inputPath,suffix<span class="op">=</span><span class="st">'_new'</span>)</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>outputPath,outputName,zoomLevels</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>dbid <span class="op">=</span> getDBID(<span class="st">'../pantograph_API/data/caseToDBID.dict'</span>,outputName)</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Opening Redis connection for db </span><span class="sc">{</span>dbid<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>redisConn <span class="op">=</span> Redis(host<span class="op">=</span><span class="st">'redis'</span>,port <span class="op">=</span> <span class="dv">6379</span>,db<span class="op">=</span>dbid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># zoomComponentLengths,zoomNodeToComponent,zoomComponentToNodes,zoomComponents,zoomCompNucleotides = \</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>exportToPantograph(inputPath<span class="op">=</span>inputPath,</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>                   outputName<span class="op">=</span>outputName,</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a>                   outputPath<span class="op">=</span>outputPath,</span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>                   isSeq<span class="op">=</span>isSeq,</span>
<span id="cb113-6"><a href="#cb113-6" aria-hidden="true" tabindex="-1"></a>                   redisConn<span class="op">=</span>redisConn,</span>
<span id="cb113-7"><a href="#cb113-7" aria-hidden="true" tabindex="-1"></a>                   GenomeGraphParams<span class="op">=</span>{<span class="st">'accessionsToRemove'</span>:[<span class="st">'Consensus'</span>]},</span>
<span id="cb113-8"><a href="#cb113-8" aria-hidden="true" tabindex="-1"></a>                   zoomLevels<span class="op">=</span>zoomLevels,</span>
<span id="cb113-9"><a href="#cb113-9" aria-hidden="true" tabindex="-1"></a>                   maxLengthChunk<span class="op">=</span>maxLengthChunk,</span>
<span id="cb113-10"><a href="#cb113-10" aria-hidden="true" tabindex="-1"></a>                   maxLengthComponent<span class="op">=</span>maxLengthComponent,</span>
<span id="cb113-11"><a href="#cb113-11" aria-hidden="true" tabindex="-1"></a>                   invertionThreshold<span class="op">=</span>invertionThreshold,)</span>
<span id="cb113-12"><a href="#cb113-12" aria-hidden="true" tabindex="-1"></a><span class="co">#                                              debug=True,returnDebugData=True)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ntfy send <span class="st">"Pantograph data generation for coregraph finished."</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exporting-to-pantograph-visualisation" class="level1">
<h1>Exporting to Pantograph visualisation</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>projectID <span class="op">=</span> <span class="st">'paths_genegraph'</span></span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>projectName <span class="op">=</span> <span class="st">'Example gene graph'</span></span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>pathToGraphs <span class="op">=</span> <span class="st">'examples/gene_graph'</span></span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>caseDict <span class="op">=</span> {<span class="st">'Main'</span>: <span class="st">'paths_genegraph.gfa'</span>}</span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>pathToIndex <span class="op">=</span> <span class="st">'examples/Visdata'</span></span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-7"><a href="#cb115-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This is if you run it in Docker compose together with active Redis image, which is named "redis".</span></span>
<span id="cb115-8"><a href="#cb115-8" aria-hidden="true" tabindex="-1"></a><span class="co"># If you have separate redis server, enter full address here.</span></span>
<span id="cb115-9"><a href="#cb115-9" aria-hidden="true" tabindex="-1"></a><span class="co"># If you do not want to add any annotation, `redisHost` should be None.</span></span>
<span id="cb115-10"><a href="#cb115-10" aria-hidden="true" tabindex="-1"></a>redisHost <span class="op">=</span> <span class="st">'redis'</span></span>
<span id="cb115-11"><a href="#cb115-11" aria-hidden="true" tabindex="-1"></a>redisPort <span class="op">=</span> <span class="dv">6379</span></span>
<span id="cb115-12"><a href="#cb115-12" aria-hidden="true" tabindex="-1"></a>redisDB <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb115-13"><a href="#cb115-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-14"><a href="#cb115-14" aria-hidden="true" tabindex="-1"></a>suffix <span class="op">=</span> <span class="st">''</span></span>
<span id="cb115-15"><a href="#cb115-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-16"><a href="#cb115-16" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb115-17"><a href="#cb115-17" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb115-18"><a href="#cb115-18" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb115-19"><a href="#cb115-19" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> <span class="va">False</span></span>
<span id="cb115-20"><a href="#cb115-20" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]</span>
<span id="cb115-21"><a href="#cb115-21" aria-hidden="true" tabindex="-1"></a>fillZoomLevel <span class="op">=</span> <span class="va">True</span></span>
<span id="cb115-22"><a href="#cb115-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb115-23"><a href="#cb115-23" aria-hidden="true" tabindex="-1"></a>exportProject(projectID, projectName, caseDict, pathToIndex, pathToGraphs,</span>
<span id="cb115-24"><a href="#cb115-24" aria-hidden="true" tabindex="-1"></a>              redisHost <span class="op">=</span> redisHost, redisPort <span class="op">=</span> redisPort, redisDB <span class="op">=</span> redisDB,</span>
<span id="cb115-25"><a href="#cb115-25" aria-hidden="true" tabindex="-1"></a>              suffix <span class="op">=</span> suffix,</span>
<span id="cb115-26"><a href="#cb115-26" aria-hidden="true" tabindex="-1"></a>              maxLengthComponent <span class="op">=</span> maxLengthComponent, maxLengthChunk <span class="op">=</span> maxLengthChunk,</span>
<span id="cb115-27"><a href="#cb115-27" aria-hidden="true" tabindex="-1"></a>              inversionThreshold <span class="op">=</span> inversionThreshold,</span>
<span id="cb115-28"><a href="#cb115-28" aria-hidden="true" tabindex="-1"></a>              isSeq <span class="op">=</span> isSeq,</span>
<span id="cb115-29"><a href="#cb115-29" aria-hidden="true" tabindex="-1"></a>              zoomLevels <span class="op">=</span> zoomLevels, fillZoomLevel <span class="op">=</span> fillZoomLevel):</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>projectID <span class="op">=</span> <span class="st">'tutorial_graph'</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>projectName <span class="op">=</span> <span class="st">'Example nucleotide graph'</span></span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>pathToGraphs <span class="op">=</span> <span class="st">'examples/nucleotide_graph'</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>caseDict <span class="op">=</span> {<span class="st">'Main'</span>: <span class="st">'paths_presentation.gfa'</span>}</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>pathToIndex <span class="op">=</span> <span class="st">'examples/Visdata'</span></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This is if you run it in Docker compose together with active Redis image, which is named "redis".</span></span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a><span class="co"># If you have separate redis server, enter full address here.</span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a><span class="co"># If you do not want to add any annotation, `redisHost` should be None.</span></span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>redisHost <span class="op">=</span> <span class="st">'redis'</span></span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>redisPort <span class="op">=</span> <span class="dv">6379</span></span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>redisDB <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>suffix <span class="op">=</span> <span class="st">''</span></span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>maxLengthComponent <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a>maxLengthChunk <span class="op">=</span> <span class="dv">6</span></span>
<span id="cb116-18"><a href="#cb116-18" aria-hidden="true" tabindex="-1"></a>inversionThreshold <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb116-19"><a href="#cb116-19" aria-hidden="true" tabindex="-1"></a>isSeq <span class="op">=</span> </span>
<span id="cb116-20"><a href="#cb116-20" aria-hidden="true" tabindex="-1"></a>zoomLevels <span class="op">=</span> [<span class="dv">1</span>]</span>
<span id="cb116-21"><a href="#cb116-21" aria-hidden="true" tabindex="-1"></a>fillZoomLevel <span class="op">=</span> <span class="va">True</span></span>
<span id="cb116-22"><a href="#cb116-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-23"><a href="#cb116-23" aria-hidden="true" tabindex="-1"></a>exportProject(projectID, projectName, caseDict, pathToIndex, pathToGraphs,</span>
<span id="cb116-24"><a href="#cb116-24" aria-hidden="true" tabindex="-1"></a>              redisHost <span class="op">=</span> redisHost, redisPort <span class="op">=</span> redisPort, redisDB <span class="op">=</span> redisDB,</span>
<span id="cb116-25"><a href="#cb116-25" aria-hidden="true" tabindex="-1"></a>              suffix <span class="op">=</span> suffix,</span>
<span id="cb116-26"><a href="#cb116-26" aria-hidden="true" tabindex="-1"></a>              maxLengthComponent <span class="op">=</span> maxLengthComponent, maxLengthChunk <span class="op">=</span> maxLengthChunk,</span>
<span id="cb116-27"><a href="#cb116-27" aria-hidden="true" tabindex="-1"></a>              inversionThreshold <span class="op">=</span> inversionThreshold,</span>
<span id="cb116-28"><a href="#cb116-28" aria-hidden="true" tabindex="-1"></a>              isSeq <span class="op">=</span> isSeq,</span>
<span id="cb116-29"><a href="#cb116-29" aria-hidden="true" tabindex="-1"></a>              zoomLevels <span class="op">=</span> zoomLevels, fillZoomLevel <span class="op">=</span> fillZoomLevel):</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>