<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="Provides functionality to construct special synteny graph from pangenome annotation (and actual sequences).">

<title>pygengraph - Synteny module</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="pygengraph - Synteny module">
<meta property="og:description" content="Provides functionality to construct special synteny graph from pangenome annotation (and actual sequences).">
<meta property="og:site-name" content="pygengraph">
<meta name="twitter:title" content="pygengraph - Synteny module">
<meta name="twitter:description" content="Provides functionality to construct special synteny graph from pangenome annotation (and actual sequences).">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">pygengraph</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./user_manual.html">
 <span class="menu-text">Pantograph user Manual</span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Synteny module</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">pygengraph</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./graph.html" class="sidebar-item-text sidebar-link">Graph module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tree.html" class="sidebar-item-text sidebar-link">Tree module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./synteny.html" class="sidebar-item-text sidebar-link active">Synteny module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./utils.html" class="sidebar-item-text sidebar-link">Utils module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./export.html" class="sidebar-item-text sidebar-link">Export module</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./scripts.html" class="sidebar-item-text sidebar-link">Script module</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#readtransmap" id="toc-readtransmap" class="nav-link active" data-scroll-target="#readtransmap">readTransMap</a></li>
  <li><a href="#generateorder" id="toc-generateorder" class="nav-link" data-scroll-target="#generateorder">generateOrder</a></li>
  <li><a href="#getids" id="toc-getids" class="nav-link" data-scroll-target="#getids">getIDs</a></li>
  <li><a href="#readgff" id="toc-readgff" class="nav-link" data-scroll-target="#readgff">readGFF</a></li>
  <li><a href="#addpangenomepositions" id="toc-addpangenomepositions" class="nav-link" data-scroll-target="#addpangenomepositions">addPangenomePositions</a></li>
  <li><a href="#processaccessions" id="toc-processaccessions" class="nav-link" data-scroll-target="#processaccessions">processAccessions</a></li>
  <li><a href="#recordsegment" id="toc-recordsegment" class="nav-link" data-scroll-target="#recordsegment">recordSegment</a></li>
  <li><a href="#recordannotation" id="toc-recordannotation" class="nav-link" data-scroll-target="#recordannotation">recordAnnotation</a></li>
  <li><a href="#recordaltchr" id="toc-recordaltchr" class="nav-link" data-scroll-target="#recordaltchr">recordAltChr</a></li>
  <li><a href="#addlink" id="toc-addlink" class="nav-link" data-scroll-target="#addlink">addLink</a></li>
  <li><a href="#generatepathslinks" id="toc-generatepathslinks" class="nav-link" data-scroll-target="#generatepathslinks">generatePathsLinks</a></li>
  <li><a href="#readsegmentids" id="toc-readsegmentids" class="nav-link" data-scroll-target="#readsegmentids">readSegmentIDs</a></li>
  <li><a href="#writesegmentids" id="toc-writesegmentids" class="nav-link" data-scroll-target="#writesegmentids">writeSegmentIDs</a></li>
  <li><a href="#writepath" id="toc-writepath" class="nav-link" data-scroll-target="#writepath">writePath</a></li>
  <li><a href="#writelinks" id="toc-writelinks" class="nav-link" data-scroll-target="#writelinks">writeLinks</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/pigrenok/pygengraph/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Synteny module</h1>
</div>

<div>
  <div class="description">
    Provides functionality to construct special synteny graph from pangenome annotation (and actual sequences).
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<hr>
<section id="readtransmap" class="level3">
<h3 class="anchored" data-anchor-id="readtransmap">readTransMap</h3>
<blockquote class="blockquote">
<pre><code> readTransMap (transMapFile, ATaccessionName='araport',
               similarityGroupColName='Orthogroup')</code></pre>
</blockquote>
<p>Read trandmap file which is pretty much a tsv file with first column containing orthogroup, or similarity group ID, and the rest of columns correspond to each accession with unique gene/interval IDs in each columns for the given orthogroup. If one accession has several genes in the given orthogroup, then they will be in a comma separated list.</p>
<hr>
</section>
<section id="generateorder" class="level3">
<h3 class="anchored" data-anchor-id="generateorder">generateOrder</h3>
<blockquote class="blockquote">
<pre><code> generateOrder (files, priorityAccession='TIAR10')</code></pre>
</blockquote>
<p>Genrate a list of files and float a file which contains specific priority accession name (or any given string) on top of the list.</p>
<p>It is not used any more, but left if will be needed in the future.</p>
<hr>
</section>
<section id="getids" class="level3">
<h3 class="anchored" data-anchor-id="getids">getIDs</h3>
<blockquote class="blockquote">
<pre><code> getIDs (iterator)</code></pre>
</blockquote>
<hr>
</section>
<section id="readgff" class="level3">
<h3 class="anchored" data-anchor-id="readgff">readGFF</h3>
<blockquote class="blockquote">
<pre><code> readGFF (gffFile:str)</code></pre>
</blockquote>
<p>Function which reads GFF3 file into a dict structure. For large GFF files random search is extremely slow and is impossible to use. This greatly speed up the process (from hours to seconds).</p>
<hr>
</section>
<section id="addpangenomepositions" class="level3">
<h3 class="anchored" data-anchor-id="addpangenomepositions">addPangenomePositions</h3>
<blockquote class="blockquote">
<pre><code> addPangenomePositions (pangenomeFiles)</code></pre>
</blockquote>
<hr>
</section>
<section id="processaccessions" class="level3">
<h3 class="anchored" data-anchor-id="processaccessions">processAccessions</h3>
<blockquote class="blockquote">
<pre><code> processAccessions (annotationFiles, ATmap=None, pangenomeDict=None,
                    similarityIDKey=None, similarityIDAssignment=None,
                    pangenomeFiles=None, sequenceFilesDict=None,
                    seqidJoinSym='_', ATsplitSym=',')</code></pre>
</blockquote>
<p>This function process only custom annotation for a group of accession with similarity IDs. It also creates separate ATmap and returns the last number of unmatchedID for consistency. There is a separate function for processing reference annotation.</p>
<p><code>sequenceFilesDict</code> can either be None or dict with accession IDs as keys and paths to FASTA files with sequences of annotated elements as values.</p>
<p><code>pangenomeFiles</code>: list[str] or None. If list of strings is provided, then its length should be the same length as <code>annotationFiles</code> and should be in the same order as in <code>annotationFiles</code>, i.e.&nbsp;<code>pangenomeFiles[i]</code> should correspond to <code>annotationFiles[i]</code></p>
<p><code>similarityIDAssignment</code>: str. Can be either ‘gene’ or ‘mRNA’. If it is not any of these an error will be raised.</p>
<hr>
</section>
<section id="recordsegment" class="level3">
<h3 class="anchored" data-anchor-id="recordsegment">recordSegment</h3>
<blockquote class="blockquote">
<pre><code> recordSegment (name, segmentIDs, segmentIDToNumDict, sequence=None,
                gfaFile=None, segmentData=None)</code></pre>
</blockquote>
<hr>
</section>
<section id="recordannotation" class="level3">
<h3 class="anchored" data-anchor-id="recordannotation">recordAnnotation</h3>
<blockquote class="blockquote">
<pre><code> recordAnnotation (nodeID, accessionID, sequenceID, chrID, start, end, og,
                   atList, sequence, nodesMetadata, pstart=-1, pend=-1)</code></pre>
</blockquote>
<hr>
</section>
<section id="recordaltchr" class="level3">
<h3 class="anchored" data-anchor-id="recordaltchr">recordAltChr</h3>
<blockquote class="blockquote">
<pre><code> recordAltChr (nodeID, accessionID, chrID, start, end, nodesMetadata)</code></pre>
</blockquote>
<hr>
</section>
<section id="addlink" class="level3">
<h3 class="anchored" data-anchor-id="addlink">addLink</h3>
<blockquote class="blockquote">
<pre><code> addLink (links, prevPathSegment, name, forward)</code></pre>
</blockquote>
<p><code>links</code>: mutable <code>prevPathSegment</code>: mutable</p>
<hr>
</section>
<section id="generatepathslinks" class="level3">
<h3 class="anchored" data-anchor-id="generatepathslinks">generatePathsLinks</h3>
<blockquote class="blockquote">
<pre><code> generatePathsLinks (genesAll, ATmap, accessionID, sequences, OGList,
                     segmentIDs, nodesMetadata, segmentIDToNumDict, links,
                     usCounter, chromosomeID=None, doUS=True,
                     segmentData=None, gfaFile=None)</code></pre>
</blockquote>
<p>This function takes a list of genes in specific format (<code>genesAll</code>) and some extra data and pretty much generates a graph (gene graph from annotations).</p>
<p><code>gfaFile</code>: file handle to write segments to GFA file <code>OGList</code>: mutable <code>links</code>: mutable <code>usCounter</code>: mutable</p>
<hr>
</section>
<section id="readsegmentids" class="level3">
<h3 class="anchored" data-anchor-id="readsegmentids">readSegmentIDs</h3>
<blockquote class="blockquote">
<pre><code> readSegmentIDs (path)</code></pre>
</blockquote>
<hr>
</section>
<section id="writesegmentids" class="level3">
<h3 class="anchored" data-anchor-id="writesegmentids">writeSegmentIDs</h3>
<blockquote class="blockquote">
<pre><code> writeSegmentIDs (path, segmentIDs)</code></pre>
</blockquote>
<hr>
</section>
<section id="writepath" class="level3">
<h3 class="anchored" data-anchor-id="writepath">writePath</h3>
<blockquote class="blockquote">
<pre><code> writePath (gfaFile, AccessionID, path, cigar, doCigars)</code></pre>
</blockquote>
<hr>
</section>
<section id="writelinks" class="level3">
<h3 class="anchored" data-anchor-id="writelinks">writeLinks</h3>
<blockquote class="blockquote">
<pre><code> writeLinks (gfaFile, links, doCigars=True)</code></pre>
</blockquote>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>